--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
@@ -5,6 +_,7 @@
 import com.google.common.io.BaseEncoding;
 import com.mojang.authlib.GameProfile;
 import com.mojang.datafixers.util.Pair;
+import de.erethon.papyrus.events.PlayerDataSaveEvent;
 import io.netty.buffer.Unpooled;
 import io.papermc.paper.FeatureHooks;
 import io.papermc.paper.entity.LookAnchor;
@@ -1530,6 +_,10 @@
 
     @Override
     public void saveData() {
+        // Papyrus start - Player saving
+        PlayerDataSaveEvent event = new PlayerDataSaveEvent(getHandle());
+        event.callEvent();
+        // Papyrus end
         this.server.getHandle().playerIo.save(this.getHandle());
     }
 
