--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftEnderDragonPart.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftEnderDragonPart.java
@@ -1,10 +_,11 @@
 package org.bukkit.craftbukkit.entity;
 
+import de.erethon.papyrus.CraftPDamageType;
+import de.erethon.papyrus.PDamageType;
+import net.minecraft.server.level.ServerLevel;
 import org.bukkit.craftbukkit.CraftServer;
 import org.bukkit.damage.DamageSource;
-import org.bukkit.entity.EnderDragon;
-import org.bukkit.entity.EnderDragonPart;
-import org.bukkit.entity.Entity;
+import org.bukkit.entity.*;
 
 public class CraftEnderDragonPart extends CraftComplexPart implements EnderDragonPart {
     public CraftEnderDragonPart(CraftServer server, net.minecraft.world.entity.boss.EnderDragonPart entity) {
@@ -82,4 +_,23 @@
     public void resetMaxHealth() {
         this.getParent().resetMaxHealth();
     }
+
+    // Papyrus start - Dragon isn't alive, so just deal normal damage I guess
+    @Override
+    public void damage(double amount, PDamageType type) {
+        getHandle().hurtServer((ServerLevel) getHandle().level(), getHandle().damageSources().generic(), (float) amount);
+    }
+
+    @Override
+    public void damage(double amount, Entity source, PDamageType type) {
+        net.minecraft.world.damagesource.DamageSource reason = this.getHandle().damageSources().generic();
+
+        if (source instanceof HumanEntity) {
+            reason = this.getHandle().damageSources().playerAttack(((CraftHumanEntity) source).getHandle());
+        } else if (source instanceof LivingEntity) {
+            reason = this.getHandle().damageSources().mobAttack(((CraftLivingEntity) source).getHandle());
+        }
+        getHandle().hurtServer((ServerLevel) getHandle().level(), reason, (float) amount);
+    }
+    // Papyrus end
 }
