--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
@@ -222,6 +_,8 @@
 import org.jspecify.annotations.NonNull;
 import org.jspecify.annotations.Nullable;
 
+import de.erethon.papyrus.events.PlayerDataSaveEvent;
+
 @DelegateDeserialization(CraftOfflinePlayer.class)
 public class CraftPlayer extends CraftHumanEntity implements Player, PluginMessageBridgeImpl {
     private static final org.slf4j.Logger LOGGER = LogUtils.getClassLogger();
@@ -1349,6 +_,10 @@
 
     @Override
     public void saveData() {
+        // Papyrus start - Player saving
+        PlayerDataSaveEvent event = new PlayerDataSaveEvent(getHandle());
+        event.callEvent();
+        // Papyrus end
         this.server.getHandle().playerIo.save(this.getHandle());
     }
 
