--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
@@ -204,6 +_,8 @@
 import org.bukkit.scoreboard.Scoreboard;
 import org.jetbrains.annotations.NotNull;
 
+import de.erethon.papyrus.events.PlayerDataSaveEvent;
+
 @DelegateDeserialization(CraftOfflinePlayer.class)
 public class CraftPlayer extends CraftHumanEntity implements Player {
 
@@ -1532,6 +_,10 @@
 
     @Override
     public void saveData() {
+        // Papyrus start - Player saving
+        PlayerDataSaveEvent event = new PlayerDataSaveEvent(getHandle());
+        event.callEvent();
+        // Papyrus end
         this.server.getHandle().playerIo.save(this.getHandle());
     }
 
