--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
@@ -208,6 +_,8 @@
 
 import net.md_5.bungee.api.chat.BaseComponent; // Spigot
 
+import de.erethon.papyrus.events.PlayerDataSaveEvent;
+
 @DelegateDeserialization(CraftOfflinePlayer.class)
 public class CraftPlayer extends CraftHumanEntity implements Player {
     private long firstPlayed = 0;
@@ -1540,6 +_,10 @@
 
     @Override
     public void saveData() {
+        // Papyrus start - Player saving
+        PlayerDataSaveEvent event = new PlayerDataSaveEvent(getHandle());
+        event.callEvent();
+        // Papyrus end
         this.server.getHandle().playerIo.save(this.getHandle());
     }
 
