--- a/src/main/java/org/bukkit/entity/Damageable.java
+++ b/src/main/java/org/bukkit/entity/Damageable.java
@@ -1,5 +_,6 @@
 package org.bukkit.entity;
 
+import de.erethon.papyrus.PDamageType;
 import org.bukkit.attribute.Attribute;
 import org.bukkit.damage.DamageSource;
 import org.jetbrains.annotations.ApiStatus;
@@ -120,4 +_,9 @@
      */
     @Deprecated(since = "1.11")
     void resetMaxHealth();
+
+    // Papyrus start
+    void damage(double amount, PDamageType type);
+    void damage(double amount, Entity source, PDamageType type);
+    // Papyrus end
 }
