--- a/src/main/java/org/bukkit/entity/LivingEntity.java
+++ b/src/main/java/org/bukkit/entity/LivingEntity.java
@@ -29,7 +_,7 @@
 /**
  * Represents a living entity, such as a monster or player
  */
-public interface LivingEntity extends Attributable, Damageable, ProjectileSource, io.papermc.paper.entity.Frictional { // Paper
+public interface LivingEntity extends Attributable, Damageable, ProjectileSource, io.papermc.paper.entity.Frictional, de.erethon.spellbook.api.SpellCaster { // Papyrus
 
     /**
      * Gets the height of the living entity's eyes above its Location.
@@ -1466,6 +_,12 @@
      */
     boolean canUseEquipmentSlot(org.bukkit.inventory.@NotNull EquipmentSlot slot);
     // Paper end - Expose canUseSlot
+
+    // Papyrus start - Actionbar utils
+    void sendParsedActionBar(String text);
+    void sendParsedMessage(String text);
+    // Papyrus end
+
 
     /**
      * Gets the entity's combat tracker
