--- a/src/main/java/org/bukkit/entity/LivingEntity.java
+++ b/src/main/java/org/bukkit/entity/LivingEntity.java
@@ -4,6 +_,8 @@
 import java.util.List;
 import java.util.Set;
 import java.util.UUID;
+
+import de.erethon.spellbook.api.SpellCaster;
 import org.bukkit.FluidCollisionMode;
 import org.bukkit.Location;
 import org.bukkit.Material;
@@ -27,7 +_,7 @@
 /**
  * Represents a living entity, such as a monster or player
  */
-public interface LivingEntity extends Attributable, Damageable, ProjectileSource, io.papermc.paper.entity.Frictional { // Paper
+public interface LivingEntity extends Attributable, Damageable, ProjectileSource, io.papermc.paper.entity.Frictional, SpellCaster { // Papyrus
 
     /**
      * Gets the height of the living entity's eyes above its Location.
