--- a/src/main/java/org/bukkit/Server.java
+++ b/src/main/java/org/bukkit/Server.java
@@ -14,6 +_,8 @@
 import java.util.UUID;
 import java.util.function.Consumer;
 import java.util.logging.Logger;
+
+import de.erethon.spellbook.api.SpellbookAPI;
 import net.kyori.adventure.text.Component;
 import net.kyori.adventure.text.minimessage.MiniMessage;
 import net.kyori.adventure.text.minimessage.tag.resolver.TagResolver;
@@ -2698,4 +_,6 @@
      */
     void allowPausing(@NotNull org.bukkit.plugin.Plugin plugin, boolean value);
     // Paper end - API to check if the server is sleeping
+
+    SpellbookAPI getSpellbookAPI(); // Papyrus - Spellbook API
 }
