--- a/net/minecraft/world/level/storage/PlayerDataStorage.java
+++ b/net/minecraft/world/level/storage/PlayerDataStorage.java
@@ -8,6 +_,8 @@
 import java.nio.file.StandardCopyOption;
 import java.time.ZonedDateTime;
 import java.util.Optional;
+
+import de.erethon.papyrus.events.PlayerDataRequestEvent;
 import net.minecraft.nbt.CompoundTag;
 import net.minecraft.nbt.NbtAccounter;
 import net.minecraft.nbt.NbtIo;
@@ -91,6 +_,13 @@
     }
 
     public Optional<CompoundTag> load(NameAndId nameAndId) {
+        // Papyrus start - Player database
+        PlayerDataRequestEvent requestEvent = new PlayerDataRequestEvent(nameAndId.id().toString());
+        requestEvent.callEvent();
+        if (requestEvent.getData() != null) { // Null -> No database present. Empty -> Database has no data for that player. We still don't want to run the Bukkit logic though.
+            return requestEvent.getData();
+        }
+        // Papyrus end
         Optional<CompoundTag> optional = this.load(nameAndId, ".dat");
         if (optional.isEmpty()) {
             this.backup(nameAndId, ".dat");
