--- a/net/minecraft/server/players/PlayerList.java
+++ b/net/minecraft/server/players/PlayerList.java
@@ -19,6 +_,8 @@
 import java.util.function.Function;
 import java.util.function.Predicate;
 import javax.annotation.Nullable;
+
+import de.erethon.papyrus.events.PlayerDataSaveEvent;
 import net.minecraft.ChatFormatting;
 import net.minecraft.FileUtil;
 import net.minecraft.commands.CommandSourceStack;
@@ -406,7 +_,11 @@
     protected void save(ServerPlayer player) {
         if (!player.getBukkitEntity().isPersistent()) return; // CraftBukkit
         player.lastSave = MinecraftServer.currentTick; // Paper - Incremental chunk and player saving
-        this.playerIo.save(player);
+        // Papyrus start - Saving
+        PlayerDataSaveEvent event = new PlayerDataSaveEvent(player);
+        event.callEvent();
+        //this.playerIo.save(player);
+        // Papyrus end
         ServerStatsCounter serverStatsCounter = player.getStats(); // CraftBukkit
         if (serverStatsCounter != null) {
             serverStatsCounter.save();
