--- a/src/main/java/org/bukkit/craftbukkit/CraftServer.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
@@ -12,10 +_,12 @@
 import com.mojang.brigadier.exceptions.CommandSyntaxException;
 import com.mojang.serialization.Dynamic;
 import com.mojang.serialization.Lifecycle;
+import de.erethon.papyrus.world.ErethonVoidGenerator;
 import io.papermc.paper.configuration.GlobalConfiguration;
 import io.papermc.paper.configuration.PaperServerConfiguration;
 import io.papermc.paper.configuration.ServerConfiguration;
 import io.papermc.paper.world.PaperWorldLoader;
+import de.erethon.spellbook.api.SpellbookAPI;
 import it.unimi.dsi.fastutil.objects.Object2IntOpenHashMap;
 import java.awt.image.BufferedImage;
 import java.io.ByteArrayOutputStream;
@@ -388,6 +_,10 @@
     }
     // Paper end - Folia reagion threading API
 
+    // Papyrus start
+    private SpellbookAPI spellbookAPI;
+    // Papyrus end
+
     static {
         ConfigurationSerialization.registerClass(CraftOfflinePlayer.class);
         ConfigurationSerialization.registerClass(CraftPlayerProfile.class);
@@ -487,6 +_,15 @@
         this.spark = new io.papermc.paper.SparksFly(this); // Paper - spark
     }
 
+    @Override
+    public SpellbookAPI getSpellbookAPI() {
+        return spellbookAPI;
+    }
+
+    public void setSpellbookAPI(SpellbookAPI api) {
+        this.spellbookAPI = api;
+    }
+
     public boolean getCommandBlockOverride(String command) {
         return this.overrideAllCommandBlockCommands || this.commandsConfiguration.getStringList("command-block-overrides").contains(command);
     }
@@ -1750,6 +_,12 @@
     }
 
     public ChunkGenerator getGenerator(String world) {
+        // Papyrus start - Void gen
+        if (world.equalsIgnoreCase("Erethon")) {
+            this.getLogger().info("Using ErethonVoidGenerator for out-of-bounds generation...");
+            return new ErethonVoidGenerator();
+        }
+        // Papyrus end
         ConfigurationSection section = this.configuration.getConfigurationSection("worlds");
         ChunkGenerator result = null;
 
