--- a/net/minecraft/world/ContainerHelper.java
+++ b/net/minecraft/world/ContainerHelper.java
@@ -6,6 +_,7 @@
 import net.minecraft.world.item.ItemStack;
 import net.minecraft.world.level.storage.ValueInput;
 import net.minecraft.world.level.storage.ValueOutput;
+import org.bukkit.Bukkit;
 
 public class ContainerHelper {
     public static final String TAG_ITEMS = "Items";
@@ -43,6 +_,8 @@
                 items.set(itemStackWithSlot.slot(), itemStackWithSlot.stack());
             }
         }
+        de.erethon.papyrus.events.ContainerLoadEvent event = new de.erethon.papyrus.events.ContainerLoadEvent(items, !Bukkit.isPrimaryThread());
+        Bukkit.getPluginManager().callEvent(event);
     }
 
     public static int clearOrCountMatchingItems(Container container, Predicate<ItemStack> itemPredicate, int maxItems, boolean simulate) {
