--- a/src/main/java/org/bukkit/event/entity/EntityDamageByEntityEvent.java
+++ b/src/main/java/org/bukkit/event/entity/EntityDamageByEntityEvent.java
@@ -2,6 +_,8 @@
 
 import com.google.common.base.Function;
 import java.util.Map;
+
+import de.erethon.papyrus.PDamageType;
 import org.bukkit.damage.DamageSource;
 import org.bukkit.damage.DamageType;
 import org.bukkit.entity.Entity;
@@ -16,6 +_,8 @@
     private final Entity damager;
     private final boolean critical;
 
+    private PDamageType damageType; // Papyrus
+
     @ApiStatus.Internal
     @Deprecated(since = "1.20.4", forRemoval = true)
     public EntityDamageByEntityEvent(@NotNull final Entity damager, @NotNull final Entity damagee, @NotNull final DamageCause cause, final double damage) {
@@ -51,6 +_,15 @@
         this.critical = critical;
     }
 
+    // Papyrus start - add damage type
+    public EntityDamageByEntityEvent(@NotNull final Entity damager, @NotNull final Entity damagee, @NotNull final DamageCause cause, @NotNull final DamageSource damageSource, final double damage, PDamageType type) {
+        super(damagee, cause, damageSource, damage);
+        this.damager = damager;
+        this.critical = false; // Paper - add critical damage API
+        this.damageType = type;
+    }
+    // Papyrus end
+
     /**
      * Returns the entity that damaged the defender.
      *
@@ -84,4 +_,14 @@
     public @NotNull DamageSource getDamageSource() {
         return super.getDamageSource();
     }
+
+    // Papyrus start - add damage type
+    public PDamageType getDamageType() {
+       return damageType;
+    }
+
+    public void setDamageType(PDamageType type) {
+        damageType = type;
+    }
+    // Papyrus end
 }
