--- a/net/minecraft/network/protocol/game/ClientboundContainerSetSlotPacket.java
+++ b/net/minecraft/network/protocol/game/ClientboundContainerSetSlotPacket.java
@@ -33,7 +_,7 @@
         output.writeContainerId(this.containerId);
         output.writeVarInt(this.stateId);
         output.writeShort(this.slot);
-        ItemStack.OPTIONAL_STREAM_CODEC.encode(output, this.itemStack);
+        ItemStack.OPTIONAL_TRANSLATED_STREAM_CODEC.encode(output, this.itemStack);
     }
 
     @Override
