--- a/net/minecraft/network/protocol/game/ClientboundContainerSetContentPacket.java
+++ b/net/minecraft/network/protocol/game/ClientboundContainerSetContentPacket.java
@@ -15,9 +_,9 @@
         ClientboundContainerSetContentPacket::containerId,
         ByteBufCodecs.VAR_INT,
         ClientboundContainerSetContentPacket::stateId,
-        ItemStack.OPTIONAL_LIST_STREAM_CODEC,
+        ItemStack.OPTIONAL_TRANSLATED_LIST_STREAM_CODEC,
         ClientboundContainerSetContentPacket::items,
-        ItemStack.OPTIONAL_STREAM_CODEC,
+        ItemStack.OPTIONAL_TRANSLATED_STREAM_CODEC,
         ClientboundContainerSetContentPacket::carriedItem,
         ClientboundContainerSetContentPacket::new
     );
