--- a/net/minecraft/world/level/pathfinder/Path.java
+++ b/net/minecraft/world/level/pathfinder/Path.java
@@ -12,7 +_,7 @@
 import net.minecraft.world.phys.Vec3;
 import org.jspecify.annotations.Nullable;
 
-public final class Path {
+public class Path { // Papyrus - Allow pending asynchronous paths
     public static final StreamCodec<ByteBuf, Path> DEBUG_STREAM_CODEC = StreamCodec.composite(
         ByteBufCodecs.BOOL,
         p -> p.reached,
