From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Jason Penilla <11360596+jpenilla@users.noreply.github.com>
Date: Sun, 15 Dec 2024 00:07:29 -0700
Subject: [PATCH] feature 1


diff --git a/src/main/java/com/destroystokyo/paper/block/CraftBlockSoundGroup.java b/src/main/java/com/destroystokyo/paper/block/CraftBlockSoundGroup.java
index 054ee606e7c36712e474e2b7093b436e043e6262..9d395898511ed5b408dacfb9c796baacb18b9430 100644
--- a/src/main/java/com/destroystokyo/paper/block/CraftBlockSoundGroup.java
+++ b/src/main/java/com/destroystokyo/paper/block/CraftBlockSoundGroup.java
@@ -20,6 +20,7 @@ public class CraftBlockSoundGroup implements BlockSoundGroup {
     @Override
     public Sound getStepSound() {
         // file patch
+        // feature patch
         return CraftSound.minecraftToBukkit(soundEffectType.getStepSound());
     }
 
