--- a/net/minecraft/world/entity/player/Inventory.java
+++ b/net/minecraft/world/entity/player/Inventory.java
@@ -4,6 +_,7 @@
 import it.unimi.dsi.fastutil.ints.Int2ObjectMap;
 import java.util.Map;
 import java.util.function.Predicate;
+
 import net.minecraft.CrashReport;
 import net.minecraft.CrashReportCategory;
 import net.minecraft.ReportedException;
@@ -24,6 +_,7 @@
 import net.minecraft.world.item.ItemStack;
 import net.minecraft.world.level.storage.ValueInput;
 import net.minecraft.world.level.storage.ValueOutput;
+import org.bukkit.Bukkit;
 
 public class Inventory implements Container, Nameable {
     public static final int POP_TIME_DURATION = 5;
@@ -506,6 +_,10 @@
                 this.setItem(itemStackWithSlot.slot(), itemStackWithSlot.stack());
             }
         }
+        // Papyrus start - Inventory API
+        de.erethon.papyrus.events.PlayerInventoryLoadEvent loadEvent = new de.erethon.papyrus.events.PlayerInventoryLoadEvent(player, items, equipment);
+        Bukkit.getPluginManager().callEvent(loadEvent);
+        // Papyrus end
     }
 
     @Override
