--- a/net/minecraft/world/level/NaturalSpawner.java
+++ b/net/minecraft/world/level/NaturalSpawner.java
@@ -1,6 +_,9 @@
 package net.minecraft.world.level;
 
 import com.mojang.logging.LogUtils;
+import de.erethon.papyrus.events.MobSpawnForBiomeEvent;
+import de.erethon.papyrus.events.PreCategorySpawnEvent;
+import de.erethon.papyrus.events.MobSpawnFinalizedEvent;
 import it.unimi.dsi.fastutil.objects.Object2IntMap;
 import it.unimi.dsi.fastutil.objects.Object2IntMaps;
 import it.unimi.dsi.fastutil.objects.Object2IntOpenHashMap;
@@ -87,7 +_,11 @@
 
         for (Entity entity : entities) {
             if (!(entity instanceof Mob mob && (mob.isPersistenceRequired() || mob.requiresCustomPersistence()))) {
-                MobCategory category = entity.getType().getCategory();
+                // Papyrus start
+                final MobCategory category = (entity instanceof de.erethon.papyrus.entities.MobSpawnCategoryOverrideProvider custom)
+                ? custom.getEffectiveMobCategory()
+                : entity.getType().getCategory();
+                // Papyrus end
                 if (category != MobCategory.MISC) {
                     // Paper start - Only count natural spawns
                     if (!entity.level().paperConfig().entities.spawning.countAllMobsForSpawning &&
@@ -255,17 +_,20 @@
         // Paper start - Optional per player mob spawns
         spawnCategoryForPosition(mobCategory, level, chunk, start, extraTest, spawnCallback, Integer.MAX_VALUE, null);
     }
-    public static void spawnCategoryForPosition(
-        final MobCategory mobCategory,
-        final ServerLevel level,
-        final ChunkAccess chunk,
-        final BlockPos start,
-        final NaturalSpawner.SpawnPredicate extraTest,
-        final NaturalSpawner.AfterSpawnCallback spawnCallback,
-        final int maxSpawns,
-        final @Nullable Consumer<Entity> trackEntity
+    public static void spawnCategoryForPosition(MobCategory mobCategory, ServerLevel level, ChunkAccess chunk, BlockPos start, NaturalSpawner.SpawnPredicate extraTest, NaturalSpawner.AfterSpawnCallback spawnCallback, int maxSpawns, @Nullable Consumer<Entity> trackEntity
     ) {
-        // Paper end - Optional per player mob spawns
+        PreCategorySpawnEvent event = new PreCategorySpawnEvent(mobCategory, level, chunk, start, extraTest, spawnCallback, maxSpawns, trackEntity);
+        event.callEvent();
+        if (event.isShouldAbortVanillaLogic()) {
+            return;
+        }
+        level = event.getLevel();
+        chunk = event.getChunk();
+        start = event.getPos();
+        extraTest = event.getFilter();
+        spawnCallback = event.getCallback();
+        maxSpawns = event.getMaxSpawns();
+        trackEntity = event.getTrackEntity();
         StructureManager structureManager = level.structureManager();
         ChunkGenerator generator = level.getChunkSource().getGenerator();
         int yStart = start.getY();
@@ -328,6 +_,9 @@
                                     groupData = mob.finalizeSpawn(
                                         level, level.getCurrentDifficultyAt(mob.blockPosition()), EntitySpawnReason.NATURAL, groupData
                                     );
+                                    MobSpawnFinalizedEvent spawnEvent = new MobSpawnFinalizedEvent(mobCategory, level, chunk, pos, mob);
+                                    spawnEvent.callEvent();
+                                    mob = spawnEvent.getMob();
                                     // CraftBukkit start
                                     // SPIGOT-7045: Give ocelot babies back their special spawn reason. Note: This is the only modification required as ocelots count as monsters which means they only spawn during normal chunk ticking and do not spawn during chunk generation as starter mobs.
                                     level.addFreshEntityWithPassengers(mob, (mob instanceof net.minecraft.world.entity.animal.feline.Ocelot && !((org.bukkit.entity.Ageable) mob.getBukkitEntity()).isAdult()) ? org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason.OCELOT_BABY : org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason.NATURAL);
@@ -472,9 +_,14 @@
         final BlockPos pos,
         final @Nullable Holder<Biome> biome
     ) {
-        return isInNetherFortressBounds(pos, level, mobCategory, structureManager)
-            ? NetherFortressStructure.FORTRESS_ENEMIES
-            : generator.getMobsAt(biome != null ? biome : level.getBiome(pos), structureManager, mobCategory, pos);
+        // Papyrus start - mob spawn event
+        WeightedList<MobSpawnSettings.SpawnerData> spawnerData = isInNetherFortressBounds(pos, level, mobCategory, structureManager)
+                ? NetherFortressStructure.FORTRESS_ENEMIES
+                : generator.getMobsAt(biome != null ? biome : level.getBiome(pos), structureManager, mobCategory, pos);
+        MobSpawnForBiomeEvent mobSpawnForBiomeEvent = new MobSpawnForBiomeEvent(level, structureManager, generator, mobCategory, pos, biome, spawnerData);
+        mobSpawnForBiomeEvent.callEvent();
+        return mobSpawnForBiomeEvent.getSpawnerData();
+        // Papyrus end
     }
 
     public static boolean isInNetherFortressBounds(
@@ -672,7 +_,11 @@
             }
 
             this.spawnPotential.addCharge(pos, charge);
-            MobCategory category = type.getCategory();
+            // Papyrus start
+            final MobCategory category = (mob instanceof de.erethon.papyrus.entities.MobSpawnCategoryOverrideProvider custom)
+                    ? custom.getEffectiveMobCategory()
+                    : mob.getType().getCategory();
+            // Papyrus end
             this.mobCategoryCounts.addTo(category, 1);
             if (this.localMobCapCalculator != null) this.localMobCapCalculator.addMob(ChunkPos.containing(pos), category); // Paper - Optional per player mob spawns
         }
