--- a/net/minecraft/world/level/block/BaseCoralWallFanBlock.java
+++ b/net/minecraft/world/level/block/BaseCoralWallFanBlock.java
@@ -65,6 +_,7 @@
         final BlockState neighbourState,
         final RandomSource random
     ) {
+        if (true) return state; // Papyrus - Disable coral plant death
         if (state.getValue(WATERLOGGED)) {
             ticks.scheduleTick(pos, Fluids.WATER, Fluids.WATER.getTickDelay(level));
         }
@@ -74,6 +_,7 @@
 
     @Override
     protected boolean canSurvive(final BlockState state, final LevelReader level, final BlockPos pos) {
+        if (true) return true; // Papyrus - Disable coral plant death
         Direction facing = state.getValue(FACING);
         BlockPos relativePos = pos.relative(facing.getOpposite());
         BlockState relativeState = level.getBlockState(relativePos);
