--- a/net/minecraft/world/level/block/BaseCoralPlantTypeBlock.java
+++ b/net/minecraft/world/level/block/BaseCoralPlantTypeBlock.java
@@ -33,12 +_,14 @@
     protected abstract MapCodec<? extends BaseCoralPlantTypeBlock> codec();
 
     protected void tryScheduleDieTick(BlockState state, BlockGetter level, ScheduledTickAccess scheduledTickAccess, RandomSource random, BlockPos pos) {
+        if (true) return; // Papyrus - Disable coral plant death
         if (!scanForWater(state, level, pos)) {
             scheduledTickAccess.scheduleTick(pos, this, 60 + random.nextInt(40));
         }
     }
 
     protected static boolean scanForWater(BlockState state, BlockGetter level, BlockPos pos) {
+        if (true) return true; // Papyrus - Disable coral plant death
         if (state.getValue(WATERLOGGED)) {
             return true;
         } else {
@@ -75,6 +_,7 @@
         BlockState neighborState,
         RandomSource random
     ) {
+        if (true) return state; // Papyrus - Disable coral plant death
         if (state.getValue(WATERLOGGED)) {
             scheduledTickAccess.scheduleTick(pos, Fluids.WATER, Fluids.WATER.getTickDelay(level));
         }
@@ -86,6 +_,7 @@
 
     @Override
     protected boolean canSurvive(BlockState state, LevelReader level, BlockPos pos) {
+        if (true) return true; // Papyrus - Disable state change
         BlockPos blockPos = pos.below();
         return level.getBlockState(blockPos).isFaceSturdy(level, blockPos, Direction.UP);
     }
