--- a/net/minecraft/world/level/block/ChorusFlowerBlock.java
+++ b/net/minecraft/world/level/block/ChorusFlowerBlock.java
@@ -45,6 +_,7 @@
 
     @Override
     protected void tick(final BlockState state, final ServerLevel level, final BlockPos pos, final RandomSource random) {
+        if (true) return; // Papyrus - Disable state update
         if (!state.canSurvive(level, pos)) {
             level.destroyBlock(pos, true);
         }
@@ -52,7 +_,8 @@
 
     @Override
     protected boolean isRandomlyTicking(final BlockState state) {
-        return state.getValue(AGE) < 5;
+        //return state.getValue(AGE) < 5;
+        return false;
     }
 
     @Override
@@ -62,6 +_,7 @@
 
     @Override
     protected void randomTick(final BlockState state, final ServerLevel level, final BlockPos pos, final RandomSource random) {
+        if (true) return; // Papyrus - Disable state update
         BlockPos above = pos.above();
         if (level.isEmptyBlock(above) && above.getY() <= level.getMaxY()) {
             int currentAge = state.getValue(AGE);
@@ -164,6 +_,7 @@
         final BlockState neighbourState,
         final RandomSource random
     ) {
+        if (true) return state; // Papyrus - Disable state update
         if (directionToNeighbour != Direction.UP && !state.canSurvive(level, pos)) {
             ticks.scheduleTick(pos, this, 1);
         }
@@ -173,6 +_,7 @@
 
     @Override
     protected boolean canSurvive(final BlockState state, final LevelReader level, final BlockPos pos) {
+        if (true) return true; // Papyrus - Disable state update
         BlockState belowState = level.getBlockState(pos.below());
         if (!belowState.is(this.plant) && !belowState.is(BlockTags.SUPPORTS_CHORUS_FLOWER)) {
             if (!belowState.isAir()) {
@@ -263,6 +_,7 @@
 
     @Override
     protected void onProjectileHit(final Level level, final BlockState state, final BlockHitResult blockHit, final Projectile projectile) {
+        if (true) return; // Papyrus - Disable state update
         BlockPos pos = blockHit.getBlockPos();
         if (level instanceof ServerLevel serverLevel && projectile.mayInteract(serverLevel, pos) && projectile.mayBreak(serverLevel)) {
             // CraftBukkit start
