--- a/net/minecraft/world/level/block/CoralBlock.java
+++ b/net/minecraft/world/level/block/CoralBlock.java
@@ -36,6 +_,7 @@
 
     @Override
     protected void tick(final BlockState state, final ServerLevel level, final BlockPos pos, final RandomSource random) {
+        if (true) return; // Papyrus - don't tick coral
         if (!this.scanForWater(level, pos)) {
             // CraftBukkit start
             if (org.bukkit.craftbukkit.event.CraftEventFactory.callBlockFadeEvent(level, pos, this.deadBlock.defaultBlockState()).isCancelled()) {
@@ -57,6 +_,7 @@
         final BlockState neighbourState,
         final RandomSource random
     ) {
+        if (true) return state; // Papyrus - Disable state update
         if (!this.scanForWater(level, pos)) {
             ticks.scheduleTick(pos, this, 60 + random.nextInt(40));
         }
@@ -65,6 +_,7 @@
     }
 
     protected boolean scanForWater(final BlockGetter level, final BlockPos blockPos) {
+        if (true) return true; // Papyrus - Disable state update
         for (Direction direction : Direction.values()) {
             FluidState fluidState = level.getFluidState(blockPos.relative(direction));
             if (fluidState.is(FluidTags.WATER)) {
@@ -77,6 +_,7 @@
 
     @Override
     public @Nullable BlockState getStateForPlacement(final BlockPlaceContext context) {
+        if (true) return this.defaultBlockState(); // Papyrus - Disable state update
         if (!this.scanForWater(context.getLevel(), context.getClickedPos())) {
             context.getLevel().scheduleTick(context.getClickedPos(), this, 60 + context.getLevel().getRandom().nextInt(40));
         }
