--- a/net/minecraft/world/level/block/NoteBlock.java
+++ b/net/minecraft/world/level/block/NoteBlock.java
@@ -119,6 +_,7 @@
         final InteractionHand hand,
         final BlockHitResult hitResult
     ) {
+        if (true) return InteractionResult.PASS; // Papyrus - Disable state update
         return (InteractionResult)(itemStack.is(ItemTags.NOTE_BLOCK_TOP_INSTRUMENTS) && hitResult.getDirection() == Direction.UP
             ? InteractionResult.PASS
             : super.useItemOn(itemStack, state, level, pos, player, hand, hitResult));
