--- a/net/minecraft/world/level/block/ShelfBlock.java
+++ b/net/minecraft/world/level/block/ShelfBlock.java
@@ -90,7 +_,8 @@
 
     @Override
     public @Nullable BlockEntity newBlockEntity(final BlockPos worldPosition, final BlockState blockState) {
-        return new ShelfBlockEntity(worldPosition, blockState);
+        return null;
+        //return new ShelfBlockEntity(worldPosition, blockState);
     }
 
     @Override
