--- a/net/minecraft/world/entity/EntityType.java
+++ b/net/minecraft/world/entity/EntityType.java
@@ -2,6 +_,8 @@
 
 import com.mojang.logging.LogUtils;
 import com.mojang.serialization.Codec;
+import java.util.HashMap;
+import java.util.Map;
 import java.util.Optional;
 import java.util.stream.Stream;
 import net.minecraft.core.BlockPos;
@@ -48,6 +_,7 @@
 import net.minecraft.world.phys.Vec3;
 import net.minecraft.world.phys.shapes.Shapes;
 import net.minecraft.world.phys.shapes.VoxelShape;
+import org.bukkit.plugin.Plugin;
 import org.jspecify.annotations.Nullable;
 import org.slf4j.Logger;
 
@@ -57,6 +_,7 @@
     private final Holder.Reference<EntityType<?>> builtInRegistryHolder = BuiltInRegistries.ENTITY_TYPE.createIntrusiveHolder(this);
     public static final Codec<EntityType<?>> CODEC = BuiltInRegistries.ENTITY_TYPE.byNameCodec();
     public static final StreamCodec<RegistryFriendlyByteBuf, EntityType<?>> STREAM_CODEC = ByteBufCodecs.registry(Registries.ENTITY_TYPE);
+    public static final Map<String, Map.Entry<Plugin, Class<? extends Entity>>> customEntities = new HashMap<>(); // Papyrus - Custom Entities
     public static final int NO_UPDATE_INTERVAL = Integer.MAX_VALUE;
     private final EntityType.EntityFactory<T> factory;
     private final MobCategory category;
@@ -392,7 +_,21 @@
     public static Optional<Entity> create(final ValueInput input, final Level level, final EntitySpawnRequest request, final boolean generation) {
         // Paper end - Don't fire sync event during generation
         return Util.ifElse(
-            by(input).map(type -> type.create(level, request)),
+            by(input).map(type -> {
+                // Papyrus start - Custom entities
+                Optional<String> customEntityId = input.getString("papyrus-entity-id");
+                if (customEntityId.isPresent()) {
+                    Map.Entry<Plugin, Class<? extends Entity>> entry = customEntities.get(customEntityId.get());
+                    if (entry != null) {
+                        if (type.getBaseClass().isAssignableFrom(entry.getValue())) {
+                            return createCustomEntity(entry.getValue(), type, entry.getKey(), level, request);
+                        }
+                        LOGGER.warn("Entity type {} is not assignable from custom entity class {}", type, entry.getValue());
+                    }
+                }
+                // Papyrus end
+                return type.create(level, request);
+            }),
             // Paper start - Don't fire sync event during generation
             entity -> {
                 if (generation) entity.generation = true; // Paper - Don't fire sync event during generation
@@ -572,6 +_,18 @@
     public boolean onlyOpCanSetNbt() {
         return EntityTypes.OP_ONLY_CUSTOM_DATA.contains(this);
     }
+
+    // Papyrus start - Custom entities
+    public static Entity createCustomEntity(Class<? extends Entity> customEntityClass, EntityType<?> type, Plugin plugin, Level level, EntitySpawnRequest request) {
+        try {
+            return customEntityClass.getConstructor(EntityType.class, Level.class).newInstance(type, level);
+        } catch (Exception e) {
+            LOGGER.warn("Failed to create custom entity", e);
+            e.printStackTrace();
+            return type.create(level, request);
+        }
+    }
+    // Papyrus end
 
     public static class Builder<T extends Entity> {
         private final EntityType.EntityFactory<T> factory;
