--- a/net/minecraft/world/entity/Mob.java
+++ b/net/minecraft/world/entity/Mob.java
@@ -1594,26 +_,36 @@
 
     @Override
     public boolean doHurtTarget(final ServerLevel level, final Entity target) {
-        float dmg = (float)this.getAttributeValue(Attributes.ATTACK_DAMAGE);
+        /* Papyrus start - Add attack damage type, remove enchantment handling
+         float f = (float)this.getAttributeValue(Attributes.ATTACK_DAMAGE);
+         ItemStack weaponItem = this.getWeaponItem();
+         DamageSource damageSource = Optional.ofNullable(weaponItem.getItem().getDamageSource(this)).orElse(this.damageSources().mobAttack(this));
+         f = EnchantmentHelper.modifyDamage(level, weaponItem, source, damageSource, f);
+         f += weaponItem.getItem().getAttackDamageBonus(source, f, damageSource);
+        boolean flag = source.hurtServer(level, damageSource, f);
+        */
+        float f = 0;
         ItemStack weaponItem = this.getWeaponItem();
         DamageSource damageSource = weaponItem.getDamageSource(this, () -> this.damageSources().mobAttack(this));
-        dmg = EnchantmentHelper.modifyDamage(level, weaponItem, target, damageSource, dmg);
-        dmg += weaponItem.getItem().getAttackDamageBonus(target, dmg, damageSource);
-        Vec3 oldMovement = target.getDeltaMovement();
-        boolean wasHurt = target.hurtServer(level, damageSource, dmg);
-        if (wasHurt) {
-            this.causeExtraKnockback(target, this.getKnockback(target, damageSource), oldMovement);
-            if (target instanceof LivingEntity livingTarget) {
-                weaponItem.hurtEnemy(livingTarget, this);
+        boolean flag;
+        if (target instanceof LivingEntity livingEntity) {
+            f = (float) this.getAttributeValue(livingEntity.attackDamageType.advantage);
+            f = (float) getBukkitLivingEntity().onAttack(livingEntity.getBukkitLivingEntity(), f, attackDamageType.apiType);
+            flag = livingEntity.hurtServer(level, damageSource, f, attackDamageType);
+        } else {
+            f = (float) this.getAttributeValue(Attributes.ATTACK_DAMAGE);
+            flag = target.hurtServer(level, damageSource, f);
+        }
+        boolean wasHurt = false;
+        if (flag) {
+            if (target instanceof LivingEntity livingEntity) {
+                wasHurt = weaponItem.hurtEnemy(livingEntity, this);
             }
 
-            EnchantmentHelper.doPostAttackEffects(level, target, damageSource);
-            this.setLastHurtMob(target);
-            this.playAttackSound();
+            this.postPiercingAttack();
         }
-
-        this.postPiercingAttack();
         return wasHurt;
+        // Papyrus end
     }
 
     @Override
