From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: malfrador <malfrador@gmail.com>
Date: Wed, 3 Aug 2022 20:57:07 +0200
Subject: [PATCH] fixup javadoc


diff --git a/build.gradle.kts b/build.gradle.kts
index 80833c13a2d19e31d6f648e7ef5b3456025e767e..9b4694c91649cac4e3e90b837a378c78cd4fb4fb 100644
--- a/build.gradle.kts
+++ b/build.gradle.kts
@@ -6,7 +6,7 @@ plugins {
 
 java {
     withSourcesJar()
-    withJavadocJar()
+    //withJavadocJar()
 }
 
 val annotationsVersion = "24.0.1"
@@ -162,6 +162,7 @@ tasks.withType<Javadoc> {
             "sourcepath",
             apiAndDocs.resolvedConfiguration.files.joinToString(separator = File.pathSeparator, transform = File::getPath)
         )
+        options.addStringOption("--ignore-source-errors")
     }
 
     // workaround for https://github.com/gradle/gradle/issues/4046
