From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: malfrador <malfrador@gmail.com>
Date: Fri, 5 Aug 2022 21:44:15 +0200
Subject: [PATCH] Spellbook


diff --git a/src/main/java/de/erethon/papyrus/DamageType.java b/src/main/java/de/erethon/papyrus/DamageType.java
new file mode 100644
index 0000000000000000000000000000000000000000..379e96edce4520a733064044eb357f3041c4ddf0
--- /dev/null
+++ b/src/main/java/de/erethon/papyrus/DamageType.java
@@ -0,0 +1,32 @@
+package de.erethon.papyrus;
+
+
+import org.bukkit.attribute.Attribute;
+
+public enum DamageType {
+
+    AIR(Attribute.ADV_AIR, Attribute.RES_AIR),
+    EARTH(Attribute.ADV_EARTH, Attribute.RES_EARTH),
+    FIRE(Attribute.ADV_FIRE, Attribute.RES_FIRE),
+    MAGIC(Attribute.ADV_MAGIC, Attribute.RES_MAGIC),
+    PHYSICAL(Attribute.ADV_PHYSICAL, Attribute.RES_PHYSICAL),
+    WATER(Attribute.ADV_WATER, Attribute.RES_WATER),
+    PURE(Attribute.ADV_PURE, Attribute.RES_PURE);
+
+    private final Attribute advAttr;
+    private final Attribute resAttr;
+
+    DamageType(Attribute advAttr, Attribute resAttr) {
+        this.advAttr = advAttr;
+        this.resAttr = resAttr;
+    }
+
+    public Attribute getAdvAttr() {
+        return this.advAttr;
+    }
+
+
+    public Attribute getResAttr() {
+        return this.resAttr;
+    }
+}
diff --git a/src/main/java/de/erethon/spellbook/api/EffectData.java b/src/main/java/de/erethon/spellbook/api/EffectData.java
new file mode 100644
index 0000000000000000000000000000000000000000..139cb902b1385f605082974c63030c6728353c1d
--- /dev/null
+++ b/src/main/java/de/erethon/spellbook/api/EffectData.java
@@ -0,0 +1,87 @@
+package de.erethon.spellbook.api;
+
+import org.bukkit.configuration.InvalidConfigurationException;
+import org.bukkit.configuration.file.YamlConfiguration;
+import org.bukkit.entity.LivingEntity;
+import org.jetbrains.annotations.NotNull;
+
+import java.io.File;
+import java.io.IOException;
+import java.lang.reflect.InvocationTargetException;
+
+public class EffectData extends YamlConfiguration {
+
+    enum StackMode {
+        INTENSIFY,
+        PROLONG,
+    }
+
+    SpellbookAPI spellbookAPI;
+    SpellQueue queue;
+    private final String id;
+
+    private int maxDuration;
+    private int maxStacks;
+
+    private StackMode stackMode;
+
+    private boolean isPositive;
+    private Class<? extends SpellEffect> effectClass;
+    private final ClassLoader classLoader;
+
+
+    public EffectData(SpellbookAPI spellbookAPI, String id, ClassLoader classLoader) {
+        this.spellbookAPI = spellbookAPI;
+        this.id = id;
+        queue = spellbookAPI.getQueue();
+        this.classLoader = classLoader;
+        spellbookAPI.getServer().getLogger().info("Created new SpellData with " + spellbookAPI.getClass().getName());
+    }
+
+    public String getIcon() {
+        return getString("icon", "<red>E");
+    }
+
+    public int getMaxDuration() {
+        return maxDuration;
+    }
+
+    public int getMaxStacks() {
+        return maxStacks;
+    }
+
+    public StackMode getStackMode() {
+        return stackMode;
+    }
+
+    public boolean isPositive() {
+        return isPositive;
+    }
+
+    public SpellEffect getActiveEffect(LivingEntity caster, LivingEntity target, int duration, int stack) {
+        SpellEffect spellEffect;
+        try {
+            spellEffect = effectClass.getDeclaredConstructor(EffectData.class, LivingEntity.class, LivingEntity.class, int.class, int.class).newInstance(this, caster, target, duration, stack);
+        } catch (InstantiationException | IllegalAccessException | InvocationTargetException | NoSuchMethodException e) {
+            spellbookAPI.getServer().getLogger().warning("Could not create class for effect " + id + ": " + effectClass.getName());
+            throw new RuntimeException(e);
+        }
+        return spellEffect;
+    }
+
+    @Override
+    public void load(@NotNull File file) throws IOException, InvalidConfigurationException {
+        super.load(file);
+        String className = getString("class");
+        try {
+            effectClass = (Class<? extends SpellEffect>) Class.forName("de.erethon.spellbook.effects." + className, true, classLoader);
+        } catch (ClassNotFoundException e1) {
+            spellbookAPI.getServer().getLogger().warning("Could not find class for effect " + id + ": " + className);
+            throw new RuntimeException(e1);
+        }
+        maxStacks = getInt("maxStacks", 1);
+        maxDuration = getInt("maxDuration", 10);
+        isPositive = getBoolean("positive", true);
+        stackMode = StackMode.valueOf(getString("stackMode", "INTENSIFY").toUpperCase());
+    }
+}
diff --git a/src/main/java/de/erethon/spellbook/api/SpellCastEvent.java b/src/main/java/de/erethon/spellbook/api/SpellCastEvent.java
new file mode 100644
index 0000000000000000000000000000000000000000..9597c69b1fd9825abfdedd316c5874b7eca0bc09
--- /dev/null
+++ b/src/main/java/de/erethon/spellbook/api/SpellCastEvent.java
@@ -0,0 +1,55 @@
+package de.erethon.spellbook.api;
+
+import org.bukkit.entity.LivingEntity;
+import org.bukkit.event.Cancellable;
+import org.bukkit.event.Event;
+import org.bukkit.event.HandlerList;
+import org.jetbrains.annotations.NotNull;
+
+public class SpellCastEvent extends Event implements Cancellable {
+
+    private final LivingEntity caster;
+    private final SpellbookSpell activeSpell;
+    private final SpellData data;
+
+    private boolean cancelled = false;
+
+    public SpellCastEvent(LivingEntity caster, SpellbookSpell activeSpell, SpellData data) {
+        this.caster = caster;
+        this.activeSpell = activeSpell;
+        this.data = data;
+    }
+
+    public LivingEntity getCaster() {
+        return caster;
+    }
+
+    public SpellbookSpell getActiveSpell() {
+        return activeSpell;
+    }
+
+    public SpellData getData() {
+        return data;
+    }
+
+    private static final HandlerList handlers = new HandlerList();
+
+    @Override
+    public @NotNull HandlerList getHandlers() {
+        return handlers;
+    }
+
+    public static @NotNull HandlerList getHandlerList() {
+        return handlers;
+    }
+
+    @Override
+    public boolean isCancelled() {
+        return cancelled;
+    }
+
+    @Override
+    public void setCancelled(boolean b) {
+        cancelled = b;
+    }
+}
diff --git a/src/main/java/de/erethon/spellbook/api/SpellCaster.java b/src/main/java/de/erethon/spellbook/api/SpellCaster.java
new file mode 100644
index 0000000000000000000000000000000000000000..f1d9367400225cf9064ef0aa87efbee805f80afd
--- /dev/null
+++ b/src/main/java/de/erethon/spellbook/api/SpellCaster.java
@@ -0,0 +1,186 @@
+package de.erethon.spellbook.api;
+
+import de.erethon.papyrus.DamageType;
+import org.bukkit.Bukkit;
+import org.bukkit.Location;
+import org.bukkit.entity.LivingEntity;
+
+import java.util.Iterator;
+import java.util.Map;
+import java.util.Set;
+
+public interface SpellCaster {
+    default void cast(SpellData spellData) {
+        if (!canCast(spellData)) {
+            return;
+        }
+        spellData.queue((LivingEntity) this);
+    }
+
+    Location getLocation();
+
+    Map<SpellData, Long> getUsedSpells();
+
+    Set<SpellEffect> getEffects();
+
+    Set<SpellbookSpell> getPassiveSpells();
+    Set<SpellbookSpell> getActiveSpells();
+
+    Set<SpellData> getUnlockedSpells();
+    Set<SpellData> getUnlockedPassives();
+
+
+    default void tick() {
+        Iterator<SpellEffect> effectIterator = getEffects().iterator();
+        while(effectIterator.hasNext()) {
+            SpellEffect effect = effectIterator.next();
+            if (effect.shouldRemove()) {
+                effect.onRemove();
+                effectIterator.remove();
+                for (SpellEffect eventEffect: getEffects()) {
+                    eventEffect.onRemoveEffect(effect);
+                }
+            } else {
+                effect.tick();
+            }
+        }
+    }
+
+    default double onDamage(LivingEntity damager, double damage, DamageType type) {
+        for (SpellbookSpell spell : getActiveSpells()) {
+            damage = spell.onDamage(damager, damage, type);
+        }
+        for (SpellEffect effect : getEffects()) {
+            damage = effect.onDamage(damager, damage, type);
+        }
+        return Math.max(damage, 0);
+    }
+
+    default double onAttack(LivingEntity target, double damage, DamageType type) {
+        for (SpellbookSpell spell : getActiveSpells()) {
+            damage = spell.onAttack(target, damage, type);
+        }
+        for (SpellEffect effect : getEffects()) {
+            damage = effect.onAttack(target, damage, type);
+        }
+        return Math.max(damage, 0);
+    }
+
+    default void addPassiveSpell(SpellbookSpell spell) {
+        getPassiveSpells().add(spell);
+        spell.ready();
+        addActiveSpell(spell);
+    }
+
+    default void removePassiveSpell(SpellbookSpell spell) {
+        removeActiveSpell(spell);
+        getPassiveSpells().remove(spell);
+    }
+
+    default void addSpell(SpellData spell) {
+        getUnlockedSpells().add(spell);
+    }
+
+    default void removeSpell(SpellData spell) {
+        getUnlockedSpells().remove(spell);
+    }
+
+    default void addActiveSpell(SpellbookSpell spell) {
+        getActiveSpells().add(spell);
+    }
+
+    default void removeActiveSpell(SpellbookSpell spell) {
+        getActiveSpells().remove(spell);
+    }
+
+    default void addEffect(LivingEntity caster, EffectData data, int duration, int stacks) {
+        SpellEffect effect = data.getActiveEffect(caster, (LivingEntity) this, duration, stacks);
+        if (hasEffect(data)) {
+            Bukkit.getLogger().info("Found existing " + data.getClass().getName());
+            SpellEffect oldEffect = getEffects().stream().filter(e -> e.data.equals(effect.data)).findFirst().get();
+            if (oldEffect.canAdd(duration, stacks)) {
+                Bukkit.getLogger().info("Extended existing " + data.getClass().getName());
+                if (!onEffectAdd(effect, false)) {
+                    return;
+                }
+                oldEffect.add(duration, stacks);
+            }
+        } else {
+            Bukkit.getLogger().info("Added new " + data.getClass().getName());
+            if (!onEffectAdd(effect, true)) {
+                return;
+            }
+            getEffects().add(effect);
+            effect.onApply();
+        }
+    }
+
+    default void removeEffect(EffectData effect) {
+        Iterator<SpellEffect> iterator = getEffects().iterator();
+        while (iterator.hasNext()) {
+            SpellEffect itr = iterator.next();
+            if (itr.data == effect) {
+                iterator.remove();
+                onEffectRemove(itr);
+            }
+        }
+    }
+
+    default boolean hasEffect(EffectData effect) {
+        return getEffects().stream().anyMatch(e -> e.data.equals(effect));
+    }
+
+    default int calculateCooldown(SpellData spellData) {
+        long timestamp = getCooldown(spellData);
+        long now = System.currentTimeMillis();
+        int cooldown = (int) (now - timestamp) / 1000;
+        return Math.max(spellData.getCooldown() - cooldown, 1); // 1 because amount = 0 removes the item.
+    }
+
+    default boolean canCast(SpellData spellData) {
+        long time = getUsedSpells().getOrDefault(spellData, 0L);
+        double skillCD = spellData.getCooldown();
+        long now = System.currentTimeMillis();
+        return now - time > (skillCD * 1000);
+    }
+
+    default long getCooldown(SpellData spellData) {
+        return getUsedSpells().getOrDefault(spellData, 0L);
+    }
+
+    default void setCooldown(SpellData skill) {
+        getUsedSpells().put(skill, System.currentTimeMillis());
+    }
+
+    int getEnergy();
+    int setEnergy(int energy);
+    int getMaxEnergy();
+    int setMaxEnergy(int maxEnergy);
+    int addEnergy(int energy);
+    int removeEnergy(int energy);
+    
+    private boolean onEffectAdd(SpellEffect effect, boolean isNew) {
+        boolean canAdd = true;
+        for (SpellbookSpell eventSpell : getActiveSpells()) {
+            if (!eventSpell.onAddEffect(effect, isNew)) {
+                return false;
+            }
+        }
+        for (SpellEffect eventEffect : getEffects()) {
+            if (!eventEffect.onAddEffect(effect, isNew)) {
+                return false;
+            }
+        }
+        return true;
+    }
+
+    private void onEffectRemove(SpellEffect effect) {
+        for (SpellbookSpell eventSpell : getActiveSpells()) {
+            eventSpell.onRemoveEffect(effect);
+        }
+        for (SpellEffect eventEffect : getEffects()) {
+            eventEffect.onRemoveEffect(effect);
+        }
+    }
+
+}
diff --git a/src/main/java/de/erethon/spellbook/api/SpellData.java b/src/main/java/de/erethon/spellbook/api/SpellData.java
new file mode 100644
index 0000000000000000000000000000000000000000..6de36125587ea7216784a6030e70164fd5727f10
--- /dev/null
+++ b/src/main/java/de/erethon/spellbook/api/SpellData.java
@@ -0,0 +1,114 @@
+package de.erethon.spellbook.api;
+
+import net.kyori.adventure.text.Component;
+import net.kyori.adventure.text.minimessage.MiniMessage;
+import org.bukkit.configuration.InvalidConfigurationException;
+import org.bukkit.configuration.file.YamlConfiguration;
+import org.bukkit.entity.LivingEntity;
+import org.jetbrains.annotations.NotNull;
+
+import java.io.File;
+import java.io.IOException;
+import java.lang.reflect.InvocationTargetException;
+import java.util.ArrayList;
+import java.util.List;
+
+public class SpellData extends YamlConfiguration {
+
+    SpellbookAPI spellbookAPI;
+    MiniMessage miniMessage = MiniMessage.miniMessage();
+    SpellQueue queue;
+    private int cooldown;
+    private final String id;
+    private Component name;
+    private final List<Component> description = new ArrayList<>();
+
+    private Class<? extends SpellbookSpell> spellClass;
+    private final ClassLoader classLoader;
+
+    public SpellData(SpellbookAPI spellbookAPI, String id, ClassLoader classLoader) {
+        this.spellbookAPI = spellbookAPI;
+        this.id = id;
+        this.classLoader = classLoader;
+        queue = spellbookAPI.getQueue();
+        spellbookAPI.getServer().getLogger().info("Created new SpellData with " + spellbookAPI.getClass().getName());
+    }
+    /* Casting process:
+    Player clicks button -> ActiveSpell is created and queued -> queue calls ActiveSpell#ready -> precast -> cast -> afterCast
+    The precast method is called before the cast method. If it returns false, the cast method is not called and the spell
+    is not cast. In the same way, the afterCast method will only be called if the cast method returned true.
+
+
+    A queue is used so that spells are cast in a consistent order and to limit the amount of spells that can be cast per tick globally.
+    If we can ensure that all damage is handled using the queue, we could even run the queue in a separate thread. T
+    he queue adds a slight delay to the cast process, even when its empty, but that is not noticeable for players and
+    could be easily hidden by adding animations, if needed.
+
+    The current implementation only works for spells that are actively cast by a SpellCaster. Passive spells likely
+    would be implemented using some sort of "trigger" that listens to events and then runs the usual queue/precast/cast/afterCast process.
+
+    Spells should be implemented by extending this class and overriding the precast, cast and afterCast methods.
+     This class will only exist once. Casting will create an instance of ActiveSpell, which should be used to handle all casting-related logic.
+
+     */
+
+
+    public SpellbookSpell queue(LivingEntity caster) {
+        return queue.addToQueue(getActiveSpell(caster));
+    }
+
+    public SpellbookSpell getActiveSpell(LivingEntity caster) {
+        SpellbookSpell spellbookSpell;
+        try {
+            spellbookSpell = spellClass.getDeclaredConstructor(LivingEntity.class, SpellData.class).newInstance(caster, this);
+        } catch (InstantiationException | IllegalAccessException | InvocationTargetException | NoSuchMethodException e) {
+            spellbookAPI.getServer().getLogger().warning("Could not create class for spell " + id + ": " + spellClass.getName());
+            throw new RuntimeException(e);
+        }
+        return spellbookSpell;
+    }
+
+    public String getId() {
+        return id;
+    }
+
+    @NotNull
+    public Component getDisplayName() {
+        return name;
+    }
+
+    public List<Component> getDescription() {
+        return description;
+    }
+
+    public int getCooldown() {
+        return cooldown;
+    }
+
+    public SpellbookAPI getSpellbook() {
+        return spellbookAPI;
+    }
+
+    @Override
+    public void save(@NotNull File file) throws IOException {
+        super.save(file);
+    }
+
+    @Override
+    public void load(@NotNull File file) throws IOException, InvalidConfigurationException {
+        super.load(file);
+        String className = getString("class");
+        spellbookAPI.getServer().getLogger().info("de.erethon.spellbook.spells." + className);
+        try {
+            spellClass = (Class<? extends SpellbookSpell>) Class.forName("de.erethon.spellbook.spells." + className, true, classLoader);
+        } catch (ClassNotFoundException e) {
+            spellbookAPI.getServer().getLogger().warning("Could not find class for spell " + id + ": " + className);
+            throw new RuntimeException(e);
+        }
+        cooldown = getInt("cooldown", 0);
+        name = miniMessage.deserialize(getString("name", "<color:#ff0000>[Invalid name]"));
+        for (String entry : getStringList("description")) {
+            description.add(miniMessage.deserialize(entry));
+        }
+    }
+}
diff --git a/src/main/java/de/erethon/spellbook/api/SpellEffect.java b/src/main/java/de/erethon/spellbook/api/SpellEffect.java
new file mode 100644
index 0000000000000000000000000000000000000000..70cd42b4f00a21c948aa066e5d88f9190b476406
--- /dev/null
+++ b/src/main/java/de/erethon/spellbook/api/SpellEffect.java
@@ -0,0 +1,84 @@
+package de.erethon.spellbook.api;
+
+import de.erethon.papyrus.DamageType;
+import org.bukkit.entity.LivingEntity;
+
+public class SpellEffect {
+
+    protected final LivingEntity target;
+    protected final LivingEntity caster;
+    public final EffectData data;
+
+    protected int ticksLeft;
+    protected int stacks;
+
+    public SpellEffect(EffectData data, LivingEntity caster, LivingEntity target, int duration, int stacks) {
+        this.data = data;
+        this.caster = caster;
+        this.target = target;
+        this.ticksLeft = duration;
+        this.stacks = stacks;
+    }
+
+    public void tick() {
+        ticksLeft--;
+        onTick();
+    }
+
+    public void add(int duration, int stacks) {
+        if (data.getStackMode() == EffectData.StackMode.PROLONG) {
+            this.ticksLeft += duration;
+        } else if (data.getStackMode() == EffectData.StackMode.INTENSIFY) {
+            this.stacks += stacks;
+        }
+        onApply();
+    }
+
+    public boolean canAdd(int duration, int newStacks) {
+        return ticksLeft + duration <= data.getMaxDuration() && stacks + newStacks <= data.getMaxStacks();
+    }
+
+    public void onTick() {
+    }
+
+    public void onApply() { // Mostly useful for stat buffs/debuffs
+    }
+
+    public void onRemove() { // Mostly useful for stat buffs/debuffs
+    }
+
+    public double onDamage(LivingEntity damager, double damage, DamageType type) { // For effects that block 1 attack and get removed after that
+        return damage;
+    }
+
+    public double onAttack(LivingEntity target, double damage, DamageType type) { // For effects that add something to the attack
+        return damage;
+    }
+
+    public boolean onCast(SpellbookSpell spell) {
+        return true;
+    }
+
+    public boolean onSpellTick(SpellbookSpell spell) {
+        return true;
+    }
+
+    public void onRemoveEffect(SpellEffect effect) {
+    }
+
+    public boolean onAddEffect(SpellEffect effect, boolean isNew) {
+        return true;
+    }
+
+    public boolean shouldRemove() {
+        return ticksLeft <= 0;
+    }
+
+    @Override
+    public boolean equals(Object obj) {
+        if (obj instanceof SpellEffect other) {
+            return other.data.equals(data);
+        }
+        return false;
+    }
+}
diff --git a/src/main/java/de/erethon/spellbook/api/SpellLibrary.java b/src/main/java/de/erethon/spellbook/api/SpellLibrary.java
new file mode 100644
index 0000000000000000000000000000000000000000..f7dec50648bf38e62f97f3e69aea2afb0ab0b21b
--- /dev/null
+++ b/src/main/java/de/erethon/spellbook/api/SpellLibrary.java
@@ -0,0 +1,109 @@
+package de.erethon.spellbook.api;
+
+import org.bukkit.Bukkit;
+import org.bukkit.configuration.InvalidConfigurationException;
+import org.bukkit.plugin.Plugin;
+import org.bukkit.plugin.java.PluginClassLoader;
+
+import java.io.File;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.logging.Level;
+
+public class SpellLibrary {
+
+    private final SpellbookAPI spellbookAPI;
+    private final HashMap<String, SpellData> loadedSpells = new HashMap<>();
+    private final HashMap<String, EffectData> loadedEffects = new HashMap<>();
+
+    public SpellLibrary(SpellbookAPI spellbookAPI) {
+        this.spellbookAPI = spellbookAPI;
+    }
+
+    public HashMap<String, SpellData> getLoaded() {
+        return loadedSpells;
+    }
+
+    public HashMap<String, EffectData> getLoadedEffects() {
+        return loadedEffects;
+    }
+
+    public SpellData getSpellByID(String id) {
+        return loadedSpells.get(id);
+    }
+
+    public EffectData getEffectByID(String id) {
+        return loadedEffects.get(id);
+    }
+
+    public void reload() {
+        spellbookAPI.getQueue().clear();
+        loadedEffects.clear();
+        loadedSpells.clear();
+        loadSpells(SpellbookAPI.SPELLS);
+    }
+
+    public void loadSpells(File spellFolder) {
+        Plugin plugin = Bukkit.getPluginManager().getPlugin("Hecate"); // TODO: maybe don't assume its always there
+        if (plugin == null) {
+            spellbookAPI.getServer().getLogger().warning("Could not load spells due to Spellbook plugin missing.");
+            return;
+        }
+        ClassLoader classLoader = plugin.getClass().getClassLoader();
+        try {
+            for (File f : getFilesForFolder(spellFolder)) {
+                if (!f.getName().endsWith(".yml") || f.getAbsolutePath().contains("effects")) {
+                    continue;
+                }
+                String id = f.getName().replace(".yml", "");
+
+                spellbookAPI.getServer().getLogger().info("Loading spell " + id);
+                SpellData spellData = new SpellData(spellbookAPI, id, classLoader);
+                try {
+                    spellData.load(f);
+                } catch (IOException | InvalidConfigurationException e) {
+                    throw new RuntimeException(e);
+                }
+                loadedSpells.put(id, spellData);
+            }
+            spellbookAPI.getServer().getLogger().info("Loaded " + loadedSpells.size() + " spells.");
+
+            for (File f : getFilesForFolder(new File(spellFolder, "effects"))) {
+                if (!f.getName().endsWith(".yml")) {
+                    continue;
+                }
+                String id = f.getName().replace(".yml", "");
+                spellbookAPI.getServer().getLogger().info("Loading effect " + id);
+                EffectData effectData = new EffectData(spellbookAPI, id, classLoader);
+                try {
+                    effectData.load(f);
+                } catch (IOException | InvalidConfigurationException e) {
+                    throw new RuntimeException(e);
+                }
+                loadedEffects.put(id, effectData);
+            }
+            spellbookAPI.getServer().getLogger().info("Loaded " + loadedEffects.size() + " effects.");
+
+        } catch (Exception e) {
+            spellbookAPI.getServer().getLogger().log(Level.SEVERE, "An error occurred while loading spells/effects: " + e.getMessage());
+            e.printStackTrace();
+        }
+    }
+
+    public static List<File> getFilesForFolder(File folder) {
+        List<File> files = new ArrayList<>();
+        if (!folder.isDirectory()) {
+            throw new IllegalArgumentException("File \"" + folder.getName() + "\" is not a directory");
+        }
+        for (File file : folder.listFiles()) {
+            if (file.isDirectory()) {
+                files.addAll(getFilesForFolder(file));
+            } else {
+                files.add(file);
+            }
+        }
+        return files;
+    }
+}
diff --git a/src/main/java/de/erethon/spellbook/api/SpellQueue.java b/src/main/java/de/erethon/spellbook/api/SpellQueue.java
new file mode 100644
index 0000000000000000000000000000000000000000..02a0ab86799cee6c5aa3b52392aa84720998b7b5
--- /dev/null
+++ b/src/main/java/de/erethon/spellbook/api/SpellQueue.java
@@ -0,0 +1,146 @@
+package de.erethon.spellbook.api;
+
+import com.destroystokyo.paper.event.server.ServerTickEndEvent;
+import org.bukkit.Bukkit;
+import org.bukkit.Server;
+import org.bukkit.event.EventHandler;
+import org.bukkit.plugin.PluginManager;
+
+import java.util.ArrayList;
+import java.util.Iterator;
+import java.util.List;
+import java.util.logging.Level;
+
+public class SpellQueue {
+
+    SpellbookAPI spellbookAPI;
+    private final List<SpellbookSpell> queue = new ArrayList<>();
+    private final List<SpellbookSpell> activeSpells = new ArrayList<>();
+
+    private final PluginManager pluginManager = Bukkit.getServer().getPluginManager();
+
+    private int maxActiveSpellsPerTick;
+    private int currentActiveSpells;
+    private int maxSpellsPerTick;
+    private int currentSpells;
+
+    private final Server server;
+
+    public SpellQueue(SpellbookAPI spellbookAPI) {
+        this.spellbookAPI = spellbookAPI;
+        this.server = spellbookAPI.getServer();
+    }
+
+    public void run() {
+        currentSpells = 0;
+        currentActiveSpells = 0;
+        updateMaxSpellsPerTick();
+        Iterator<SpellbookSpell> activeSpellIterator = activeSpells.iterator();
+        while (activeSpellIterator.hasNext()) {
+            currentActiveSpells++;
+            if (currentActiveSpells >= maxActiveSpellsPerTick) {
+                spellbookAPI.getServer().getLogger().warning("Active spell queue overflow! " + activeSpells.size() + "/" + maxActiveSpellsPerTick + " | Tick time: " + server.getAverageTickTime());
+                break;
+            }
+            SpellbookSpell spell = activeSpellIterator.next();
+            try {
+                if (spell.shouldRemove()) {
+                    activeSpellIterator.remove();
+                } else if (canTickSpell(spell)){
+                    spell.tick();
+                }
+            } catch (Exception e) {
+                server.getLogger().log(Level.SEVERE, "An error occurred while executing " + spell.getId() + " for " + spell.getCaster().getName() + ": " + e.getMessage());
+                e.printStackTrace();
+            }
+
+        }
+
+        Iterator<SpellbookSpell> queueIterator = queue.iterator();
+        while (queueIterator.hasNext()) {
+            currentSpells++;
+            if (currentSpells >= maxSpellsPerTick) {
+                spellbookAPI.getServer().getLogger().warning("New spell queue overflow! " + queue.size() + "/" + maxSpellsPerTick + " | Tick time: " + server.getAverageTickTime());
+                break;
+            }
+            SpellbookSpell spell = queueIterator.next();
+            try {
+                SpellCastEvent event = new SpellCastEvent(spell.getCaster(), spell, spell.getData());
+                pluginManager.callEvent(event);
+                if (event.isCancelled() || !canCastSpell(spell)) {
+                    queueIterator.remove();
+                    continue;
+                }
+                spell.ready();
+                queueIterator.remove();
+                activeSpells.add(spell);
+                spell.getCaster().addActiveSpell(spell);
+            } catch (Exception e) {
+                server.getLogger().log(Level.SEVERE, "An error occurred while readying " + spell.getId() + " for " + spell.getCaster().getName() + ": " + e.getMessage());
+                e.printStackTrace();
+            }
+        }
+    }
+
+    private void updateMaxSpellsPerTick() {
+        double tickTime = server.getAverageTickTime();
+        if (tickTime > 50.0) {
+            maxActiveSpellsPerTick = 8;
+            maxSpellsPerTick = 2;
+        }
+        else if (tickTime > 40.0) {
+            maxActiveSpellsPerTick = 16;
+            maxSpellsPerTick = 4;
+        }
+        else if (tickTime > 30.0) {
+            maxActiveSpellsPerTick = 32;
+            maxSpellsPerTick = 8;
+        }
+        else if (tickTime > 20.0) {
+            maxActiveSpellsPerTick = 64;
+            maxSpellsPerTick = 16;
+        } else {
+            maxActiveSpellsPerTick = 128;
+            maxSpellsPerTick = 64;
+        }
+    }
+
+    public SpellbookSpell addToQueue(SpellbookSpell spell) {
+        queue.add(spell);
+        return spell;
+    }
+
+    public void clear() {
+        queue.clear();
+        activeSpells.clear();
+    }
+
+    private boolean canCastSpell(SpellbookSpell spell) {
+        for (SpellbookSpell spellbookSpell : spell.getCaster().getActiveSpells()) {
+            if (!spellbookSpell.onCast(spell)) {
+                return false;
+            }
+        }
+        for (SpellEffect effect : spell.getCaster().getEffects()) {
+            if (!effect.onCast(spell)) {
+                return false;
+            }
+        }
+        return true;
+    }
+
+    private boolean canTickSpell(SpellbookSpell spell) {
+        for (SpellbookSpell spellbookSpell : spell.getCaster().getActiveSpells()) {
+            if (!spellbookSpell.onSpellTick(spell)) {
+                return false;
+            }
+        }
+        for (SpellEffect effect : spell.getCaster().getEffects()) {
+            if (!effect.onSpellTick(spell)) {
+                 return false;
+            }
+        }
+        return true;
+    }
+
+}
diff --git a/src/main/java/de/erethon/spellbook/api/SpellbookAPI.java b/src/main/java/de/erethon/spellbook/api/SpellbookAPI.java
new file mode 100644
index 0000000000000000000000000000000000000000..633828d54c4b67d82bbd400870e7b73a65e5ee5a
--- /dev/null
+++ b/src/main/java/de/erethon/spellbook/api/SpellbookAPI.java
@@ -0,0 +1,45 @@
+package de.erethon.spellbook.api;
+
+import org.bukkit.Bukkit;
+import org.bukkit.Server;
+
+import java.io.File;
+
+public class SpellbookAPI {
+    private static SpellbookAPI instance;
+    public static File SPELLS;
+    private final SpellQueue queue;
+    private final SpellLibrary library;
+    private final Server server;
+
+    public SpellbookAPI(Server server) {
+        instance = this;
+        this.server = server;
+
+        SPELLS = new File(Bukkit.getPluginsFolder(), "Spellbook");
+        if (!SPELLS.exists()) {
+            SPELLS.mkdir();
+        }
+
+        queue = new SpellQueue(this);
+
+        library = new SpellLibrary(this);
+    }
+
+    public Server getServer() {
+        return server;
+    }
+
+    public static SpellbookAPI getInstance() {
+        return instance;
+    }
+
+    public SpellLibrary getLibrary() {
+        return library;
+    }
+
+    public SpellQueue getQueue() {
+        return queue;
+    }
+
+}
diff --git a/src/main/java/de/erethon/spellbook/api/SpellbookSpell.java b/src/main/java/de/erethon/spellbook/api/SpellbookSpell.java
new file mode 100644
index 0000000000000000000000000000000000000000..5c88815c0591b15fb2068586d76c32db2ed87cd9
--- /dev/null
+++ b/src/main/java/de/erethon/spellbook/api/SpellbookSpell.java
@@ -0,0 +1,139 @@
+package de.erethon.spellbook.api;
+
+import de.erethon.papyrus.DamageType;
+import org.bukkit.entity.LivingEntity;
+
+import java.util.UUID;
+
+public abstract class SpellbookSpell {
+
+    SpellbookAPI spellbookAPI = SpellbookAPI.getInstance();
+    protected final UUID uuid;
+
+    protected final SpellData data;
+    protected final LivingEntity caster;
+
+    protected int keepAliveTicks = 0;
+    private int currentTicks = 0;
+
+    protected int tickInterval = 1;
+    private int currentTickInterval = 0;
+
+    private boolean failed = false;
+
+    public SpellbookSpell(LivingEntity caster, SpellData spellData) {
+        this.data = spellData;
+        this.caster = caster;
+        uuid = UUID.randomUUID();
+    }
+
+
+    /**
+     * This should be used to check for prerequisites for the spell, such as mana, target, location, etc.
+     *
+     * @return true if the spell can be cast, false otherwise
+     */
+    protected boolean onPrecast() {
+        return true;
+    }
+
+    /**
+     * This should be used to implement the spell itself.
+     *
+     * @return true if the spell was successfully cast, false otherwise
+     */
+    protected boolean onCast() {
+        return true;
+    }
+
+    /**
+     * This should be used do execute code after the spell was cast, like removing mana.
+     */
+    protected void onAfterCast() {
+    }
+
+    protected void onTick() {
+    }
+
+    protected void onTickFinish() {
+    }
+
+    public double onDamage(LivingEntity attacker, double damage, DamageType type) {
+        return damage;
+    }
+
+    public double onAttack(LivingEntity target, double damage, DamageType type) {
+        return damage;
+    }
+    
+    public boolean onCast(SpellbookSpell spell) {
+        return true;
+    }
+
+    public boolean onSpellTick(SpellbookSpell spell) {
+        return true;
+    }
+    
+    public void onRemoveEffect(SpellEffect effect) {
+    }
+    
+    public boolean onAddEffect(SpellEffect effect, boolean isNew) {
+        return true;
+    }
+
+    public void ready() {
+        if (onPrecast()) {
+            if (onCast()) {
+                onAfterCast();
+            } else {
+                failed = true;
+            }
+        } else {
+            failed = true;
+        }
+    }
+
+    public void tick() {
+        if (failed) {
+            return;
+        }
+        currentTicks++;
+        if (currentTickInterval >= tickInterval) {
+            currentTickInterval = 0;
+            onTick();
+        } else {
+            currentTickInterval++;
+        }
+        if (shouldRemove()) {
+            onTickFinish();
+            caster.getActiveSpells().remove(this);
+        }
+    }
+
+
+    public boolean shouldRemove() {
+        if (failed) {
+            return true;
+        }
+        if (keepAliveTicks < 0) {
+            return false;
+        }
+        return currentTicks >= keepAliveTicks;
+    }
+
+    public SpellData getData() {
+        return data;
+    }
+
+    public LivingEntity getCaster() {
+        return caster;
+    }
+
+    public UUID getUuid() {
+        return uuid;
+    }
+
+    public String getId() {
+        return data.getId();
+    }
+}
diff --git a/src/main/java/org/bukkit/Server.java b/src/main/java/org/bukkit/Server.java
index 56e261efa654e4a6872ccea28f0461df13845d13..62c5606ffdd3fb824141d868eaece2b404f3b22d 100644
--- a/src/main/java/org/bukkit/Server.java
+++ b/src/main/java/org/bukkit/Server.java
@@ -13,6 +13,7 @@ import java.util.Set;
 import java.util.UUID;
 import java.util.function.Consumer;
 import java.util.logging.Logger;
+import de.erethon.spellbook.api.SpellbookAPI;
 import org.bukkit.Warning.WarningState;
 import org.bukkit.advancement.Advancement;
 import org.bukkit.block.data.BlockData;
@@ -2105,4 +2106,7 @@ public interface Server extends PluginMessageRecipient, net.kyori.adventure.audi
      */
     @NotNull org.bukkit.potion.PotionBrewer getPotionBrewer();
     // Paper end
+    // Papyrus start
+    SpellbookAPI getSpellbookAPI();
+    // Papyrus end
 }
diff --git a/src/main/java/org/bukkit/attribute/Attribute.java b/src/main/java/org/bukkit/attribute/Attribute.java
index dc820ba494e92a0fc770703c7d5020cb2af549fd..f7b822600099f614bd4c8ee3d755c25c9cc2c4d5 100644
--- a/src/main/java/org/bukkit/attribute/Attribute.java
+++ b/src/main/java/org/bukkit/attribute/Attribute.java
@@ -80,9 +80,8 @@ public enum Attribute implements Keyed, net.kyori.adventure.translation.Translat
     ADV_FIRE("advantage.fire"),
     ADV_WATER("advantage.water"),
 
-    STAT_INT("stat.intelligence"),
-    STAT_AGI("stat.agility"),
-    STAT_STR("stat.strength"),
+    STAT_CRIT_CHANCE("stat.crit.chance"),
+    STAT_CRIT_DMG("stat.crit.dmg"),
     STAT_CDR("stat.cooldownreduction"),
 
     COMBAT_HURTINVUL("combat.hurtinvul");
diff --git a/src/main/java/org/bukkit/entity/Damageable.java b/src/main/java/org/bukkit/entity/Damageable.java
index fc4d3bcd9b16097086fef7975274d825b65adb10..0ada5e125ee88a606399b6437baa1af45ae25b19 100644
--- a/src/main/java/org/bukkit/entity/Damageable.java
+++ b/src/main/java/org/bukkit/entity/Damageable.java
@@ -1,5 +1,6 @@
 package org.bukkit.entity;
 
+import de.erethon.papyrus.DamageType;
 import org.bukkit.attribute.Attribute;
 import org.jetbrains.annotations.Nullable;
 
@@ -13,6 +14,8 @@ public interface Damageable extends Entity {
      * @param amount Amount of damage to deal
      */
     void damage(double amount);
+    void damage(double amount, DamageType type);
+
 
     /**
      * Deals the given amount of damage to this entity, from a specified
@@ -23,6 +26,8 @@ public interface Damageable extends Entity {
      */
     void damage(double amount, @Nullable Entity source);
 
+    void damage(double amount, Entity source, DamageType type);
+
     /**
      * Gets the entity's health from 0 to {@link #getMaxHealth()}, where 0 is dead.
      *
diff --git a/src/main/java/org/bukkit/entity/LivingEntity.java b/src/main/java/org/bukkit/entity/LivingEntity.java
index 8fa8922ad2fb0ea8f770368faff61e56e9761df9..0c560c57f09329435b9de06b9698a5f651151a30 100644
--- a/src/main/java/org/bukkit/entity/LivingEntity.java
+++ b/src/main/java/org/bukkit/entity/LivingEntity.java
@@ -4,6 +4,8 @@ import java.util.Collection;
 import java.util.List;
 import java.util.Set;
 import java.util.UUID;
+
+import de.erethon.spellbook.api.SpellCaster;
 import org.bukkit.FluidCollisionMode;
 import org.bukkit.Location;
 import org.bukkit.Material;
@@ -23,7 +25,7 @@ import org.jetbrains.annotations.Nullable;
 /**
  * Represents a living entity, such as a monster or player
  */
-public interface LivingEntity extends Attributable, Damageable, ProjectileSource {
+public interface LivingEntity extends Attributable, Damageable, ProjectileSource, SpellCaster {
 
     /**
      * Gets the height of the living entity's eyes above its Location.
@@ -909,4 +911,7 @@ public interface LivingEntity extends Attributable, Damageable, ProjectileSource
      */
     void setHurtDirection(float hurtDirection);
     // Paper end
+    // Papyrus start
+    void sendMessage(String msg);
+    void sendActionbar(String msg);
 }
diff --git a/src/main/java/org/bukkit/entity/Player.java b/src/main/java/org/bukkit/entity/Player.java
index 3fcfe8651a9c422fa9c8ff77556477f1461424cf..dc7269032c49a703559c8061144d973a7cfdc706 100644
--- a/src/main/java/org/bukkit/entity/Player.java
+++ b/src/main/java/org/bukkit/entity/Player.java
@@ -944,29 +944,6 @@ public interface Player extends HumanEntity, Conversable, OfflinePlayer, PluginM
         return banEntry;
     }
 
-    /**
-     * Sends an Action Bar message to the client.
-     *
-     * Use Section symbols for legacy color codes to send formatting.
-     *
-     * @param message The message to send
-     * @deprecated use {@link #sendActionBar(Component)}
-     */
-    @Deprecated
-    public void sendActionBar(@NotNull String message);
-
-    /**
-     * Sends an Action Bar message to the client.
-     *
-     * Use supplied alternative character to the section symbol to represent legacy color codes.
-     *
-     * @param alternateChar Alternate symbol such as '&amp;'
-     * @param message The message to send
-     * @deprecated use {@link #sendActionBar(Component)}
-     */
-    @Deprecated
-    public void sendActionBar(char alternateChar, @NotNull String message);
-
     /**
      * Sends an Action Bar message to the client.
      *
@@ -1003,7 +980,7 @@ public interface Player extends HumanEntity, Conversable, OfflinePlayer, PluginM
     /**
      * Sends an array of components as a single message to the specified screen position of this player
      *
-     * @deprecated This is unlikely the API you want to use. See {@link #sendActionBar(String)} for a more proper Action Bar API. This deprecated API may send unsafe items to the client.
+     * @deprecated This is unlikely the API you want to use. See or a more proper Action Bar API. This deprecated API may send unsafe items to the client.
      * @param position the screen position
      * @param components the components to send
      */
@@ -2641,7 +2618,7 @@ public interface Player extends HumanEntity, Conversable, OfflinePlayer, PluginM
         /**
          * Sends the component to the specified screen position of this player
          *
-         * @deprecated This is unlikely the API you want to use. See {@link #sendActionBar(String)} for a more proper Action Bar API. This deprecated API may send unsafe items to the client.
+         * @deprecated This is unlikely the API you want to use. for a more proper Action Bar API. This deprecated API may send unsafe items to the client.
          * @param position the screen position
          * @param component the components to send
          * @deprecated use {@code sendMessage} methods that accept {@link net.kyori.adventure.text.Component}
@@ -2654,7 +2631,7 @@ public interface Player extends HumanEntity, Conversable, OfflinePlayer, PluginM
         /**
          * Sends an array of components as a single message to the specified screen position of this player
          *
-         * @deprecated This is unlikely the API you want to use. See {@link #sendActionBar(String)} for a more proper Action Bar API. This deprecated API may send unsafe items to the client.
+         * @deprecated This is unlikely the API you want to use. See  for a more proper Action Bar API. This deprecated API may send unsafe items to the client.
          * @param position the screen position
          * @param components the components to send
          * @deprecated use {@code sendMessage} methods that accept {@link net.kyori.adventure.text.Component}
