From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Malfrador <malfrador@gmail.com>
Date: Sat, 6 Aug 2022 00:56:47 +0200
Subject: [PATCH] Spellbook


diff --git a/src/main/java/de/erethon/spellbook/api/EffectData.java b/src/main/java/de/erethon/spellbook/api/EffectData.java
index 531da0b20a7414326fed12987b10c4715b94df56..2de5a56b7cbfc56e99a405765cb38f2bdea14ffb 100644
--- a/src/main/java/de/erethon/spellbook/api/EffectData.java
+++ b/src/main/java/de/erethon/spellbook/api/EffectData.java
@@ -2,6 +2,7 @@ package de.erethon.spellbook.api;
 
 import org.bukkit.configuration.InvalidConfigurationException;
 import org.bukkit.configuration.file.YamlConfiguration;
+import org.bukkit.entity.LivingEntity;
 import org.jetbrains.annotations.NotNull;
 
 import java.io.File;
@@ -26,11 +27,14 @@ public class EffectData extends YamlConfiguration {
 
     private boolean isPositive;
     private Class<? extends SpellEffect> effectClass;
+    private final ClassLoader classLoader;
 
-    public EffectData(SpellbookAPI spellbookAPI, String id) {
+
+    public EffectData(SpellbookAPI spellbookAPI, String id, ClassLoader classLoader) {
         this.spellbookAPI = spellbookAPI;
         this.id = id;
         queue = spellbookAPI.getQueue();
+        this.classLoader = classLoader;
         spellbookAPI.getServer().getLogger().info("Created new SpellData with " + spellbookAPI.getClass().getName());
     }
 
@@ -54,10 +58,10 @@ public class EffectData extends YamlConfiguration {
         return isPositive;
     }
 
-    public SpellEffect getActiveEffect(SpellCaster target, int duration, int stack) {
+    public SpellEffect getActiveEffect(LivingEntity target, int duration, int stack) {
         SpellEffect spellEffect;
         try {
-            spellEffect = effectClass.getDeclaredConstructor(EffectData.class, SpellCaster.class, Integer.class, Integer.class).newInstance(this, target, duration, stack);
+            spellEffect = effectClass.getDeclaredConstructor(EffectData.class, LivingEntity.class, Integer.class, Integer.class).newInstance(this, target, duration, stack);
         } catch (InstantiationException | IllegalAccessException | InvocationTargetException | NoSuchMethodException e) {
             spellbookAPI.getServer().getLogger().warning("Could not create class for effect " + id + ": " + effectClass.getName());
             throw new RuntimeException(e);
@@ -70,7 +74,7 @@ public class EffectData extends YamlConfiguration {
         super.load(file);
         String className = getString("class");
         try {
-            effectClass = (Class<? extends SpellEffect>) Class.forName("de.erethon.spellbook.spells." + className);
+            effectClass = (Class<? extends SpellEffect>) Class.forName("de.erethon.spellbook.spells." + className, true, classLoader);
         } catch (ClassNotFoundException e1) {
             spellbookAPI.getServer().getLogger().warning("Could not find class for effect " + id + ": " + className);
             throw new RuntimeException(e1);
diff --git a/src/main/java/de/erethon/spellbook/api/SpellCaster.java b/src/main/java/de/erethon/spellbook/api/SpellCaster.java
index 4ebb9aaa10368f63ef4cd1b063c1bafedc81cab1..20a488df1bb2fea09d61120a786c86de694ccc56 100644
--- a/src/main/java/de/erethon/spellbook/api/SpellCaster.java
+++ b/src/main/java/de/erethon/spellbook/api/SpellCaster.java
@@ -1,6 +1,7 @@
 package de.erethon.spellbook.api;
 
 import org.bukkit.Location;
+import org.bukkit.entity.LivingEntity;
 
 import java.util.Iterator;
 import java.util.Map;
@@ -9,9 +10,9 @@ import java.util.Set;
 public interface SpellCaster {
     default void cast(SpellData spellData) {
         if (!canCast(spellData)) {
-            spellData.queue(this);
+            return;
         }
-        spellData.queue(this);
+        spellData.queue((LivingEntity) this);
     }
 
     Location getLocation();
@@ -53,7 +54,7 @@ public interface SpellCaster {
     }
 
     default void addEffect(EffectData data, int duration, int stacks) {
-        SpellEffect effect = data.getActiveEffect(this, duration, stacks);
+        SpellEffect effect = data.getActiveEffect((LivingEntity) this, duration, stacks);
         if (getEffects().contains(effect)) {
             SpellEffect oldEffect = getEffects().stream().filter(e -> e.getData().equals(effect.getData())).findFirst().get();
             if (oldEffect.canAdd(duration, stacks)) {
diff --git a/src/main/java/de/erethon/spellbook/api/SpellData.java b/src/main/java/de/erethon/spellbook/api/SpellData.java
index 0a38c0e41a3e0ded01e312e31e6acfc026174b46..d8453120f69e7e998be4c6aef5a12df086817773 100644
--- a/src/main/java/de/erethon/spellbook/api/SpellData.java
+++ b/src/main/java/de/erethon/spellbook/api/SpellData.java
@@ -2,6 +2,7 @@ package de.erethon.spellbook.api;
 
 import org.bukkit.configuration.InvalidConfigurationException;
 import org.bukkit.configuration.file.YamlConfiguration;
+import org.bukkit.entity.LivingEntity;
 import org.jetbrains.annotations.NotNull;
 
 import java.io.File;
@@ -20,10 +21,12 @@ public class SpellData extends YamlConfiguration {
     private List<String> description = new ArrayList<>();
 
     private Class<? extends SpellbookSpell> spellClass;
+    private final ClassLoader classLoader;
 
-    public SpellData(SpellbookAPI spellbookAPI, String id) {
+    public SpellData(SpellbookAPI spellbookAPI, String id, ClassLoader classLoader) {
         this.spellbookAPI = spellbookAPI;
         this.id = id;
+        this.classLoader = classLoader;
         queue = spellbookAPI.getQueue();
         spellbookAPI.getServer().getLogger().info("Created new SpellData with " + spellbookAPI.getClass().getName());
     }
@@ -47,14 +50,14 @@ public class SpellData extends YamlConfiguration {
      */
 
 
-    public SpellbookSpell queue(SpellCaster caster) {
+    public SpellbookSpell queue(LivingEntity caster) {
         return queue.addToQueue(getActiveSpell(caster));
     }
 
-    public SpellbookSpell getActiveSpell(SpellCaster caster) {
+    public SpellbookSpell getActiveSpell(LivingEntity caster) {
         SpellbookSpell spellbookSpell;
         try {
-            spellbookSpell = spellClass.getDeclaredConstructor(SpellCaster.class, SpellData.class).newInstance(caster, this);
+            spellbookSpell = spellClass.getDeclaredConstructor(LivingEntity.class, SpellData.class).newInstance(caster, this);
         } catch (InstantiationException | IllegalAccessException | InvocationTargetException | NoSuchMethodException e) {
             spellbookAPI.getServer().getLogger().warning("Could not create class for spell " + id + ": " + spellClass.getName());
             throw new RuntimeException(e);
@@ -95,7 +98,7 @@ public class SpellData extends YamlConfiguration {
         String className = getString("class");
         spellbookAPI.getServer().getLogger().info("de.erethon.spellbook.spells." + className);
         try {
-            spellClass = (Class<? extends SpellbookSpell>) Class.forName("de.erethon.spellbook.spells." + className);
+            spellClass = (Class<? extends SpellbookSpell>) Class.forName("de.erethon.spellbook.spells." + className, true, classLoader);
         } catch (ClassNotFoundException e) {
             spellbookAPI.getServer().getLogger().warning("Could not find class for spell " + id + ": " + className);
             throw new RuntimeException(e);
diff --git a/src/main/java/de/erethon/spellbook/api/SpellLibrary.java b/src/main/java/de/erethon/spellbook/api/SpellLibrary.java
index 99b89cb680f8e90d7db48f1a019431d298e58cb2..53e397104959b75fb37a00ebaac7d16eadd35777 100644
--- a/src/main/java/de/erethon/spellbook/api/SpellLibrary.java
+++ b/src/main/java/de/erethon/spellbook/api/SpellLibrary.java
@@ -1,6 +1,9 @@
 package de.erethon.spellbook.api;
 
+import org.bukkit.Bukkit;
 import org.bukkit.configuration.InvalidConfigurationException;
+import org.bukkit.plugin.Plugin;
+import org.bukkit.plugin.java.PluginClassLoader;
 
 import java.io.File;
 import java.io.IOException;
@@ -27,6 +30,12 @@ public class SpellLibrary {
     }
 
     public void loadSpells(File spellFolder) {
+        Plugin plugin = Bukkit.getPluginManager().getPlugin("Hecate"); // TODO: maybe don't assume its always there
+        if (plugin == null) {
+            spellbookAPI.getServer().getLogger().warning("Could not load spells due to Spellbook plugin missing.");
+            return;
+        }
+        ClassLoader classLoader = plugin.getClass().getClassLoader();
         for (File f : getFilesForFolder(spellFolder)) {
             if (!f.getName().endsWith(".yml")) {
                 continue;
@@ -34,7 +43,7 @@ public class SpellLibrary {
             String id = f.getName().replace(".yml", "");
             if (f.getName().contains("effect")) {
                 spellbookAPI.getServer().getLogger().info("Loading effect " + id);
-                EffectData effectData = new EffectData(spellbookAPI, id);
+                EffectData effectData = new EffectData(spellbookAPI, id, classLoader);
                 try {
                     effectData.load(f);
                 } catch (IOException | InvalidConfigurationException e) {
@@ -44,7 +53,7 @@ public class SpellLibrary {
                 continue;
             }
             spellbookAPI.getServer().getLogger().info("Loading spell " + id);
-            SpellData spellData = new SpellData(spellbookAPI, id);
+            SpellData spellData = new SpellData(spellbookAPI, id, classLoader);
             try {
                 spellData.load(f);
             } catch (IOException | InvalidConfigurationException e) {
diff --git a/src/main/java/de/erethon/spellbook/api/SpellbookAPI.java b/src/main/java/de/erethon/spellbook/api/SpellbookAPI.java
index 7cef585882e884b2334e8753f0d39b81c9ab3a0a..633828d54c4b67d82bbd400870e7b73a65e5ee5a 100644
--- a/src/main/java/de/erethon/spellbook/api/SpellbookAPI.java
+++ b/src/main/java/de/erethon/spellbook/api/SpellbookAPI.java
@@ -24,8 +24,6 @@ public class SpellbookAPI {
         queue = new SpellQueue(this);
 
         library = new SpellLibrary(this);
-        library.loadSpells(SPELLS);
-
     }
 
     public Server getServer() {
diff --git a/src/main/java/org/bukkit/Server.java b/src/main/java/org/bukkit/Server.java
index 56e261efa654e4a6872ccea28f0461df13845d13..62c5606ffdd3fb824141d868eaece2b404f3b22d 100644
--- a/src/main/java/org/bukkit/Server.java
+++ b/src/main/java/org/bukkit/Server.java
@@ -13,6 +13,7 @@ import java.util.Set;
 import java.util.UUID;
 import java.util.function.Consumer;
 import java.util.logging.Logger;
+import de.erethon.spellbook.api.SpellbookAPI;
 import org.bukkit.Warning.WarningState;
 import org.bukkit.advancement.Advancement;
 import org.bukkit.block.data.BlockData;
@@ -2105,4 +2106,7 @@ public interface Server extends PluginMessageRecipient, net.kyori.adventure.audi
      */
     @NotNull org.bukkit.potion.PotionBrewer getPotionBrewer();
     // Paper end
+    // Papyrus start
+    SpellbookAPI getSpellbookAPI();
+    // Papyrus end
 }
diff --git a/src/main/java/org/bukkit/entity/LivingEntity.java b/src/main/java/org/bukkit/entity/LivingEntity.java
index f9894d4d9a5317da1dfaff4761ec81ea583ceadf..0c560c57f09329435b9de06b9698a5f651151a30 100644
--- a/src/main/java/org/bukkit/entity/LivingEntity.java
+++ b/src/main/java/org/bukkit/entity/LivingEntity.java
@@ -911,4 +911,7 @@ public interface LivingEntity extends Attributable, Damageable, ProjectileSource
      */
     void setHurtDirection(float hurtDirection);
     // Paper end
+    // Papyrus start
+    void sendMessage(String msg);
+    void sendActionbar(String msg);
 }
diff --git a/src/main/java/org/bukkit/entity/Player.java b/src/main/java/org/bukkit/entity/Player.java
index 3fcfe8651a9c422fa9c8ff77556477f1461424cf..dc7269032c49a703559c8061144d973a7cfdc706 100644
--- a/src/main/java/org/bukkit/entity/Player.java
+++ b/src/main/java/org/bukkit/entity/Player.java
@@ -944,29 +944,6 @@ public interface Player extends HumanEntity, Conversable, OfflinePlayer, PluginM
         return banEntry;
     }
 
-    /**
-     * Sends an Action Bar message to the client.
-     *
-     * Use Section symbols for legacy color codes to send formatting.
-     *
-     * @param message The message to send
-     * @deprecated use {@link #sendActionBar(Component)}
-     */
-    @Deprecated
-    public void sendActionBar(@NotNull String message);
-
-    /**
-     * Sends an Action Bar message to the client.
-     *
-     * Use supplied alternative character to the section symbol to represent legacy color codes.
-     *
-     * @param alternateChar Alternate symbol such as '&amp;'
-     * @param message The message to send
-     * @deprecated use {@link #sendActionBar(Component)}
-     */
-    @Deprecated
-    public void sendActionBar(char alternateChar, @NotNull String message);
-
     /**
      * Sends an Action Bar message to the client.
      *
@@ -1003,7 +980,7 @@ public interface Player extends HumanEntity, Conversable, OfflinePlayer, PluginM
     /**
      * Sends an array of components as a single message to the specified screen position of this player
      *
-     * @deprecated This is unlikely the API you want to use. See {@link #sendActionBar(String)} for a more proper Action Bar API. This deprecated API may send unsafe items to the client.
+     * @deprecated This is unlikely the API you want to use. See or a more proper Action Bar API. This deprecated API may send unsafe items to the client.
      * @param position the screen position
      * @param components the components to send
      */
@@ -2641,7 +2618,7 @@ public interface Player extends HumanEntity, Conversable, OfflinePlayer, PluginM
         /**
          * Sends the component to the specified screen position of this player
          *
-         * @deprecated This is unlikely the API you want to use. See {@link #sendActionBar(String)} for a more proper Action Bar API. This deprecated API may send unsafe items to the client.
+         * @deprecated This is unlikely the API you want to use. for a more proper Action Bar API. This deprecated API may send unsafe items to the client.
          * @param position the screen position
          * @param component the components to send
          * @deprecated use {@code sendMessage} methods that accept {@link net.kyori.adventure.text.Component}
@@ -2654,7 +2631,7 @@ public interface Player extends HumanEntity, Conversable, OfflinePlayer, PluginM
         /**
          * Sends an array of components as a single message to the specified screen position of this player
          *
-         * @deprecated This is unlikely the API you want to use. See {@link #sendActionBar(String)} for a more proper Action Bar API. This deprecated API may send unsafe items to the client.
+         * @deprecated This is unlikely the API you want to use. See  for a more proper Action Bar API. This deprecated API may send unsafe items to the client.
          * @param position the screen position
          * @param components the components to send
          * @deprecated use {@code sendMessage} methods that accept {@link net.kyori.adventure.text.Component}
