From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Malfrador <malfrador@gmail.com>
Date: Wed, 26 Apr 2023 21:15:24 +0200
Subject: [PATCH] Profile switching


diff --git a/src/main/java/de/erethon/papyrus/DamageType.java b/src/main/java/de/erethon/papyrus/PDamageType.java
similarity index 89%
rename from src/main/java/de/erethon/papyrus/DamageType.java
rename to src/main/java/de/erethon/papyrus/PDamageType.java
index 379e96edce4520a733064044eb357f3041c4ddf0..efa090fd79875852b30e8af27b9bcbc64487b9b1 100644
--- a/src/main/java/de/erethon/papyrus/DamageType.java
+++ b/src/main/java/de/erethon/papyrus/PDamageType.java
@@ -3,7 +3,7 @@ package de.erethon.papyrus;
 
 import org.bukkit.attribute.Attribute;
 
-public enum DamageType {
+public enum PDamageType {
 
     AIR(Attribute.ADV_AIR, Attribute.RES_AIR),
     EARTH(Attribute.ADV_EARTH, Attribute.RES_EARTH),
@@ -16,7 +16,7 @@ public enum DamageType {
     private final Attribute advAttr;
     private final Attribute resAttr;
 
-    DamageType(Attribute advAttr, Attribute resAttr) {
+    PDamageType(Attribute advAttr, Attribute resAttr) {
         this.advAttr = advAttr;
         this.resAttr = resAttr;
     }
diff --git a/src/main/java/de/erethon/papyrus/PlayerSwitchProfileEvent.java b/src/main/java/de/erethon/papyrus/PlayerSwitchProfileEvent.java
new file mode 100644
index 0000000000000000000000000000000000000000..1f20fd2c6128cd58ec160e95ca59d276168497a2
--- /dev/null
+++ b/src/main/java/de/erethon/papyrus/PlayerSwitchProfileEvent.java
@@ -0,0 +1,39 @@
+package de.erethon.papyrus;
+
+import org.bukkit.entity.Player;
+import org.bukkit.event.HandlerList;
+import org.bukkit.event.player.PlayerEvent;
+import org.jetbrains.annotations.NotNull;
+
+public class PlayerSwitchProfileEvent extends PlayerEvent {
+
+    private static final HandlerList handlers = new HandlerList();
+    private int oldProfileID;
+    private int newProfileID;
+
+
+    public PlayerSwitchProfileEvent(@NotNull Player who, int oldProfileID, int newProfileID) {
+        super(who);
+        this.oldProfileID = oldProfileID;
+        this.newProfileID = newProfileID;
+    }
+
+    public int getOldProfileID() {
+        return oldProfileID;
+    }
+
+    public int getNewProfileID() {
+        return newProfileID;
+    }
+
+    @NotNull
+    @Override
+    public HandlerList getHandlers() {
+        return handlers;
+    }
+
+    public static HandlerList getHandlerList() {
+        return handlers;
+    }
+
+}
diff --git a/src/main/java/de/erethon/spellbook/api/SpellCaster.java b/src/main/java/de/erethon/spellbook/api/SpellCaster.java
index 7260506a0830a69278e6276174857f588c1c6125..3650f4e7206a0de6cf5da42b6d66fd7ceb9cb5ee 100644
--- a/src/main/java/de/erethon/spellbook/api/SpellCaster.java
+++ b/src/main/java/de/erethon/spellbook/api/SpellCaster.java
@@ -1,6 +1,6 @@
 package de.erethon.spellbook.api;
 
-import de.erethon.papyrus.DamageType;
+import de.erethon.papyrus.PDamageType;
 import org.bukkit.Bukkit;
 import org.bukkit.Location;
 import org.bukkit.entity.LivingEntity;
@@ -58,6 +58,10 @@ public interface SpellCaster {
      */
     Set<SpellTrait> getActiveTraits();
 
+    boolean isChanneling();
+
+    void setChanneling(boolean channeling);
+
     default void tick() {
         Iterator<SpellEffect> effectIterator = getEffects().iterator();
         while(effectIterator.hasNext()) {
@@ -82,7 +86,7 @@ public interface SpellCaster {
         }
     }
 
-    default double onDamage(LivingEntity damager, double damage, DamageType type) {
+    default double onDamage(LivingEntity damager, double damage, PDamageType type) {
         for (SpellbookSpell spell : getActiveSpells()) {
             damage = spell.onDamage(damager, damage, type);
         }
@@ -95,7 +99,7 @@ public interface SpellCaster {
         return Math.max(damage, 0);
     }
 
-    default double onAttack(LivingEntity target, double damage, DamageType type) {
+    default double onAttack(LivingEntity target, double damage, PDamageType type) {
         for (SpellbookSpell spell : getActiveSpells()) {
             damage = spell.onAttack(target, damage, type);
         }
@@ -318,5 +322,26 @@ public interface SpellCaster {
         }
     }
 
+    /**
+     * This will only interrupt channeling spells for now.
+     */
+    default void interrupt() {
+        for (SpellbookSpell spell : getActiveSpells()) {
+            if (spell.isChanneling) {
+                spell.interrupt();
+                setChanneling(false);
+            }
+        }
+    }
+
+    default void interruptFromMovement() {
+        for (SpellbookSpell spell : getActiveSpells()) {
+            if (spell.isChanneling && spell.isMovementInterrupted) {
+                spell.interrupt();
+                setChanneling(false);
+            }
+        }
+    }
+
 
 }
diff --git a/src/main/java/de/erethon/spellbook/api/SpellChannelFinishEvent.java b/src/main/java/de/erethon/spellbook/api/SpellChannelFinishEvent.java
new file mode 100644
index 0000000000000000000000000000000000000000..3ab73e1736b09e3b1ed053463848993b81259bc0
--- /dev/null
+++ b/src/main/java/de/erethon/spellbook/api/SpellChannelFinishEvent.java
@@ -0,0 +1,41 @@
+package de.erethon.spellbook.api;
+
+import org.bukkit.event.Event;
+import org.bukkit.event.HandlerList;
+import org.jetbrains.annotations.NotNull;
+
+public class SpellChannelFinishEvent extends Event {
+
+    private final SpellCaster caster;
+    private final SpellbookSpell spell;
+    private final boolean interrupted;
+
+    private static final HandlerList handlers = new HandlerList();
+
+    public SpellChannelFinishEvent(SpellCaster caster, SpellbookSpell spell, boolean interrupted) {
+        this.caster = caster;
+        this.spell = spell;
+        this.interrupted = interrupted;
+    }
+
+    public SpellCaster getCaster() {
+        return caster;
+    }
+
+    public SpellbookSpell getSpell() {
+        return spell;
+    }
+
+    public boolean wasInterrupted() {
+        return interrupted;
+    }
+
+    @Override
+    public @NotNull HandlerList getHandlers() {
+        return handlers;
+    }
+
+    public static @NotNull HandlerList getHandlerList() {
+        return handlers;
+    }
+}
diff --git a/src/main/java/de/erethon/spellbook/api/SpellChannelStartEvent.java b/src/main/java/de/erethon/spellbook/api/SpellChannelStartEvent.java
new file mode 100644
index 0000000000000000000000000000000000000000..ff0571c54408568ef66cc149e10a9daa2528b1d4
--- /dev/null
+++ b/src/main/java/de/erethon/spellbook/api/SpellChannelStartEvent.java
@@ -0,0 +1,34 @@
+package de.erethon.spellbook.api;
+
+import org.bukkit.event.Event;
+import org.bukkit.event.HandlerList;
+import org.jetbrains.annotations.NotNull;
+
+public class SpellChannelStartEvent extends Event {
+    private final SpellCaster caster;
+    private final SpellbookSpell spell;
+
+    private static final HandlerList handlers = new HandlerList();
+
+    public SpellChannelStartEvent(SpellCaster caster, SpellbookSpell spell) {
+        this.caster = caster;
+        this.spell = spell;
+    }
+
+    public SpellCaster getCaster() {
+        return caster;
+    }
+
+    public SpellbookSpell getSpell() {
+        return spell;
+    }
+
+    @Override
+    public @NotNull HandlerList getHandlers() {
+        return handlers;
+    }
+
+    public static @NotNull HandlerList getHandlerList() {
+        return handlers;
+    }
+}
diff --git a/src/main/java/de/erethon/spellbook/api/SpellEffect.java b/src/main/java/de/erethon/spellbook/api/SpellEffect.java
index 00e3dc889b0c4978fa20a76f233ede7bea34f878..2b28cf882250df1836b5b5b9c8a0a9f9864e1430 100644
--- a/src/main/java/de/erethon/spellbook/api/SpellEffect.java
+++ b/src/main/java/de/erethon/spellbook/api/SpellEffect.java
@@ -1,6 +1,6 @@
 package de.erethon.spellbook.api;
 
-import de.erethon.papyrus.DamageType;
+import de.erethon.papyrus.PDamageType;
 import org.bukkit.entity.LivingEntity;
 
 public class SpellEffect {
@@ -47,11 +47,11 @@ public class SpellEffect {
     public void onRemove() { // Mostly useful for stat buffs/debuffs
     }
 
-    public double onDamage(LivingEntity damager, double damage, DamageType type) { // For effects that block 1 attack and get removed after that
+    public double onDamage(LivingEntity damager, double damage, PDamageType type) { // For effects that block 1 attack and get removed after that
         return damage;
     }
 
-    public double onAttack(LivingEntity target, double damage, DamageType type) { // For effects that add something to the attack
+    public double onAttack(LivingEntity target, double damage, PDamageType type) { // For effects that add something to the attack
         return damage;
     }
 
diff --git a/src/main/java/de/erethon/spellbook/api/SpellTrait.java b/src/main/java/de/erethon/spellbook/api/SpellTrait.java
index 96ec2b714f8e9f8c69f26811ee2878d20dfee5b6..828b24e26444f98b423d79a4f2775ea02348f609 100644
--- a/src/main/java/de/erethon/spellbook/api/SpellTrait.java
+++ b/src/main/java/de/erethon/spellbook/api/SpellTrait.java
@@ -1,6 +1,6 @@
 package de.erethon.spellbook.api;
 
-import de.erethon.papyrus.DamageType;
+import de.erethon.papyrus.PDamageType;
 import org.bukkit.entity.LivingEntity;
 
 import java.util.Set;
@@ -81,7 +81,7 @@ public class SpellTrait {
      * This method is called when the caster takes damage.
      * @param attacker the attacker
      */
-    public double onDamage(LivingEntity attacker, double damage, DamageType type) {
+    public double onDamage(LivingEntity attacker, double damage, PDamageType type) {
         return damage;
     }
 
@@ -89,7 +89,7 @@ public class SpellTrait {
      * This method is called when the caster deals damage.
      * @param target the target the caster is attacking
      */
-    public double onAttack(LivingEntity target, double damage, DamageType type) {
+    public double onAttack(LivingEntity target, double damage, PDamageType type) {
         return damage;
     }
 
diff --git a/src/main/java/de/erethon/spellbook/api/SpellbookSpell.java b/src/main/java/de/erethon/spellbook/api/SpellbookSpell.java
index c4a07139e108bbd54e33d499bb883b6ec6eb8baa..a4ed5621278d74f04a543c8f0405b33347a16704 100644
--- a/src/main/java/de/erethon/spellbook/api/SpellbookSpell.java
+++ b/src/main/java/de/erethon/spellbook/api/SpellbookSpell.java
@@ -1,6 +1,7 @@
 package de.erethon.spellbook.api;
 
-import de.erethon.papyrus.DamageType;
+import de.erethon.papyrus.PDamageType;
+import org.bukkit.Bukkit;
 import org.bukkit.entity.LivingEntity;
 
 import java.util.Set;
@@ -16,12 +17,16 @@ public abstract class SpellbookSpell {
 
     protected int keepAliveTicks = 0;
     protected int currentTicks = 0;
+    public int channelDuration = 0;
+    public boolean isMovementInterrupted = false;
+    public int currentChannelTicks = 0;
 
     protected int tickInterval = 1;
     protected int currentTickInterval = 0;
 
     protected boolean failed = false;
     protected boolean interrupted = false;
+    public boolean isChanneling = false;
 
     public SpellbookSpell(LivingEntity caster, SpellData spellData) {
         this.data = spellData;
@@ -67,6 +72,19 @@ public abstract class SpellbookSpell {
         cleanup();
     }
 
+    /**
+     * This is method is called after channeling was finished.
+     */
+    protected void onChannelFinish() {
+    }
+
+    /**
+     * This method is called when channeling is interrupted for whatever reason. cleanup() will be called afterwards
+     */
+    protected void onChannelInterrupt() {
+
+    }
+
     /**
      * This method should be used for cleaning up after a Spell cast. For example spawned entities should be
      * removed here.
@@ -78,7 +96,7 @@ public abstract class SpellbookSpell {
      * This method is called when the caster takes damage.
      * @param attacker the attacker
      */
-    public double onDamage(LivingEntity attacker, double damage, DamageType type) {
+    public double onDamage(LivingEntity attacker, double damage, PDamageType type) {
         return damage;
     }
 
@@ -86,7 +104,7 @@ public abstract class SpellbookSpell {
      * This method is called when the caster deals damage.
      * @param target the target the caster is attacking
      */
-    public double onAttack(LivingEntity target, double damage, DamageType type) {
+    public double onAttack(LivingEntity target, double damage, PDamageType type) {
         return damage;
     }
 
@@ -197,9 +215,15 @@ public abstract class SpellbookSpell {
 
     /**
      * Interrupts the spell. It will stop ticking, but cleanup() will still
-     * be called.
+     * be called. onChannelInterrupt() will be called first if the Spell is currently channeling.
      */
     public void interrupt() {
+        if (isChanneling) {
+            SpellChannelFinishEvent finishEvent = new SpellChannelFinishEvent(caster, this, true);
+            Bukkit.getPluginManager().callEvent(finishEvent);
+            onChannelInterrupt();
+            channelDuration = -1; // Don't continue channeling
+        }
         interrupted = true;
     }
 
@@ -235,6 +259,22 @@ public abstract class SpellbookSpell {
             cleanup();
             return;
         }
+        if (channelDuration > 0) {
+            currentChannelTicks++;
+            if (!isChanneling && currentChannelTicks < channelDuration) {
+                isChanneling = true;
+                caster.setChanneling(true);
+                SpellChannelStartEvent startEvent = new SpellChannelStartEvent(caster, this);
+                Bukkit.getPluginManager().callEvent(startEvent);
+            }
+            if (currentChannelTicks == channelDuration) {
+                SpellChannelFinishEvent finishEvent = new SpellChannelFinishEvent(caster, this, false);
+                Bukkit.getPluginManager().callEvent(finishEvent);
+                onChannelFinish();
+                isChanneling = false;
+                caster.setChanneling(false);
+            }
+        }
         currentTicks++;
         if (currentTickInterval >= tickInterval) {
             currentTickInterval = 0;
diff --git a/src/main/java/org/bukkit/entity/Damageable.java b/src/main/java/org/bukkit/entity/Damageable.java
index 01b179822b7b405b07695d60269bd315f6117e56..70e50f1ee559a9eabf4fa96199f2e86a0eba143e 100644
--- a/src/main/java/org/bukkit/entity/Damageable.java
+++ b/src/main/java/org/bukkit/entity/Damageable.java
@@ -1,6 +1,6 @@
 package org.bukkit.entity;
 
-import de.erethon.papyrus.DamageType;
+import de.erethon.papyrus.PDamageType;
 import org.bukkit.attribute.Attribute;
 import org.bukkit.damage.DamageSource;
 import org.jetbrains.annotations.ApiStatus;
@@ -17,7 +17,7 @@ public interface Damageable extends Entity {
      * @param amount Amount of damage to deal
      */
     void damage(double amount);
-    void damage(double amount, DamageType type);
+    void damage(double amount, PDamageType type);
 
 
     /**
@@ -29,7 +29,7 @@ public interface Damageable extends Entity {
      */
     void damage(double amount, @Nullable Entity source);
 
-    void damage(double amount, Entity source, DamageType type);
+    void damage(double amount, Entity source, PDamageType type);
 
     /**
      * Deals the given amount of damage to this entity from a specified
diff --git a/src/main/java/org/bukkit/entity/Projectile.java b/src/main/java/org/bukkit/entity/Projectile.java
index 9fe73c563e015c37d2056218599ba21f283ae980..8816c19764150ed1dd406c4f98394b4383672d24 100644
--- a/src/main/java/org/bukkit/entity/Projectile.java
+++ b/src/main/java/org/bukkit/entity/Projectile.java
@@ -1,6 +1,6 @@
 package org.bukkit.entity;
 
-import de.erethon.papyrus.DamageType;
+import de.erethon.papyrus.PDamageType;
 import org.bukkit.projectiles.ProjectileSource;
 import org.jetbrains.annotations.Nullable;
 
@@ -128,5 +128,5 @@ public interface Projectile extends Entity {
     @Nullable
     java.util.UUID getOwnerUniqueId();
     // Paper end
-    void setDamageType(DamageType type);
+    void setDamageType(PDamageType type);
 }
diff --git a/src/main/java/org/bukkit/event/entity/EntityDamageByBlockEvent.java b/src/main/java/org/bukkit/event/entity/EntityDamageByBlockEvent.java
index 1fb15e2ade8ff3c4d662eca87b078b4577f786e1..01045c40b7d1ed3c80c50abb4537cc0c083727e6 100644
--- a/src/main/java/org/bukkit/event/entity/EntityDamageByBlockEvent.java
+++ b/src/main/java/org/bukkit/event/entity/EntityDamageByBlockEvent.java
@@ -2,6 +2,8 @@ package org.bukkit.event.entity;
 
 import com.google.common.base.Function;
 import java.util.Map;
+
+import de.erethon.papyrus.PDamageType;
 import org.bukkit.block.Block;
 import org.bukkit.damage.DamageSource;
 import org.bukkit.damage.DamageType;
@@ -37,7 +39,7 @@ public class EntityDamageByBlockEvent extends EntityDamageEvent {
     }
 
     public EntityDamageByBlockEvent(@Nullable final Block damager, @NotNull final Entity damagee, @NotNull final DamageCause cause, @NotNull final DamageSource damageSource, @NotNull final Map<DamageModifier, Double> modifiers, @NotNull final Map<DamageModifier, ? extends Function<? super Double, Double>> modifierFunctions, final @Nullable org.bukkit.block.BlockState damagerBlockState) { // Paper
-        super(damagee, cause, damageSource, modifiers, modifierFunctions);
+        super(damagee, cause, damageSource, modifiers, modifierFunctions, PDamageType.PURE);
         this.damager = damager;
         this.damagerBlockState = damagerBlockState; // Paper
     }
diff --git a/src/main/java/org/bukkit/event/entity/EntityDamageByEntityEvent.java b/src/main/java/org/bukkit/event/entity/EntityDamageByEntityEvent.java
index 6b24d1281cb8f0253430c9c1a1323e2670bb9c93..ba0c7115553e15be10531b711f4faaea5aea9689 100644
--- a/src/main/java/org/bukkit/event/entity/EntityDamageByEntityEvent.java
+++ b/src/main/java/org/bukkit/event/entity/EntityDamageByEntityEvent.java
@@ -2,8 +2,11 @@ package org.bukkit.event.entity;
 
 import com.google.common.base.Function;
 import java.util.Map;
+
+import de.erethon.papyrus.PDamageType;
 import org.bukkit.damage.DamageSource;
 import org.bukkit.damage.DamageType;
+
 import org.bukkit.entity.Entity;
 import org.jetbrains.annotations.ApiStatus;
 import org.jetbrains.annotations.NotNull;
@@ -27,24 +30,30 @@ public class EntityDamageByEntityEvent extends EntityDamageEvent {
         this.critical = false; // Paper - add critical damage API
     }
 
-    @Deprecated(forRemoval = true)
+    public EntityDamageByEntityEvent(@NotNull final Entity damager, @NotNull final Entity damagee, @NotNull final DamageCause cause, final double damage, DamageType type) {
+        super(damagee, cause, damage);
+        this.damager = damager;
+        this.critical = false; // Paper - add critical damage API
+    }
+
     public EntityDamageByEntityEvent(@NotNull final Entity damager, @NotNull final Entity damagee, @NotNull final DamageCause cause, @NotNull final Map<DamageModifier, Double> modifiers, @NotNull final Map<DamageModifier, ? extends Function<? super Double, Double>> modifierFunctions) {
         this(damager, damagee, cause, DamageSource.builder(DamageType.GENERIC).withCausingEntity(damager).withDirectEntity(damager).build(), modifiers, modifierFunctions);
     }
 
     @Deprecated
     public EntityDamageByEntityEvent(@NotNull final Entity damager, @NotNull final Entity damagee, @NotNull final DamageCause cause, @NotNull final DamageSource damageSource, @NotNull final Map<DamageModifier, Double> modifiers, @NotNull final Map<DamageModifier, ? extends Function<? super Double, Double>> modifierFunctions) {
-        super(damagee, cause, damageSource, modifiers, modifierFunctions);
+        super(damagee, cause, damageSource, modifiers, modifierFunctions, PDamageType.PURE);
         this.damager = damager;
         // Paper start
         this.critical = false;
     }
 
     @ApiStatus.Internal
-    public EntityDamageByEntityEvent(@NotNull final Entity damager, @NotNull final Entity damagee, @NotNull final DamageCause cause, @NotNull final DamageSource damageSource, @NotNull final Map<DamageModifier, Double> modifiers, @NotNull final Map<DamageModifier, ? extends Function<? super Double, Double>> modifierFunctions, boolean critical) {
-        super(damagee, cause, damageSource, modifiers, modifierFunctions);
+    public EntityDamageByEntityEvent(@NotNull final Entity damager, @NotNull final Entity damagee, @NotNull final DamageCause cause, @NotNull final DamageSource damageSource, @NotNull final Map<DamageModifier, Double> modifiers, @NotNull final Map<DamageModifier, ? extends Function<? super Double, Double>> modifierFunctions, boolean critical, PDamageType pType) {
+        super(damagee, cause, damageSource, modifiers, modifierFunctions, pType);
         this.damager = damager;
         this.critical = critical;
+        this.pDamageType = pType;
     }
 
     /**
diff --git a/src/main/java/org/bukkit/event/entity/EntityDamageEvent.java b/src/main/java/org/bukkit/event/entity/EntityDamageEvent.java
index 446b3ffd5caca5344be1c250475679834cd0d4a2..86e66829a073e6bc1c8c38005e6ff78732f83da2 100644
--- a/src/main/java/org/bukkit/event/entity/EntityDamageEvent.java
+++ b/src/main/java/org/bukkit/event/entity/EntityDamageEvent.java
@@ -7,6 +7,8 @@ import com.google.common.collect.ImmutableMap;
 import java.util.EnumMap;
 import java.util.Map;
 import java.util.Objects;
+
+import de.erethon.papyrus.PDamageType;
 import org.bukkit.Material;
 import org.bukkit.WorldBorder;
 import org.bukkit.damage.DamageSource;
@@ -31,21 +33,22 @@ public class EntityDamageEvent extends EntityEvent implements Cancellable {
     private final DamageCause cause;
     private final DamageSource damageSource;
 
-    @Deprecated(forRemoval = true)
+    protected PDamageType pDamageType; // Papyrus
+
     public EntityDamageEvent(@NotNull final Entity damagee, @NotNull final DamageCause cause, final double damage) {
         this(damagee, cause, DamageSource.builder(DamageType.GENERIC).build(), damage);
     }
 
     public EntityDamageEvent(@NotNull final Entity damagee, @NotNull final DamageCause cause, @NotNull final DamageSource damageSource, final double damage) {
-        this(damagee, cause, damageSource, new EnumMap<DamageModifier, Double>(ImmutableMap.of(DamageModifier.BASE, damage)), new EnumMap<DamageModifier, Function<? super Double, Double>>(ImmutableMap.of(DamageModifier.BASE, ZERO)));
+        this(damagee, cause, damageSource, new EnumMap<DamageModifier, Double>(ImmutableMap.of(DamageModifier.BASE, damage)), new EnumMap<DamageModifier, Function<? super Double, Double>>(ImmutableMap.of(DamageModifier.BASE, ZERO)), PDamageType.PURE);
     }
 
-    @Deprecated(forRemoval = true)
     public EntityDamageEvent(@NotNull final Entity damagee, @NotNull final DamageCause cause, @NotNull final Map<DamageModifier, Double> modifiers, @NotNull final Map<DamageModifier, ? extends Function<? super Double, Double>> modifierFunctions) {
-        this(damagee, cause, DamageSource.builder(DamageType.GENERIC).build(), modifiers, modifierFunctions);
+        this(damagee, cause, DamageSource.builder(DamageType.GENERIC).build(), modifiers, modifierFunctions, PDamageType.PURE);
     }
 
-    public EntityDamageEvent(@NotNull final Entity damagee, @NotNull final DamageCause cause, @NotNull final DamageSource damageSource, @NotNull final Map<DamageModifier, Double> modifiers, @NotNull final Map<DamageModifier, ? extends Function<? super Double, Double>> modifierFunctions) {
+
+    public EntityDamageEvent(@NotNull final Entity damagee, @NotNull final DamageCause cause, DamageSource source, @NotNull final Map<DamageModifier, Double> modifiers, @NotNull final Map<DamageModifier, ? extends Function<? super Double, Double>> modifierFunctions, PDamageType pType) {
         super(damagee);
         Preconditions.checkArgument(modifiers.containsKey(DamageModifier.BASE), "BASE DamageModifier missing");
         Preconditions.checkArgument(!modifiers.containsKey(null), "Cannot have null DamageModifier");
@@ -56,7 +59,8 @@ public class EntityDamageEvent extends EntityEvent implements Cancellable {
         this.cause = cause;
         this.modifiers = modifiers;
         this.modifierFunctions = modifierFunctions;
-        this.damageSource = damageSource;
+        this.damageSource = source;
+        this.pDamageType = pType; // Papyrus
     }
 
     @Override
@@ -232,6 +236,14 @@ public class EntityDamageEvent extends EntityEvent implements Cancellable {
         return handlers;
     }
 
+    public PDamageType getDamageType() {
+        return pDamageType;
+    }
+
+    public void setDamageType(PDamageType type) {
+        pDamageType = type;
+    }
+
     /**
      * An enum to specify the types of modifier
      *
