From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Malfrador <malfrador@gmail.com>
Date: Wed, 26 Apr 2023 21:15:24 +0200
Subject: [PATCH] Profile switching


diff --git a/src/main/java/de/erethon/papyrus/PlayerSwitchProfileEvent.java b/src/main/java/de/erethon/papyrus/PlayerSwitchProfileEvent.java
new file mode 100644
index 0000000000000000000000000000000000000000..1f20fd2c6128cd58ec160e95ca59d276168497a2
--- /dev/null
+++ b/src/main/java/de/erethon/papyrus/PlayerSwitchProfileEvent.java
@@ -0,0 +1,39 @@
+package de.erethon.papyrus;
+
+import org.bukkit.entity.Player;
+import org.bukkit.event.HandlerList;
+import org.bukkit.event.player.PlayerEvent;
+import org.jetbrains.annotations.NotNull;
+
+public class PlayerSwitchProfileEvent extends PlayerEvent {
+
+    private static final HandlerList handlers = new HandlerList();
+    private int oldProfileID;
+    private int newProfileID;
+
+
+    public PlayerSwitchProfileEvent(@NotNull Player who, int oldProfileID, int newProfileID) {
+        super(who);
+        this.oldProfileID = oldProfileID;
+        this.newProfileID = newProfileID;
+    }
+
+    public int getOldProfileID() {
+        return oldProfileID;
+    }
+
+    public int getNewProfileID() {
+        return newProfileID;
+    }
+
+    @NotNull
+    @Override
+    public HandlerList getHandlers() {
+        return handlers;
+    }
+
+    public static HandlerList getHandlerList() {
+        return handlers;
+    }
+
+}
diff --git a/src/main/java/de/erethon/spellbook/api/SpellCaster.java b/src/main/java/de/erethon/spellbook/api/SpellCaster.java
index 9db25aed60af480255c42cb06d07308cfb7ee080..dbd2c348cf94d687f7153cfc8e5e8d08694689b6 100644
--- a/src/main/java/de/erethon/spellbook/api/SpellCaster.java
+++ b/src/main/java/de/erethon/spellbook/api/SpellCaster.java
@@ -58,6 +58,10 @@ public interface SpellCaster {
      */
     Set<SpellTrait> getActiveTraits();
 
+    boolean isChanneling();
+
+    void setChanneling(boolean channeling);
+
     default void tick() {
         Iterator<SpellEffect> effectIterator = getEffects().iterator();
         while(effectIterator.hasNext()) {
@@ -318,5 +322,26 @@ public interface SpellCaster {
         }
     }
 
+    /**
+     * This will only interrupt channeling spells for now.
+     */
+    default void interrupt() {
+        for (SpellbookSpell spell : getActiveSpells()) {
+            if (spell.isChanneling) {
+                spell.interrupt();
+                setChanneling(false);
+            }
+        }
+    }
+
+    default void interruptFromMovement() {
+        for (SpellbookSpell spell : getActiveSpells()) {
+            if (spell.isChanneling && spell.isMovementInterrupted) {
+                spell.interrupt();
+                setChanneling(false);
+            }
+        }
+    }
+
 
 }
diff --git a/src/main/java/de/erethon/spellbook/api/SpellChannelFinishEvent.java b/src/main/java/de/erethon/spellbook/api/SpellChannelFinishEvent.java
new file mode 100644
index 0000000000000000000000000000000000000000..3ab73e1736b09e3b1ed053463848993b81259bc0
--- /dev/null
+++ b/src/main/java/de/erethon/spellbook/api/SpellChannelFinishEvent.java
@@ -0,0 +1,41 @@
+package de.erethon.spellbook.api;
+
+import org.bukkit.event.Event;
+import org.bukkit.event.HandlerList;
+import org.jetbrains.annotations.NotNull;
+
+public class SpellChannelFinishEvent extends Event {
+
+    private final SpellCaster caster;
+    private final SpellbookSpell spell;
+    private final boolean interrupted;
+
+    private static final HandlerList handlers = new HandlerList();
+
+    public SpellChannelFinishEvent(SpellCaster caster, SpellbookSpell spell, boolean interrupted) {
+        this.caster = caster;
+        this.spell = spell;
+        this.interrupted = interrupted;
+    }
+
+    public SpellCaster getCaster() {
+        return caster;
+    }
+
+    public SpellbookSpell getSpell() {
+        return spell;
+    }
+
+    public boolean wasInterrupted() {
+        return interrupted;
+    }
+
+    @Override
+    public @NotNull HandlerList getHandlers() {
+        return handlers;
+    }
+
+    public static @NotNull HandlerList getHandlerList() {
+        return handlers;
+    }
+}
diff --git a/src/main/java/de/erethon/spellbook/api/SpellChannelStartEvent.java b/src/main/java/de/erethon/spellbook/api/SpellChannelStartEvent.java
new file mode 100644
index 0000000000000000000000000000000000000000..ff0571c54408568ef66cc149e10a9daa2528b1d4
--- /dev/null
+++ b/src/main/java/de/erethon/spellbook/api/SpellChannelStartEvent.java
@@ -0,0 +1,34 @@
+package de.erethon.spellbook.api;
+
+import org.bukkit.event.Event;
+import org.bukkit.event.HandlerList;
+import org.jetbrains.annotations.NotNull;
+
+public class SpellChannelStartEvent extends Event {
+    private final SpellCaster caster;
+    private final SpellbookSpell spell;
+
+    private static final HandlerList handlers = new HandlerList();
+
+    public SpellChannelStartEvent(SpellCaster caster, SpellbookSpell spell) {
+        this.caster = caster;
+        this.spell = spell;
+    }
+
+    public SpellCaster getCaster() {
+        return caster;
+    }
+
+    public SpellbookSpell getSpell() {
+        return spell;
+    }
+
+    @Override
+    public @NotNull HandlerList getHandlers() {
+        return handlers;
+    }
+
+    public static @NotNull HandlerList getHandlerList() {
+        return handlers;
+    }
+}
diff --git a/src/main/java/de/erethon/spellbook/api/SpellbookSpell.java b/src/main/java/de/erethon/spellbook/api/SpellbookSpell.java
index a755287c75d9e4236c68cf70caa500d1dc530103..855c8da2b94424a9af550e81128124ca2114d47a 100644
--- a/src/main/java/de/erethon/spellbook/api/SpellbookSpell.java
+++ b/src/main/java/de/erethon/spellbook/api/SpellbookSpell.java
@@ -1,6 +1,7 @@
 package de.erethon.spellbook.api;
 
 import de.erethon.papyrus.DamageType;
+import org.bukkit.Bukkit;
 import org.bukkit.entity.LivingEntity;
 
 import java.util.Set;
@@ -16,12 +17,16 @@ public abstract class SpellbookSpell {
 
     protected int keepAliveTicks = 0;
     protected int currentTicks = 0;
+    public int channelDuration = 0;
+    public boolean isMovementInterrupted = false;
+    public int currentChannelTicks = 0;
 
     protected int tickInterval = 1;
     protected int currentTickInterval = 0;
 
     protected boolean failed = false;
     protected boolean interrupted = false;
+    public boolean isChanneling = false;
 
     public SpellbookSpell(LivingEntity caster, SpellData spellData) {
         this.data = spellData;
@@ -67,6 +72,19 @@ public abstract class SpellbookSpell {
         cleanup();
     }
 
+    /**
+     * This is method is called after channeling was finished.
+     */
+    protected void onChannelFinish() {
+    }
+
+    /**
+     * This method is called when channeling is interrupted for whatever reason. cleanup() will be called afterwards
+     */
+    protected void onChannelInterrupt() {
+
+    }
+
     /**
      * This method should be used for cleaning up after a Spell cast. For example spawned entities should be
      * removed here.
@@ -170,9 +188,15 @@ public abstract class SpellbookSpell {
 
     /**
      * Interrupts the spell. It will stop ticking, but cleanup() will still
-     * be called.
+     * be called. onChannelInterrupt() will be called first if the Spell is currently channeling.
      */
     public void interrupt() {
+        if (isChanneling) {
+            SpellChannelFinishEvent finishEvent = new SpellChannelFinishEvent(caster, this, true);
+            Bukkit.getPluginManager().callEvent(finishEvent);
+            onChannelInterrupt();
+            channelDuration = -1; // Don't continue channeling
+        }
         interrupted = true;
     }
 
@@ -208,6 +232,22 @@ public abstract class SpellbookSpell {
             cleanup();
             return;
         }
+        if (channelDuration > 0) {
+            currentChannelTicks++;
+            if (!isChanneling && currentChannelTicks < channelDuration) {
+                isChanneling = true;
+                caster.setChanneling(true);
+                SpellChannelStartEvent startEvent = new SpellChannelStartEvent(caster, this);
+                Bukkit.getPluginManager().callEvent(startEvent);
+            }
+            if (currentChannelTicks == channelDuration) {
+                SpellChannelFinishEvent finishEvent = new SpellChannelFinishEvent(caster, this, false);
+                Bukkit.getPluginManager().callEvent(finishEvent);
+                onChannelFinish();
+                isChanneling = false;
+                caster.setChanneling(false);
+            }
+        }
         currentTicks++;
         if (currentTickInterval >= tickInterval) {
             currentTickInterval = 0;
diff --git a/src/main/java/org/bukkit/event/entity/EntityDamageByEntityEvent.java b/src/main/java/org/bukkit/event/entity/EntityDamageByEntityEvent.java
index 7ce8f1a26c1b33dd0eb6e6435952fd73abf49879..5f94cb25a5137b4e6a93c1b2850f55d4f9290a78 100644
--- a/src/main/java/org/bukkit/event/entity/EntityDamageByEntityEvent.java
+++ b/src/main/java/org/bukkit/event/entity/EntityDamageByEntityEvent.java
@@ -2,6 +2,8 @@ package org.bukkit.event.entity;
 
 import com.google.common.base.Function;
 import java.util.Map;
+
+import de.erethon.papyrus.DamageType;
 import org.bukkit.entity.Entity;
 import org.jetbrains.annotations.NotNull;
 
@@ -18,6 +20,13 @@ public class EntityDamageByEntityEvent extends EntityDamageEvent {
         this.critical = false; // Paper - add critical damage API
     }
 
+    public EntityDamageByEntityEvent(@NotNull final Entity damager, @NotNull final Entity damagee, @NotNull final DamageCause cause, final double damage, DamageType type) {
+        super(damagee, cause, damage);
+        this.damager = damager;
+        this.critical = false; // Paper - add critical damage API
+        this.damageType = type;
+    }
+
     @Deprecated // Paper - add critical damage API
     public EntityDamageByEntityEvent(@NotNull final Entity damager, @NotNull final Entity damagee, @NotNull final DamageCause cause, @NotNull final Map<DamageModifier, Double> modifiers, @NotNull final Map<DamageModifier, ? extends Function<? super Double, Double>> modifierFunctions) {
         // Paper start - add critical damage API
diff --git a/src/main/java/org/bukkit/event/entity/EntityDamageEvent.java b/src/main/java/org/bukkit/event/entity/EntityDamageEvent.java
index 4773f537dec20d6ebd82e4b145a1cdea0077fe90..589b2781d14e84aaba11ee5bf8a2de0a7cc7fdaf 100644
--- a/src/main/java/org/bukkit/event/entity/EntityDamageEvent.java
+++ b/src/main/java/org/bukkit/event/entity/EntityDamageEvent.java
@@ -7,6 +7,8 @@ import com.google.common.collect.ImmutableMap;
 import java.util.EnumMap;
 import java.util.Map;
 import java.util.Objects;
+
+import de.erethon.papyrus.DamageType;
 import org.bukkit.Material;
 import org.bukkit.WorldBorder;
 import org.bukkit.entity.Entity;
@@ -28,11 +30,19 @@ public class EntityDamageEvent extends EntityEvent implements Cancellable {
     private boolean cancelled;
     private final DamageCause cause;
 
+    protected DamageType damageType; // Papyrus
+
     public EntityDamageEvent(@NotNull final Entity damagee, @NotNull final DamageCause cause, final double damage) {
         this(damagee, cause, new EnumMap<DamageModifier, Double>(ImmutableMap.of(DamageModifier.BASE, damage)), new EnumMap<DamageModifier, Function<? super Double, Double>>(ImmutableMap.of(DamageModifier.BASE, ZERO)));
     }
 
+    // Papyrus start - Damage types
     public EntityDamageEvent(@NotNull final Entity damagee, @NotNull final DamageCause cause, @NotNull final Map<DamageModifier, Double> modifiers, @NotNull final Map<DamageModifier, ? extends Function<? super Double, Double>> modifierFunctions) {
+        this(damagee, cause, modifiers, modifierFunctions, DamageType.PURE);
+    }
+    // Papyrus end
+
+    public EntityDamageEvent(@NotNull final Entity damagee, @NotNull final DamageCause cause, @NotNull final Map<DamageModifier, Double> modifiers, @NotNull final Map<DamageModifier, ? extends Function<? super Double, Double>> modifierFunctions, DamageType type) {
         super(damagee);
         Preconditions.checkArgument(modifiers.containsKey(DamageModifier.BASE), "BASE DamageModifier missing");
         Preconditions.checkArgument(!modifiers.containsKey(null), "Cannot have null DamageModifier");
@@ -43,6 +53,7 @@ public class EntityDamageEvent extends EntityEvent implements Cancellable {
         this.cause = cause;
         this.modifiers = modifiers;
         this.modifierFunctions = modifierFunctions;
+        this.damageType = type; // Papyrus
     }
 
     @Override
@@ -204,6 +215,14 @@ public class EntityDamageEvent extends EntityEvent implements Cancellable {
         return handlers;
     }
 
+    public DamageType getDamageType() {
+        return damageType;
+    }
+
+    public void setDamageType(DamageType type) {
+        damageType = type;
+    }
+
     /**
      * An enum to specify the types of modifier
      *
