From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Fyreum <luka.vogel@gmx.de>
Date: Wed, 17 Aug 2022 14:36:07 +0200
Subject: [PATCH] Make entity block damage optional


diff --git a/src/main/java/org/bukkit/entity/Explosive.java b/src/main/java/org/bukkit/entity/Explosive.java
index 48650f6ab82e602e8a08f69c31e0c2c38e27ff98..e43b9fc5c63840cc81fa9a5e2d76ef93fadf647e 100644
--- a/src/main/java/org/bukkit/entity/Explosive.java
+++ b/src/main/java/org/bukkit/entity/Explosive.java
@@ -32,4 +32,22 @@ public interface Explosive extends Entity {
      * @return true if the explosive creates fire, false otherwise
      */
     public boolean isIncendiary();
+
+    // Papyrus start - Make block damage optional
+
+    /**
+     * Set whether this explosive's explosion causes block damage
+     *
+     * @param blockDamage Whether it should cause block damage
+     */
+    public void setBlockDamage(boolean blockDamage);
+
+    /**
+     * Return whether this explosive causes block damage when exploding
+     *
+     * @return true if the explosive causes block damage, false otherwise
+     */
+    public boolean isBlockDamage();
+
+    // Papyrus end
 }
diff --git a/src/main/java/org/bukkit/event/entity/ExplosionPrimeEvent.java b/src/main/java/org/bukkit/event/entity/ExplosionPrimeEvent.java
index 6db5dad719de5566f3b27666ff6006c1dbb189e9..054fc9a26102d5739e71bb5a3029787388e67915 100644
--- a/src/main/java/org/bukkit/event/entity/ExplosionPrimeEvent.java
+++ b/src/main/java/org/bukkit/event/entity/ExplosionPrimeEvent.java
@@ -14,16 +14,18 @@ public class ExplosionPrimeEvent extends EntityEvent implements Cancellable {
     private boolean cancel;
     private float radius;
     private boolean fire;
+    private boolean blockDamage; // Papyrus
 
-    public ExplosionPrimeEvent(@NotNull final Entity what, final float radius, final boolean fire) {
+    public ExplosionPrimeEvent(@NotNull final Entity what, final float radius, final boolean fire, /* Papyrus start */ final boolean blockDamage /* Papyrus end*/) {
         super(what);
         this.cancel = false;
         this.radius = radius;
         this.fire = fire;
+        this.blockDamage = blockDamage; // Papyrus
     }
 
     public ExplosionPrimeEvent(@NotNull final Explosive explosive) {
-        this(explosive, explosive.getYield(), explosive.isIncendiary());
+        this(explosive, explosive.getYield(), explosive.isIncendiary(), explosive.isBlockDamage());
     }
 
     @Override
@@ -72,6 +74,32 @@ public class ExplosionPrimeEvent extends EntityEvent implements Cancellable {
         this.fire = fire;
     }
 
+    // Papyrus start - Make block damage optional
+
+    /**
+     * Return whether this explosive causes block damage when exploding.
+     * If {@link org.bukkit.GameRule#MOB_GRIEFING} is disabled, this value
+     * won't change anything.
+     *
+     * @return true if the explosive causes block damage, false otherwise
+     */
+    public boolean isBlockDamage() {
+        return blockDamage;
+    }
+
+    /**
+     * Set whether this explosive's explosion causes block damage.
+     * If {@link org.bukkit.GameRule#MOB_GRIEFING} is disabled, this value
+     * won't change anything.
+     *
+     * @param blockDamage Whether it should cause block damage
+     */
+    public void setBlockDamage(boolean blockDamage) {
+        this.blockDamage = blockDamage;
+    }
+
+    // Papyrus end
+
     @NotNull
     @Override
     public HandlerList getHandlers() {
