From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Malfrador <malfrador@gmail.com>
Date: Sat, 19 Mar 2022 15:19:37 +0100
Subject: [PATCH] remove heightmap spam


diff --git a/src/main/java/net/minecraft/world/level/levelgen/Heightmap.java b/src/main/java/net/minecraft/world/level/levelgen/Heightmap.java
index 0181cef8b6c9e7d1b5c08a43b429f5db40cc8d07..06c1a11b5a1ac314cd8f8e50b7c3ef97932e3792 100644
--- a/src/main/java/net/minecraft/world/level/levelgen/Heightmap.java
+++ b/src/main/java/net/minecraft/world/level/levelgen/Heightmap.java
@@ -129,7 +129,7 @@ public class Heightmap {
         if (ms.length == ls.length) {
             System.arraycopy(ls, 0, ms, 0, ls.length);
         } else {
-            LOGGER.warn("Ignoring heightmap data for chunk " + chunk.getPos() + ", size does not match; expected: " + ms.length + ", got: " + ls.length);
+            // Papyrus - remove spam - LOGGER.warn("Ignoring heightmap data for chunk " + chunk.getPos() + ", size does not match; expected: " + ms.length + ", got: " + ls.length);
             primeHeightmaps(chunk, EnumSet.of(type));
         }
     }
