From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: malfrador <malfrador@gmail.com>
Date: Sat, 11 Jun 2022 18:07:35 +0200
Subject: [PATCH] combat changes


diff --git a/src/main/java/de/erethon/papyrus/DamageType.java b/src/main/java/de/erethon/papyrus/DamageType.java
new file mode 100644
index 0000000000000000000000000000000000000000..0cbc67c143dbbe8d8bc1ebf2ea384f6851cf8658
--- /dev/null
+++ b/src/main/java/de/erethon/papyrus/DamageType.java
@@ -0,0 +1,32 @@
+package de.erethon.papyrus;
+
+import net.minecraft.world.entity.ai.attributes.Attribute;
+import net.minecraft.world.entity.ai.attributes.Attributes;
+
+public enum DamageType {
+
+    AIR(Attributes.ADV_AIR, Attributes.RES_AIR),
+    EARTH(Attributes.ADV_EARTH, Attributes.RES_EARTH),
+    FIRE(Attributes.ADV_FIRE, Attributes.RES_FIRE),
+    MAGIC(Attributes.ADV_MAGIC, Attributes.RES_MAGIC),
+    PHYSICAL(Attributes.ADV_PHYSICAL, Attributes.RES_PHYSICAL),
+    WATER(Attributes.ADV_WATER, Attributes.RES_WATER),
+    PURE(Attributes.ADV_PURE, Attributes.RES_PURE);
+
+    private final Attribute advAttr;
+    private final Attribute resAttr;
+
+    DamageType(Attribute advAttr, Attribute resAttr) {
+        this.advAttr = advAttr;
+        this.resAttr = resAttr;
+    }
+
+    public Attribute getAdvAttr() {
+        return this.advAttr;
+    }
+
+
+    public Attribute getResAttr() {
+        return this.resAttr;
+    }
+}
diff --git a/src/main/java/de/erethon/papyrus/PapyrusUtils.java b/src/main/java/de/erethon/papyrus/PapyrusUtils.java
new file mode 100644
index 0000000000000000000000000000000000000000..afe4ed93f622cc46c0bb1faa12976e44297f4ea9
--- /dev/null
+++ b/src/main/java/de/erethon/papyrus/PapyrusUtils.java
@@ -0,0 +1,14 @@
+package de.erethon.papyrus;
+
+import io.papermc.paper.adventure.PaperAdventure;
+import net.kyori.adventure.text.minimessage.MiniMessage;
+import net.minecraft.network.chat.Component;
+
+public class PapyrusUtils {
+
+    private static final char healthBar = '|';
+
+    public static Component getHealthMessage(double health, double max) { // maybe with color based on damage type?
+        return PaperAdventure.asVanilla(MiniMessage.miniMessage().deserialize("<red>" + Math.round(health) + "<dark_gray>/" + "<dark_red>" + Math.round(max)));
+    }
+}
diff --git a/src/main/java/net/minecraft/server/level/ServerPlayer.java b/src/main/java/net/minecraft/server/level/ServerPlayer.java
index 3615576c24d5d6790a6894a91180de25fa0e5a9c..805408e24a3c7700f6c3d44e8df6b3e72e8dc238 100644
--- a/src/main/java/net/minecraft/server/level/ServerPlayer.java
+++ b/src/main/java/net/minecraft/server/level/ServerPlayer.java
@@ -13,6 +13,9 @@ import java.util.Objects;
 import java.util.Optional;
 import java.util.OptionalInt;
 import javax.annotation.Nullable;
+
+import de.erethon.papyrus.DamageType;
+import net.kyori.adventure.text.minimessage.MiniMessage;
 import net.minecraft.BlockUtil;
 import net.minecraft.ChatFormatting;
 import net.minecraft.CrashReport;
@@ -2064,7 +2067,7 @@ public class ServerPlayer extends Player {
         if (this.gameMode.getGameModeForPlayer() == GameType.SPECTATOR) {
             this.setCamera(target);
         } else {
-            super.attack(target);
+            super.attack(target, DamageType.PHYSICAL); // Papyrus
         }
 
     }
diff --git a/src/main/java/net/minecraft/server/network/ServerGamePacketListenerImpl.java b/src/main/java/net/minecraft/server/network/ServerGamePacketListenerImpl.java
index cc55cb14f2d0f656213be25a1e428132871ac5dd..442a18f3f3593b13bf190473383c763ccd0403f2 100644
--- a/src/main/java/net/minecraft/server/network/ServerGamePacketListenerImpl.java
+++ b/src/main/java/net/minecraft/server/network/ServerGamePacketListenerImpl.java
@@ -323,6 +323,7 @@ public class ServerGamePacketListenerImpl implements ServerPlayerConnection, Tic
 
         // CraftBukkit start - add fields and methods
         this.cserver = server.server;
+        LOGGER.info("Initialized packet listener for " + player.getScoreboardName());
     }
 
     private final org.bukkit.craftbukkit.CraftServer cserver;
diff --git a/src/main/java/net/minecraft/world/entity/Entity.java b/src/main/java/net/minecraft/world/entity/Entity.java
index 417570c2ad71bc3727e7d5eb841a92ec815d12f6..b97e8a18ca2c6d6ca673a6aaabaaa924a2d12464 100644
--- a/src/main/java/net/minecraft/world/entity/Entity.java
+++ b/src/main/java/net/minecraft/world/entity/Entity.java
@@ -7,6 +7,7 @@ import com.google.common.collect.Lists;
 import com.google.common.collect.Sets;
 import com.google.common.collect.UnmodifiableIterator;
 import com.mojang.logging.LogUtils;
+import de.erethon.papyrus.DamageType;
 import it.unimi.dsi.fastutil.objects.Object2DoubleArrayMap;
 import it.unimi.dsi.fastutil.objects.Object2DoubleMap;
 import java.util.Arrays;
@@ -2019,6 +2020,16 @@ public abstract class Entity implements Nameable, EntityAccess, CommandSource {
         }
     }
 
+    // Papyrus - combat rework
+    public boolean hurt(DamageSource source, float amount, DamageType type) {
+        if (this.isInvulnerableTo(source)) {
+            return false;
+        } else {
+            this.markHurt();
+            return false;
+        }
+    }
+
     public final Vec3 getViewVector(float tickDelta) {
         return this.calculateViewVector(this.getViewXRot(tickDelta), this.getViewYRot(tickDelta));
     }
diff --git a/src/main/java/net/minecraft/world/entity/LivingEntity.java b/src/main/java/net/minecraft/world/entity/LivingEntity.java
index ded69bc892aaf35cc57a562116bb623c07b08550..8321f089f23812fa40ebf93a4a65356c637f18fb 100644
--- a/src/main/java/net/minecraft/world/entity/LivingEntity.java
+++ b/src/main/java/net/minecraft/world/entity/LivingEntity.java
@@ -10,15 +10,14 @@ import com.mojang.datafixers.util.Pair;
 import com.mojang.logging.LogUtils;
 import com.mojang.serialization.DataResult;
 import com.mojang.serialization.Dynamic;
-import java.util.Collection;
-import java.util.ConcurrentModificationException;
-import java.util.Iterator;
-import java.util.List;
-import java.util.Map;
-import java.util.Optional;
-import java.util.UUID;
+
+import java.util.*;
 import java.util.function.Predicate;
 import javax.annotation.Nullable;
+
+import de.erethon.papyrus.DamageType;
+import de.erethon.papyrus.PapyrusUtils;
+import net.kyori.adventure.text.Component;
 import net.minecraft.BlockUtil;
 import net.minecraft.advancements.CriteriaTriggers;
 import net.minecraft.commands.arguments.EntityAnchorArgument;
@@ -84,14 +83,7 @@ import net.minecraft.world.entity.boss.wither.WitherBoss;
 import net.minecraft.world.entity.item.ItemEntity;
 import net.minecraft.world.entity.projectile.AbstractArrow;
 import net.minecraft.world.food.FoodProperties;
-import net.minecraft.world.item.ArmorItem;
-import net.minecraft.world.item.AxeItem;
-import net.minecraft.world.item.BlockItem;
-import net.minecraft.world.item.ElytraItem;
-import net.minecraft.world.item.Item;
-import net.minecraft.world.item.ItemStack;
-import net.minecraft.world.item.Items;
-import net.minecraft.world.item.UseAnim;
+import net.minecraft.world.item.*;
 import net.minecraft.world.item.alchemy.PotionUtils;
 import net.minecraft.world.item.enchantment.EnchantmentHelper;
 import net.minecraft.world.item.enchantment.Enchantments;
@@ -111,6 +103,7 @@ import net.minecraft.world.level.block.state.BlockState;
 import net.minecraft.world.level.gameevent.GameEvent;
 import net.minecraft.world.level.material.Fluid;
 import net.minecraft.world.level.material.FluidState;
+import net.minecraft.world.level.material.Material;
 import net.minecraft.world.level.storage.loot.LootContext;
 import net.minecraft.world.level.storage.loot.LootTable;
 import net.minecraft.world.level.storage.loot.parameters.LootContextParamSets;
@@ -120,25 +113,18 @@ import net.minecraft.world.phys.EntityHitResult;
 import net.minecraft.world.phys.HitResult;
 import net.minecraft.world.phys.Vec3;
 import net.minecraft.world.scores.PlayerTeam;
+import org.bukkit.Bukkit;
+import org.bukkit.event.entity.*;
 import org.slf4j.Logger;
 
 // CraftBukkit start
-import java.util.ArrayList;
-import java.util.HashSet;
-import java.util.Set;
 import com.google.common.base.Function;
 import org.bukkit.Location;
 import org.bukkit.craftbukkit.attribute.CraftAttributeMap;
 import org.bukkit.craftbukkit.event.CraftEventFactory;
 import org.bukkit.craftbukkit.inventory.CraftItemStack;
 import org.bukkit.entity.Player;
-import org.bukkit.event.entity.ArrowBodyCountChangeEvent;
-import org.bukkit.event.entity.EntityDamageEvent;
 import org.bukkit.event.entity.EntityDamageEvent.DamageModifier;
-import org.bukkit.event.entity.EntityPotionEffectEvent;
-import org.bukkit.event.entity.EntityRegainHealthEvent;
-import org.bukkit.event.entity.EntityResurrectEvent;
-import org.bukkit.event.entity.EntityTeleportEvent;
 import org.bukkit.event.player.PlayerItemConsumeEvent;
 // CraftBukkit end
 
@@ -264,6 +250,8 @@ public abstract class LivingEntity extends Entity {
     public org.bukkit.craftbukkit.entity.CraftLivingEntity getBukkitLivingEntity() { return (org.bukkit.craftbukkit.entity.CraftLivingEntity) super.getBukkitEntity(); } // Paper
     public boolean silentDeath = false; // Paper - mark entity as dying silently for cancellable death event
 
+    public DamageType attackDamageType = DamageType.PHYSICAL; // Papyrus - combat rework
+
     @Override
     public float getBukkitYaw() {
         return this.getYHeadRot();
@@ -336,9 +324,31 @@ public abstract class LivingEntity extends Entity {
     }
 
     public static AttributeSupplier.Builder createLivingAttributes() {
-        return AttributeSupplier.builder().add(Attributes.MAX_HEALTH).add(Attributes.KNOCKBACK_RESISTANCE).add(Attributes.MOVEMENT_SPEED).add(Attributes.ARMOR).add(Attributes.ARMOR_TOUGHNESS);
+        return AttributeSupplier.builder()
+            .add(Attributes.MAX_HEALTH)
+            .add(Attributes.KNOCKBACK_RESISTANCE)
+            .add(Attributes.MOVEMENT_SPEED)
+            .add(Attributes.ARMOR)
+            .add(Attributes.ARMOR_TOUGHNESS)
+            .add(Attributes.COMBAT_HURTINVUL)
+            .add(Attributes.STAT_CRITCHANCE)
+            .add(Attributes.STAT_CRITDMG)
+            .add(Attributes.STAT_ATTACKSPEED)
+            .add(Attributes.RES_AIR)
+            .add(Attributes.RES_EARTH)
+            .add(Attributes.RES_FIRE)
+            .add(Attributes.RES_WATER)
+            .add(Attributes.RES_PHYSICAL)
+            .add(Attributes.RES_MAGIC)
+            .add(Attributes.RES_PURE)
+            .add(Attributes.ADV_AIR)
+            .add(Attributes.ADV_EARTH)
+            .add(Attributes.ADV_FIRE)
+            .add(Attributes.ADV_WATER)
+            .add(Attributes.ADV_PHYSICAL)
+            .add(Attributes.ADV_MAGIC)
+            .add(Attributes.ADV_PURE);
     }
-
     @Override
     protected void checkFallDamage(double heightDifference, boolean onGround, BlockState state, BlockPos landedPosition) {
         if (!this.isInWater()) {
@@ -1283,6 +1293,7 @@ public abstract class LivingEntity extends Entity {
         return (Float) this.entityData.get(LivingEntity.DATA_HEALTH_ID);
     }
 
+
     public void setHealth(float health) {
         // Paper start
         if (Float.isNaN(health)) { health = getMaxHealth(); if (this.valid) {
@@ -1303,7 +1314,11 @@ public abstract class LivingEntity extends Entity {
             player.updateScaledHealth(false);
             return;
         }
-        // CraftBukkit end
+        // Papyrus start
+        if (!(this instanceof net.minecraft.world.entity.player.Player)) {
+            setCustomName(PapyrusUtils.getHealthMessage(health, getMaxHealth()));
+        }
+        // Papyrus end
         this.entityData.set(LivingEntity.DATA_HEALTH_ID, Mth.clamp(health, 0.0F, this.getMaxHealth()));
     }
 
@@ -1320,9 +1335,78 @@ public abstract class LivingEntity extends Entity {
     }
     // Pufferfish end
 
+    // Papyrus start - rewrite damage system
+
+    @Override
+    public boolean hurt(DamageSource source, float amount, DamageType type) {
+        if (isInvulnerable() || isRemoved() || dead || getHealth() <= 0F || isInvulnerableTo(source) || level.isClientSide) {
+            return false;
+        }
+        markHurt();
+        if (invulnerableTime > 0) {
+            return false;
+        }
+        double finalDmg = amount - getAttributeValue(type.getResAttr());
+        Entity attacker = source.getEntity();
+        if (attacker != null) {
+            EntityDamageByEntityEvent event = new EntityDamageByEntityEvent(source.getEntity().getBukkitEntity(), this.getBukkitLivingEntity(), EntityDamageEvent.DamageCause.ENTITY_ATTACK,amount);
+            Bukkit.getServer().getPluginManager().callEvent(event);
+            if (event.isCancelled()) {
+                return false;
+            }
+            if (source.getEntity() instanceof net.minecraft.world.entity.player.Player player) {
+                player.getBukkitEntity().sendActionBar(Component.text(type + ": " + amount + " | Res: " + getAttributeValue(type.getResAttr()) + " | = " + finalDmg));
+            }
+            if (isDamageSourceBlocked(source)) { // TODO: Maybe change isDamageSourceBlocked logic
+                hurtCurrentlyUsedShield(amount);
+                if (!source.isProjectile()) {
+                    Entity entity = source.getDirectEntity();
+                    if (entity instanceof LivingEntity && entity.distanceToSqr(this) <= (200.0D * 200.0D)) {
+                        blockUsingShield((LivingEntity) entity);
+                    }
+                }
+                finalDmg = finalDmg - getAttributeValue(type.getResAttr());
+                if (useItem.is(Items.SHIELD)) {
+                    // Use the HEAD slot modifier from shields for the damage
+                    double attrRes = 0;
+                    for (AttributeModifier modifier : useItem.getAttributeModifiers(EquipmentSlot.HEAD).get(type.getResAttr())) {
+                        attrRes = attrRes + modifier.getAmount();
+                    }
+                    finalDmg = finalDmg - attrRes;
+                }
+            }
+            // TODO: This is a mess
+            final boolean isFarAway = attacker.distanceToSqr(this) > (200.0 * 200.0); // Vanilla value
+            double x = isFarAway ? (Math.random() - Math.random()) : attacker.getX() - this.getX();
+            double z;
+            for (z = isFarAway ? Math.random() - Math.random() : attacker.getZ() - this.getZ(); x * x + z * z < 1.0E-4D; z = (Math.random() - Math.random()) * 0.01D) { // Paper
+                x = (Math.random() - Math.random()) * 0.01D;
+            }
+
+            this.hurtDir = (float) (Mth.atan2(z, x) * 57.2957763671875D - (double) this.getYRot());
+            this.knockback(0.4000000059604645D, x, z, attacker);
+
+            if (finalDmg <= 0) {
+                return false;
+            }
+
+
+        } else {
+            hurtDir = (float) ((int) (Math.random() * 2.0D) * 180);
+        }
+        invulnerableTime = (int) getAttribute(Attributes.COMBAT_HURTINVUL).getValue();
+        playHurtSound(source);
+        level.broadcastEntityEvent(this, (byte) 2); // hurt animation
+        setHealth((float) (getHealth() - Math.max(finalDmg, 0)));
+        return true;
+    }
     @Override
     public boolean hurt(DamageSource source, float amount) {
-        if (this.isInvulnerableTo(source)) {
+        hurt(source, amount, DamageType.PURE);
+        return true;
+    }
+    // Papyrus end
+        /*if (this.isInvulnerableTo(source)) {
             return false;
         } else if (this.level.isClientSide) {
             return false;
@@ -1455,7 +1539,7 @@ public abstract class LivingEntity extends Entity {
                     this.markHurt();
                 }
 
-                if (entity1 != null) {
+                if (entity1 != null) { // entity1 = source
                     final boolean far = entity1.distanceToSqr(this) > (200.0 * 200.0); // Paper
                     double d0 = far ? (Math.random() - Math.random()) : entity1.getX() - this.getX(); // Paper
 
@@ -1505,8 +1589,8 @@ public abstract class LivingEntity extends Entity {
             }
 
             return flag2;
-        }
-    }
+            }
+    }*/
 
     protected void blockUsingShield(LivingEntity attacker) {
         attacker.blockedByShield(this);
@@ -2502,6 +2586,7 @@ public abstract class LivingEntity extends Entity {
 
     @Nullable
     public AttributeInstance getAttribute(Attribute attribute) {
+        LOGGER.debug(getAttributes().toString());
         return this.getAttributes().getInstance(attribute);
     }
 
diff --git a/src/main/java/net/minecraft/world/entity/Mob.java b/src/main/java/net/minecraft/world/entity/Mob.java
index 5ab1f5d36a61e7a7fe8da4e74f75963ea6432101..292ca252a4dea8e146a28300aba18e39d8fd09ba 100644
--- a/src/main/java/net/minecraft/world/entity/Mob.java
+++ b/src/main/java/net/minecraft/world/entity/Mob.java
@@ -9,6 +9,8 @@ import java.util.Map;
 import java.util.Optional;
 import java.util.UUID;
 import javax.annotation.Nullable;
+
+import de.erethon.papyrus.DamageType;
 import net.minecraft.core.BlockPos;
 import net.minecraft.core.NonNullList;
 import net.minecraft.core.Vec3i;
@@ -1593,12 +1595,16 @@ public abstract class Mob extends LivingEntity {
 
     @Override
     public boolean doHurtTarget(Entity target) {
-        float f = (float) this.getAttributeValue(Attributes.ATTACK_DAMAGE);
-        float f1 = (float) this.getAttributeValue(Attributes.ATTACK_KNOCKBACK);
+        return doHurtTarget(target, attackDamageType);
+    }
+
+    public boolean doHurtTarget(Entity target, DamageType type) {
+        float dmg = (float) this.getAttributeValue(Attributes.ATTACK_DAMAGE);
+        float knockback = (float) this.getAttributeValue(Attributes.ATTACK_KNOCKBACK);
 
         if (target instanceof LivingEntity) {
-            f += EnchantmentHelper.getDamageBonus(this.getMainHandItem(), ((LivingEntity) target).getMobType());
-            f1 += (float) EnchantmentHelper.getKnockbackBonus(this);
+            dmg += EnchantmentHelper.getDamageBonus(this.getMainHandItem(), ((LivingEntity) target).getMobType());
+            knockback += (float) EnchantmentHelper.getKnockbackBonus(this);
         }
 
         int i = EnchantmentHelper.getFireAspect(this);
@@ -1613,12 +1619,12 @@ public abstract class Mob extends LivingEntity {
             }
             // CraftBukkit end
         }
-
-        boolean flag = target.hurt(DamageSource.mobAttack(this), f);
+        dmg = (float) (dmg + getAttributeValue(type.getAdvAttr()));
+        boolean flag = target.hurt(DamageSource.mobAttack(this), dmg, type);
 
         if (flag) {
-            if (f1 > 0.0F && target instanceof LivingEntity) {
-                ((LivingEntity) target).knockback((double) (f1 * 0.5F), (double) Mth.sin(this.getYRot() * 0.017453292F), (double) (-Mth.cos(this.getYRot() * 0.017453292F)), this); // Paper
+            if (knockback > 0.0F && target instanceof LivingEntity) {
+                ((LivingEntity) target).knockback((double) (knockback * 0.5F), (double) Mth.sin(this.getYRot() * 0.017453292F), (double) (-Mth.cos(this.getYRot() * 0.017453292F)), this); // Paper
                 this.setDeltaMovement(this.getDeltaMovement().multiply(0.6D, 1.0D, 0.6D));
             }
 
diff --git a/src/main/java/net/minecraft/world/entity/ai/attributes/Attributes.java b/src/main/java/net/minecraft/world/entity/ai/attributes/Attributes.java
index dc27e53c6c8711db87a38ef678d5f328d099376f..c4d18fb80c7b20869f712cfd09b6689f4e67b000 100644
--- a/src/main/java/net/minecraft/world/entity/ai/attributes/Attributes.java
+++ b/src/main/java/net/minecraft/world/entity/ai/attributes/Attributes.java
@@ -5,12 +5,12 @@ import net.minecraft.core.Registry;
 
 public class Attributes {
 
-    public static final Attribute MAX_HEALTH = Attributes.register("generic.max_health", (new RangedAttribute("attribute.name.generic.max_health", 20.0D, 1.0D, org.spigotmc.SpigotConfig.maxHealth)).setSyncable(true));
+    public static final Attribute MAX_HEALTH = Attributes.register("generic.max_health", new RangedAttribute("attribute.name.generic.max_health", 500.0D, 1.0D, org.spigotmc.SpigotConfig.maxHealth).setSyncable(true));
     public static final Attribute FOLLOW_RANGE = Attributes.register("generic.follow_range", new RangedAttribute("attribute.name.generic.follow_range", 32.0D, 0.0D, 2048.0D));
     public static final Attribute KNOCKBACK_RESISTANCE = Attributes.register("generic.knockback_resistance", new RangedAttribute("attribute.name.generic.knockback_resistance", 0.0D, 0.0D, 1.0D));
     public static final Attribute MOVEMENT_SPEED = Attributes.register("generic.movement_speed", (new RangedAttribute("attribute.name.generic.movement_speed", 0.699999988079071D, 0.0D, org.spigotmc.SpigotConfig.movementSpeed)).setSyncable(true));
     public static final Attribute FLYING_SPEED = Attributes.register("generic.flying_speed", (new RangedAttribute("attribute.name.generic.flying_speed", 0.4000000059604645D, 0.0D, 1024.0D)).setSyncable(true));
-    public static final Attribute ATTACK_DAMAGE = Attributes.register("generic.attack_damage", new RangedAttribute("attribute.name.generic.attack_damage", 2.0D, 0.0D, org.spigotmc.SpigotConfig.attackDamage));
+    public static final Attribute ATTACK_DAMAGE = Attributes.register("generic.attack_damage", new RangedAttribute("attribute.name.generic.attack_damage", 20.0D, 0.0D, org.spigotmc.SpigotConfig.attackDamage));
     public static final Attribute ATTACK_KNOCKBACK = Attributes.register("generic.attack_knockback", new RangedAttribute("attribute.name.generic.attack_knockback", 0.0D, 0.0D, 5.0D));
     public static final Attribute ATTACK_SPEED = Attributes.register("generic.attack_speed", (new RangedAttribute("attribute.name.generic.attack_speed", 4.0D, 0.0D, 1024.0D)).setSyncable(true));
     public static final Attribute ARMOR = Attributes.register("generic.armor", (new RangedAttribute("attribute.name.generic.armor", 0.0D, 0.0D, 30.0D)).setSyncable(true));
@@ -19,6 +19,33 @@ public class Attributes {
     public static final Attribute SPAWN_REINFORCEMENTS_CHANCE = Attributes.register("zombie.spawn_reinforcements", new RangedAttribute("attribute.name.zombie.spawn_reinforcements", 0.0D, 0.0D, 1.0D));
     public static final Attribute JUMP_STRENGTH = Attributes.register("horse.jump_strength", (new RangedAttribute("attribute.name.horse.jump_strength", 0.7D, 0.0D, 2.0D)).setSyncable(true));
 
+    // Papyrus start - combat changes
+    // Element resistance
+    public static final Attribute RES_PURE = Attributes.register("resistance.pure", new RangedAttribute("attribute.name.resistance.pure", 0, 0, 4096));
+    public static final Attribute RES_PHYSICAL = Attributes.register("resistance.physical", new RangedAttribute("attribute.name.resistance.physical", 0, 0, 4096));
+    public static final Attribute RES_MAGIC = Attributes.register("resistance.magic", new RangedAttribute("attribute.name.resistance.magic", 0, 0, 4096));
+    public static final Attribute RES_AIR = Attributes.register("resistance.air", new RangedAttribute("attribute.name.resistance.air", 0, 0, 4096));
+    public static final Attribute RES_EARTH = Attributes.register("resistance.earth", new RangedAttribute("attribute.name.resistance.earth", 0, 0, 4096));
+    public static final Attribute RES_FIRE = Attributes.register("resistance.fire", new RangedAttribute("attribute.name.resistance.fire", 0, 0, 4096));
+    public static final Attribute RES_WATER = Attributes.register("resistance.water", new RangedAttribute("attribute.name.resistance.water", 0, 0, 4096));
+    // Element advantage
+    public static final Attribute ADV_PURE = Attributes.register("advantage.pure", new RangedAttribute("attribute.name.advantage.pure", 0, 0, 4096));
+    public static final Attribute ADV_PHYSICAL = Attributes.register("advantage.physical", new RangedAttribute("attribute.name.advantage.physical", 0, 0, 4096));
+    public static final Attribute ADV_MAGIC = Attributes.register("advantage.magic", new RangedAttribute("attribute.name.advantage.magic", 0, 0, 4096));
+    public static final Attribute ADV_AIR = Attributes.register("advantage.air", new RangedAttribute("attribute.name.advantage.air", 0, 0, 4096));
+    public static final Attribute ADV_EARTH = Attributes.register("advantage.earth", new RangedAttribute("attribute.name.advantage.earth", 0, 0, 4096));
+    public static final Attribute ADV_FIRE = Attributes.register("advantage.fire", new RangedAttribute("attribute.name.advantage.fire", 0, 0, 4096));
+    public static final Attribute ADV_WATER = Attributes.register("advantage.water", new RangedAttribute("attribute.name.advantage.water", 0, 0, 4096));
+    // Stats
+    public static final Attribute STAT_CRITCHANCE = Attributes.register("stat.crit.chance", new RangedAttribute("attribute.name.stat.crit.chance", 0, 0, 4096));
+    public static final Attribute STAT_CRITDMG = Attributes.register("stat.crit.dmg", new RangedAttribute("attribute.name.stat.crit.dmg", 0, 0, 4096));
+    public static final Attribute STAT_ATTACKSPEED= Attributes.register("stat.attackspeed", new RangedAttribute("attribute.name.stat.attackspeed", 0, 0, 4096));
+    public static final Attribute STAT_CDR = Attributes.register("stat.cooldownreduction", new RangedAttribute("attribute.name.stat.cooldownreduction", 0, 0, 4096));
+    // Combat
+    public static final Attribute COMBAT_HURTINVUL= Attributes.register("combat.hurtinvul", new RangedAttribute("attribute.name.combat.hurtinvul", 5, 0, 4096));
+
+    // Papyrus end
+
     public Attributes() {}
 
     private static Attribute register(String id, Attribute attribute) {
diff --git a/src/main/java/net/minecraft/world/entity/player/Player.java b/src/main/java/net/minecraft/world/entity/player/Player.java
index 9b131f0a827413e9f5d6d0f7491c5481576cb8b1..05dafb5ae30ac77b9917e31d08b49e8dd462e262 100644
--- a/src/main/java/net/minecraft/world/entity/player/Player.java
+++ b/src/main/java/net/minecraft/world/entity/player/Player.java
@@ -17,6 +17,8 @@ import java.util.Optional;
 import java.util.OptionalInt;
 import java.util.function.Predicate;
 import javax.annotation.Nullable;
+
+import de.erethon.papyrus.DamageType;
 import net.minecraft.SharedConstants;
 import net.minecraft.Util;
 import net.minecraft.advancements.CriteriaTriggers;
@@ -131,7 +133,7 @@ public abstract class Player extends LivingEntity {
 
     private static final Logger LOGGER = LogUtils.getLogger();
     public static final int MAX_NAME_LENGTH = 16;
-    public static final int MAX_HEALTH = 20;
+    public static final int MAX_HEALTH = 500;
     public static final int SLEEP_DURATION = 100;
     public static final int WAKE_UP_DURATION = 10;
     public static final int ENDER_SLOT_OFFSET = 200;
@@ -227,7 +229,8 @@ public abstract class Player extends LivingEntity {
     }
 
     public static AttributeSupplier.Builder createAttributes() {
-        return LivingEntity.createLivingAttributes().add(Attributes.ATTACK_DAMAGE, 1.0D).add(Attributes.MOVEMENT_SPEED, 0.10000000149011612D).add(Attributes.ATTACK_SPEED).add(Attributes.LUCK);
+        return LivingEntity.createLivingAttributes().add(Attributes.ATTACK_DAMAGE, 1.0D).add(Attributes.MOVEMENT_SPEED, 0.10000000149011612D).add(Attributes.ATTACK_SPEED).add(Attributes.LUCK)
+            .add(Attributes.MAX_HEALTH, 500);
     }
 
     @Override
@@ -1250,6 +1253,15 @@ public abstract class Player extends LivingEntity {
         return this.onGround || this.fallDistance < this.maxUpStep && !this.level.noCollision(this, this.getBoundingBox().move(0.0D, (double) (this.fallDistance - this.maxUpStep), 0.0D));
     }
 
+    // Papyrus - combat rework
+    public void attack(Entity target, DamageType type) {
+        if (!target.isAttackable()) {
+            return;
+        }
+        float damage = (float) (getAttributeValue(Attributes.ATTACK_DAMAGE) + getAttributeValue(type.getAdvAttr()));
+        target.hurt(DamageSource.playerAttack(this), damage, type);
+    }
+
     public void attack(Entity target) {
         if (target.isAttackable()) {
             if (!target.skipAttackInteraction(this)) {
@@ -1460,7 +1472,7 @@ public abstract class Player extends LivingEntity {
 
     @Override
     protected void doAutoAttackOnTouch(LivingEntity target) {
-        this.attack(target);
+        this.attack(target, DamageType.PHYSICAL); // Papyrus
     }
 
     public void disableShield(boolean sprinting) {
diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
index 4ba927b4585ca56337fa763d60d21de3b58bb2f7..88d4e19c71f65445b582eb577ba9cea6b5c14450 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
@@ -160,8 +160,8 @@ public class CraftPlayer extends CraftHumanEntity implements Player {
     private final Map<UUID, Set<WeakReference<Plugin>>> hiddenEntities = new HashMap<>();
     private static final WeakHashMap<Plugin, WeakReference<Plugin>> pluginWeakReferences = new WeakHashMap<>();
     private int hash = 0;
-    private double health = 20;
-    private boolean scaledHealth = false;
+    private double health = 500; // Papyrus
+    private boolean scaledHealth = true; // Papyrus
     private double healthScale = 20;
     private CraftWorldBorder clientWorldBorder = null;
     private BorderChangeListener clientWorldBorderListener = this.createWorldBorderListener();
