From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: malfrador <malfrador@gmail.com>
Date: Sat, 11 Jun 2022 18:07:35 +0200
Subject: [PATCH] combat changes


diff --git a/src/main/java/de/erethon/papyrus/DamageType.java b/src/main/java/de/erethon/papyrus/DamageType.java
new file mode 100644
index 0000000000000000000000000000000000000000..0cbc67c143dbbe8d8bc1ebf2ea384f6851cf8658
--- /dev/null
+++ b/src/main/java/de/erethon/papyrus/DamageType.java
@@ -0,0 +1,32 @@
+package de.erethon.papyrus;
+
+import net.minecraft.world.entity.ai.attributes.Attribute;
+import net.minecraft.world.entity.ai.attributes.Attributes;
+
+public enum DamageType {
+
+    AIR(Attributes.ADV_AIR, Attributes.RES_AIR),
+    EARTH(Attributes.ADV_EARTH, Attributes.RES_EARTH),
+    FIRE(Attributes.ADV_FIRE, Attributes.RES_FIRE),
+    MAGIC(Attributes.ADV_MAGIC, Attributes.RES_MAGIC),
+    PHYSICAL(Attributes.ADV_PHYSICAL, Attributes.RES_PHYSICAL),
+    WATER(Attributes.ADV_WATER, Attributes.RES_WATER),
+    PURE(Attributes.ADV_PURE, Attributes.RES_PURE);
+
+    private final Attribute advAttr;
+    private final Attribute resAttr;
+
+    DamageType(Attribute advAttr, Attribute resAttr) {
+        this.advAttr = advAttr;
+        this.resAttr = resAttr;
+    }
+
+    public Attribute getAdvAttr() {
+        return this.advAttr;
+    }
+
+
+    public Attribute getResAttr() {
+        return this.resAttr;
+    }
+}
diff --git a/src/main/java/de/erethon/papyrus/PapyrusUtils.java b/src/main/java/de/erethon/papyrus/PapyrusUtils.java
new file mode 100644
index 0000000000000000000000000000000000000000..afe4ed93f622cc46c0bb1faa12976e44297f4ea9
--- /dev/null
+++ b/src/main/java/de/erethon/papyrus/PapyrusUtils.java
@@ -0,0 +1,14 @@
+package de.erethon.papyrus;
+
+import io.papermc.paper.adventure.PaperAdventure;
+import net.kyori.adventure.text.minimessage.MiniMessage;
+import net.minecraft.network.chat.Component;
+
+public class PapyrusUtils {
+
+    private static final char healthBar = '|';
+
+    public static Component getHealthMessage(double health, double max) { // maybe with color based on damage type?
+        return PaperAdventure.asVanilla(MiniMessage.miniMessage().deserialize("<red>" + Math.round(health) + "<dark_gray>/" + "<dark_red>" + Math.round(max)));
+    }
+}
diff --git a/src/main/java/net/minecraft/server/level/ServerPlayer.java b/src/main/java/net/minecraft/server/level/ServerPlayer.java
index 37b3c142d55a88167e1685c7e0b8ca3f84c9b6e4..28305ae8d83b20406a3d6c02b230186eae49eb16 100644
--- a/src/main/java/net/minecraft/server/level/ServerPlayer.java
+++ b/src/main/java/net/minecraft/server/level/ServerPlayer.java
@@ -18,6 +18,9 @@ import java.util.Optional;
 import java.util.OptionalInt;
 import java.util.Set;
 import javax.annotation.Nullable;
+
+import de.erethon.papyrus.DamageType;
+import net.kyori.adventure.text.minimessage.MiniMessage;
 import net.minecraft.BlockUtil;
 import net.minecraft.ChatFormatting;
 import net.minecraft.CrashReport;
@@ -2167,7 +2170,7 @@ public class ServerPlayer extends Player {
         if (this.gameMode.getGameModeForPlayer() == GameType.SPECTATOR) {
             this.setCamera(target);
         } else {
-            super.attack(target);
+            super.attack(target, DamageType.PHYSICAL); // Papyrus
         }
 
     }
diff --git a/src/main/java/net/minecraft/server/network/ServerGamePacketListenerImpl.java b/src/main/java/net/minecraft/server/network/ServerGamePacketListenerImpl.java
index 776c7df81d2b71a5610fe90475f4e8044850beab..9235a0f5cc3f08e2b599101e514e67d6b8ad7c57 100644
--- a/src/main/java/net/minecraft/server/network/ServerGamePacketListenerImpl.java
+++ b/src/main/java/net/minecraft/server/network/ServerGamePacketListenerImpl.java
@@ -323,6 +323,7 @@ public class ServerGamePacketListenerImpl implements ServerPlayerConnection, Tic
         this.chatMessageChain = new FutureChain(server.chatExecutor); // CraftBukkit - async chat
         // CraftBukkit start - add fields and methods
         this.cserver = server.server;
+        LOGGER.info("Initialized packet listener for " + player.getScoreboardName());
     }
 
     private final org.bukkit.craftbukkit.CraftServer cserver;
diff --git a/src/main/java/net/minecraft/world/entity/Entity.java b/src/main/java/net/minecraft/world/entity/Entity.java
index 36795968f4c296f680f79cc5a795391ae13c64e4..f91dbe93fefaafa845bbefe1280f7a10a4819b26 100644
--- a/src/main/java/net/minecraft/world/entity/Entity.java
+++ b/src/main/java/net/minecraft/world/entity/Entity.java
@@ -7,6 +7,7 @@ import com.google.common.collect.Lists;
 import com.google.common.collect.Sets;
 import com.google.common.collect.UnmodifiableIterator;
 import com.mojang.logging.LogUtils;
+import de.erethon.papyrus.DamageType;
 import it.unimi.dsi.fastutil.objects.Object2DoubleArrayMap;
 import it.unimi.dsi.fastutil.objects.Object2DoubleMap;
 import java.util.Arrays;
@@ -2185,6 +2186,16 @@ public abstract class Entity implements Nameable, EntityAccess, CommandSource {
         }
     }
 
+    // Papyrus - combat rework
+    public boolean hurt(DamageSource source, float amount, DamageType type) {
+        if (this.isInvulnerableTo(source)) {
+            return false;
+        } else {
+            this.markHurt();
+            return false;
+        }
+    }
+
     public final Vec3 getViewVector(float tickDelta) {
         return this.calculateViewVector(this.getViewXRot(tickDelta), this.getViewYRot(tickDelta));
     }
diff --git a/src/main/java/net/minecraft/world/entity/Mob.java b/src/main/java/net/minecraft/world/entity/Mob.java
index f6eb032897c6d5d16ab5c8c287e49e189c24571c..e689c8b00dd8167d0027d261df4ae1f25ace4e6e 100644
--- a/src/main/java/net/minecraft/world/entity/Mob.java
+++ b/src/main/java/net/minecraft/world/entity/Mob.java
@@ -9,6 +9,8 @@ import java.util.Optional;
 import java.util.UUID;
 import java.util.function.Predicate;
 import javax.annotation.Nullable;
+
+import de.erethon.papyrus.DamageType;
 import net.minecraft.core.BlockPos;
 import net.minecraft.core.NonNullList;
 import net.minecraft.core.Vec3i;
@@ -1647,12 +1649,16 @@ public abstract class Mob extends LivingEntity implements Targeting {
 
     @Override
     public boolean doHurtTarget(Entity target) {
-        float f = (float) this.getAttributeValue(Attributes.ATTACK_DAMAGE);
-        float f1 = (float) this.getAttributeValue(Attributes.ATTACK_KNOCKBACK);
+        return doHurtTarget(target, attackDamageType);
+    }
+
+    public boolean doHurtTarget(Entity target, DamageType type) {
+        float dmg = (float) this.getAttributeValue(Attributes.ATTACK_DAMAGE);
+        float knockback = (float) this.getAttributeValue(Attributes.ATTACK_KNOCKBACK);
 
         if (target instanceof LivingEntity) {
-            f += EnchantmentHelper.getDamageBonus(this.getMainHandItem(), ((LivingEntity) target).getMobType());
-            f1 += (float) EnchantmentHelper.getKnockbackBonus(this);
+            dmg += EnchantmentHelper.getDamageBonus(this.getMainHandItem(), ((LivingEntity) target).getMobType());
+            knockback += (float) EnchantmentHelper.getKnockbackBonus(this);
         }
 
         int i = EnchantmentHelper.getFireAspect(this);
@@ -1667,12 +1673,12 @@ public abstract class Mob extends LivingEntity implements Targeting {
             }
             // CraftBukkit end
         }
-
-        boolean flag = target.hurt(this.damageSources().mobAttack(this), f);
+        dmg = (float) (dmg + getAttributeValue(type.getAdvAttr()));
+        boolean flag = target.hurt(DamageSource.mobAttack(this), dmg, type);
 
         if (flag) {
-            if (f1 > 0.0F && target instanceof LivingEntity) {
-                ((LivingEntity) target).knockback((double) (f1 * 0.5F), (double) Mth.sin(this.getYRot() * 0.017453292F), (double) (-Mth.cos(this.getYRot() * 0.017453292F)), this); // Paper
+            if (knockback > 0.0F && target instanceof LivingEntity) {
+                ((LivingEntity) target).knockback((double) (knockback * 0.5F), (double) Mth.sin(this.getYRot() * 0.017453292F), (double) (-Mth.cos(this.getYRot() * 0.017453292F)), this); // Paper
                 this.setDeltaMovement(this.getDeltaMovement().multiply(0.6D, 1.0D, 0.6D));
             }
 
diff --git a/src/main/java/net/minecraft/world/entity/ai/attributes/Attributes.java b/src/main/java/net/minecraft/world/entity/ai/attributes/Attributes.java
index 2878bcbfd41839249b160e67b83d5dad149c94db..57be3d8f01e53a1edf8f92cd6cf35bede099602f 100644
--- a/src/main/java/net/minecraft/world/entity/ai/attributes/Attributes.java
+++ b/src/main/java/net/minecraft/world/entity/ai/attributes/Attributes.java
@@ -6,12 +6,12 @@ import net.minecraft.core.registries.BuiltInRegistries;
 
 public class Attributes {
 
-    public static final Attribute MAX_HEALTH = Attributes.register("generic.max_health", (new RangedAttribute("attribute.name.generic.max_health", 20.0D, 1.0D, org.spigotmc.SpigotConfig.maxHealth)).setSyncable(true));
+    public static final Attribute MAX_HEALTH = Attributes.register("generic.max_health", new RangedAttribute("attribute.name.generic.max_health", 500.0D, 1.0D, org.spigotmc.SpigotConfig.maxHealth).setSyncable(true));
     public static final Attribute FOLLOW_RANGE = Attributes.register("generic.follow_range", new RangedAttribute("attribute.name.generic.follow_range", 32.0D, 0.0D, 2048.0D));
     public static final Attribute KNOCKBACK_RESISTANCE = Attributes.register("generic.knockback_resistance", new RangedAttribute("attribute.name.generic.knockback_resistance", 0.0D, 0.0D, 1.0D));
     public static final Attribute MOVEMENT_SPEED = Attributes.register("generic.movement_speed", (new RangedAttribute("attribute.name.generic.movement_speed", 0.699999988079071D, 0.0D, org.spigotmc.SpigotConfig.movementSpeed)).setSyncable(true));
     public static final Attribute FLYING_SPEED = Attributes.register("generic.flying_speed", (new RangedAttribute("attribute.name.generic.flying_speed", 0.4000000059604645D, 0.0D, 1024.0D)).setSyncable(true));
-    public static final Attribute ATTACK_DAMAGE = Attributes.register("generic.attack_damage", new RangedAttribute("attribute.name.generic.attack_damage", 2.0D, 0.0D, org.spigotmc.SpigotConfig.attackDamage));
+    public static final Attribute ATTACK_DAMAGE = Attributes.register("generic.attack_damage", new RangedAttribute("attribute.name.generic.attack_damage", 20.0D, 0.0D, org.spigotmc.SpigotConfig.attackDamage));
     public static final Attribute ATTACK_KNOCKBACK = Attributes.register("generic.attack_knockback", new RangedAttribute("attribute.name.generic.attack_knockback", 0.0D, 0.0D, 5.0D));
     public static final Attribute ATTACK_SPEED = Attributes.register("generic.attack_speed", (new RangedAttribute("attribute.name.generic.attack_speed", 4.0D, 0.0D, 1024.0D)).setSyncable(true));
     public static final Attribute ARMOR = Attributes.register("generic.armor", (new RangedAttribute("attribute.name.generic.armor", 0.0D, 0.0D, 30.0D)).setSyncable(true));
@@ -20,6 +20,33 @@ public class Attributes {
     public static final Attribute SPAWN_REINFORCEMENTS_CHANCE = Attributes.register("zombie.spawn_reinforcements", new RangedAttribute("attribute.name.zombie.spawn_reinforcements", 0.0D, 0.0D, 1.0D));
     public static final Attribute JUMP_STRENGTH = Attributes.register("horse.jump_strength", (new RangedAttribute("attribute.name.horse.jump_strength", 0.7D, 0.0D, 2.0D)).setSyncable(true));
 
+    // Papyrus start - combat changes
+    // Element resistance
+    public static final Attribute RES_PURE = Attributes.register("resistance.pure", new RangedAttribute("attribute.name.resistance.pure", 0, 0, 4096));
+    public static final Attribute RES_PHYSICAL = Attributes.register("resistance.physical", new RangedAttribute("attribute.name.resistance.physical", 0, 0, 4096));
+    public static final Attribute RES_MAGIC = Attributes.register("resistance.magic", new RangedAttribute("attribute.name.resistance.magic", 0, 0, 4096));
+    public static final Attribute RES_AIR = Attributes.register("resistance.air", new RangedAttribute("attribute.name.resistance.air", 0, 0, 4096));
+    public static final Attribute RES_EARTH = Attributes.register("resistance.earth", new RangedAttribute("attribute.name.resistance.earth", 0, 0, 4096));
+    public static final Attribute RES_FIRE = Attributes.register("resistance.fire", new RangedAttribute("attribute.name.resistance.fire", 0, 0, 4096));
+    public static final Attribute RES_WATER = Attributes.register("resistance.water", new RangedAttribute("attribute.name.resistance.water", 0, 0, 4096));
+    // Element advantage
+    public static final Attribute ADV_PURE = Attributes.register("advantage.pure", new RangedAttribute("attribute.name.advantage.pure", 0, 0, 4096));
+    public static final Attribute ADV_PHYSICAL = Attributes.register("advantage.physical", new RangedAttribute("attribute.name.advantage.physical", 0, 0, 4096));
+    public static final Attribute ADV_MAGIC = Attributes.register("advantage.magic", new RangedAttribute("attribute.name.advantage.magic", 0, 0, 4096));
+    public static final Attribute ADV_AIR = Attributes.register("advantage.air", new RangedAttribute("attribute.name.advantage.air", 0, 0, 4096));
+    public static final Attribute ADV_EARTH = Attributes.register("advantage.earth", new RangedAttribute("attribute.name.advantage.earth", 0, 0, 4096));
+    public static final Attribute ADV_FIRE = Attributes.register("advantage.fire", new RangedAttribute("attribute.name.advantage.fire", 0, 0, 4096));
+    public static final Attribute ADV_WATER = Attributes.register("advantage.water", new RangedAttribute("attribute.name.advantage.water", 0, 0, 4096));
+    // Stats
+    public static final Attribute STAT_CRITCHANCE = Attributes.register("stat.crit.chance", new RangedAttribute("attribute.name.stat.crit.chance", 0, 0, 4096));
+    public static final Attribute STAT_CRITDMG = Attributes.register("stat.crit.dmg", new RangedAttribute("attribute.name.stat.crit.dmg", 0, 0, 4096));
+    public static final Attribute STAT_ATTACKSPEED= Attributes.register("stat.attackspeed", new RangedAttribute("attribute.name.stat.attackspeed", 0, 0, 4096));
+    public static final Attribute STAT_CDR = Attributes.register("stat.cooldownreduction", new RangedAttribute("attribute.name.stat.cooldownreduction", 0, 0, 4096));
+    // Combat
+    public static final Attribute COMBAT_HURTINVUL= Attributes.register("combat.hurtinvul", new RangedAttribute("attribute.name.combat.hurtinvul", 5, 0, 4096));
+
+    // Papyrus end
+
     public Attributes() {}
 
     private static Attribute register(String id, Attribute attribute) {
diff --git a/src/main/java/net/minecraft/world/entity/player/Player.java b/src/main/java/net/minecraft/world/entity/player/Player.java
index 58152160d609d0e9d105153aeb166a56a7955603..1a977285f028ee6082b810da75582c2f5968186b 100644
--- a/src/main/java/net/minecraft/world/entity/player/Player.java
+++ b/src/main/java/net/minecraft/world/entity/player/Player.java
@@ -16,6 +16,9 @@ import java.util.Optional;
 import java.util.OptionalInt;
 import java.util.function.Predicate;
 import javax.annotation.Nullable;
+import net.minecraft.SharedConstants;
+import de.erethon.papyrus.DamageType;
+import net.minecraft.SharedConstants;
 import net.minecraft.Util;
 import net.minecraft.advancements.CriteriaTriggers;
 import net.minecraft.core.BlockPos;
@@ -132,7 +135,7 @@ public abstract class Player extends LivingEntity {
 
     private static final Logger LOGGER = LogUtils.getLogger();
     public static final int MAX_NAME_LENGTH = 16;
-    public static final int MAX_HEALTH = 20;
+    public static final int MAX_HEALTH = 500;
     public static final int SLEEP_DURATION = 100;
     public static final int WAKE_UP_DURATION = 10;
     public static final int ENDER_SLOT_OFFSET = 200;
@@ -226,7 +229,8 @@ public abstract class Player extends LivingEntity {
     }
 
     public static AttributeSupplier.Builder createAttributes() {
-        return LivingEntity.createLivingAttributes().add(Attributes.ATTACK_DAMAGE, 1.0D).add(Attributes.MOVEMENT_SPEED, 0.10000000149011612D).add(Attributes.ATTACK_SPEED).add(Attributes.LUCK);
+        return LivingEntity.createLivingAttributes().add(Attributes.ATTACK_DAMAGE, 1.0D).add(Attributes.MOVEMENT_SPEED, 0.10000000149011612D).add(Attributes.ATTACK_SPEED).add(Attributes.LUCK)
+            .add(Attributes.MAX_HEALTH, 500);
     }
 
     @Override
@@ -1239,6 +1243,15 @@ public abstract class Player extends LivingEntity {
         return this.onGround() || this.fallDistance < this.maxUpStep() && !this.level().noCollision(this, this.getBoundingBox().move(0.0D, (double) (this.fallDistance - this.maxUpStep()), 0.0D));
     }
 
+    // Papyrus - combat rework
+    public void attack(Entity target, DamageType type) {
+        if (!target.isAttackable()) {
+            return;
+        }
+        float damage = (float) (getAttributeValue(Attributes.ATTACK_DAMAGE) + getAttributeValue(type.getAdvAttr()));
+        target.hurt(DamageSource.playerAttack(this), damage, type);
+    }
+
     public void attack(Entity target) {
         // Paper start - PlayerAttackEntityEvent
         boolean willAttack = target.isAttackable() && !target.skipAttackInteraction(this); // Vanilla logic
@@ -1458,7 +1471,7 @@ public abstract class Player extends LivingEntity {
 
     @Override
     protected void doAutoAttackOnTouch(LivingEntity target) {
-        this.attack(target);
+        this.attack(target, DamageType.PHYSICAL); // Papyrus
     }
 
     public void disableShield(boolean sprinting) {
diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
index 4e6fea7cf11b1e29ae7c7098a6f5d06bb5f93cc2..b5bcc80b460dbb58db633ad55d25d92d3e3cad3d 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
@@ -185,8 +185,8 @@ public class CraftPlayer extends CraftHumanEntity implements Player {
     private final Set<UUID> unlistedEntities = new HashSet<>(); // Paper
     private static final WeakHashMap<Plugin, WeakReference<Plugin>> pluginWeakReferences = new WeakHashMap<>();
     private int hash = 0;
-    private double health = 20;
-    private boolean scaledHealth = false;
+    private double health = 500; // Papyrus
+    private boolean scaledHealth = true; // Papyrus
     private double healthScale = 20;
     private CraftWorldBorder clientWorldBorder = null;
     private BorderChangeListener clientWorldBorderListener = this.createWorldBorderListener();
