From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: malfrador <malfrador@gmail.com>
Date: Sat, 11 Jun 2022 18:07:35 +0200
Subject: [PATCH] combat changes


diff --git a/src/main/java/de/erethon/papyrus/DamageType.java b/src/main/java/de/erethon/papyrus/DamageType.java
new file mode 100644
index 0000000000000000000000000000000000000000..0cbc67c143dbbe8d8bc1ebf2ea384f6851cf8658
--- /dev/null
+++ b/src/main/java/de/erethon/papyrus/DamageType.java
@@ -0,0 +1,32 @@
+package de.erethon.papyrus;
+
+import net.minecraft.world.entity.ai.attributes.Attribute;
+import net.minecraft.world.entity.ai.attributes.Attributes;
+
+public enum DamageType {
+
+    AIR(Attributes.ADV_AIR, Attributes.RES_AIR),
+    EARTH(Attributes.ADV_EARTH, Attributes.RES_EARTH),
+    FIRE(Attributes.ADV_FIRE, Attributes.RES_FIRE),
+    MAGIC(Attributes.ADV_MAGIC, Attributes.RES_MAGIC),
+    PHYSICAL(Attributes.ADV_PHYSICAL, Attributes.RES_PHYSICAL),
+    WATER(Attributes.ADV_WATER, Attributes.RES_WATER),
+    PURE(Attributes.ADV_PURE, Attributes.RES_PURE);
+
+    private final Attribute advAttr;
+    private final Attribute resAttr;
+
+    DamageType(Attribute advAttr, Attribute resAttr) {
+        this.advAttr = advAttr;
+        this.resAttr = resAttr;
+    }
+
+    public Attribute getAdvAttr() {
+        return this.advAttr;
+    }
+
+
+    public Attribute getResAttr() {
+        return this.resAttr;
+    }
+}
diff --git a/src/main/java/de/erethon/papyrus/PapyrusUtils.java b/src/main/java/de/erethon/papyrus/PapyrusUtils.java
new file mode 100644
index 0000000000000000000000000000000000000000..afe4ed93f622cc46c0bb1faa12976e44297f4ea9
--- /dev/null
+++ b/src/main/java/de/erethon/papyrus/PapyrusUtils.java
@@ -0,0 +1,14 @@
+package de.erethon.papyrus;
+
+import io.papermc.paper.adventure.PaperAdventure;
+import net.kyori.adventure.text.minimessage.MiniMessage;
+import net.minecraft.network.chat.Component;
+
+public class PapyrusUtils {
+
+    private static final char healthBar = '|';
+
+    public static Component getHealthMessage(double health, double max) { // maybe with color based on damage type?
+        return PaperAdventure.asVanilla(MiniMessage.miniMessage().deserialize("<red>" + Math.round(health) + "<dark_gray>/" + "<dark_red>" + Math.round(max)));
+    }
+}
diff --git a/src/main/java/net/minecraft/server/level/ServerPlayer.java b/src/main/java/net/minecraft/server/level/ServerPlayer.java
index 48ae5c5c48a1ad016acee57468973aaab687ab72..beeaa5e67520be1f5e02516f91e84b32b8db5151 100644
--- a/src/main/java/net/minecraft/server/level/ServerPlayer.java
+++ b/src/main/java/net/minecraft/server/level/ServerPlayer.java
@@ -6,6 +6,7 @@ import com.mojang.datafixers.util.Either;
 import com.mojang.logging.LogUtils;
 import com.mojang.serialization.DataResult;
 import com.mojang.serialization.Dynamic;
+import de.erethon.papyrus.DamageType;
 import io.papermc.paper.adventure.PaperAdventure;
 import net.minecraft.*;
 import net.minecraft.advancements.CriteriaTriggers;
@@ -2203,7 +2204,7 @@ public class ServerPlayer extends Player {
         if (this.gameMode.getGameModeForPlayer() == GameType.SPECTATOR) {
             this.setCamera(target);
         } else {
-            super.attack(target);
+            super.attack(target, DamageType.PHYSICAL); // Papyrus
         }
 
     }
diff --git a/src/main/java/net/minecraft/server/network/ServerGamePacketListenerImpl.java b/src/main/java/net/minecraft/server/network/ServerGamePacketListenerImpl.java
index df07e24ad5e1eaf5211d307965107ec1776cfa19..7645c68694b64f6161ade4e2fc929ca4dee0628d 100644
--- a/src/main/java/net/minecraft/server/network/ServerGamePacketListenerImpl.java
+++ b/src/main/java/net/minecraft/server/network/ServerGamePacketListenerImpl.java
@@ -307,6 +307,7 @@ public class ServerGamePacketListenerImpl extends ServerCommonPacketListenerImpl
         Objects.requireNonNull(server);
         this.signedMessageDecoder = SignedMessageChain.Decoder.unsigned(uuid, server::enforceSecureProfile);
         this.chatMessageChain = new FutureChain(server.chatExecutor); // CraftBukkit - async chat
+        LOGGER.info("Initialized packet listener for " + player.getScoreboardName());
     }
 
     // CraftBukkit start - add fields
diff --git a/src/main/java/net/minecraft/world/entity/Entity.java b/src/main/java/net/minecraft/world/entity/Entity.java
index 5cbe6cb0676ac23e184e7586e2dacbbc1d5fb218..3345e06351d6daa4cad0971a738a36b0f97165f4 100644
--- a/src/main/java/net/minecraft/world/entity/Entity.java
+++ b/src/main/java/net/minecraft/world/entity/Entity.java
@@ -7,6 +7,7 @@ import com.google.common.collect.Lists;
 import com.google.common.collect.Sets;
 import com.google.common.collect.UnmodifiableIterator;
 import com.mojang.logging.LogUtils;
+import de.erethon.papyrus.DamageType;
 import it.unimi.dsi.fastutil.objects.Object2DoubleArrayMap;
 import it.unimi.dsi.fastutil.objects.Object2DoubleMap;
 import java.util.Arrays;
@@ -2183,6 +2184,16 @@ public abstract class Entity implements Nameable, EntityAccess, CommandSource, S
         }
     }
 
+    // Papyrus - combat rework
+    public boolean hurt(DamageSource source, float amount, DamageType type) {
+        if (this.isInvulnerableTo(source)) {
+            return false;
+        } else {
+            this.markHurt();
+            return false;
+        }
+    }
+
     public final Vec3 getViewVector(float tickDelta) {
         return this.calculateViewVector(this.getViewXRot(tickDelta), this.getViewYRot(tickDelta));
     }
diff --git a/src/main/java/net/minecraft/world/entity/Mob.java b/src/main/java/net/minecraft/world/entity/Mob.java
index 40fbf1bc4409c289fe9ca154ff306be52708540d..b5f85491721d81ccda13593a7fbbd57af92a8d32 100644
--- a/src/main/java/net/minecraft/world/entity/Mob.java
+++ b/src/main/java/net/minecraft/world/entity/Mob.java
@@ -10,6 +10,8 @@ import java.util.Optional;
 import java.util.UUID;
 import java.util.function.Predicate;
 import javax.annotation.Nullable;
+
+import de.erethon.papyrus.DamageType;
 import net.minecraft.core.BlockPos;
 import net.minecraft.core.NonNullList;
 import net.minecraft.core.Vec3i;
@@ -1662,12 +1664,16 @@ public abstract class Mob extends LivingEntity implements Targeting {
 
     @Override
     public boolean doHurtTarget(Entity target) {
-        float f = (float) this.getAttributeValue(Attributes.ATTACK_DAMAGE);
-        float f1 = (float) this.getAttributeValue(Attributes.ATTACK_KNOCKBACK);
+        return doHurtTarget(target, attackDamageType);
+    }
+
+    public boolean doHurtTarget(Entity target, DamageType type) {
+        float dmg = (float) this.getAttributeValue(Attributes.ATTACK_DAMAGE);
+        float knockback = (float) this.getAttributeValue(Attributes.ATTACK_KNOCKBACK);
 
         if (target instanceof LivingEntity) {
-            f += EnchantmentHelper.getDamageBonus(this.getMainHandItem(), ((LivingEntity) target).getMobType());
-            f1 += (float) EnchantmentHelper.getKnockbackBonus(this);
+            dmg += EnchantmentHelper.getDamageBonus(this.getMainHandItem(), ((LivingEntity) target).getMobType());
+            knockback += (float) EnchantmentHelper.getKnockbackBonus(this);
         }
 
         int i = EnchantmentHelper.getFireAspect(this);
@@ -1682,12 +1688,12 @@ public abstract class Mob extends LivingEntity implements Targeting {
             }
             // CraftBukkit end
         }
-
-        boolean flag = target.hurt(this.damageSources().mobAttack(this), f);
+        dmg = (float) (dmg + getAttributeValue(type.getAdvAttr()));
+        boolean flag = target.hurt(DamageSource.mobAttack(this), dmg, type);
 
         if (flag) {
-            if (f1 > 0.0F && target instanceof LivingEntity) {
-                ((LivingEntity) target).knockback((double) (f1 * 0.5F), (double) Mth.sin(this.getYRot() * 0.017453292F), (double) (-Mth.cos(this.getYRot() * 0.017453292F)), this); // Paper
+            if (knockback > 0.0F && target instanceof LivingEntity) {
+                ((LivingEntity) target).knockback((double) (knockback * 0.5F), (double) Mth.sin(this.getYRot() * 0.017453292F), (double) (-Mth.cos(this.getYRot() * 0.017453292F)), this); // Paper
                 this.setDeltaMovement(this.getDeltaMovement().multiply(0.6D, 1.0D, 0.6D));
             }
 
diff --git a/src/main/java/net/minecraft/world/entity/ai/attributes/Attributes.java b/src/main/java/net/minecraft/world/entity/ai/attributes/Attributes.java
index 80a83bb6123bbacadcd84674686235f322910779..89acc14db04496b4273ce4db8d2cd37cd998e72c 100644
--- a/src/main/java/net/minecraft/world/entity/ai/attributes/Attributes.java
+++ b/src/main/java/net/minecraft/world/entity/ai/attributes/Attributes.java
@@ -6,21 +6,49 @@ import net.minecraft.core.registries.BuiltInRegistries;
 
 public class Attributes {
 
+
     public static final Attribute ARMOR = Attributes.register("generic.armor", (new RangedAttribute("attribute.name.generic.armor", 0.0D, 0.0D, 30.0D)).setSyncable(true));
     public static final Attribute ARMOR_TOUGHNESS = Attributes.register("generic.armor_toughness", (new RangedAttribute("attribute.name.generic.armor_toughness", 0.0D, 0.0D, 20.0D)).setSyncable(true));
-    public static final Attribute ATTACK_DAMAGE = Attributes.register("generic.attack_damage", new RangedAttribute("attribute.name.generic.attack_damage", 2.0D, 0.0D, org.spigotmc.SpigotConfig.attackDamage));
+    public static final Attribute MAX_HEALTH = Attributes.register("generic.max_health", (new RangedAttribute("attribute.name.generic.max_health", 20.0D, 1.0D, org.spigotmc.SpigotConfig.maxHealth)).setSyncable(true));
+    public static final Attribute FOLLOW_RANGE = Attributes.register("generic.follow_range", new RangedAttribute("attribute.name.generic.follow_range", 32.0D, 0.0D, 2048.0D));
+    public static final Attribute KNOCKBACK_RESISTANCE = Attributes.register("generic.knockback_resistance", new RangedAttribute("attribute.name.generic.knockback_resistance", 0.0D, 0.0D, 1.0D));
+    public static final Attribute MOVEMENT_SPEED = Attributes.register("generic.movement_speed", (new RangedAttribute("attribute.name.generic.movement_speed", 0.699999988079071D, 0.0D, org.spigotmc.SpigotConfig.movementSpeed)).setSyncable(true));
+    public static final Attribute FLYING_SPEED = Attributes.register("generic.flying_speed", (new RangedAttribute("attribute.name.generic.flying_speed", 0.4000000059604645D, 0.0D, 1024.0D)).setSyncable(true));
+    public static final Attribute ATTACK_DAMAGE = Attributes.register("generic.attack_damage", new RangedAttribute("attribute.name.generic.attack_damage", 20.0D, 0.0D, org.spigotmc.SpigotConfig.attackDamage));
     public static final Attribute ATTACK_KNOCKBACK = Attributes.register("generic.attack_knockback", new RangedAttribute("attribute.name.generic.attack_knockback", 0.0D, 0.0D, 5.0D));
     public static final Attribute ATTACK_SPEED = Attributes.register("generic.attack_speed", (new RangedAttribute("attribute.name.generic.attack_speed", 4.0D, 0.0D, 1024.0D)).setSyncable(true));
-    public static final Attribute FLYING_SPEED = Attributes.register("generic.flying_speed", (new RangedAttribute("attribute.name.generic.flying_speed", 0.4000000059604645D, 0.0D, 1024.0D)).setSyncable(true));
-    public static final Attribute FOLLOW_RANGE = Attributes.register("generic.follow_range", new RangedAttribute("attribute.name.generic.follow_range", 32.0D, 0.0D, 2048.0D));
     public static final Attribute JUMP_STRENGTH = Attributes.register("horse.jump_strength", (new RangedAttribute("attribute.name.horse.jump_strength", 0.7D, 0.0D, 2.0D)).setSyncable(true));
-    public static final Attribute KNOCKBACK_RESISTANCE = Attributes.register("generic.knockback_resistance", new RangedAttribute("attribute.name.generic.knockback_resistance", 0.0D, 0.0D, 1.0D));
     public static final Attribute LUCK = Attributes.register("generic.luck", (new RangedAttribute("attribute.name.generic.luck", 0.0D, -1024.0D, 1024.0D)).setSyncable(true));
     public static final Attribute MAX_ABSORPTION = Attributes.register("generic.max_absorption", new RangedAttribute("attribute.name.generic.max_absorption", 0.0D, 0.0D, 2048.0D)).setSyncable(true);
-    public static final Attribute MAX_HEALTH = Attributes.register("generic.max_health", (new RangedAttribute("attribute.name.generic.max_health", 20.0D, 1.0D, org.spigotmc.SpigotConfig.maxHealth)).setSyncable(true));
-    public static final Attribute MOVEMENT_SPEED = Attributes.register("generic.movement_speed", (new RangedAttribute("attribute.name.generic.movement_speed", 0.699999988079071D, 0.0D, org.spigotmc.SpigotConfig.movementSpeed)).setSyncable(true));
     public static final Attribute SPAWN_REINFORCEMENTS_CHANCE = Attributes.register("zombie.spawn_reinforcements", new RangedAttribute("attribute.name.zombie.spawn_reinforcements", 0.0D, 0.0D, 1.0D));
 
+    // Papyrus start - combat changes
+    // Element resistance
+    public static final Attribute RES_PURE = Attributes.register("resistance.pure", new RangedAttribute("attribute.name.resistance.pure", 0, 0, 4096));
+    public static final Attribute RES_PHYSICAL = Attributes.register("resistance.physical", new RangedAttribute("attribute.name.resistance.physical", 0, 0, 4096));
+    public static final Attribute RES_MAGIC = Attributes.register("resistance.magic", new RangedAttribute("attribute.name.resistance.magic", 0, 0, 4096));
+    public static final Attribute RES_AIR = Attributes.register("resistance.air", new RangedAttribute("attribute.name.resistance.air", 0, 0, 4096));
+    public static final Attribute RES_EARTH = Attributes.register("resistance.earth", new RangedAttribute("attribute.name.resistance.earth", 0, 0, 4096));
+    public static final Attribute RES_FIRE = Attributes.register("resistance.fire", new RangedAttribute("attribute.name.resistance.fire", 0, 0, 4096));
+    public static final Attribute RES_WATER = Attributes.register("resistance.water", new RangedAttribute("attribute.name.resistance.water", 0, 0, 4096));
+    // Element advantage
+    public static final Attribute ADV_PURE = Attributes.register("advantage.pure", new RangedAttribute("attribute.name.advantage.pure", 0, 0, 4096));
+    public static final Attribute ADV_PHYSICAL = Attributes.register("advantage.physical", new RangedAttribute("attribute.name.advantage.physical", 0, 0, 4096));
+    public static final Attribute ADV_MAGIC = Attributes.register("advantage.magic", new RangedAttribute("attribute.name.advantage.magic", 0, 0, 4096));
+    public static final Attribute ADV_AIR = Attributes.register("advantage.air", new RangedAttribute("attribute.name.advantage.air", 0, 0, 4096));
+    public static final Attribute ADV_EARTH = Attributes.register("advantage.earth", new RangedAttribute("attribute.name.advantage.earth", 0, 0, 4096));
+    public static final Attribute ADV_FIRE = Attributes.register("advantage.fire", new RangedAttribute("attribute.name.advantage.fire", 0, 0, 4096));
+    public static final Attribute ADV_WATER = Attributes.register("advantage.water", new RangedAttribute("attribute.name.advantage.water", 0, 0, 4096));
+    // Stats
+    public static final Attribute STAT_CRITCHANCE = Attributes.register("stat.crit.chance", new RangedAttribute("attribute.name.stat.crit.chance", 0, 0, 4096));
+    public static final Attribute STAT_CRITDMG = Attributes.register("stat.crit.dmg", new RangedAttribute("attribute.name.stat.crit.dmg", 0, 0, 4096));
+    public static final Attribute STAT_ATTACKSPEED= Attributes.register("stat.attackspeed", new RangedAttribute("attribute.name.stat.attackspeed", 0, 0, 4096));
+    public static final Attribute STAT_CDR = Attributes.register("stat.cooldownreduction", new RangedAttribute("attribute.name.stat.cooldownreduction", 0, 0, 4096));
+    // Combat
+    public static final Attribute COMBAT_HURTINVUL= Attributes.register("combat.hurtinvul", new RangedAttribute("attribute.name.combat.hurtinvul", 5, 0, 4096));
+
+    // Papyrus end
+
     public Attributes() {}
 
     private static Attribute register(String id, Attribute attribute) {
diff --git a/src/main/java/net/minecraft/world/entity/player/Player.java b/src/main/java/net/minecraft/world/entity/player/Player.java
index ccc1caafb0ada52c7b99b7358253826f5390843e..84bc777cdb43f8f4b038deec46e4538fa03dfadb 100644
--- a/src/main/java/net/minecraft/world/entity/player/Player.java
+++ b/src/main/java/net/minecraft/world/entity/player/Player.java
@@ -7,15 +7,7 @@ import com.mojang.authlib.GameProfile;
 import com.mojang.datafixers.util.Either;
 import com.mojang.logging.LogUtils;
 import com.mojang.serialization.DataResult;
-import java.util.Collection;
-import java.util.Iterator;
-import java.util.List;
-import java.util.Map;
-import java.util.Objects;
-import java.util.Optional;
-import java.util.OptionalInt;
-import java.util.function.Predicate;
-import javax.annotation.Nullable;
+import de.erethon.papyrus.DamageType;
 import net.minecraft.Util;
 import net.minecraft.advancements.CriteriaTriggers;
 import net.minecraft.core.BlockPos;
@@ -24,11 +16,7 @@ import net.minecraft.core.GlobalPos;
 import net.minecraft.core.particles.ParticleOptions;
 import net.minecraft.core.particles.ParticleTypes;
 import net.minecraft.core.registries.Registries;
-import net.minecraft.nbt.CompoundTag;
-import net.minecraft.nbt.ListTag;
-import net.minecraft.nbt.NbtOps;
-import net.minecraft.nbt.NbtUtils;
-import net.minecraft.nbt.Tag;
+import net.minecraft.nbt.*;
 import net.minecraft.network.chat.ClickEvent;
 import net.minecraft.network.chat.Component;
 import net.minecraft.network.chat.MutableComponent;
@@ -49,27 +37,12 @@ import net.minecraft.tags.DamageTypeTags;
 import net.minecraft.tags.FluidTags;
 import net.minecraft.util.Mth;
 import net.minecraft.util.Unit;
-import net.minecraft.world.Container;
-import net.minecraft.world.Difficulty;
-import net.minecraft.world.InteractionHand;
-import net.minecraft.world.InteractionResult;
-import net.minecraft.world.MenuProvider;
+import net.minecraft.world.*;
 import net.minecraft.world.damagesource.DamageSource;
 import net.minecraft.world.effect.MobEffectInstance;
 import net.minecraft.world.effect.MobEffectUtil;
 import net.minecraft.world.effect.MobEffects;
-import net.minecraft.world.entity.Entity;
-import net.minecraft.world.entity.EntityDimensions;
-import net.minecraft.world.entity.EntityType;
-import net.minecraft.world.entity.EquipmentSlot;
-import net.minecraft.world.entity.HumanoidArm;
-import net.minecraft.world.entity.LivingEntity;
-import net.minecraft.world.entity.Mob;
-import net.minecraft.world.entity.MobType;
-import net.minecraft.world.entity.MoverType;
-import net.minecraft.world.entity.Pose;
-import net.minecraft.world.entity.SlotAccess;
-import net.minecraft.world.entity.TamableAnimal;
+import net.minecraft.world.entity.*;
 import net.minecraft.world.entity.ai.attributes.AttributeSupplier;
 import net.minecraft.world.entity.ai.attributes.Attributes;
 import net.minecraft.world.entity.animal.Parrot;
@@ -84,12 +57,7 @@ import net.minecraft.world.inventory.AbstractContainerMenu;
 import net.minecraft.world.inventory.ClickAction;
 import net.minecraft.world.inventory.InventoryMenu;
 import net.minecraft.world.inventory.PlayerEnderChestContainer;
-import net.minecraft.world.item.ElytraItem;
-import net.minecraft.world.item.ItemCooldowns;
-import net.minecraft.world.item.ItemStack;
-import net.minecraft.world.item.Items;
-import net.minecraft.world.item.ProjectileWeaponItem;
-import net.minecraft.world.item.SwordItem;
+import net.minecraft.world.item.*;
 import net.minecraft.world.item.crafting.RecipeHolder;
 import net.minecraft.world.item.enchantment.EnchantmentHelper;
 import net.minecraft.world.item.trading.MerchantOffers;
@@ -111,7 +79,6 @@ import net.minecraft.world.phys.AABB;
 import net.minecraft.world.phys.Vec3;
 import net.minecraft.world.scores.PlayerTeam;
 import net.minecraft.world.scores.Scoreboard;
-import org.slf4j.Logger;
 import org.bukkit.craftbukkit.entity.CraftHumanEntity;
 import org.bukkit.craftbukkit.event.CraftEventFactory;
 import org.bukkit.craftbukkit.util.CraftVector;
@@ -121,6 +88,11 @@ import org.bukkit.event.entity.EntityCombustByEntityEvent;
 import org.bukkit.event.entity.EntityExhaustionEvent;
 import org.bukkit.event.player.PlayerDropItemEvent;
 import org.bukkit.event.player.PlayerVelocityEvent;
+import org.slf4j.Logger;
+
+import javax.annotation.Nullable;
+import java.util.*;
+import java.util.function.Predicate;
 // CraftBukkit end
 
 public abstract class Player extends LivingEntity {
@@ -129,7 +101,7 @@ public abstract class Player extends LivingEntity {
     public static final int MAX_NAME_LENGTH = 16;
     public static final HumanoidArm DEFAULT_MAIN_HAND = HumanoidArm.RIGHT;
     public static final int DEFAULT_MODEL_CUSTOMIZATION = 0;
-    public static final int MAX_HEALTH = 20;
+    public static final int MAX_HEALTH = 500;
     public static final int SLEEP_DURATION = 100;
     public static final int WAKE_UP_DURATION = 10;
     public static final int ENDER_SLOT_OFFSET = 200;
@@ -222,7 +194,8 @@ public abstract class Player extends LivingEntity {
     }
 
     public static AttributeSupplier.Builder createAttributes() {
-        return LivingEntity.createLivingAttributes().add(Attributes.ATTACK_DAMAGE, 1.0D).add(Attributes.MOVEMENT_SPEED, 0.10000000149011612D).add(Attributes.ATTACK_SPEED).add(Attributes.LUCK);
+        return LivingEntity.createLivingAttributes().add(Attributes.ATTACK_DAMAGE, 1.0D).add(Attributes.MOVEMENT_SPEED, 0.10000000149011612D).add(Attributes.ATTACK_SPEED).add(Attributes.LUCK)
+            .add(Attributes.MAX_HEALTH, 500);
     }
 
     @Override
@@ -1240,6 +1213,15 @@ public abstract class Player extends LivingEntity {
         return this.onGround() || this.fallDistance < this.maxUpStep() && !this.level().noCollision(this, this.getBoundingBox().move(0.0D, (double) (this.fallDistance - this.maxUpStep()), 0.0D));
     }
 
+    // Papyrus - combat rework
+    public void attack(Entity target, DamageType type) {
+        if (!target.isAttackable()) {
+            return;
+        }
+        float damage = (float) (getAttributeValue(Attributes.ATTACK_DAMAGE) + getAttributeValue(type.getAdvAttr()));
+        target.hurt(damageSources().playerAttack(this), damage, type);
+    }
+
     public void attack(Entity target) {
         // Paper start - PlayerAttackEntityEvent
         boolean willAttack = target.isAttackable() && !target.skipAttackInteraction(this); // Vanilla logic
@@ -1459,7 +1441,7 @@ public abstract class Player extends LivingEntity {
 
     @Override
     protected void doAutoAttackOnTouch(LivingEntity target) {
-        this.attack(target);
+        this.attack(target, DamageType.PHYSICAL); // Papyrus
     }
 
     @io.papermc.paper.annotation.DoNotUse // Paper - add player shield disable event
diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
index da63b4050be25dcb91d04df8c2fcc643cbb0751d..715337c292e2b9f3c18ec606f62daefabd40210f 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
@@ -185,8 +185,8 @@ public class CraftPlayer extends CraftHumanEntity implements Player {
     private final Set<UUID> unlistedEntities = new HashSet<>(); // Paper
     private static final WeakHashMap<Plugin, WeakReference<Plugin>> pluginWeakReferences = new WeakHashMap<>();
     private int hash = 0;
-    private double health = 20;
-    private boolean scaledHealth = false;
+    private double health = 500; // Papyrus
+    private boolean scaledHealth = true; // Papyrus
     private double healthScale = 20;
     private CraftWorldBorder clientWorldBorder = null;
     private BorderChangeListener clientWorldBorderListener = this.createWorldBorderListener();
