From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: malfrador <malfrador@gmail.com>
Date: Fri, 5 Aug 2022 22:03:37 +0200
Subject: [PATCH] Spellbook


diff --git a/src/main/java/de/erethon/papyrus/CraftDamageType.java b/src/main/java/de/erethon/papyrus/CraftDamageType.java
new file mode 100644
index 0000000000000000000000000000000000000000..3004a15bd02ff5557beba43407e883a20e66bf0c
--- /dev/null
+++ b/src/main/java/de/erethon/papyrus/CraftDamageType.java
@@ -0,0 +1,44 @@
+package de.erethon.papyrus;
+
+import net.minecraft.world.entity.ai.attributes.Attribute;
+import net.minecraft.world.entity.ai.attributes.Attributes;
+
+import java.util.Arrays;
+
+public enum CraftDamageType {
+
+    AIR(Attributes.ADV_AIR, Attributes.RES_AIR, DamageType.AIR),
+    EARTH(Attributes.ADV_EARTH, Attributes.RES_EARTH, DamageType.EARTH),
+    FIRE(Attributes.ADV_FIRE, Attributes.RES_FIRE, DamageType.FIRE),
+    MAGIC(Attributes.ADV_MAGIC, Attributes.RES_MAGIC, DamageType.FIRE),
+    PHYSICAL(Attributes.ADV_PHYSICAL, Attributes.RES_PHYSICAL, DamageType.PHYSICAL),
+    WATER(Attributes.ADV_WATER, Attributes.RES_WATER, DamageType.WATER),
+    PURE(Attributes.ADV_PURE, Attributes.RES_PURE, DamageType.PURE);
+
+    private final Attribute advAttr;
+    private final Attribute resAttr;
+    private final DamageType apiDamageType;
+
+    CraftDamageType(Attribute advAttr, Attribute resAttr, DamageType apiDamageType) {
+        this.advAttr = advAttr;
+        this.resAttr = resAttr;
+        this.apiDamageType = apiDamageType;
+    }
+
+    public Attribute getAdvAttr() {
+        return this.advAttr;
+    }
+
+
+    public Attribute getResAttr() {
+        return this.resAttr;
+    }
+
+    public static CraftDamageType ofAPI(DamageType type) {
+        return Arrays.stream(values()).filter(crafttype -> crafttype.apiDamageType == type).findFirst().orElse(null);
+    }
+
+    public DamageType getApiDamageType() {
+        return apiDamageType;
+    }
+}
diff --git a/src/main/java/de/erethon/papyrus/DamageType.java b/src/main/java/de/erethon/papyrus/DamageType.java
deleted file mode 100644
index 0cbc67c143dbbe8d8bc1ebf2ea384f6851cf8658..0000000000000000000000000000000000000000
--- a/src/main/java/de/erethon/papyrus/DamageType.java
+++ /dev/null
@@ -1,32 +0,0 @@
-package de.erethon.papyrus;
-
-import net.minecraft.world.entity.ai.attributes.Attribute;
-import net.minecraft.world.entity.ai.attributes.Attributes;
-
-public enum DamageType {
-
-    AIR(Attributes.ADV_AIR, Attributes.RES_AIR),
-    EARTH(Attributes.ADV_EARTH, Attributes.RES_EARTH),
-    FIRE(Attributes.ADV_FIRE, Attributes.RES_FIRE),
-    MAGIC(Attributes.ADV_MAGIC, Attributes.RES_MAGIC),
-    PHYSICAL(Attributes.ADV_PHYSICAL, Attributes.RES_PHYSICAL),
-    WATER(Attributes.ADV_WATER, Attributes.RES_WATER),
-    PURE(Attributes.ADV_PURE, Attributes.RES_PURE);
-
-    private final Attribute advAttr;
-    private final Attribute resAttr;
-
-    DamageType(Attribute advAttr, Attribute resAttr) {
-        this.advAttr = advAttr;
-        this.resAttr = resAttr;
-    }
-
-    public Attribute getAdvAttr() {
-        return this.advAttr;
-    }
-
-
-    public Attribute getResAttr() {
-        return this.resAttr;
-    }
-}
diff --git a/src/main/java/net/minecraft/server/MinecraftServer.java b/src/main/java/net/minecraft/server/MinecraftServer.java
index 29354a30d11fa49fa2ae924de7e53d111391fe54..d6a1c4cb488a4d9a743ae55bbab92f8d78d2ad13 100644
--- a/src/main/java/net/minecraft/server/MinecraftServer.java
+++ b/src/main/java/net/minecraft/server/MinecraftServer.java
@@ -1468,6 +1468,9 @@ public abstract class MinecraftServer extends ReentrantBlockableEventLoop<TickTa
         // Paper start
         long endTime = System.nanoTime();
         long remaining = (TICK_TIME - (endTime - lastTick)) - catchupTime;
+        if (remaining > 0) {
+            server.getSpellbookAPI().getQueue().run(); // Papyrus - Spellbook
+        }
         new com.destroystokyo.paper.event.server.ServerTickEndEvent(this.tickCount, ((double)(endTime - lastTick) / 1000000D), remaining).callEvent();
         // Paper end
         this.profiler.push("tallying");
diff --git a/src/main/java/net/minecraft/server/level/ServerPlayer.java b/src/main/java/net/minecraft/server/level/ServerPlayer.java
index 1f1c986b34bec3ccee969674786454a2b7ac20e3..4a0165a7efe666d3df07e39347d7e46d578ba833 100644
--- a/src/main/java/net/minecraft/server/level/ServerPlayer.java
+++ b/src/main/java/net/minecraft/server/level/ServerPlayer.java
@@ -14,8 +14,7 @@ import java.util.Optional;
 import java.util.OptionalInt;
 import javax.annotation.Nullable;
 
-import de.erethon.papyrus.DamageType;
-import net.kyori.adventure.text.minimessage.MiniMessage;
+import de.erethon.papyrus.CraftDamageType;
 import net.minecraft.BlockUtil;
 import net.minecraft.ChatFormatting;
 import net.minecraft.CrashReport;
@@ -2062,7 +2061,7 @@ public class ServerPlayer extends Player {
         if (this.gameMode.getGameModeForPlayer() == GameType.SPECTATOR) {
             this.setCamera(target);
         } else {
-            super.attack(target, DamageType.PHYSICAL); // Papyrus
+            super.attack(target, CraftDamageType.PHYSICAL); // Papyrus
         }
 
     }
diff --git a/src/main/java/net/minecraft/world/entity/Entity.java b/src/main/java/net/minecraft/world/entity/Entity.java
index c7e2081fbd557417daa707da1c2e8604fc9faa1e..50a80b766c5dedb1b961642f2f0cba58f1e0b163 100644
--- a/src/main/java/net/minecraft/world/entity/Entity.java
+++ b/src/main/java/net/minecraft/world/entity/Entity.java
@@ -7,7 +7,7 @@ import com.google.common.collect.Lists;
 import com.google.common.collect.Sets;
 import com.google.common.collect.UnmodifiableIterator;
 import com.mojang.logging.LogUtils;
-import de.erethon.papyrus.DamageType;
+import de.erethon.papyrus.CraftDamageType;
 import it.unimi.dsi.fastutil.objects.Object2DoubleArrayMap;
 import it.unimi.dsi.fastutil.objects.Object2DoubleMap;
 import java.util.Arrays;
@@ -2007,7 +2007,7 @@ public abstract class Entity implements Nameable, EntityAccess, CommandSource {
     }
 
     // Papyrus - combat rework
-    public boolean hurt(DamageSource source, float amount, DamageType type) {
+    public boolean hurt(DamageSource source, float amount, CraftDamageType type) {
         if (this.isInvulnerableTo(source)) {
             return false;
         } else {
diff --git a/src/main/java/net/minecraft/world/entity/LivingEntity.java b/src/main/java/net/minecraft/world/entity/LivingEntity.java
index 1ac5edbb58e7ba831f6c67b719cdf67c5d317c91..a5c25e45b8cc534edb067d64a81ab6ef391dce9c 100644
--- a/src/main/java/net/minecraft/world/entity/LivingEntity.java
+++ b/src/main/java/net/minecraft/world/entity/LivingEntity.java
@@ -15,8 +15,9 @@ import java.util.*;
 import java.util.function.Predicate;
 import javax.annotation.Nullable;
 
-import de.erethon.papyrus.DamageType;
+import de.erethon.papyrus.CraftDamageType;
 import de.erethon.papyrus.PapyrusUtils;
+import de.erethon.spellbook.api.SpellCaster;
 import net.kyori.adventure.text.Component;
 import net.minecraft.BlockUtil;
 import net.minecraft.advancements.CriteriaTriggers;
@@ -62,7 +63,6 @@ import net.minecraft.world.InteractionHand;
 import net.minecraft.world.damagesource.CombatRules;
 import net.minecraft.world.damagesource.CombatTracker;
 import net.minecraft.world.damagesource.DamageSource;
-import net.minecraft.world.damagesource.EntityDamageSource;
 import net.minecraft.world.effect.MobEffect;
 import net.minecraft.world.effect.MobEffectInstance;
 import net.minecraft.world.effect.MobEffectUtil;
@@ -76,9 +76,7 @@ import net.minecraft.world.entity.ai.attributes.AttributeSupplier;
 import net.minecraft.world.entity.ai.attributes.Attributes;
 import net.minecraft.world.entity.ai.attributes.DefaultAttributes;
 import net.minecraft.world.entity.ai.targeting.TargetingConditions;
-import net.minecraft.world.entity.animal.Animal;
 import net.minecraft.world.entity.animal.FlyingAnimal;
-import net.minecraft.world.entity.animal.Wolf;
 import net.minecraft.world.entity.boss.wither.WitherBoss;
 import net.minecraft.world.entity.item.ItemEntity;
 import net.minecraft.world.entity.projectile.AbstractArrow;
@@ -103,7 +101,6 @@ import net.minecraft.world.level.block.state.BlockState;
 import net.minecraft.world.level.gameevent.GameEvent;
 import net.minecraft.world.level.material.Fluid;
 import net.minecraft.world.level.material.FluidState;
-import net.minecraft.world.level.material.Material;
 import net.minecraft.world.level.storage.loot.LootContext;
 import net.minecraft.world.level.storage.loot.LootTable;
 import net.minecraft.world.level.storage.loot.parameters.LootContextParamSets;
@@ -250,7 +247,7 @@ public abstract class LivingEntity extends Entity {
     public org.bukkit.craftbukkit.entity.CraftLivingEntity getBukkitLivingEntity() { return (org.bukkit.craftbukkit.entity.CraftLivingEntity) super.getBukkitEntity(); } // Paper
     public boolean silentDeath = false; // Paper - mark entity as dying silently for cancellable death event
 
-    public DamageType attackDamageType = DamageType.PHYSICAL; // Papyrus - combat rework
+    public CraftDamageType attackDamageType = CraftDamageType.PHYSICAL; // Papyrus - combat rework
 
     @Override
     public float getBukkitYaw() {
@@ -498,6 +495,7 @@ public abstract class LivingEntity extends Entity {
         }
 
         this.tickEffects();
+        getBukkitLivingEntity().tick(); // Papyrus - Spellbook tick
         this.animStepO = this.animStep;
         this.yBodyRotO = this.yBodyRot;
         this.yHeadRotO = this.yHeadRot;
@@ -1338,7 +1336,7 @@ public abstract class LivingEntity extends Entity {
     // Papyrus start - rewrite damage system
 
     @Override
-    public boolean hurt(DamageSource source, float amount, DamageType type) {
+    public boolean hurt(DamageSource source, float amount, CraftDamageType type) {
         if (isInvulnerable() || isRemoved() || dead || getHealth() <= 0F || isInvulnerableTo(source) || level.isClientSide) {
             return false;
         }
@@ -1397,12 +1395,15 @@ public abstract class LivingEntity extends Entity {
         invulnerableTime = (int) getAttribute(Attributes.COMBAT_HURTINVUL).getValue();
         playHurtSound(source);
         level.broadcastEntityEvent(this, (byte) 2); // hurt animation
+        if (source.getEntity() instanceof LivingEntity entity) {
+            finalDmg = getBukkitLivingEntity().onDamage(entity.getBukkitLivingEntity(), finalDmg,type.getApiDamageType());
+        }
         setHealth((float) (getHealth() - Math.max(finalDmg, 0)));
         return true;
     }
     @Override
     public boolean hurt(DamageSource source, float amount) {
-        hurt(source, amount, DamageType.PURE);
+        hurt(source, amount, CraftDamageType.PURE);
         return true;
     }
     // Papyrus end
diff --git a/src/main/java/net/minecraft/world/entity/Mob.java b/src/main/java/net/minecraft/world/entity/Mob.java
index a92795d2a3d0ecdd7c1e3046a2cddb88b62c88ec..347d4f49718f169ab1db3caa5352c6e3d753d03d 100644
--- a/src/main/java/net/minecraft/world/entity/Mob.java
+++ b/src/main/java/net/minecraft/world/entity/Mob.java
@@ -10,7 +10,8 @@ import java.util.Optional;
 import java.util.UUID;
 import javax.annotation.Nullable;
 
-import de.erethon.papyrus.DamageType;
+import de.erethon.papyrus.CraftDamageType;
+import de.erethon.spellbook.api.SpellCaster;
 import net.minecraft.core.BlockPos;
 import net.minecraft.core.NonNullList;
 import net.minecraft.core.Vec3i;
@@ -1592,7 +1593,7 @@ public abstract class Mob extends LivingEntity {
         return doHurtTarget(target, attackDamageType);
     }
 
-    public boolean doHurtTarget(Entity target, DamageType type) {
+    public boolean doHurtTarget(Entity target, CraftDamageType type) {
         float dmg = (float) this.getAttributeValue(Attributes.ATTACK_DAMAGE);
         float knockback = (float) this.getAttributeValue(Attributes.ATTACK_KNOCKBACK);
 
@@ -1614,6 +1615,9 @@ public abstract class Mob extends LivingEntity {
             // CraftBukkit end
         }
         dmg = (float) (dmg + getAttributeValue(type.getAdvAttr()));
+        if (target instanceof LivingEntity entity) {
+            dmg = (float) getBukkitLivingEntity().onAttack(entity.getBukkitLivingEntity(), dmg, type.getApiDamageType());
+        }
         boolean flag = target.hurt(DamageSource.mobAttack(this), dmg, type);
 
         if (flag) {
diff --git a/src/main/java/net/minecraft/world/entity/player/Player.java b/src/main/java/net/minecraft/world/entity/player/Player.java
index 05dafb5ae30ac77b9917e31d08b49e8dd462e262..d483087d50f31fdf1dcfb03e0e868120e0b6fa47 100644
--- a/src/main/java/net/minecraft/world/entity/player/Player.java
+++ b/src/main/java/net/minecraft/world/entity/player/Player.java
@@ -18,7 +18,8 @@ import java.util.OptionalInt;
 import java.util.function.Predicate;
 import javax.annotation.Nullable;
 
-import de.erethon.papyrus.DamageType;
+import de.erethon.papyrus.CraftDamageType;
+import de.erethon.spellbook.api.SpellCaster;
 import net.minecraft.SharedConstants;
 import net.minecraft.Util;
 import net.minecraft.advancements.CriteriaTriggers;
@@ -1254,11 +1255,14 @@ public abstract class Player extends LivingEntity {
     }
 
     // Papyrus - combat rework
-    public void attack(Entity target, DamageType type) {
+    public void attack(Entity target, CraftDamageType type) {
         if (!target.isAttackable()) {
             return;
         }
         float damage = (float) (getAttributeValue(Attributes.ATTACK_DAMAGE) + getAttributeValue(type.getAdvAttr()));
+        if (target instanceof LivingEntity entity) {
+            damage = (float) getBukkitEntity().onAttack(entity.getBukkitLivingEntity(), damage, type.getApiDamageType());
+        }
         target.hurt(DamageSource.playerAttack(this), damage, type);
     }
 
@@ -1472,7 +1476,7 @@ public abstract class Player extends LivingEntity {
 
     @Override
     protected void doAutoAttackOnTouch(LivingEntity target) {
-        this.attack(target, DamageType.PHYSICAL); // Papyrus
+        this.attack(target, CraftDamageType.PHYSICAL); // Papyrus
     }
 
     public void disableShield(boolean sprinting) {
diff --git a/src/main/java/org/bukkit/craftbukkit/CraftServer.java b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
index 81fd066ddcdc3eadb1a531f00bea7b6ce1ed4bbe..378818feb7b24c8d48adcfa0996784a768b31f05 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftServer.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
@@ -13,6 +13,7 @@ import com.mojang.brigadier.exceptions.CommandSyntaxException;
 import com.mojang.brigadier.tree.CommandNode;
 import com.mojang.brigadier.tree.LiteralCommandNode;
 import com.mojang.serialization.Lifecycle;
+import de.erethon.spellbook.api.SpellbookAPI;
 import io.netty.buffer.ByteBuf;
 import io.netty.buffer.ByteBufOutputStream;
 import io.netty.buffer.Unpooled;
@@ -285,6 +286,9 @@ public final class CraftServer implements Server {
     private final io.papermc.paper.logging.SysoutCatcher sysoutCatcher = new io.papermc.paper.logging.SysoutCatcher(); // Paper
     private final CraftPotionBrewer potionBrewer = new CraftPotionBrewer(); // Paper
 
+    // Papyrus start
+    private final SpellbookAPI spellbookAPI;
+
     static {
         ConfigurationSerialization.registerClass(CraftOfflinePlayer.class);
         ConfigurationSerialization.registerClass(CraftPlayerProfile.class);
@@ -370,7 +374,14 @@ public final class CraftServer implements Server {
             MapPalette.setMapColorCache(new CraftMapColorCache(this.logger));
         }
         datapackManager = new io.papermc.paper.datapack.PaperDatapackManager(console.getPackRepository()); // Paper
+        spellbookAPI = new SpellbookAPI(this); // Papyrus start - Spellbook
+    }
+
+    @Override
+    public SpellbookAPI getSpellbookAPI() {
+        return spellbookAPI;
     }
+    // Papyrus end
 
     public boolean getCommandBlockOverride(String command) {
         return this.overrideAllCommandBlockCommands || this.commandsConfiguration.getStringList("command-block-overrides").contains(command);
@@ -432,6 +443,7 @@ public final class CraftServer implements Server {
         } else {
             pluginFolder.mkdir();
         }
+        spellbookAPI.getLibrary().loadSpells(SpellbookAPI.SPELLS); // Papyrus - Load spells after plugins
     }
 
     // Paper start
diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftEnderDragonPart.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftEnderDragonPart.java
index a407e802164603c74ec1f4ad3c98bc50398954bc..c7fdecbcd6eb7571a09883637bee4150dcb4eb72 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftEnderDragonPart.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftEnderDragonPart.java
@@ -1,5 +1,6 @@
 package org.bukkit.craftbukkit.entity;
 
+import de.erethon.papyrus.DamageType;
 import org.bukkit.craftbukkit.CraftServer;
 import org.bukkit.entity.EnderDragon;
 import org.bukkit.entity.EnderDragonPart;
@@ -30,11 +31,21 @@ public class CraftEnderDragonPart extends CraftComplexPart implements EnderDrago
         this.getParent().damage(amount);
     }
 
+    @Override
+    public void damage(double amount, DamageType type) {
+        this.getParent().damage(amount,type);
+    }
+
     @Override
     public void damage(double amount, Entity source) {
         this.getParent().damage(amount, source);
     }
 
+    @Override
+    public void damage(double amount, Entity source, DamageType type) {
+        this.getParent().damage(amount, source, type);
+    }
+
     @Override
     public double getHealth() {
         return this.getParent().getHealth();
diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftLivingEntity.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftLivingEntity.java
index c022751e3b45469cc0ad6732e2d6ff08918bafa4..2dcc2add6171a775f7846151d990bb193a48fa6f 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftLivingEntity.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftLivingEntity.java
@@ -3,12 +3,15 @@ package org.bukkit.craftbukkit.entity;
 import com.destroystokyo.paper.entity.TargetEntityInfo;
 import com.google.common.base.Preconditions;
 import com.google.common.collect.Sets;
-import java.util.ArrayList;
-import java.util.Collection;
-import java.util.Iterator;
-import java.util.List;
-import java.util.Set;
-import java.util.UUID;
+
+import java.util.*;
+
+import de.erethon.papyrus.CraftDamageType;
+import de.erethon.papyrus.DamageType;
+import de.erethon.spellbook.api.SpellData;
+import de.erethon.spellbook.api.SpellEffect;
+import de.erethon.spellbook.api.SpellbookSpell;
+import net.kyori.adventure.text.minimessage.MiniMessage;
 import net.minecraft.server.level.ServerPlayer;
 import net.minecraft.world.InteractionHand;
 import net.minecraft.world.damagesource.DamageSource;
@@ -47,6 +50,7 @@ import org.bukkit.craftbukkit.inventory.CraftEntityEquipment;
 import org.bukkit.craftbukkit.inventory.CraftItemStack;
 import org.bukkit.craftbukkit.potion.CraftPotionUtil;
 import org.bukkit.entity.AbstractArrow;
+import org.bukkit.entity.Damageable;
 import org.bukkit.entity.DragonFireball;
 import org.bukkit.entity.Egg;
 import org.bukkit.entity.EnderPearl;
@@ -87,6 +91,21 @@ import org.bukkit.util.Vector;
 public class CraftLivingEntity extends CraftEntity implements LivingEntity {
     private CraftEntityEquipment equipment;
 
+
+    // Papyrus start
+    private Set<SpellData> unlockedSpellData = new HashSet<>();
+    private Set<SpellData> passiveSpells = new HashSet<>();
+    private Set<SpellbookSpell> activeSpells = new HashSet<>();
+    private final SpellData[] assignedSlots = new SpellData[8];
+
+    Map<SpellData, Long> usedSpells = new LinkedHashMap<>();
+    Set<SpellEffect> effects = new HashSet<>();
+    Set<SpellbookSpell> runningPassiveSpells = new HashSet<>();
+    private int maxEnergy = 0;
+    private int energy = 0;
+    private MiniMessage miniMessage = MiniMessage.miniMessage();
+    // Papyrus end
+
     public CraftLivingEntity(final CraftServer server, final net.minecraft.world.entity.LivingEntity entity) {
         super(server, entity);
 
@@ -345,11 +364,21 @@ public class CraftLivingEntity extends CraftEntity implements LivingEntity {
     // Paper End - Bee Stinger API
     @Override
     public void damage(double amount) {
-        this.damage(amount, null);
+        this.damage(amount, (Entity) null);
     }
 
     @Override
     public void damage(double amount, org.bukkit.entity.Entity source) {
+        damage(amount, source, DamageType.PURE);
+    }
+
+    @Override
+    public void damage(double amount, DamageType type) {
+        damage(amount, null, type);
+    }
+
+    @Override
+    public void damage(double amount, org.bukkit.entity.Entity source, DamageType type) {
         Preconditions.checkState(!this.getHandle().generation, "Cannot damage entity during world generation");
 
         DamageSource reason = DamageSource.GENERIC;
@@ -359,8 +388,12 @@ public class CraftLivingEntity extends CraftEntity implements LivingEntity {
         } else if (source instanceof LivingEntity) {
             reason = DamageSource.mobAttack(((CraftLivingEntity) source).getHandle());
         }
+        if (entity instanceof net.minecraft.world.entity.LivingEntity) {
+            entity.hurt(reason, (float) amount, CraftDamageType.ofAPI(type));
+        } else {
+            entity.hurt(reason, (float) amount);
+        }
 
-        entity.hurt(reason, (float) amount);
     }
 
     @Override
@@ -904,6 +937,20 @@ public class CraftLivingEntity extends CraftEntity implements LivingEntity {
         getHandle().hurtDir = hurtDirection;
     }
 
+    @Override
+    public void sendActionbar(String msg) {
+        if (this instanceof Player player) {
+            player.sendActionBar(miniMessage.deserialize(msg));
+        }
+    }
+
+    @Override
+    public void sendMessage(String msg) {
+        if (this instanceof Player player) {
+            player.sendMessage(miniMessage.deserialize(msg));
+        }
+    }
+
     public static MobType fromBukkitEntityCategory(EntityCategory entityCategory) {
         switch (entityCategory) {
             case NONE:
@@ -920,4 +967,69 @@ public class CraftLivingEntity extends CraftEntity implements LivingEntity {
         throw new IllegalArgumentException(entityCategory + " is an unrecognized entity category");
     }
     // Paper end
+
+    // Papyrus start - Spellbook
+
+    @Override
+    public Map<SpellData, Long> getUsedSpells() {
+        return usedSpells;
+    }
+
+    @Override
+    public Set<SpellEffect> getEffects() {
+        return effects;
+    }
+
+    @Override
+    public Set<SpellbookSpell> getPassiveSpells() {
+        return runningPassiveSpells;
+    }
+
+    @Override
+    public Set<SpellbookSpell> getActiveSpells() {
+        return activeSpells;
+    }
+
+    @Override
+    public Set<SpellData> getUnlockedSpells() {
+        return unlockedSpellData;
+    }
+
+    @Override
+    public Set<SpellData> getUnlockedPassives() {
+        return passiveSpells;
+    }
+
+    @Override
+    public int getEnergy() {
+        return energy;
+    }
+
+    @Override
+    public int setEnergy(int e) {
+        return energy = Math.min(e, maxEnergy);
+    }
+
+    @Override
+    public int getMaxEnergy() {
+        return maxEnergy;
+    }
+
+    @Override
+    public int setMaxEnergy(int e) {
+        return maxEnergy = e;
+    }
+
+    @Override
+    public int addEnergy(int e) {
+        energy = Math.min(energy + e, maxEnergy);
+        return energy;
+    }
+
+    @Override
+    public int removeEnergy(int e) {
+        energy = Math.max(energy - e, 0);
+        return energy;
+    }
+    // Papyrus end
 }
diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
index a1b101093e775ac8444b98aff9df977e9d4521ad..914f30ffd9c114ab3e5f2385357e1655431d7547 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
@@ -295,18 +295,6 @@ public class CraftPlayer extends CraftHumanEntity implements Player {
         getHandle().connection.send(packet);
     }
 
-    @Override
-    public void sendActionBar(String message) {
-        if (getHandle().connection == null || message == null || message.isEmpty()) return;
-        getHandle().connection.send(new net.minecraft.network.protocol.game.ClientboundSetActionBarTextPacket(CraftChatMessage.fromStringOrNull(message)));
-    }
-
-    @Override
-    public void sendActionBar(char alternateChar, String message) {
-        if (message == null || message.isEmpty()) return;
-        sendActionBar(org.bukkit.ChatColor.translateAlternateColorCodes(alternateChar, message));
-    }
-
     @Override
     public void setPlayerListHeaderFooter(BaseComponent[] header, BaseComponent[] footer) {
          if (header != null) {
