From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: malfrador <malfrador@gmail.com>
Date: Fri, 5 Aug 2022 22:03:37 +0200
Subject: [PATCH] Spellbook


diff --git a/src/main/java/de/erethon/papyrus/CraftDamageType.java b/src/main/java/de/erethon/papyrus/CraftDamageType.java
new file mode 100644
index 0000000000000000000000000000000000000000..20184e08449c0f6fc37b68093973ac592c1feb93
--- /dev/null
+++ b/src/main/java/de/erethon/papyrus/CraftDamageType.java
@@ -0,0 +1,44 @@
+package de.erethon.papyrus;
+
+import net.minecraft.world.entity.ai.attributes.Attribute;
+import net.minecraft.world.entity.ai.attributes.Attributes;
+
+import java.util.Arrays;
+
+public enum CraftDamageType {
+
+    AIR(Attributes.ADV_AIR, Attributes.RES_AIR, DamageType.AIR),
+    EARTH(Attributes.ADV_EARTH, Attributes.RES_EARTH, DamageType.EARTH),
+    FIRE(Attributes.ADV_FIRE, Attributes.RES_FIRE, DamageType.FIRE),
+    MAGIC(Attributes.ADV_MAGIC, Attributes.RES_MAGIC, DamageType.MAGIC),
+    PHYSICAL(Attributes.ADV_PHYSICAL, Attributes.RES_PHYSICAL, DamageType.PHYSICAL),
+    WATER(Attributes.ADV_WATER, Attributes.RES_WATER, DamageType.WATER),
+    PURE(Attributes.ADV_PURE, Attributes.RES_PURE, DamageType.PURE);
+
+    private final Attribute advAttr;
+    private final Attribute resAttr;
+    private final DamageType apiDamageType;
+
+    CraftDamageType(Attribute advAttr, Attribute resAttr, DamageType apiDamageType) {
+        this.advAttr = advAttr;
+        this.resAttr = resAttr;
+        this.apiDamageType = apiDamageType;
+    }
+
+    public Attribute getAdvAttr() {
+        return this.advAttr;
+    }
+
+
+    public Attribute getResAttr() {
+        return this.resAttr;
+    }
+
+    public static CraftDamageType ofAPI(DamageType type) {
+        return Arrays.stream(values()).filter(crafttype -> crafttype.apiDamageType == type).findFirst().orElse(null);
+    }
+
+    public DamageType getApiDamageType() {
+        return apiDamageType;
+    }
+}
diff --git a/src/main/java/de/erethon/papyrus/DamageType.java b/src/main/java/de/erethon/papyrus/DamageType.java
deleted file mode 100644
index 0cbc67c143dbbe8d8bc1ebf2ea384f6851cf8658..0000000000000000000000000000000000000000
--- a/src/main/java/de/erethon/papyrus/DamageType.java
+++ /dev/null
@@ -1,32 +0,0 @@
-package de.erethon.papyrus;
-
-import net.minecraft.world.entity.ai.attributes.Attribute;
-import net.minecraft.world.entity.ai.attributes.Attributes;
-
-public enum DamageType {
-
-    AIR(Attributes.ADV_AIR, Attributes.RES_AIR),
-    EARTH(Attributes.ADV_EARTH, Attributes.RES_EARTH),
-    FIRE(Attributes.ADV_FIRE, Attributes.RES_FIRE),
-    MAGIC(Attributes.ADV_MAGIC, Attributes.RES_MAGIC),
-    PHYSICAL(Attributes.ADV_PHYSICAL, Attributes.RES_PHYSICAL),
-    WATER(Attributes.ADV_WATER, Attributes.RES_WATER),
-    PURE(Attributes.ADV_PURE, Attributes.RES_PURE);
-
-    private final Attribute advAttr;
-    private final Attribute resAttr;
-
-    DamageType(Attribute advAttr, Attribute resAttr) {
-        this.advAttr = advAttr;
-        this.resAttr = resAttr;
-    }
-
-    public Attribute getAdvAttr() {
-        return this.advAttr;
-    }
-
-
-    public Attribute getResAttr() {
-        return this.resAttr;
-    }
-}
diff --git a/src/main/java/de/erethon/papyrus/EntityAttachment.java b/src/main/java/de/erethon/papyrus/EntityAttachment.java
new file mode 100644
index 0000000000000000000000000000000000000000..6650bb2ffb961035a7e598990097621c80b2c7ae
--- /dev/null
+++ b/src/main/java/de/erethon/papyrus/EntityAttachment.java
@@ -0,0 +1,22 @@
+package de.erethon.papyrus;
+
+import net.minecraft.world.entity.Entity;
+
+public class EntityAttachment {
+
+    public String id;
+    public double dx;
+    public double dy;
+    public double dz;
+    public float dyaw;
+    public float dpitch;
+    public Entity entity;
+
+    public EntityAttachment(double dx, double dy, double dz, Entity entity, String id) {
+        this.dx = dx;
+        this.dy = dy;
+        this.dz = dz;
+        this.entity = entity;
+        this.id = id;
+    }
+}
diff --git a/src/main/java/net/minecraft/server/MinecraftServer.java b/src/main/java/net/minecraft/server/MinecraftServer.java
index 0489e97a49aa39cce8c101d52dac14c6e5afa6d9..c0d4906be32ce8a99b9eaa515f3bbfbb3b8812f8 100644
--- a/src/main/java/net/minecraft/server/MinecraftServer.java
+++ b/src/main/java/net/minecraft/server/MinecraftServer.java
@@ -1450,6 +1450,9 @@ public abstract class MinecraftServer extends ReentrantBlockableEventLoop<TickTa
         // Paper start
         long endTime = System.nanoTime();
         long remaining = (TICK_TIME - (endTime - lastTick)) - catchupTime;
+        if (remaining > 0) {
+            server.getSpellbookAPI().getQueue().run(); // Papyrus - Spellbook
+        }
         new com.destroystokyo.paper.event.server.ServerTickEndEvent(this.tickCount, ((double)(endTime - lastTick) / 1000000D), remaining).callEvent();
         // Paper end
         this.profiler.push("tallying");
diff --git a/src/main/java/net/minecraft/server/level/ServerPlayer.java b/src/main/java/net/minecraft/server/level/ServerPlayer.java
index b1f0187b917c724cfef8f53e9f503b4f4afc4f27..bb5473440d935e81ef532fbab9e37de562c90f0c 100644
--- a/src/main/java/net/minecraft/server/level/ServerPlayer.java
+++ b/src/main/java/net/minecraft/server/level/ServerPlayer.java
@@ -15,8 +15,7 @@ import java.util.Optional;
 import java.util.OptionalInt;
 import javax.annotation.Nullable;
 
-import de.erethon.papyrus.DamageType;
-import net.kyori.adventure.text.minimessage.MiniMessage;
+import de.erethon.papyrus.CraftDamageType;
 import net.minecraft.BlockUtil;
 import net.minecraft.ChatFormatting;
 import net.minecraft.CrashReport;
@@ -2118,7 +2117,7 @@ public class ServerPlayer extends Player {
         if (this.gameMode.getGameModeForPlayer() == GameType.SPECTATOR) {
             this.setCamera(target);
         } else {
-            super.attack(target, DamageType.PHYSICAL); // Papyrus
+            super.attack(target, CraftDamageType.PHYSICAL); // Papyrus
         }
 
     }
diff --git a/src/main/java/net/minecraft/world/entity/Entity.java b/src/main/java/net/minecraft/world/entity/Entity.java
index 43492b861556d0e1f70d92f72a505d3d880a81b5..5e52a12bf4f43e5adbedd3a57eec4ca0f4cd0321 100644
--- a/src/main/java/net/minecraft/world/entity/Entity.java
+++ b/src/main/java/net/minecraft/world/entity/Entity.java
@@ -7,7 +7,8 @@ import com.google.common.collect.Lists;
 import com.google.common.collect.Sets;
 import com.google.common.collect.UnmodifiableIterator;
 import com.mojang.logging.LogUtils;
-import de.erethon.papyrus.DamageType;
+import de.erethon.papyrus.CraftDamageType;
+import de.erethon.papyrus.EntityAttachment;
 import it.unimi.dsi.fastutil.objects.Object2DoubleArrayMap;
 import it.unimi.dsi.fastutil.objects.Object2DoubleMap;
 import java.util.Arrays;
@@ -400,6 +401,8 @@ public abstract class Entity implements Nameable, EntityAccess, CommandSource {
     public boolean freezeLocked = false; // Paper - Freeze Tick Lock API
     public boolean collidingWithWorldBorder; // Paper
 
+    public Set<EntityAttachment> attachedEntities = new HashSet<>(); // Papyrus - Entity attachments
+
     public void setOrigin(@javax.annotation.Nonnull Location location) {
         this.origin = location.toVector();
         this.originWorld = location.getWorld().getUID();
@@ -704,6 +707,9 @@ public abstract class Entity implements Nameable, EntityAccess, CommandSource {
 
     public void remove(Entity.RemovalReason reason) {
         this.setRemoved(reason);
+        for (EntityAttachment attachment : attachedEntities) {
+            attachment.entity.remove(reason);
+        }
     }
 
     public void onClientRemoval() {}
@@ -768,6 +774,11 @@ public abstract class Entity implements Nameable, EntityAccess, CommandSource {
 
         this.setYRot(yaw % 360.0F);
         this.setXRot(pitch % 360.0F);
+        // Papyrus - Attachments
+        for (EntityAttachment attached : attachedEntities) {
+            attached.entity.setYRot((yaw + attached.dyaw) % 360.0F);
+            attached.entity.setXRot((pitch + attached.dpitch) % 360.0F);
+        }
     }
 
     public final void setPos(Vec3 pos) {
@@ -805,6 +816,9 @@ public abstract class Entity implements Nameable, EntityAccess, CommandSource {
 
     public void tick() {
         this.baseTick();
+        for (EntityAttachment attached : attachedEntities) {
+            attached.entity.tick();
+        }
     }
 
     // CraftBukkit start
@@ -885,6 +899,9 @@ public abstract class Entity implements Nameable, EntityAccess, CommandSource {
 
         this.firstTick = false;
         this.level.getProfiler().pop();
+        for (EntityAttachment attached : attachedEntities) {
+            attached.entity.baseTick();
+        }
     }
 
     public void setSharedFlagOnFire(boolean onFire) {
@@ -1043,6 +1060,11 @@ public abstract class Entity implements Nameable, EntityAccess, CommandSource {
             this.moveStartZ = this.getZ();
             this.moveVector = movement;
         }
+        // Papyrus - attachments
+        for (EntityAttachment attached : attachedEntities) {
+            Vec3 rel = new Vec3(movement.x + attached.dx, movement.y + attached.dy, movement.z + attached.dz);
+            attached.entity.move(movementType, rel);
+        }
         try {
         // Paper end - detailed watchdog information
         if (this.noPhysics) {
@@ -1869,6 +1891,10 @@ public abstract class Entity implements Nameable, EntityAccess, CommandSource {
         Vec3 vec3d1 = Entity.getInputVector(movementInput, speed, this.getYRot());
 
         this.setDeltaMovement(this.getDeltaMovement().add(vec3d1));
+        // Papyrus - Attachments
+        for (EntityAttachment attached : attachedEntities) {
+            attached.entity.setDeltaMovement(this.getDeltaMovement().add(vec3d1));
+        }
     }
 
     private static Vec3 getInputVector(Vec3 movementInput, float speed, float yaw) {
@@ -1898,6 +1924,10 @@ public abstract class Entity implements Nameable, EntityAccess, CommandSource {
         this.yRotO = this.getYRot();
         this.xRotO = this.getXRot();
         this.setYHeadRot(yaw); // Paper - Update head rotation
+        // Papyrus - Attachments
+        for (EntityAttachment attached : attachedEntities) {
+            attached.entity.absMoveTo(x + attached.dx, y + attached.dy, z + attached.dz, yaw + attached.dyaw, pitch + attached.dpitch);
+        }
     }
 
     public void absMoveTo(double x, double y, double z) {
@@ -1909,6 +1939,10 @@ public abstract class Entity implements Nameable, EntityAccess, CommandSource {
         this.zo = d4;
         this.setPos(d3, y, d4);
         if (this.valid) this.level.getChunk((int) Math.floor(this.getX()) >> 4, (int) Math.floor(this.getZ()) >> 4); // CraftBukkit
+        // Papyrus - Attachments
+        for (EntityAttachment attached : attachedEntities) {
+            attached.entity.absMoveTo(x + attached.dx, y + attached.dy, z + attached.dz);
+        }
     }
 
     public void moveTo(Vec3 pos) {
@@ -1937,6 +1971,10 @@ public abstract class Entity implements Nameable, EntityAccess, CommandSource {
         this.setOldPosAndRot();
         this.reapplyPosition();
         this.setYHeadRot(yaw); // Paper - Update head rotation
+        // Papyrus - Attachments
+        for (EntityAttachment attached : attachedEntities) {
+            attached.entity.moveTo(x + attached.dx, y + attached.dy, z + attached.dz, yaw + attached.dyaw, pitch + attached.dpitch);
+        }
     }
 
     public final void setOldPosAndRot() {
@@ -2047,7 +2085,7 @@ public abstract class Entity implements Nameable, EntityAccess, CommandSource {
     }
 
     // Papyrus - combat rework
-    public boolean hurt(DamageSource source, float amount, DamageType type) {
+    public boolean hurt(DamageSource source, float amount, CraftDamageType type) {
         if (this.isInvulnerableTo(source)) {
             return false;
         } else {
@@ -4648,4 +4686,71 @@ public abstract class Entity implements Nameable, EntityAccess, CommandSource {
         return ((net.minecraft.server.level.ServerChunkCache) level.getChunkSource()).isPositionTicking(this);
     }
     // Paper end
+
+    // Papyrus start
+    public void addAttachment(Entity entity, String id) {
+        entity.noPhysics = true;
+        entity.setInvulnerable(true);
+        entity.setSilent(true);
+        if (entity instanceof net.minecraft.world.entity.LivingEntity living) {
+            living.collides = false;
+            if (living instanceof Mob mob) {
+                mob.setNoAi(true);
+            }
+        }
+        EntityAttachment attachment = new EntityAttachment(0, 0, 0, entity, id);
+        attachedEntities.add(attachment);
+    }
+
+    public void addAttachment(Entity entity, String id, double dx, double dy, double dz) {
+        entity.noPhysics = true;
+        entity.setInvulnerable(true);
+        entity.setSilent(true);
+        if (entity instanceof net.minecraft.world.entity.LivingEntity living) {
+            living.collides = false;
+            if (living instanceof Mob mob) {
+                mob.setNoAi(true);
+            }
+        }
+        EntityAttachment attachment = new EntityAttachment(dx, dy, dz, entity, id);
+        attachedEntities.add(attachment);
+    }
+
+    public void removeAttachment(Entity entity) {
+        Iterator<EntityAttachment> itr = attachedEntities.iterator();
+        while (itr.hasNext()) {
+            EntityAttachment attachment = itr.next();
+            if (attachment.entity == entity) {
+                itr.remove();
+            }
+        }
+    }
+
+    public void removeAttachment(String id) {
+        Iterator<EntityAttachment> itr = attachedEntities.iterator();
+        while (itr.hasNext()) {
+            EntityAttachment attachment = itr.next();
+            if (attachment.id.equals(id)) {
+                itr.remove();
+            }
+        }
+    }
+
+    public EntityAttachment getAttachment(Entity entity) {
+        for (EntityAttachment attached : attachedEntities) {
+            if (attached.entity == entity) {
+                return attached;
+            }
+        }
+        return null;
+    }
+
+    public EntityAttachment getAttachment(String id) {
+        for (EntityAttachment attached : attachedEntities) {
+            if (attached.id.equals(id)) {
+                return attached;
+            }
+        }
+        return null;
+    }
 }
diff --git a/src/main/java/net/minecraft/world/entity/LivingEntity.java b/src/main/java/net/minecraft/world/entity/LivingEntity.java
index 413652284e229a51e3eefe47f239e8fa9a09ccb2..e534b0564a61be7d3e0e5146e07aa5f680b6cc8e 100644
--- a/src/main/java/net/minecraft/world/entity/LivingEntity.java
+++ b/src/main/java/net/minecraft/world/entity/LivingEntity.java
@@ -19,6 +19,8 @@ import java.util.Optional;
 import java.util.UUID;
 import java.util.function.Predicate;
 import javax.annotation.Nullable;
+
+import de.erethon.papyrus.CraftDamageType;
 import net.minecraft.BlockUtil;
 import net.minecraft.advancements.CriteriaTriggers;
 import net.minecraft.commands.arguments.EntityAnchorArgument;
@@ -278,6 +280,8 @@ public abstract class LivingEntity extends Entity {
         ++this.noActionTime; // Above all the floats
     }
     // Spigot end
+    // Papyrus start
+    public CraftDamageType attackDamageType = CraftDamageType.PHYSICAL;
 
     protected LivingEntity(EntityType<? extends LivingEntity> type, Level world) {
         super(type, world);
@@ -338,7 +342,27 @@ public abstract class LivingEntity extends Entity {
     }
 
     public static AttributeSupplier.Builder createLivingAttributes() {
-        return AttributeSupplier.builder().add(Attributes.MAX_HEALTH).add(Attributes.KNOCKBACK_RESISTANCE).add(Attributes.MOVEMENT_SPEED).add(Attributes.ARMOR).add(Attributes.ARMOR_TOUGHNESS);
+        return AttributeSupplier.builder().add(Attributes.MAX_HEALTH).add(Attributes.KNOCKBACK_RESISTANCE).add(Attributes.MOVEMENT_SPEED).add(Attributes.ARMOR).add(Attributes.ARMOR_TOUGHNESS)
+            .add(Attributes.ADV_AIR)
+            .add(Attributes.ADV_EARTH)
+            .add(Attributes.ADV_FIRE)
+            .add(Attributes.ADV_WATER)
+            .add(Attributes.ADV_MAGIC)
+            .add(Attributes.ADV_PURE)
+            .add(Attributes.ADV_PHYSICAL)
+            .add(Attributes.RES_AIR)
+            .add(Attributes.RES_EARTH)
+            .add(Attributes.RES_FIRE)
+            .add(Attributes.RES_WATER)
+            .add(Attributes.RES_MAGIC)
+            .add(Attributes.RES_PURE)
+            .add(Attributes.RES_PHYSICAL)
+            .add(Attributes.STAT_CDR)
+            .add(Attributes.STAT_CRITDMG)
+            .add(Attributes.STAT_CRITCHANCE)
+            .add(Attributes.STAT_TENACITY)
+            .add(Attributes.STAT_ENERGYREGEN)
+            .add(Attributes.STAT_HEALTHREGEN);
     }
 
     @Override
@@ -489,6 +513,7 @@ public abstract class LivingEntity extends Entity {
         }
 
         this.tickEffects();
+        getBukkitLivingEntity().tick(); // Papyrus - Spellbook tick
         this.animStepO = this.animStep;
         this.yBodyRotO = this.yBodyRot;
         this.yHeadRotO = this.yHeadRot;
diff --git a/src/main/java/net/minecraft/world/entity/LivingEntity.javaDVl8e7 b/src/main/java/net/minecraft/world/entity/LivingEntity.javaDVl8e7
new file mode 100644
index 0000000000000000000000000000000000000000..e69de29bb2d1d6434b8b29ae775ad8c2e48c5391
diff --git a/src/main/java/net/minecraft/world/entity/Mob.java b/src/main/java/net/minecraft/world/entity/Mob.java
index ce81bac7bd761637143a2153e24a2d2d58ce8598..f1fd32bd439813ddcfba9789d1e12f064309eaa0 100644
--- a/src/main/java/net/minecraft/world/entity/Mob.java
+++ b/src/main/java/net/minecraft/world/entity/Mob.java
@@ -10,7 +10,8 @@ import java.util.UUID;
 import java.util.function.Predicate;
 import javax.annotation.Nullable;
 
-import de.erethon.papyrus.DamageType;
+import de.erethon.papyrus.CraftDamageType;
+import de.erethon.spellbook.api.SpellCaster;
 import net.minecraft.core.BlockPos;
 import net.minecraft.core.NonNullList;
 import net.minecraft.core.Vec3i;
@@ -171,7 +172,7 @@ public abstract class Mob extends LivingEntity {
     protected void registerGoals() {}
 
     public static AttributeSupplier.Builder createMobAttributes() {
-        return LivingEntity.createLivingAttributes().add(Attributes.FOLLOW_RANGE, 16.0D).add(Attributes.ATTACK_KNOCKBACK);
+        return LivingEntity.createLivingAttributes().add(Attributes.FOLLOW_RANGE, 16.0D).add(Attributes.ATTACK_KNOCKBACK).add(Attributes.ATTACK_SPEED, 80.0);
     }
 
     protected PathNavigation createNavigation(Level world) {
@@ -1624,7 +1625,7 @@ public abstract class Mob extends LivingEntity {
         return doHurtTarget(target, attackDamageType);
     }
 
-    public boolean doHurtTarget(Entity target, DamageType type) {
+    public boolean doHurtTarget(Entity target, CraftDamageType type) {
         float dmg = (float) this.getAttributeValue(Attributes.ATTACK_DAMAGE);
         float knockback = (float) this.getAttributeValue(Attributes.ATTACK_KNOCKBACK);
 
@@ -1646,6 +1647,13 @@ public abstract class Mob extends LivingEntity {
             // CraftBukkit end
         }
         dmg = (float) (dmg + getAttributeValue(type.getAdvAttr()));
+        if (target instanceof LivingEntity entity) {
+            dmg = (float) getBukkitLivingEntity().onAttack(entity.getBukkitLivingEntity(), dmg, type.getApiDamageType());
+        }
+        int crit = random.nextIntBetweenInclusive(0, 100);
+        if (crit < getAttributeValue(Attributes.STAT_CRITCHANCE)) {
+            dmg = (float) (dmg + getAttributeValue(Attributes.STAT_CRITDMG));
+        }
         boolean flag = target.hurt(DamageSource.mobAttack(this), dmg, type);
 
         if (flag) {
diff --git a/src/main/java/net/minecraft/world/entity/ai/attributes/Attributes.java b/src/main/java/net/minecraft/world/entity/ai/attributes/Attributes.java
index 57be3d8f01e53a1edf8f92cd6cf35bede099602f..ddb9388cb94ee9f880f7dc979c460b180b643bcb 100644
--- a/src/main/java/net/minecraft/world/entity/ai/attributes/Attributes.java
+++ b/src/main/java/net/minecraft/world/entity/ai/attributes/Attributes.java
@@ -39,9 +39,14 @@ public class Attributes {
     public static final Attribute ADV_WATER = Attributes.register("advantage.water", new RangedAttribute("attribute.name.advantage.water", 0, 0, 4096));
     // Stats
     public static final Attribute STAT_CRITCHANCE = Attributes.register("stat.crit.chance", new RangedAttribute("attribute.name.stat.crit.chance", 0, 0, 4096));
-    public static final Attribute STAT_CRITDMG = Attributes.register("stat.crit.dmg", new RangedAttribute("attribute.name.stat.crit.dmg", 0, 0, 4096));
-    public static final Attribute STAT_ATTACKSPEED= Attributes.register("stat.attackspeed", new RangedAttribute("attribute.name.stat.attackspeed", 0, 0, 4096));
-    public static final Attribute STAT_CDR = Attributes.register("stat.cooldownreduction", new RangedAttribute("attribute.name.stat.cooldownreduction", 0, 0, 4096));
+    public static final Attribute STAT_CRITDMG = Attributes.register("stat.crit.dmg", new RangedAttribute("attribute.name.stat.crit.dmg", 0, -4096, 4096));
+    public static final Attribute STAT_CDR = Attributes.register("stat.cooldownreduction", new RangedAttribute("attribute.name.stat.cooldownreduction", 0, -4096, 4096));
+    public static final Attribute STAT_HEALTHREGEN = Attributes.register("stat.healthregen", new RangedAttribute("attribute.name.stat.healthregen", 0, 0, 4096));
+    public static final Attribute STAT_ENERGYREGEN = Attributes.register("stat.energyregen", new RangedAttribute("attribute.name.stat.energyregen", 0, 0, 4096));
+    public static final Attribute STAT_TENACITY = Attributes.register("stat.tenacity", new RangedAttribute("attribute.name.stat.tenacity", 0, 0, 4096));
+
+
+
     // Combat
     public static final Attribute COMBAT_HURTINVUL= Attributes.register("combat.hurtinvul", new RangedAttribute("attribute.name.combat.hurtinvul", 5, 0, 4096));
 
diff --git a/src/main/java/net/minecraft/world/entity/ai/goal/MeleeAttackGoal.java b/src/main/java/net/minecraft/world/entity/ai/goal/MeleeAttackGoal.java
index 1875f9f4ab4ebd182e10f17b89d493a4584db87e..2fcf00beccd992019ddbbcae5cfa0d5720d933bc 100644
--- a/src/main/java/net/minecraft/world/entity/ai/goal/MeleeAttackGoal.java
+++ b/src/main/java/net/minecraft/world/entity/ai/goal/MeleeAttackGoal.java
@@ -5,6 +5,7 @@ import net.minecraft.world.InteractionHand;
 import net.minecraft.world.entity.EntitySelector;
 import net.minecraft.world.entity.LivingEntity;
 import net.minecraft.world.entity.PathfinderMob;
+import net.minecraft.world.entity.ai.attributes.Attributes;
 import net.minecraft.world.entity.player.Player;
 import net.minecraft.world.level.pathfinder.Path;
 
@@ -18,7 +19,7 @@ public class MeleeAttackGoal extends Goal {
     private double pathedTargetZ;
     private int ticksUntilNextPathRecalculation;
     private int ticksUntilNextAttack;
-    private final int attackInterval = 20;
+    private int attackInterval = 20; // Papyrus - use attackspeed
     private long lastCanUseCheck;
     private static final long COOLDOWN_BETWEEN_CAN_USE_CHECKS = 20L;
 
@@ -96,6 +97,7 @@ public class MeleeAttackGoal extends Goal {
     public void tick() {
         LivingEntity livingEntity = this.mob.getTarget();
         if (livingEntity != null) {
+            attackInterval = Math.max(100 - (int) mob.getAttribute(Attributes.ATTACK_SPEED).getValue(), 0); // Papyrus - attackspeed
             this.mob.getLookControl().setLookAt(livingEntity, 30.0F, 30.0F);
             double d = this.mob.getPerceivedTargetDistanceSquareForMeleeAttack(livingEntity);
             this.ticksUntilNextPathRecalculation = Math.max(this.ticksUntilNextPathRecalculation - 1, 0);
@@ -133,7 +135,7 @@ public class MeleeAttackGoal extends Goal {
     }
 
     protected void resetAttackCooldown() {
-        this.ticksUntilNextAttack = this.adjustedTickDelay(20);
+        this.ticksUntilNextAttack = attackInterval;
     }
 
     protected boolean isTimeToAttack() {
@@ -145,7 +147,7 @@ public class MeleeAttackGoal extends Goal {
     }
 
     protected int getAttackInterval() {
-        return this.adjustedTickDelay(20);
+        return attackInterval;
     }
 
     protected double getAttackReachSqr(LivingEntity entity) {
diff --git a/src/main/java/net/minecraft/world/entity/ai/goal/RangedCrossbowAttackGoal.java b/src/main/java/net/minecraft/world/entity/ai/goal/RangedCrossbowAttackGoal.java
index 749ab8537ac30a3c6536b703989d5f22bace0f1e..d256474e099f93f03d66b7e492c637e38cb25f42 100644
--- a/src/main/java/net/minecraft/world/entity/ai/goal/RangedCrossbowAttackGoal.java
+++ b/src/main/java/net/minecraft/world/entity/ai/goal/RangedCrossbowAttackGoal.java
@@ -4,6 +4,7 @@ import java.util.EnumSet;
 import net.minecraft.util.TimeUtil;
 import net.minecraft.util.valueproviders.UniformInt;
 import net.minecraft.world.entity.LivingEntity;
+import net.minecraft.world.entity.ai.attributes.Attributes;
 import net.minecraft.world.entity.monster.CrossbowAttackMob;
 import net.minecraft.world.entity.monster.Monster;
 import net.minecraft.world.entity.monster.RangedAttackMob;
@@ -112,7 +113,7 @@ public class RangedCrossbowAttackGoal<T extends Monster & RangedAttackMob & Cros
                 if (i >= CrossbowItem.getChargeDuration(itemStack)) {
                     this.mob.releaseUsingItem();
                     this.crossbowState = RangedCrossbowAttackGoal.CrossbowState.CHARGED;
-                    this.attackDelay = 20 + this.mob.getRandom().nextInt(20);
+                    this.attackDelay = Math.max(100 - (int) mob.getAttribute(Attributes.ATTACK_SPEED).getValue(), 0); // Papyrus - attack speed
                     this.mob.setChargingCrossbow(false);
                 }
             } else if (this.crossbowState == RangedCrossbowAttackGoal.CrossbowState.CHARGED) {
diff --git a/src/main/java/net/minecraft/world/entity/animal/SnowGolem.java b/src/main/java/net/minecraft/world/entity/animal/SnowGolem.java
index 35e53663e4a6c4d56ec4577d08e7b040cc0c720f..50188050e709dcbc5783f88e685b39705db41869 100644
--- a/src/main/java/net/minecraft/world/entity/animal/SnowGolem.java
+++ b/src/main/java/net/minecraft/world/entity/animal/SnowGolem.java
@@ -56,7 +56,7 @@ public class SnowGolem extends AbstractGolem implements Shearable, RangedAttackM
 
     @Override
     protected void registerGoals() {
-        this.goalSelector.addGoal(1, new RangedAttackGoal(this, 1.25D, 20, 10.0F));
+        this.goalSelector.addGoal(1, new RangedAttackGoal(this, 1.25D, Math.max(100 - (int) getAttribute(Attributes.ATTACK_SPEED).getValue(), 0), 10.0F)); // Papyrus - attack speed
         this.goalSelector.addGoal(2, new WaterAvoidingRandomStrollGoal(this, 1.0D, 1.0000001E-5F));
         this.goalSelector.addGoal(3, new LookAtPlayerGoal(this, Player.class, 6.0F));
         this.goalSelector.addGoal(4, new RandomLookAroundGoal(this));
diff --git a/src/main/java/net/minecraft/world/entity/animal/horse/Llama.java b/src/main/java/net/minecraft/world/entity/animal/horse/Llama.java
index 7ae0e4b3aa8e861500ddc7b38aa671258b532fcd..05bf37bd8d17f3bc4e0583891a3007f94608656d 100644
--- a/src/main/java/net/minecraft/world/entity/animal/horse/Llama.java
+++ b/src/main/java/net/minecraft/world/entity/animal/horse/Llama.java
@@ -130,7 +130,7 @@ public class Llama extends AbstractChestedHorse implements VariantHolder<Llama.V
         this.goalSelector.addGoal(0, new FloatGoal(this));
         this.goalSelector.addGoal(1, new RunAroundLikeCrazyGoal(this, 1.2D));
         this.goalSelector.addGoal(2, new LlamaFollowCaravanGoal(this, 2.0999999046325684D));
-        this.goalSelector.addGoal(3, new RangedAttackGoal(this, 1.25D, 40, 20.0F));
+        this.goalSelector.addGoal(3, new RangedAttackGoal(this, 1.25D, Math.max(100 - (int) getAttribute(Attributes.ATTACK_SPEED).getValue(), 0), 20.0F)); // Papyrus - attack speed
         this.goalSelector.addGoal(3, new PanicGoal(this, 1.2D));
         this.goalSelector.addGoal(4, new BreedGoal(this, 1.0D));
         this.goalSelector.addGoal(5, new TemptGoal(this, 1.25D, Ingredient.of(Items.HAY_BLOCK), false));
diff --git a/src/main/java/net/minecraft/world/entity/boss/wither/WitherBoss.java b/src/main/java/net/minecraft/world/entity/boss/wither/WitherBoss.java
index b781c849737225c00f1dd8d45a6d08cebc000263..6bc59bc649936b3119d383fbe0f6251842df9d81 100644
--- a/src/main/java/net/minecraft/world/entity/boss/wither/WitherBoss.java
+++ b/src/main/java/net/minecraft/world/entity/boss/wither/WitherBoss.java
@@ -110,7 +110,7 @@ public class WitherBoss extends Monster implements PowerableMob, RangedAttackMob
     @Override
     protected void registerGoals() {
         this.goalSelector.addGoal(0, new WitherBoss.WitherDoNothingGoal());
-        this.goalSelector.addGoal(2, new RangedAttackGoal(this, 1.0D, 40, 20.0F));
+        this.goalSelector.addGoal(2, new RangedAttackGoal(this, 1.0D, Math.max(100 - (int) getAttribute(Attributes.ATTACK_SPEED).getValue(), 0), 20.0F)); // Papyrus - attack speed
         this.goalSelector.addGoal(5, new WaterAvoidingRandomFlyingGoal(this, 1.0D));
         this.goalSelector.addGoal(6, new LookAtPlayerGoal(this, Player.class, 8.0F));
         this.goalSelector.addGoal(7, new RandomLookAroundGoal(this));
diff --git a/src/main/java/net/minecraft/world/entity/monster/AbstractSkeleton.java b/src/main/java/net/minecraft/world/entity/monster/AbstractSkeleton.java
index b8abee145fc92faddef98da913eca7715b6bfc03..e9b8f2ff35a03422fdec8efa41b48ad5ba36d163 100644
--- a/src/main/java/net/minecraft/world/entity/monster/AbstractSkeleton.java
+++ b/src/main/java/net/minecraft/world/entity/monster/AbstractSkeleton.java
@@ -47,7 +47,7 @@ import net.minecraft.world.level.block.state.BlockState;
 
 public abstract class AbstractSkeleton extends Monster implements RangedAttackMob {
 
-    private final RangedBowAttackGoal<AbstractSkeleton> bowGoal = new RangedBowAttackGoal<>(this, 1.0D, 20, 15.0F);
+    private final RangedBowAttackGoal<AbstractSkeleton> bowGoal = new RangedBowAttackGoal<>(this, 1.0D, Math.max(100 - (int) getAttribute(Attributes.ATTACK_SPEED).getValue(), 0), 15.0F); // Papyrus - attack speed
     private final MeleeAttackGoal meleeGoal = new MeleeAttackGoal(this, 1.2D, false) {
         @Override
         public void stop() {
@@ -184,7 +184,7 @@ public abstract class AbstractSkeleton extends Monster implements RangedAttackMo
                     b0 = 40;
                 }
 
-                this.bowGoal.setMinAttackInterval(b0);
+                //this.bowGoal.setMinAttackInterval(b0); Papyrus - don't change interval
                 this.goalSelector.addGoal(4, this.bowGoal);
             } else {
                 this.goalSelector.addGoal(4, this.meleeGoal);
diff --git a/src/main/java/net/minecraft/world/entity/monster/Illusioner.java b/src/main/java/net/minecraft/world/entity/monster/Illusioner.java
index 86f7fdd42461db151221d2c0d5cff6953392fa80..98a30c0540f69d3fd2df8174012a49fb02a21bd8 100644
--- a/src/main/java/net/minecraft/world/entity/monster/Illusioner.java
+++ b/src/main/java/net/minecraft/world/entity/monster/Illusioner.java
@@ -66,7 +66,7 @@ public class Illusioner extends SpellcasterIllager implements RangedAttackMob {
         this.goalSelector.addGoal(1, new SpellcasterIllager.SpellcasterCastingSpellGoal());
         this.goalSelector.addGoal(4, new Illusioner.IllusionerMirrorSpellGoal());
         this.goalSelector.addGoal(5, new Illusioner.IllusionerBlindnessSpellGoal());
-        this.goalSelector.addGoal(6, new RangedBowAttackGoal<>(this, 0.5D, 20, 15.0F));
+        this.goalSelector.addGoal(6, new RangedBowAttackGoal<>(this, 0.5D, Math.max(100 - (int) getAttribute(Attributes.ATTACK_SPEED).getValue(), 0), 15.0F)); // Papyrus - attack speed
         this.goalSelector.addGoal(8, new RandomStrollGoal(this, 0.6D));
         this.goalSelector.addGoal(9, new LookAtPlayerGoal(this, Player.class, 3.0F, 1.0F));
         this.goalSelector.addGoal(10, new LookAtPlayerGoal(this, Mob.class, 8.0F));
diff --git a/src/main/java/net/minecraft/world/entity/monster/Witch.java b/src/main/java/net/minecraft/world/entity/monster/Witch.java
index b7bc64818387288955d0723cd071d4203bd2f121..90720fb4d41e066dbebd4d755bd182bc5d59f81b 100644
--- a/src/main/java/net/minecraft/world/entity/monster/Witch.java
+++ b/src/main/java/net/minecraft/world/entity/monster/Witch.java
@@ -64,7 +64,7 @@ public class Witch extends Raider implements RangedAttackMob {
         });
         this.attackPlayersGoal = new NearestAttackableWitchTargetGoal<>(this, Player.class, 10, true, false, (Predicate) null);
         this.goalSelector.addGoal(1, new FloatGoal(this));
-        this.goalSelector.addGoal(2, new RangedAttackGoal(this, 1.0D, 60, 10.0F));
+        this.goalSelector.addGoal(2, new RangedAttackGoal(this, 1.0D, Math.max(100 - (int) getAttribute(Attributes.ATTACK_SPEED).getValue(), 0), 10.0F)); // Papyrus - attack speed
         this.goalSelector.addGoal(2, new WaterAvoidingRandomStrollGoal(this, 1.0D));
         this.goalSelector.addGoal(3, new LookAtPlayerGoal(this, Player.class, 8.0F));
         this.goalSelector.addGoal(3, new RandomLookAroundGoal(this));
diff --git a/src/main/java/net/minecraft/world/entity/player/Player.java b/src/main/java/net/minecraft/world/entity/player/Player.java
index c9c90347a30c8c4aa6be7c1f9cdf255caa076ae7..afb6f2defba69555d8c8aab6979ea020fd5c8374 100644
--- a/src/main/java/net/minecraft/world/entity/player/Player.java
+++ b/src/main/java/net/minecraft/world/entity/player/Player.java
@@ -17,7 +17,8 @@ import java.util.OptionalInt;
 import java.util.function.Predicate;
 import javax.annotation.Nullable;
 
-import de.erethon.papyrus.DamageType;
+import de.erethon.papyrus.CraftDamageType;
+import de.erethon.spellbook.api.SpellCaster;
 import net.minecraft.SharedConstants;
 import net.minecraft.Util;
 import net.minecraft.advancements.CriteriaTriggers;
@@ -1237,12 +1238,21 @@ public abstract class Player extends LivingEntity {
     }
 
     // Papyrus - combat rework
-    public void attack(Entity target, DamageType type) {
+    public void attack(Entity target, CraftDamageType type) {
         if (!target.isAttackable()) {
             return;
         }
         float damage = (float) (getAttributeValue(Attributes.ATTACK_DAMAGE) + getAttributeValue(type.getAdvAttr()));
+        damage = damage * getAttackStrengthScale(0.5f);
+        if (target instanceof LivingEntity entity) {
+            damage = (float) getBukkitEntity().onAttack(entity.getBukkitLivingEntity(), damage, type.getApiDamageType());
+        }
+        int crit = random.nextIntBetweenInclusive(0, 100);
+        if (crit < getAttributeValue(Attributes.STAT_CRITCHANCE)) {
+            damage = (float) (damage + getAttributeValue(Attributes.STAT_CRITDMG));
+        }
         target.hurt(DamageSource.playerAttack(this), damage, type);
+        resetAttackStrengthTicker();
     }
 
     public void attack(Entity target) {
@@ -1266,13 +1276,13 @@ public abstract class Player extends LivingEntity {
                     f1 = EnchantmentHelper.getDamageBonus(this.getMainHandItem(), MobType.UNDEFINED);
                 }
 
-                float f2 = this.getAttackStrengthScale(0.5F);
+                float scaled = this.getAttackStrengthScale(0.5F);
 
-                f *= 0.2F + f2 * f2 * 0.8F;
-                f1 *= f2;
+                f *= 0.2F + scaled * scaled * 0.8F;
+                f1 *= scaled;
                 // this.resetAttackCooldown(); // CraftBukkit - Moved to EntityLiving to reset the cooldown after the damage is dealt
                 if (f > 0.0F || f1 > 0.0F) {
-                    boolean flag = f2 > 0.9F;
+                    boolean flag = scaled > 0.9F;
                     boolean flag1 = false;
                     byte b0 = 0;
                     int i = b0 + EnchantmentHelper.getKnockbackBonus(this);
@@ -1323,7 +1333,7 @@ public abstract class Player extends LivingEntity {
                     }
 
                     Vec3 vec3d = target.getDeltaMovement();
-                    boolean flag5 = target.hurt(DamageSource.playerAttack(this).critical(flag2), f); // Paper - add critical damage API
+                    boolean flag5 = target.hurt(DamageSource.playerAttack(this).critical(flag2), f, attackDamageType); // Paper - add critical damage API
 
                     if (flag5) {
                         if (i > 0) {
@@ -1351,7 +1361,7 @@ public abstract class Player extends LivingEntity {
 
                                 if (entityliving != this && entityliving != target && !this.isAlliedTo((Entity) entityliving) && (!(entityliving instanceof ArmorStand) || !((ArmorStand) entityliving).isMarker()) && this.distanceToSqr((Entity) entityliving) < 9.0D) {
                                     // CraftBukkit start - Only apply knockback if the damage hits
-                                    if (entityliving.hurt(DamageSource.playerAttack(this).sweep().critical(flag2), f4)) { // Paper - add critical damage API
+                                    if (entityliving.hurt(DamageSource.playerAttack(this).sweep().critical(flag2), f4, attackDamageType)) { // Paper - add critical damage API
                                     entityliving.knockback(0.4000000059604645D, (double) Mth.sin(this.getYRot() * 0.017453292F), (double) (-Mth.cos(this.getYRot() * 0.017453292F)), this); // Paper
                                     }
                                     // CraftBukkit end
@@ -1464,7 +1474,7 @@ public abstract class Player extends LivingEntity {
 
     @Override
     protected void doAutoAttackOnTouch(LivingEntity target) {
-        this.attack(target, DamageType.PHYSICAL); // Papyrus
+        this.attack(target, CraftDamageType.PHYSICAL); // Papyrus
     }
 
     public void disableShield(boolean sprinting) {
diff --git a/src/main/java/net/minecraft/world/entity/projectile/AbstractArrow.java b/src/main/java/net/minecraft/world/entity/projectile/AbstractArrow.java
index 9788e477ff1446ad2ea3669922cc7dfc09900ce8..2d3b021f30032b6363834535a96c6101af6bcadb 100644
--- a/src/main/java/net/minecraft/world/entity/projectile/AbstractArrow.java
+++ b/src/main/java/net/minecraft/world/entity/projectile/AbstractArrow.java
@@ -1,6 +1,7 @@
 package net.minecraft.world.entity.projectile;
 
 import com.google.common.collect.Lists;
+import de.erethon.papyrus.CraftDamageType;
 import it.unimi.dsi.fastutil.ints.IntOpenHashSet;
 import java.util.Arrays;
 import java.util.Collection;
@@ -399,7 +400,7 @@ public abstract class AbstractArrow extends Projectile {
             // CraftBukkit end
         }
 
-        if (entity.hurt(damagesource, (float) i)) {
+        if (entity.hurt(damagesource, (float) i, damageType)) { // Papyrus - add damage type to arrows
             if (flag) {
                 return;
             }
diff --git a/src/main/java/net/minecraft/world/entity/projectile/FireworkRocketEntity.java b/src/main/java/net/minecraft/world/entity/projectile/FireworkRocketEntity.java
index 5406925cd66f46ab8744123c670d72cea7bfc3a1..b1694698070ca3658614f9c13059c1018c053343 100644
--- a/src/main/java/net/minecraft/world/entity/projectile/FireworkRocketEntity.java
+++ b/src/main/java/net/minecraft/world/entity/projectile/FireworkRocketEntity.java
@@ -242,7 +242,7 @@ public class FireworkRocketEntity extends Projectile implements ItemSupplier {
         if (f > 0.0F) {
             if (this.attachedToEntity != null) {
                 CraftEventFactory.entityDamage = this; // CraftBukkit
-                this.attachedToEntity.hurt(DamageSource.fireworks(this, this.getOwner()), 5.0F + (float) (nbttaglist.size() * 2));
+                this.attachedToEntity.hurt(DamageSource.fireworks(this, this.getOwner()), 5.0F + (float) (nbttaglist.size() * 2), damageType);
                 CraftEventFactory.entityDamage = null; // CraftBukkit
             }
 
@@ -271,7 +271,7 @@ public class FireworkRocketEntity extends Projectile implements ItemSupplier {
                         float f1 = f * (float) Math.sqrt((5.0D - (double) this.distanceTo(entityliving)) / 5.0D);
 
                         CraftEventFactory.entityDamage = this; // CraftBukkit
-                        entityliving.hurt(DamageSource.fireworks(this, this.getOwner()), f1);
+                        entityliving.hurt(DamageSource.fireworks(this, this.getOwner()), f1, damageType);
                         CraftEventFactory.entityDamage = null; // CraftBukkit
                     }
                 }
diff --git a/src/main/java/net/minecraft/world/entity/projectile/LargeFireball.java b/src/main/java/net/minecraft/world/entity/projectile/LargeFireball.java
index 3a4b95288d390e72c0d97671ecc2e2ef2f976de1..7f5dc7d56abd939e2a3a6fbe4547e3888f4deff3 100644
--- a/src/main/java/net/minecraft/world/entity/projectile/LargeFireball.java
+++ b/src/main/java/net/minecraft/world/entity/projectile/LargeFireball.java
@@ -53,7 +53,7 @@ public class LargeFireball extends Fireball {
             Entity entity = entityHitResult.getEntity();
             Entity entity1 = this.getOwner();
 
-            entity.hurt(DamageSource.fireball(this, entity1), 6.0F);
+            entity.hurt(DamageSource.fireball(this, entity1), 6.0F, damageType);
             if (entity1 instanceof LivingEntity) {
                 this.doEnchantDamageEffects((LivingEntity) entity1, entity);
             }
diff --git a/src/main/java/net/minecraft/world/entity/projectile/LlamaSpit.java b/src/main/java/net/minecraft/world/entity/projectile/LlamaSpit.java
index 4132c1113f5437a776e5e3c1cb306904775aed88..252bd3fbe1fac0565a188c8794c7fc1449aa11eb 100644
--- a/src/main/java/net/minecraft/world/entity/projectile/LlamaSpit.java
+++ b/src/main/java/net/minecraft/world/entity/projectile/LlamaSpit.java
@@ -62,7 +62,7 @@ public class LlamaSpit extends Projectile {
         Entity entity = this.getOwner();
 
         if (entity instanceof LivingEntity) {
-            entityHitResult.getEntity().hurt(DamageSource.indirectMobAttack(this, (LivingEntity) entity).setProjectile(), 1.0F);
+            entityHitResult.getEntity().hurt(DamageSource.indirectMobAttack(this, (LivingEntity) entity).setProjectile(), 1.0F, damageType);
         }
 
     }
diff --git a/src/main/java/net/minecraft/world/entity/projectile/Projectile.java b/src/main/java/net/minecraft/world/entity/projectile/Projectile.java
index 5ccd7dc0a093277d3ea09c0fe8be03c0073ee1a8..97cc67b11021c3bbc4b6c119d2de921480ca6f69 100644
--- a/src/main/java/net/minecraft/world/entity/projectile/Projectile.java
+++ b/src/main/java/net/minecraft/world/entity/projectile/Projectile.java
@@ -4,6 +4,8 @@ import com.google.common.base.MoreObjects;
 import java.util.Iterator;
 import java.util.UUID;
 import javax.annotation.Nullable;
+
+import de.erethon.papyrus.CraftDamageType;
 import net.minecraft.core.BlockPos;
 import net.minecraft.nbt.CompoundTag;
 import net.minecraft.network.protocol.Packet;
@@ -48,6 +50,10 @@ public abstract class Projectile extends Entity {
     private static int loadedTick;
 
     private int loadedLifetime = 0;
+
+    // Papyrus start
+    public CraftDamageType damageType = CraftDamageType.PHYSICAL;
+
     @Override
     public void setPos(double x, double y, double z) {
         int currentTick = net.minecraft.server.MinecraftServer.currentTick;
diff --git a/src/main/java/net/minecraft/world/entity/projectile/ShulkerBullet.java b/src/main/java/net/minecraft/world/entity/projectile/ShulkerBullet.java
index a89cbcb913628eaaf107f49520f6aea7ffcc9446..97459044a28bf749ee7aa48f15a02e63a84feb06 100644
--- a/src/main/java/net/minecraft/world/entity/projectile/ShulkerBullet.java
+++ b/src/main/java/net/minecraft/world/entity/projectile/ShulkerBullet.java
@@ -299,7 +299,7 @@ public class ShulkerBullet extends Projectile {
         Entity entity = entityHitResult.getEntity();
         Entity entity1 = this.getOwner();
         LivingEntity entityliving = entity1 instanceof LivingEntity ? (LivingEntity) entity1 : null;
-        boolean flag = entity.hurt(DamageSource.indirectMobAttack(this, entityliving).setProjectile(), 4.0F);
+        boolean flag = entity.hurt(DamageSource.indirectMobAttack(this, entityliving).setProjectile(), 4.0F, damageType);
 
         if (flag) {
             this.doEnchantDamageEffects(entityliving, entity);
diff --git a/src/main/java/net/minecraft/world/entity/projectile/SmallFireball.java b/src/main/java/net/minecraft/world/entity/projectile/SmallFireball.java
index 00ac1cdc4734cc57f15433c5c6e7a3a545739d33..671b4400c0dc3fe4c1ae95765b1aa583a8f416c5 100644
--- a/src/main/java/net/minecraft/world/entity/projectile/SmallFireball.java
+++ b/src/main/java/net/minecraft/world/entity/projectile/SmallFireball.java
@@ -49,7 +49,7 @@ public class SmallFireball extends Fireball {
                 entity.setSecondsOnFire(event.getDuration(), false);
             }
             // CraftBukkit end
-            if (!entity.hurt(DamageSource.fireball(this, entity1), 5.0F)) {
+            if (!entity.hurt(DamageSource.fireball(this, entity1), 5.0F, damageType)) {
                 entity.setRemainingFireTicks(i);
             } else if (entity1 instanceof LivingEntity) {
                 this.doEnchantDamageEffects((LivingEntity) entity1, entity);
diff --git a/src/main/java/net/minecraft/world/entity/projectile/ThrownEgg.java b/src/main/java/net/minecraft/world/entity/projectile/ThrownEgg.java
index bdddec9510d84c8940b6116ede389bcac3680ab1..94f894f3fd93f84b8d8c9a053e2c4194396f8f71 100644
--- a/src/main/java/net/minecraft/world/entity/projectile/ThrownEgg.java
+++ b/src/main/java/net/minecraft/world/entity/projectile/ThrownEgg.java
@@ -46,7 +46,7 @@ public class ThrownEgg extends ThrowableItemProjectile {
     @Override
     protected void onHitEntity(EntityHitResult entityHitResult) {
         super.onHitEntity(entityHitResult);
-        entityHitResult.getEntity().hurt(DamageSource.thrown(this, this.getOwner()), 0.0F);
+        entityHitResult.getEntity().hurt(DamageSource.thrown(this, this.getOwner()), 0.0F, damageType);
     }
 
     @Override
diff --git a/src/main/java/net/minecraft/world/entity/projectile/ThrownEnderpearl.java b/src/main/java/net/minecraft/world/entity/projectile/ThrownEnderpearl.java
index f224ebbc0efefddede43d87f0300c014077b9931..de495cbfa41eb4e8f25562b57b5494e307ee5b0d 100644
--- a/src/main/java/net/minecraft/world/entity/projectile/ThrownEnderpearl.java
+++ b/src/main/java/net/minecraft/world/entity/projectile/ThrownEnderpearl.java
@@ -41,7 +41,7 @@ public class ThrownEnderpearl extends ThrowableItemProjectile {
     @Override
     protected void onHitEntity(EntityHitResult entityHitResult) {
         super.onHitEntity(entityHitResult);
-        entityHitResult.getEntity().hurt(DamageSource.thrown(this, this.getOwner()), 0.0F);
+        entityHitResult.getEntity().hurt(DamageSource.thrown(this, this.getOwner()), 0.0F, damageType);
     }
 
     @Override
diff --git a/src/main/java/net/minecraft/world/entity/projectile/ThrownTrident.java b/src/main/java/net/minecraft/world/entity/projectile/ThrownTrident.java
index fb6e590e4613f59aaa8278932134aa0ca3d9da8f..618a008b0df0d8710228d7aa9eb3a9509f1377fc 100644
--- a/src/main/java/net/minecraft/world/entity/projectile/ThrownTrident.java
+++ b/src/main/java/net/minecraft/world/entity/projectile/ThrownTrident.java
@@ -142,7 +142,7 @@ public class ThrownTrident extends AbstractArrow {
         this.dealtDamage = true;
         SoundEvent soundeffect = SoundEvents.TRIDENT_HIT;
 
-        if (entity.hurt(damagesource, f)) {
+        if (entity.hurt(damagesource, f, damageType)) {
             if (entity.getType() == EntityType.ENDERMAN) {
                 return;
             }
diff --git a/src/main/java/net/minecraft/world/entity/projectile/WitherSkull.java b/src/main/java/net/minecraft/world/entity/projectile/WitherSkull.java
index 9665095240a370983878350aed41badacfb6f261..72934b52e94615d05c080beac26d2db428adadc2 100644
--- a/src/main/java/net/minecraft/world/entity/projectile/WitherSkull.java
+++ b/src/main/java/net/minecraft/world/entity/projectile/WitherSkull.java
@@ -59,7 +59,7 @@ public class WitherSkull extends AbstractHurtingProjectile {
             if (entity1 instanceof LivingEntity) {
                 LivingEntity entityliving = (LivingEntity) entity1;
 
-                flag = entity.hurt(DamageSource.witherSkull(this, entityliving), 8.0F);
+                flag = entity.hurt(DamageSource.witherSkull(this, entityliving), 8.0F, damageType);
                 if (flag) {
                     if (entity.isAlive()) {
                         this.doEnchantDamageEffects(entityliving, entity);
@@ -68,7 +68,7 @@ public class WitherSkull extends AbstractHurtingProjectile {
                     }
                 }
             } else {
-                flag = entity.hurt(DamageSource.MAGIC, 5.0F);
+                flag = entity.hurt(DamageSource.MAGIC, 5.0F, damageType);
             }
 
             if (flag && entity instanceof LivingEntity) {
diff --git a/src/main/java/org/bukkit/craftbukkit/CraftServer.java b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
index aa091998b2cc71b78f6d8e99556f00b3a315b91a..58e907163b7aa15d181061c7c21c4cd82d5a8e83 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftServer.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
@@ -16,6 +16,7 @@ import com.mojang.brigadier.tree.LiteralCommandNode;
 import com.mojang.datafixers.util.Pair;
 import com.mojang.serialization.DynamicOps;
 import com.mojang.serialization.Lifecycle;
+import de.erethon.spellbook.api.SpellbookAPI;
 import io.netty.buffer.ByteBuf;
 import io.netty.buffer.ByteBufOutputStream;
 import io.netty.buffer.Unpooled;
@@ -300,6 +301,9 @@ public final class CraftServer implements Server {
     private final io.papermc.paper.logging.SysoutCatcher sysoutCatcher = new io.papermc.paper.logging.SysoutCatcher(); // Paper
     private final CraftPotionBrewer potionBrewer = new CraftPotionBrewer(); // Paper
 
+    // Papyrus start
+    private final SpellbookAPI spellbookAPI;
+
     static {
         ConfigurationSerialization.registerClass(CraftOfflinePlayer.class);
         ConfigurationSerialization.registerClass(CraftPlayerProfile.class);
@@ -385,7 +389,14 @@ public final class CraftServer implements Server {
             MapPalette.setMapColorCache(new CraftMapColorCache(this.logger));
         }
         datapackManager = new io.papermc.paper.datapack.PaperDatapackManager(console.getPackRepository()); // Paper
+        spellbookAPI = new SpellbookAPI(this); // Papyrus start - Spellbook
+    }
+
+    @Override
+    public SpellbookAPI getSpellbookAPI() {
+        return spellbookAPI;
     }
+    // Papyrus end
 
     public boolean getCommandBlockOverride(String command) {
         return this.overrideAllCommandBlockCommands || this.commandsConfiguration.getStringList("command-block-overrides").contains(command);
@@ -447,6 +458,7 @@ public final class CraftServer implements Server {
         } else {
             pluginFolder.mkdir();
         }
+        spellbookAPI.getLibrary().loadSpells(SpellbookAPI.SPELLS); // Papyrus - Load spells after plugins
     }
 
     // Paper start
diff --git a/src/main/java/org/bukkit/craftbukkit/entity/AbstractProjectile.java b/src/main/java/org/bukkit/craftbukkit/entity/AbstractProjectile.java
index 825fdc6162797ade8e76e1ca3a863ed5fb48f936..ea62bcb4237ac89bdad42ff702968311c1e3525e 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/AbstractProjectile.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/AbstractProjectile.java
@@ -1,5 +1,7 @@
 package org.bukkit.craftbukkit.entity;
 
+import de.erethon.papyrus.CraftDamageType;
+import de.erethon.papyrus.DamageType;
 import org.bukkit.craftbukkit.CraftServer;
 import org.bukkit.entity.Projectile;
 
@@ -48,4 +50,8 @@ public abstract class AbstractProjectile extends CraftEntity implements Projecti
     }
     // Paper end
 
+    @Override
+    public void setDamageType(DamageType type) {
+        ((net.minecraft.world.entity.projectile.Projectile) getHandle()).damageType = CraftDamageType.ofAPI(type);
+    }
 }
diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftEnderDragonPart.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftEnderDragonPart.java
index a407e802164603c74ec1f4ad3c98bc50398954bc..c7fdecbcd6eb7571a09883637bee4150dcb4eb72 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftEnderDragonPart.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftEnderDragonPart.java
@@ -1,5 +1,6 @@
 package org.bukkit.craftbukkit.entity;
 
+import de.erethon.papyrus.DamageType;
 import org.bukkit.craftbukkit.CraftServer;
 import org.bukkit.entity.EnderDragon;
 import org.bukkit.entity.EnderDragonPart;
@@ -30,11 +31,21 @@ public class CraftEnderDragonPart extends CraftComplexPart implements EnderDrago
         this.getParent().damage(amount);
     }
 
+    @Override
+    public void damage(double amount, DamageType type) {
+        this.getParent().damage(amount,type);
+    }
+
     @Override
     public void damage(double amount, Entity source) {
         this.getParent().damage(amount, source);
     }
 
+    @Override
+    public void damage(double amount, Entity source, DamageType type) {
+        this.getParent().damage(amount, source, type);
+    }
+
     @Override
     public double getHealth() {
         return this.getParent().getHealth();
diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftLivingEntity.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftLivingEntity.java
index 09da9d3c85e36c2a78663f58a97963dbc795a367..4173769af13142cd2384cd4f6951a9cd2a65aeaa 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftLivingEntity.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftLivingEntity.java
@@ -3,12 +3,13 @@ package org.bukkit.craftbukkit.entity;
 import com.destroystokyo.paper.entity.TargetEntityInfo;
 import com.google.common.base.Preconditions;
 import com.google.common.collect.Sets;
-import java.util.ArrayList;
-import java.util.Collection;
-import java.util.Iterator;
-import java.util.List;
-import java.util.Set;
-import java.util.UUID;
+
+import java.util.*;
+
+import de.erethon.papyrus.DamageType;
+import de.erethon.spellbook.api.SpellData;
+import de.erethon.spellbook.api.SpellEffect;
+import de.erethon.spellbook.api.SpellbookSpell;
 import net.minecraft.sounds.SoundEvent;
 import net.minecraft.world.InteractionHand;
 import net.minecraft.world.damagesource.DamageSource;
@@ -89,6 +90,19 @@ import org.bukkit.util.Vector;
 public class CraftLivingEntity extends CraftEntity implements LivingEntity {
     private CraftEntityEquipment equipment;
 
+    // Papyrus start
+    private Set<SpellData> unlockedSpellData = new HashSet<>();
+    private Set<SpellData> passiveSpells = new HashSet<>();
+    private Set<SpellbookSpell> activeSpells = new HashSet<>();
+    private final SpellData[] assignedSlots = new SpellData[8];
+
+    Map<SpellData, Long> usedSpells = new LinkedHashMap<>();
+    Set<SpellEffect> effects = new HashSet<>();
+    Set<SpellbookSpell> runningPassiveSpells = new HashSet<>();
+    private int maxEnergy = 0;
+    private int energy = 0;
+    // Papyrus end
+
     public CraftLivingEntity(final CraftServer server, final net.minecraft.world.entity.LivingEntity entity) {
         super(server, entity);
 
@@ -358,7 +372,12 @@ public class CraftLivingEntity extends CraftEntity implements LivingEntity {
     // Paper End - Bee Stinger API
     @Override
     public void damage(double amount) {
-        this.damage(amount, null);
+        this.damage(amount, DamageType.PURE);
+    }
+
+    @Override
+    public void damage(double amount, DamageType type) {
+
     }
 
     @Override
@@ -376,6 +395,11 @@ public class CraftLivingEntity extends CraftEntity implements LivingEntity {
         entity.hurt(reason, (float) amount);
     }
 
+    @Override
+    public void damage(double amount, Entity source, DamageType type) {
+
+    }
+
     @Override
     public Location getEyeLocation() {
         Location loc = getLocation();
@@ -1052,6 +1076,71 @@ public class CraftLivingEntity extends CraftEntity implements LivingEntity {
     public void knockback(double strength, double directionX, double directionZ) {
         Preconditions.checkArgument(strength > 0, "Knockback strength must be > 0");
         getHandle().knockback(strength, directionX, directionZ);
-    };
-    // Paper end
+    }
+
+
+         // Papyrus start - Spellbook
+
+    @Override
+    public Map<SpellData, Long> getUsedSpells() {
+        return usedSpells;
+    }
+
+    @Override
+    public Set<SpellEffect> getEffects() {
+         return effects;
+     }
+
+    @Override
+    public Set<SpellbookSpell> getPassiveSpells() {
+        return runningPassiveSpells;
+    }
+
+    @Override
+    public Set<SpellbookSpell> getActiveSpells() {
+         return activeSpells;
+    }
+
+    @Override
+    public Set<SpellData> getUnlockedSpells() {
+         return unlockedSpellData;
+    }
+
+    @Override
+    public Set<SpellData> getUnlockedPassives() {
+        return passiveSpells;
+    }
+
+    @Override
+    public int getEnergy() {
+        return energy;
+    }
+
+    @Override
+    public int setEnergy(int e) {
+        return energy = Math.min(e, maxEnergy);
+    }
+
+    @Override
+    public int getMaxEnergy() {
+        return maxEnergy;
+    }
+
+    @Override
+    public int setMaxEnergy(int e) {
+        return maxEnergy = e;
+    }
+
+    @Override
+    public int addEnergy(int e) {
+        energy = Math.min(energy + e, maxEnergy);
+        return energy;
+    }
+
+    @Override
+    public int removeEnergy(int e) {
+        energy = Math.max(energy - e, 0);
+        return energy;
+    }
+    // Papyrus end
 }
diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
index 68606c94ba353b83703e6495836ddada876d7d83..f48959fb50083fafbd7d5871e5b4ed500b6bc3ac 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
@@ -377,18 +377,6 @@ public class CraftPlayer extends CraftHumanEntity implements Player {
         getHandle().connection.send(packet);
     }
 
-    @Override
-    public void sendActionBar(String message) {
-        if (getHandle().connection == null || message == null || message.isEmpty()) return;
-        getHandle().connection.send(new net.minecraft.network.protocol.game.ClientboundSetActionBarTextPacket(CraftChatMessage.fromStringOrNull(message)));
-    }
-
-    @Override
-    public void sendActionBar(char alternateChar, String message) {
-        if (message == null || message.isEmpty()) return;
-        sendActionBar(org.bukkit.ChatColor.translateAlternateColorCodes(alternateChar, message));
-    }
-
     @Override
     public void setPlayerListHeaderFooter(BaseComponent[] header, BaseComponent[] footer) {
          if (header != null) {
diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftProjectile.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftProjectile.java
index 9a68f4ef68870d0baab5b6464d6c0a82a8fd105d..52ac90a890c73eec5dc7164f05e83ba884f91472 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftProjectile.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftProjectile.java
@@ -1,5 +1,7 @@
 package org.bukkit.craftbukkit.entity;
 
+import de.erethon.papyrus.CraftDamageType;
+import de.erethon.papyrus.DamageType;
 import net.minecraft.world.entity.LivingEntity;
 import org.bukkit.craftbukkit.CraftServer;
 import org.bukkit.entity.Projectile;
@@ -34,4 +36,9 @@ public abstract class CraftProjectile extends AbstractProjectile implements Proj
     public String toString() {
         return "CraftProjectile";
     }
+
+    @Override
+    public void setDamageType(DamageType type) {
+        getHandle().damageType = CraftDamageType.ofAPI(type);
+    }
 }
