From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Malfrador <malfrador@gmail.com>
Date: Sat, 19 Mar 2022 15:52:36 +0100
Subject: [PATCH] disable gui


diff --git a/src/main/java/net/minecraft/server/Main.java b/src/main/java/net/minecraft/server/Main.java
index 1e0d261439255091a6f61485c0747231fbd5b1db..1e3629169439a0df14c6ba48871f1b03e72e0711 100644
--- a/src/main/java/net/minecraft/server/Main.java
+++ b/src/main/java/net/minecraft/server/Main.java
@@ -271,7 +271,7 @@ public class Main {
                 dedicatedserver1.setDemo(optionset.has(optionspec2));
                 dedicatedserver1.setId((String) optionset.valueOf(optionspec12));
                 */
-                boolean flag1 = !optionset.has("nogui") && !optionset.nonOptionArguments().contains("nogui");
+                boolean flag1 = false; // Papyrus - disable gui
 
                 if(!Boolean.parseBoolean(System.getenv().getOrDefault("PAPER_DISABLE_SERVER_GUI", String.valueOf(false)))) // Paper
                 if (flag1 && !GraphicsEnvironment.isHeadless()) {
