From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Malfrador <malfrador@gmail.com>
Date: Fri, 12 Jan 2024 13:22:05 +0100
Subject: [PATCH] Hephaestus


diff --git a/src/main/java/de/erethon/hephaestus/HBukkitHandler.java b/src/main/java/de/erethon/hephaestus/HBukkitHandler.java
new file mode 100644
index 0000000000000000000000000000000000000000..522ebf35e2088bee34bcd51bb18a3f9691cfb482
--- /dev/null
+++ b/src/main/java/de/erethon/hephaestus/HBukkitHandler.java
@@ -0,0 +1,174 @@
+package de.erethon.hephaestus;
+
+import com.destroystokyo.paper.event.player.PlayerArmorChangeEvent;
+import de.erethon.spellbook.api.SpellCastEvent;
+import net.minecraft.world.entity.player.Player;
+import org.bukkit.Bukkit;
+import org.bukkit.Material;
+import org.bukkit.craftbukkit.inventory.CraftItemStack;
+import org.bukkit.entity.Item;
+import org.bukkit.entity.LivingEntity;
+import org.bukkit.event.EventHandler;
+import org.bukkit.event.Listener;
+import org.bukkit.event.block.BlockDropItemEvent;
+import org.bukkit.event.entity.EntityDamageByEntityEvent;
+import org.bukkit.event.entity.EntityDropItemEvent;
+import org.bukkit.event.entity.EntityPickupItemEvent;
+import org.bukkit.event.inventory.InventoryClickEvent;
+import org.bukkit.event.player.PlayerAttemptPickupItemEvent;
+import org.bukkit.event.player.PlayerDropItemEvent;
+import org.bukkit.event.player.PlayerInteractEvent;
+import org.bukkit.inventory.ItemStack;
+import org.bukkit.plugin.Plugin;
+import org.bukkit.plugin.java.JavaPlugin;
+
+public class HBukkitHandler implements Listener {
+
+    private final Plugin plugin;
+    private final ItemLibrary lib;
+
+    public HBukkitHandler(JavaPlugin pl, ItemLibrary library) {
+        this.lib = library;
+        this.plugin = pl;
+        Bukkit.getPluginManager().registerEvents(this, plugin);
+    }
+
+    @EventHandler
+    private void onInvClick(InventoryClickEvent event) {
+        if (event.getCurrentItem() == null) {
+            return;
+        }
+        HItemStack item = ItemLibrary.getHItemFromBukkit(event.getCurrentItem());
+        if (item != null && item.getBehaviour() != null) {
+            item.getBehaviour().onInventoryClick(CraftItemStack.unwrap(event.getCurrentItem()), (Player) event.getWhoClicked(), event.getClick(), event.getAction(), event);
+        }
+    }
+
+    @EventHandler
+    private void onDamage(EntityDamageByEntityEvent event) {
+        if (event.getEntity() instanceof LivingEntity living && living.getEquipment() != null) {
+            ItemStack mainHand = living.getEquipment().getItemInMainHand();
+            HItemStack item = ItemLibrary.getHItemFromBukkit(mainHand);
+            if (item != null && item.getBehaviour() != null) {
+                item.getBehaviour().onDamaged(CraftItemStack.unwrap(mainHand), living, event);
+            }
+            ItemStack offHand = living.getEquipment().getItemInOffHand();
+            item = ItemLibrary.getHItemFromBukkit(mainHand);
+            if (item != null && item.getBehaviour() != null) {
+                item.getBehaviour().onDamaged(CraftItemStack.unwrap(offHand), living, event);
+            }
+            for (ItemStack armor : living.getEquipment().getArmorContents()) {
+                item = ItemLibrary.getHItemFromBukkit(armor);
+                if (item != null && item.getBehaviour() != null) {
+                    item.getBehaviour().onDamaged(CraftItemStack.unwrap(armor), living, event);
+                }
+            }
+        }
+    }
+
+    @EventHandler
+    private void onDamage(SpellCastEvent event) {
+        LivingEntity living = event.getCaster();
+        if (living.getEquipment() != null) {
+            ItemStack mainHand = living.getEquipment().getItemInMainHand();
+            HItemStack item = ItemLibrary.getHItemFromBukkit(mainHand);
+            if (item != null && item.getBehaviour() != null) {
+                item.getBehaviour().onSpellCast(CraftItemStack.unwrap(mainHand), living, event);
+            }
+            ItemStack offHand = living.getEquipment().getItemInOffHand();
+            item = ItemLibrary.getHItemFromBukkit(mainHand);
+            if (item != null && item.getBehaviour() != null) {
+                item.getBehaviour().onSpellCast(CraftItemStack.unwrap(offHand), living, event);
+            }
+            for (ItemStack armor : living.getEquipment().getArmorContents()) {
+                item = ItemLibrary.getHItemFromBukkit(armor);
+                if (item != null && item.getBehaviour() != null) {
+                    item.getBehaviour().onSpellCast(CraftItemStack.unwrap(armor), living, event);
+                }
+            }
+        }
+    }
+
+    @EventHandler
+    private void onPlayerDrop(PlayerDropItemEvent event) {
+        ItemStack stack = event.getItemDrop().getItemStack();
+        HItemStack item = ItemLibrary.getHItemFromBukkit(stack);
+        if (item != null && item.getBehaviour() != null) {
+            item.getBehaviour().onPlayerDrop(CraftItemStack.unwrap(stack), event.getPlayer(), event);
+        }
+    }
+
+    @EventHandler
+    private void onEntityDrop(EntityDropItemEvent event) {
+        ItemStack stack = event.getItemDrop().getItemStack();
+        HItemStack item = ItemLibrary.getHItemFromBukkit(stack);
+        if (item != null && item.getBehaviour() != null) {
+            item.getBehaviour().onEntityDrop(CraftItemStack.unwrap(stack), event.getEntity(), event);
+        }
+    }
+
+    @EventHandler
+    private void onBlockDrop(BlockDropItemEvent event) {
+        for (Item stack : event.getItems()) {
+            HItemStack item = ItemLibrary.getHItemFromBukkit(stack.getItemStack());
+            if (item != null && item.getBehaviour() != null) {
+                item.getBehaviour().onBlockDrop(CraftItemStack.unwrap(stack.getItemStack()), event.getPlayer(), event);
+            }
+        }
+    }
+
+    @EventHandler
+    private void onPickup(PlayerAttemptPickupItemEvent event) {
+        ItemStack stack = event.getItem().getItemStack();
+        HItemStack item = ItemLibrary.getHItemFromBukkit(stack);
+        if (item != null && item.getBehaviour() != null) {
+            item.getBehaviour().onPickup(CraftItemStack.unwrap(stack), event.getPlayer(), event);
+        }
+    }
+
+    @EventHandler
+    private void onEntityPickup(EntityPickupItemEvent event) {
+        ItemStack stack = event.getItem().getItemStack();
+        HItemStack item = ItemLibrary.getHItemFromBukkit(stack);
+        if (item != null && item.getBehaviour() != null) {
+            item.getBehaviour().onEntityPickup(CraftItemStack.unwrap(stack), event.getEntity(), event);
+        }
+    }
+
+    @EventHandler
+    private void onInteract(PlayerInteractEvent event) {
+        if (!event.hasItem()) {
+            return;
+        }
+        HItemStack item = ItemLibrary.getHItemFromBukkit(event.getItem());
+        if (event.getAction().isRightClick()) {
+            if (item != null && item.getBehaviour() != null) {
+                item.getBehaviour().onRightClick(CraftItemStack.unwrap(event.getItem()), event.getPlayer(), event);
+            }
+        }
+        if (event.getAction().isLeftClick()) {
+            if (item != null && item.getBehaviour() != null) {
+                item.getBehaviour().onLeftClick(CraftItemStack.unwrap(event.getItem()), event.getPlayer(), event);
+            }
+        }
+    }
+
+    @EventHandler
+    private void onEquip(PlayerArmorChangeEvent event) {
+        if (event.getNewItem() == event.getOldItem()) {
+            return;
+        }
+        ItemStack stack = event.getNewItem();
+        HItemStack item = ItemLibrary.getHItemFromBukkit(stack);
+        if (item != null && item.getBehaviour() != null && stack.getType() != Material.AIR) {
+            item.getBehaviour().onEquip(CraftItemStack.unwrap(stack), event.getPlayer(), event);
+            return;
+        }
+        stack = event.getOldItem();
+        item = ItemLibrary.getHItemFromBukkit(stack);
+        if (item != null && item.getBehaviour() != null) {
+            item.getBehaviour().onUnequip(CraftItemStack.unwrap(stack), event.getPlayer(), event);
+        }
+    }
+
+}
diff --git a/src/main/java/de/erethon/hephaestus/HItem.java b/src/main/java/de/erethon/hephaestus/HItem.java
new file mode 100644
index 0000000000000000000000000000000000000000..3c6c35f38236d2b6fd7efe1570b5abf98e413e81
--- /dev/null
+++ b/src/main/java/de/erethon/hephaestus/HItem.java
@@ -0,0 +1,240 @@
+package de.erethon.hephaestus;
+
+import io.papermc.paper.adventure.PaperAdventure;
+import net.kyori.adventure.text.Component;
+import net.kyori.adventure.text.minimessage.MiniMessage;
+import net.minecraft.core.registries.BuiltInRegistries;
+import net.minecraft.resources.ResourceLocation;
+import net.minecraft.server.Main;
+import net.minecraft.world.entity.EquipmentSlot;
+import net.minecraft.world.entity.ai.attributes.AttributeModifier;
+import net.minecraft.world.item.Item;
+import net.minecraft.world.item.ItemStack;
+import org.bukkit.Bukkit;
+import org.bukkit.Material;
+import org.bukkit.configuration.InvalidConfigurationException;
+import org.bukkit.configuration.file.YamlConfiguration;
+import org.bukkit.craftbukkit.util.CraftMagicNumbers;
+import org.bukkit.plugin.Plugin;
+import org.checkerframework.checker.nullness.qual.Nullable;
+import org.jetbrains.annotations.NotNull;
+
+import java.io.File;
+import java.io.IOException;
+import java.lang.reflect.InvocationTargetException;
+import java.util.List;
+import java.util.Map;
+import java.util.logging.Level;
+
+public class HItem extends YamlConfiguration {
+
+    private final MiniMessage mm = MiniMessage.miniMessage();
+
+    private ResourceLocation id;
+    private Item baseItem;
+    private HItemBehaviour behaviour;
+    private Component displayName;
+    private List<Component> lore;
+    private String mojangID;
+    private int modelData;
+    private Map<EquipmentSlot, AttributeModifier> modifierMap;
+    private int maxStackSize;
+    private int maxDurability;
+    private String classURL;
+    private Plugin plugin;
+    private Class<? extends HItemBehaviour> behaviourClass;
+    private ClassLoader classLoader;
+
+    private boolean isEnabled = true;
+    private HItemStack replaceWith = null;
+
+    public HItem(Plugin plugin) {
+    }
+
+    public HItem(File file) {
+        super();
+        try {
+            load(file);
+        } catch (IOException | InvalidConfigurationException e) {
+            Bukkit.getLogger().log(Level.WARNING, "[Hephaestus] Failed to load " + file.getName());
+            Bukkit.getLogger().log(Level.WARNING, e.getMessage());
+            throw new RuntimeException(e);
+        }
+    }
+
+    public ResourceLocation getId() {
+        return id;
+    }
+
+    public void setId(ResourceLocation location) {
+        this.id = location;
+    }
+
+    public Component getDisplayName() {
+        return displayName;
+    }
+
+    public List<Component> getLore() {
+        return lore;
+    }
+
+    public String getMojangID() {
+        return mojangID;
+    }
+
+    public void setMojangID(String mojangID) {
+        this.mojangID = mojangID;
+    }
+
+    public Item getBaseItem() {
+        return baseItem;
+    }
+
+    public void setClassURL(String url) {
+        this.classURL = url;
+    }
+
+    public int getModelData() {
+        return modelData;
+    }
+
+    public Map<EquipmentSlot, AttributeModifier> getModifierMap() {
+        return modifierMap;
+    }
+
+    public int getMaxStackSize() {
+        return maxStackSize;
+    }
+
+    public int getMaxDurability() {
+        return maxDurability;
+    }
+
+    public boolean isEnabled() {
+        return isEnabled;
+    }
+
+    public HItemStack getReplaceWith() {
+        return replaceWith;
+    }
+    public @Nullable String getClassURL() {
+        return classURL;
+    }
+
+    public void setPlugin(Plugin plugin) {
+        this.plugin = plugin;
+    }
+
+    public HItemStack getItem() {
+        return getItem(1);
+    }
+
+    public HItemStack getItem(int count) {
+        HItemStack item = prepareNewStack().hItemStack;
+        if (behaviour != null) {
+            item.setBehaviour(behaviour);
+        }
+        item.getStack().setCount(count);
+        return item;
+    }
+
+    private ItemStack prepareNewStack() {
+        ItemStack baseItem = new ItemStack(BuiltInRegistries.ITEM.get(new ResourceLocation(getMojangID())));
+        for (EquipmentSlot slot : EquipmentSlot.values()) {
+            baseItem.getAttributeModifiers(slot).clear();
+        }
+        baseItem.getTag().putInt("CustomModelData", getModelData());
+        baseItem.getTag().putString("HephaestusID", id.toString());
+        baseItem.setHoverName(PaperAdventure.asVanilla(displayName));
+        baseItem.getBukkitStack().editMeta(m -> m.lore(lore)); // Annoying but oh well.
+        return baseItem;
+    }
+
+    private void loadBehaviour() {
+        if (plugin == null) {
+            throw new RuntimeException("Tried loading behaviours for item without plugin: " + getId().toString());
+        }
+        classLoader = plugin.getClass().getClassLoader();
+        try {
+            behaviourClass = ((Class<? extends HItemBehaviour>) Class.forName(getClassURL(), true, classLoader));
+        } catch (ClassNotFoundException e) {
+            throw new RuntimeException(e);
+        }
+        try {
+            behaviour = behaviourClass.getDeclaredConstructor(HItemStack.class).newInstance(this);
+        } catch (InstantiationException | IllegalAccessException | InvocationTargetException | NoSuchMethodException e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    @Override
+    public void load(@NotNull File file) throws IOException, InvalidConfigurationException {
+        super.load(file);
+        id = ResourceLocation.tryParse(getString("id", "minecraft:stone"));
+        if (contains("class")) {
+            classURL = getString("class");
+            loadBehaviour();
+        }
+        isEnabled = getBoolean("enabled", true);
+        displayName = mm.deserialize(getString("displayName", ""));
+        for (String string : getStringList("lore")) {
+            lore.add(mm.deserialize(string));
+        }
+        mojangID = getString("mojangID", "minecraft:stone");
+        baseItem = BuiltInRegistries.ITEM.get(new ResourceLocation(mojangID));
+        modelData = getInt("modelData", 0);
+        maxStackSize = getInt("maxStackSize", 1);
+        maxDurability = getInt("maxDurability", 0);
+        if (contains("modifiers")) {
+            for (String slot : getConfigurationSection("modifiers").getKeys(false)) {
+                EquipmentSlot equipmentSlot = EquipmentSlot.byName(slot);
+                AttributeModifier modifier = new AttributeModifier(
+                    getString("modifiers." + slot + ".name", "HephaestusModifier"),
+                    getDouble("modifiers." + slot + ".amount", 0),
+                    AttributeModifier.Operation.valueOf(getString("modifiers." + slot + ".operation", "ADD_NUMBER")));
+                modifierMap.put(equipmentSlot, modifier);
+            }
+        }
+    }
+
+
+    public void save(File file) {
+        set("id", id);
+        set("class", classURL);
+        set("enabled", isEnabled);
+        set("mojangID", mojangID);
+        save(file);
+    }
+
+    public static class Builder {
+        private final HItem item;
+        private final Plugin plugin;
+        private HItemBehaviour behaviour;
+
+        public Builder(Plugin plugin, ResourceLocation location) {
+            this.item = new HItem(plugin);
+            this.plugin = plugin;
+            item.setId(location);
+        }
+
+        public Builder behaviour(HItemBehaviour behaviour) {
+            this.behaviour = behaviour;
+            item.setClassURL(behaviour.getClass().getName());
+            return this;
+        }
+        public Builder baseItem(Material material) {
+            Item item1 = CraftMagicNumbers.getItem(material);
+            item.setMojangID(BuiltInRegistries.ITEM.getKey(item1).toString());
+            return this;
+        }
+
+        public HItem register() {
+            Main.itemLibrary.registerItem(plugin, item);
+            File folder = Main.itemLibrary.getFolder();
+            File saveFile = new File(folder, item.getId().getNamespace() + "/" + item.getId().getPath() + ".yml");
+            item.save(saveFile);
+            return item;
+        }
+    }
+
+}
diff --git a/src/main/java/de/erethon/hephaestus/HItemBehaviour.java b/src/main/java/de/erethon/hephaestus/HItemBehaviour.java
new file mode 100644
index 0000000000000000000000000000000000000000..df678ecffd9a70a44107024df0699e38e2111cb9
--- /dev/null
+++ b/src/main/java/de/erethon/hephaestus/HItemBehaviour.java
@@ -0,0 +1,82 @@
+package de.erethon.hephaestus;
+
+import com.destroystokyo.paper.event.player.PlayerArmorChangeEvent;
+import de.erethon.spellbook.api.SpellCastEvent;
+import net.minecraft.core.BlockPos;
+import net.minecraft.world.InteractionHand;
+import net.minecraft.world.InteractionResult;
+import net.minecraft.world.entity.Entity;
+import net.minecraft.world.entity.LivingEntity;
+import net.minecraft.world.entity.player.Player;
+import net.minecraft.world.item.ItemStack;
+import net.minecraft.world.level.Level;
+import net.minecraft.world.level.block.state.BlockState;
+import org.bukkit.event.block.BlockDropItemEvent;
+import org.bukkit.event.entity.EntityDamageByEntityEvent;
+import org.bukkit.event.entity.EntityDropItemEvent;
+import org.bukkit.event.entity.EntityPickupItemEvent;
+import org.bukkit.event.inventory.ClickType;
+import org.bukkit.event.inventory.InventoryAction;
+import org.bukkit.event.inventory.InventoryClickEvent;
+import org.bukkit.event.player.PlayerAttemptPickupItemEvent;
+import org.bukkit.event.player.PlayerDropItemEvent;
+import org.bukkit.event.player.PlayerInteractEvent;
+
+public class HItemBehaviour {
+
+    public void onLeftClick(ItemStack stack, org.bukkit.entity.Player player, PlayerInteractEvent event) {
+    }
+
+    public void onRightClick(ItemStack stack, org.bukkit.entity.Player player, PlayerInteractEvent event) {
+    }
+
+    public void onPlayerDrop(ItemStack stack, org.bukkit.entity.Player player, PlayerDropItemEvent event) {
+    }
+
+    public void onEntityDrop(ItemStack stack, org.bukkit.entity.Entity entity, EntityDropItemEvent event) {
+    }
+
+    public void onBlockDrop(ItemStack stack, org.bukkit.entity.Player player, BlockDropItemEvent event) {
+    }
+
+    public void onPickup(ItemStack stack, org.bukkit.entity.Player player, PlayerAttemptPickupItemEvent event) {
+    }
+
+    public void onEntityPickup(ItemStack stack, org.bukkit.entity.Entity entity, EntityPickupItemEvent event) {
+    }
+
+    public void onEquip(ItemStack stack, org.bukkit.entity.Player player, PlayerArmorChangeEvent event) {
+    }
+
+    public void onUnequip(ItemStack stack, org.bukkit.entity.Player player, PlayerArmorChangeEvent event) {
+    }
+
+    public boolean onAttack(ItemStack stack, LivingEntity target, Player attacker) {
+        return true;
+    }
+
+    public void onDamaged(ItemStack stack, org.bukkit.entity.LivingEntity damaged, EntityDamageByEntityEvent event) {
+    }
+
+    public int onItemHurt(ItemStack stack, int amount) {
+        return amount;
+    }
+
+    public boolean onMine(ItemStack stack, Level world, BlockState state, BlockPos pos, Player miner) {
+        return true;
+    }
+
+    public InteractionResult onInteractLiving(ItemStack stack, Player player, LivingEntity living, InteractionHand hand) {
+        return stack.getItem().interactLivingEntity(stack, player, living, hand);
+    }
+
+    public void onSpellCast(ItemStack stack, org.bukkit.entity.LivingEntity damaged, SpellCastEvent event) {
+    }
+
+    public void onTick(ItemStack stack, Entity entity, int slot, boolean selected) {
+    }
+
+    public void onInventoryClick(ItemStack stack, Player player, ClickType type, InventoryAction action, InventoryClickEvent event) {
+    }
+
+}
diff --git a/src/main/java/de/erethon/hephaestus/HItemStack.java b/src/main/java/de/erethon/hephaestus/HItemStack.java
new file mode 100644
index 0000000000000000000000000000000000000000..0eddb94a5f24089091e9dee4690d29a099e19182
--- /dev/null
+++ b/src/main/java/de/erethon/hephaestus/HItemStack.java
@@ -0,0 +1,63 @@
+package de.erethon.hephaestus;
+
+import net.minecraft.resources.ResourceLocation;
+import net.minecraft.world.item.ItemStack;
+
+public class HItemStack {
+
+    private HItem item;
+    private HItemBehaviour behaviour;
+    private ItemStack stack;
+
+    public HItemStack(ItemStack stack, HItem config) {
+        this.stack = stack;
+        this.item = config;
+    }
+
+    public ResourceLocation getId() {
+        return item.getId();
+    }
+
+    public boolean needsUpdating(ItemStack stack) {
+        return !ItemStack.matches(stack, this.stack);
+    }
+
+    public ItemStack getStack() {
+        return stack;
+    }
+
+    public org.bukkit.inventory.ItemStack getBukkitStack() {
+        return stack.getBukkitStack();
+    }
+
+    public void setBehaviour(HItemBehaviour behaviour) {
+        this.behaviour = behaviour;
+    }
+
+    public HItemBehaviour getBehaviour() {
+        return behaviour;
+    }
+
+    public HItem config() {
+        return item;
+    }
+
+    public boolean isEnabled() {
+        return item.isEnabled();
+    }
+
+    public boolean isVanilla() {
+        return item.getId().getNamespace().equals("minecraft");
+    }
+
+    public HItemStack replaceWith() {
+        return item.getReplaceWith();
+    }
+
+    public void setStack(ItemStack stack) {
+        this.stack = stack;
+    }
+
+    public void validate() {
+    }
+}
diff --git a/src/main/java/de/erethon/hephaestus/ItemLibrary.java b/src/main/java/de/erethon/hephaestus/ItemLibrary.java
new file mode 100644
index 0000000000000000000000000000000000000000..acb455c1e689ff4700fed4062113691e269d8e76
--- /dev/null
+++ b/src/main/java/de/erethon/hephaestus/ItemLibrary.java
@@ -0,0 +1,145 @@
+package de.erethon.hephaestus;
+
+import de.erethon.spellbook.api.SpellLibrary;
+import net.minecraft.core.registries.BuiltInRegistries;
+import net.minecraft.resources.ResourceLocation;
+import net.minecraft.world.item.Item;
+import org.bukkit.Bukkit;
+import org.bukkit.Material;
+import org.bukkit.configuration.file.YamlConfiguration;
+import org.bukkit.craftbukkit.inventory.CraftItemStack;
+import org.bukkit.craftbukkit.util.CraftMagicNumbers;
+import org.bukkit.inventory.ItemStack;
+import org.bukkit.plugin.Plugin;
+import org.bukkit.plugin.java.JavaPlugin;
+import org.jetbrains.annotations.Nullable;
+
+import java.io.File;
+import java.io.IOException;
+import java.util.HashMap;
+import java.util.logging.Level;
+
+import static net.minecraft.server.MinecraftServer.LOGGER;
+
+public class ItemLibrary {
+
+    private final File folder;
+
+    private final HashMap<ResourceLocation, HItem> items = new HashMap<>();
+    private final HashMap<Material, HItem> vanillaMaterials = new HashMap<>();
+    private HBukkitHandler bukkitHandler = null;
+
+    public ItemLibrary(File folder) {
+        this.folder = new File(folder + "/plugins/Hephaestus/items");
+        if (!folder.exists()) {
+            folder.mkdirs();
+        }
+        generateVanillaItemFiles();
+        loadItems();
+    }
+
+    @Nullable
+    public HItem get(ResourceLocation key) {
+        return items.get(key);
+    }
+
+    @Nullable
+    public HItem get(Material material) {
+        return vanillaMaterials.get(material);
+    }
+
+    @Nullable
+    public HItem get(ItemStack bukkitStack) {
+        CraftItemStack craftItemStack = (CraftItemStack) bukkitStack;
+        if (craftItemStack.handle.hItemStack == null) {
+            return null;
+        }
+        return craftItemStack.handle.hItemStack.config();
+    }
+
+    @Nullable
+    public HItem get(net.minecraft.world.item.ItemStack stack) {
+        return stack.hItemStack.config();
+    }
+
+    public boolean registerItem(Plugin plugin, HItem item) {
+        if (bukkitHandler == null) { // Use the first plugin we get to register the bukkit handler, because we need a Plugin for listeners.
+            bukkitHandler = new HBukkitHandler((JavaPlugin) plugin, this);
+        }
+        if (item.getId().getNamespace().equals("minecraft")) {
+            throw new RuntimeException(plugin.getName() + " tried registering a custom item in the vanilla namespace: " + item.getId().toString());
+        }
+        item.setPlugin(plugin);
+        if (items.containsKey(item.getId())) {
+            Bukkit.getLogger().log(Level.WARNING, "[Hephaestus] Item with ID " + item.getId() + " is already registered.");
+            return false;
+        }
+        items.put(item.getId(), item);
+        return true;
+    }
+
+    public boolean overrideVanillaBehaviour(Plugin plugin, HItem item) {
+        if (bukkitHandler == null) {
+            bukkitHandler = new HBukkitHandler((JavaPlugin) plugin, this);
+        }
+        item.setPlugin(plugin);
+        return true;
+    }
+
+
+    public boolean unregisterItem(HItem item) {
+        if (!items.containsKey(item.getId())) {
+            Bukkit.getLogger().log(Level.WARNING, "[Hephaestus] Item with ID " + item.getId() + " is not registered.");
+            return false;
+        }
+        if (item.getId().getNamespace().equals("minecraft")) {
+            throw new RuntimeException("Vanilla items can not be unregistered.");
+        }
+        items.remove(item.getId());
+        return true;
+    }
+
+    private void generateVanillaItemFiles() {
+        for (Item item : BuiltInRegistries.ITEM.stream().toList()) {
+            File vanillaFolder = new File(folder, "vanilla");
+            if (!vanillaFolder.exists()) {
+                vanillaFolder.mkdirs();
+            }
+            YamlConfiguration configuration = new YamlConfiguration();
+            ResourceLocation location = BuiltInRegistries.ITEM.getKey(item);
+            configuration.set("id", location.toString());
+            configuration.set("mojangID", location.toString());
+            configuration.set("maxStackSize", item.getMaxStackSize());
+            configuration.set("maxDurability", item.getMaxDamage());
+            File file = new File(vanillaFolder, location.getPath() + ".yml");
+            if (file.exists()) {
+                return;
+            }
+            try {
+                configuration.save(file);
+            } catch (IOException e) {
+                throw new RuntimeException(e);
+            }
+        }
+    }
+
+    private void loadItems() {
+        for (File file : SpellLibrary.getFilesForFolder(folder)) {
+            if (!file.getName().contains(".yml")) {
+                continue;
+            }
+            HItem config = new HItem(file);
+            items.put(config.getId(), config);
+            vanillaMaterials.put(CraftMagicNumbers.getMaterial(config.getBaseItem()), config);
+        }
+        LOGGER.info("[Hephaestus] Loaded " + items.keySet().size() + " items.");
+    }
+
+    public File getFolder() {
+        return folder;
+    }
+
+    public static HItemStack getHItemFromBukkit(ItemStack stack) {
+        return ((CraftItemStack) stack).handle.hItemStack;
+     }
+}
diff --git a/src/main/java/de/erethon/papyrus/ContainerLoadEvent.java b/src/main/java/de/erethon/papyrus/ContainerLoadEvent.java
index e672340310a8863a627c083cbea2c08a8d6049b9..b6e1158585a4a8971a2fcc5fcea55e9d467b03e6 100644
--- a/src/main/java/de/erethon/papyrus/ContainerLoadEvent.java
+++ b/src/main/java/de/erethon/papyrus/ContainerLoadEvent.java
@@ -12,7 +12,12 @@ public class ContainerLoadEvent extends Event {
 
     private NonNullList<ItemStack> itemStacks;
 
-    public ContainerLoadEvent(NonNullList<ItemStack> items) {
+    /**
+     * Note: This event may be triggered async during feature placement.
+     * This shouldn't happen on Erethon, but who knows.
+     */
+    public ContainerLoadEvent(NonNullList<ItemStack> items, boolean async) {
+        super(async);
         this.itemStacks = items;
     }
 
diff --git a/src/main/java/net/minecraft/server/Bootstrap.java b/src/main/java/net/minecraft/server/Bootstrap.java
index 5816998cde8504d58732e63eb2179a5d828f35bb..96ac8c11b80abeaee1a218fc4ab53571f8eec995 100644
--- a/src/main/java/net/minecraft/server/Bootstrap.java
+++ b/src/main/java/net/minecraft/server/Bootstrap.java
@@ -1,7 +1,10 @@
 package net.minecraft.server;
 
 import com.mojang.logging.LogUtils;
+
+import java.io.File;
 import java.io.PrintStream;
+import java.nio.file.Paths;
 import java.time.Duration;
 import java.time.Instant;
 import java.util.Set;
@@ -9,6 +12,8 @@ import java.util.TreeSet;
 import java.util.concurrent.atomic.AtomicLong;
 import java.util.function.Function;
 import java.util.function.Supplier;
+
+import de.erethon.hephaestus.ItemLibrary;
 import net.minecraft.SharedConstants;
 import net.minecraft.commands.Commands;
 import net.minecraft.commands.arguments.selector.options.EntitySelectorOptions;
@@ -69,6 +74,7 @@ public class Bootstrap {
             } else {
                 FireBlock.bootStrap();
                 ComposterBlock.bootStrap();
+                Main.itemLibrary = new ItemLibrary(new File(Paths.get("").toUri())); // Papyrus
                 if (EntityType.getKey(EntityType.PLAYER) == null) {
                     throw new IllegalStateException("Failed loading EntityTypes");
                 } else {
diff --git a/src/main/java/net/minecraft/server/Main.java b/src/main/java/net/minecraft/server/Main.java
index 7573c12a77797146c51ef2dfe4b2a636df45e21a..e86db50efb6196db4ecae1d8b38b18cc1ae1507d 100644
--- a/src/main/java/net/minecraft/server/Main.java
+++ b/src/main/java/net/minecraft/server/Main.java
@@ -16,6 +16,8 @@ import java.nio.file.Paths;
 import java.util.Optional;
 import java.util.function.BooleanSupplier;
 import javax.annotation.Nullable;
+
+import de.erethon.hephaestus.ItemLibrary;
 import joptsimple.NonOptionArgumentSpec;
 import joptsimple.OptionParser;
 import joptsimple.OptionSet;
@@ -71,6 +73,7 @@ import org.bukkit.configuration.file.YamlConfiguration;
 public class Main {
 
     private static final Logger LOGGER = LogUtils.getLogger();
+    public static ItemLibrary itemLibrary; // Papyrus
 
     public Main() {}
 
diff --git a/src/main/java/net/minecraft/server/MinecraftServer.java b/src/main/java/net/minecraft/server/MinecraftServer.java
index c535eb33ef8e820d11d75702059613043982deee..3dac663daf000e443b141d11c1432ec2524d33de 100644
--- a/src/main/java/net/minecraft/server/MinecraftServer.java
+++ b/src/main/java/net/minecraft/server/MinecraftServer.java
@@ -14,6 +14,7 @@ import com.mojang.authlib.GameProfileRepository;
 import com.mojang.authlib.minecraft.MinecraftSessionService;
 import com.mojang.datafixers.DataFixer;
 import com.mojang.logging.LogUtils;
+import de.erethon.hephaestus.ItemLibrary;
 import it.unimi.dsi.fastutil.longs.LongIterator;
 import it.unimi.dsi.fastutil.objects.ObjectArrayList;
 import java.awt.image.BufferedImage;
diff --git a/src/main/java/net/minecraft/world/ContainerHelper.java b/src/main/java/net/minecraft/world/ContainerHelper.java
index 01063c3bc672eecdf8fe1c5b5a89fbf576c940f3..949207f366aa0659c563fa123047112d30b6a4ec 100644
--- a/src/main/java/net/minecraft/world/ContainerHelper.java
+++ b/src/main/java/net/minecraft/world/ContainerHelper.java
@@ -53,8 +53,13 @@ public class ContainerHelper {
                 stacks.set(j, ItemStack.of(compoundTag));
             }
         }
-        ContainerLoadEvent loadEvent = new ContainerLoadEvent(stacks);
-        Bukkit.getPluginManager().callEvent(loadEvent);
+        ContainerLoadEvent loadEvent = new ContainerLoadEvent(stacks, false);
+        try {
+            Bukkit.getPluginManager().callEvent(loadEvent);
+        } catch (IllegalStateException e) { // This really only happens during feature placement which shouldn't happen in production anyways
+            loadEvent = new ContainerLoadEvent(stacks, true);
+            Bukkit.getPluginManager().callEvent(loadEvent);
+        }
 
     }
 
diff --git a/src/main/java/net/minecraft/world/item/ItemStack.java b/src/main/java/net/minecraft/world/item/ItemStack.java
index 4ab0452af6eeae9316311d473be23de9a74d9e4b..2eff89662e9fb4ac5f99686be69356ad47b7cdca 100644
--- a/src/main/java/net/minecraft/world/item/ItemStack.java
+++ b/src/main/java/net/minecraft/world/item/ItemStack.java
@@ -23,12 +23,15 @@ import java.util.function.Predicate;
 import java.util.stream.Collectors;
 import java.util.stream.Stream;
 import javax.annotation.Nullable;
+
+import de.erethon.hephaestus.HItemStack;
+import de.erethon.hephaestus.HItemBehaviour;
+import de.erethon.hephaestus.HItem;
 import net.minecraft.ChatFormatting;
 import net.minecraft.Util;
 import net.minecraft.advancements.CriteriaTriggers;
 import net.minecraft.commands.arguments.blocks.BlockStateParser;
 import net.minecraft.core.BlockPos;
-import net.minecraft.core.Direction;
 import net.minecraft.core.Holder;
 import net.minecraft.core.HolderSet;
 import net.minecraft.core.Registry;
@@ -45,8 +48,8 @@ import net.minecraft.network.chat.ComponentUtils;
 import net.minecraft.network.chat.HoverEvent;
 import net.minecraft.network.chat.MutableComponent;
 import net.minecraft.network.chat.Style;
-import net.minecraft.network.protocol.game.ClientboundBlockUpdatePacket;
 import net.minecraft.resources.ResourceLocation;
+import net.minecraft.server.Main;
 import net.minecraft.tags.TagKey;
 import net.minecraft.util.ExtraCodecs;
 import net.minecraft.util.RandomSource;
@@ -84,7 +87,7 @@ import org.slf4j.Logger;
 // CraftBukkit start
 import com.mojang.serialization.Dynamic;
 import java.util.Map;
-import java.util.Objects;
+
 import net.minecraft.server.MinecraftServer;
 import net.minecraft.server.level.ServerLevel;
 import net.minecraft.server.level.ServerPlayer;
@@ -234,6 +237,9 @@ public final class ItemStack {
         return net.minecraft.nbt.StringTag.valueOf(org.bukkit.craftbukkit.util.CraftChatMessage.toJSON(component));
     }
     // Paper end
+    // Papyrus start
+    public HItemStack hItemStack;
+    public HItemBehaviour behaviour;
 
     public ItemStack(ItemLike item) {
         this(item, 1);
@@ -258,6 +264,16 @@ public final class ItemStack {
         if (this.item.canBeDepleted()) {
             this.setDamageValue(this.getDamageValue());
         }
+        HItem hConfig = Main.itemLibrary.get(BuiltInRegistries.ITEM.getKey(item.asItem()));
+        if (hConfig != null) {
+            hItemStack = new HItemStack(this, hConfig);
+            hItemStack.setStack(this);
+            if (hItemStack.getBehaviour() != null) {
+                behaviour = hItemStack.getBehaviour();
+            }
+        } else {
+            LOGGER.info("[Hephaestus] Unable to find vanilla hItem for " + item.asItem());
+        }
 
     }
 
@@ -289,7 +305,30 @@ public final class ItemStack {
         if (this.getItem().canBeDepleted()) {
             this.setDamageValue(this.getDamageValue());
         }
-
+        if (tag != null && tag.contains("HephaestusID")) { // Papyrus - Hephaestus
+            String id = tag.getString("HephaestusID");
+            HItem hConfig = Main.itemLibrary.get(new ResourceLocation(id));
+            if (hConfig != null) {
+                hItemStack = new HItemStack(this, hConfig);
+                hItemStack.setStack(this);
+                if (hItemStack.getBehaviour() != null) {
+                    behaviour = hItemStack.getBehaviour();
+                }
+            } else {
+                LOGGER.info("[Hephaestus] Unable to find hItem for " + id);
+            }
+        } else {
+            HItem hConfig = Main.itemLibrary.get(BuiltInRegistries.ITEM.getKey(item.asItem()));
+            if (hConfig != null) {
+                hItemStack = new HItemStack(this, hConfig);
+                hItemStack.setStack(this);
+                if (hItemStack.getBehaviour() != null) {
+                    behaviour = hItemStack.getBehaviour();
+                }
+            } else {
+                LOGGER.info("[Hephaestus] Unable to find vanilla hItem for " + item.asItem().toString());
+            }
+        }
     }
 
     private ItemStack(CompoundTag nbt) {
@@ -311,7 +350,7 @@ public final class ItemStack {
     }
 
     public boolean isItemEnabled(FeatureFlagSet enabledFeatures) {
-        return this.isEmpty() || this.getItem().isEnabled(enabledFeatures);
+        return this.isEmpty() || (this.getItem().isEnabled(enabledFeatures) && hItemStack.isEnabled()); // Papyrus
     }
 
     public ItemStack split(int amount) {
@@ -584,14 +623,19 @@ public final class ItemStack {
         nbt.putString("id", minecraftkey == null ? "minecraft:air" : minecraftkey.toString());
         nbt.putByte("Count", (byte) this.count);
         if (this.tag != null) {
+            if (hItemStack != null) { // Papyrus
+                tag.putString("HephaestusID", hItemStack.getId().toString());
+            }
             nbt.put("tag", this.tag.copy());
         }
-
         return nbt;
     }
 
     public int getMaxStackSize() {
-        return this.getItem().getMaxStackSize();
+        if (hItemStack == null) {
+            getItem().getMaxStackSize();
+        }
+        return hItemStack.config().getMaxStackSize(); // Papyrus
     }
 
     public boolean isStackable() {
@@ -621,7 +665,10 @@ public final class ItemStack {
     }
 
     public int getMaxDamage() {
-        return this.getItem().getMaxDamage();
+        if (hItemStack == null) {
+            return getItem().getMaxDamage();
+        }
+        return hItemStack.config().getMaxDurability();
     }
 
     public boolean hurt(int amount, RandomSource random, @Nullable LivingEntity player) { // Paper - allow any living entity instead of only ServerPlayers
@@ -675,6 +722,9 @@ public final class ItemStack {
             }
 
             j = this.getDamageValue() + amount;
+            if (behaviour != null) {
+                j = behaviour.onItemHurt(this, j);
+            }
             this.setDamageValue(j);
             return j >= this.getMaxDamage();
         }
@@ -727,7 +777,7 @@ public final class ItemStack {
     public void hurtEnemy(LivingEntity target, net.minecraft.world.entity.player.Player attacker) {
         Item item = this.getItem();
 
-        if (item.hurtEnemy(this, target, attacker)) {
+        if (item.hurtEnemy(this, target, attacker) && (behaviour != null && behaviour.onAttack(this, target, attacker)) ) {
             attacker.awardStat(Stats.ITEM_USED.get(item));
         }
 
@@ -736,7 +786,7 @@ public final class ItemStack {
     public void mineBlock(Level world, net.minecraft.world.level.block.state.BlockState state, BlockPos pos, net.minecraft.world.entity.player.Player miner) {
         Item item = this.getItem();
 
-        if (item.mineBlock(this, world, state, pos, miner)) {
+        if (item.mineBlock(this, world, state, pos, miner) && (behaviour != null && behaviour.onMine(this, world, state, pos, miner))) {
             miner.awardStat(Stats.ITEM_USED.get(item));
         }
 
@@ -747,6 +797,9 @@ public final class ItemStack {
     }
 
     public InteractionResult interactLivingEntity(net.minecraft.world.entity.player.Player user, LivingEntity entity, InteractionHand hand) {
+        if (behaviour != null) {
+            return behaviour.onInteractLiving(this, user, entity, hand);
+        }
         return this.getItem().interactLivingEntity(this, user, entity, hand);
     }
 
@@ -811,6 +864,9 @@ public final class ItemStack {
 
         if (this.getItem() != null) {
             this.getItem().inventoryTick(this, world, entity, slot, selected);
+            if (behaviour != null) {
+                behaviour.onTick(this, entity, slot, selected);
+            }
         }
 
     }
diff --git a/src/main/java/net/minecraft/world/level/block/Block.java b/src/main/java/net/minecraft/world/level/block/Block.java
index e21867d3956078bb0db4ceed45e5811e9acd7377..9e1ae79ad84045933e435cc5cb08e7137dfa0790 100644
--- a/src/main/java/net/minecraft/world/level/block/Block.java
+++ b/src/main/java/net/minecraft/world/level/block/Block.java
@@ -304,7 +304,7 @@ public class Block extends BlockBehaviour implements ItemLike {
     public static List<ItemStack> getDrops(BlockState state, ServerLevel world, BlockPos pos, @Nullable BlockEntity blockEntity, @Nullable Entity entity, ItemStack stack) {
         LootParams.Builder lootparams_a = (new LootParams.Builder(world)).withParameter(LootContextParams.ORIGIN, Vec3.atCenterOf(pos)).withParameter(LootContextParams.TOOL, stack).withOptionalParameter(LootContextParams.THIS_ENTITY, entity).withOptionalParameter(LootContextParams.BLOCK_ENTITY, blockEntity);
 
-        return state.getDrops(lootparams_a);
+        return state.   getDrops(lootparams_a);
     }
 
     public static void dropResources(BlockState state, Level world, BlockPos pos) {
diff --git a/src/main/java/org/bukkit/craftbukkit/CraftServer.java b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
index c0cb4358e013372aa880aac9e40639f865246e6b..07a373f7ccbe298e579e7db268b1662a0df5d393 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftServer.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
@@ -14,6 +14,7 @@ import com.mojang.brigadier.tree.CommandNode;
 import com.mojang.brigadier.tree.LiteralCommandNode;
 import com.mojang.serialization.Dynamic;
 import com.mojang.serialization.Lifecycle;
+import de.erethon.hephaestus.ItemLibrary;
 import de.erethon.spellbook.api.SpellbookAPI;
 import it.unimi.dsi.fastutil.objects.Object2IntOpenHashMap;
 import java.awt.image.BufferedImage;
