From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Malfrador <malfrador@gmail.com>
Date: Fri, 12 Jan 2024 13:22:05 +0100
Subject: [PATCH] Hephaestus


diff --git a/src/main/java/de/erethon/hephaestus/HBukkitHandler.java b/src/main/java/de/erethon/hephaestus/HBukkitHandler.java
new file mode 100644
index 0000000000000000000000000000000000000000..b30d9ecee3736c1311893bde357b9dc0a2a485b8
--- /dev/null
+++ b/src/main/java/de/erethon/hephaestus/HBukkitHandler.java
@@ -0,0 +1,189 @@
+package de.erethon.hephaestus;
+
+import com.destroystokyo.paper.event.player.PlayerArmorChangeEvent;
+import de.erethon.spellbook.api.SpellCastEvent;
+import org.bukkit.Bukkit;
+import org.bukkit.Material;
+import org.bukkit.craftbukkit.inventory.CraftItemStack;
+import org.bukkit.entity.Item;
+import org.bukkit.entity.LivingEntity;
+import org.bukkit.entity.Player;
+import org.bukkit.event.EventHandler;
+import org.bukkit.event.Listener;
+import org.bukkit.event.block.BlockDropItemEvent;
+import org.bukkit.event.entity.EntityDamageByEntityEvent;
+import org.bukkit.event.entity.EntityDropItemEvent;
+import org.bukkit.event.entity.EntityPickupItemEvent;
+import org.bukkit.event.inventory.InventoryClickEvent;
+import org.bukkit.event.player.PlayerAttemptPickupItemEvent;
+import org.bukkit.event.player.PlayerDropItemEvent;
+import org.bukkit.event.player.PlayerInteractEvent;
+import org.bukkit.event.server.ServerLoadEvent;
+import org.bukkit.inventory.ItemStack;
+import org.bukkit.plugin.Plugin;
+import org.bukkit.plugin.java.JavaPlugin;
+
+import static net.minecraft.server.MinecraftServer.LOGGER;
+
+public class HBukkitHandler implements Listener {
+
+    private final Plugin plugin;
+    private final ItemLibrary lib;
+
+    public HBukkitHandler(JavaPlugin pl, ItemLibrary library) {
+        this.lib = library;
+        this.plugin = pl;
+        Bukkit.getPluginManager().registerEvents(this, plugin);
+        LOGGER.info("Initialised HBukkitHandler using " + plugin.getName());
+    }
+
+    @EventHandler
+    private void onInvClick(InventoryClickEvent event) {
+        if (event.getCurrentItem() == null) {
+            return;
+        }
+        HItemStack item = ItemLibrary.getHItemFromBukkit(event.getCurrentItem());
+        if (item != null && item.getBehaviour() != null) {
+            item.getBehaviour().onInventoryClick(CraftItemStack.unwrap(event.getCurrentItem()), (Player) event.getWhoClicked(), event.getClick(), event.getAction(), event);
+        }
+    }
+
+    @EventHandler
+    private void onDamage(EntityDamageByEntityEvent event) {
+        if (event.getEntity() instanceof LivingEntity living && living.getEquipment() != null) {
+            ItemStack mainHand = living.getEquipment().getItemInMainHand();
+            HItemStack item = ItemLibrary.getHItemFromBukkit(mainHand);
+            if (item != null && item.getBehaviour() != null) {
+                item.getBehaviour().onDamaged(CraftItemStack.unwrap(mainHand), living, event);
+            }
+            ItemStack offHand = living.getEquipment().getItemInOffHand();
+            item = ItemLibrary.getHItemFromBukkit(mainHand);
+            if (item != null && item.getBehaviour() != null) {
+                item.getBehaviour().onDamaged(CraftItemStack.unwrap(offHand), living, event);
+            }
+            for (ItemStack armor : living.getEquipment().getArmorContents()) {
+                if (armor == null) {
+                    continue;
+                }
+                item = ItemLibrary.getHItemFromBukkit(armor);
+                if (item != null && item.getBehaviour() != null) {
+                    item.getBehaviour().onDamaged(CraftItemStack.unwrap(armor), living, event);
+                }
+            }
+        }
+    }
+
+    @EventHandler
+    private void onDamage(SpellCastEvent event) {
+        LivingEntity living = event.getCaster();
+        if (living.getEquipment() != null) {
+            ItemStack mainHand = living.getEquipment().getItemInMainHand();
+            HItemStack item = ItemLibrary.getHItemFromBukkit(mainHand);
+            if (item != null && item.getBehaviour() != null) {
+                item.getBehaviour().onSpellCast(CraftItemStack.unwrap(mainHand), living, event);
+            }
+            ItemStack offHand = living.getEquipment().getItemInOffHand();
+            item = ItemLibrary.getHItemFromBukkit(mainHand);
+            if (item != null && item.getBehaviour() != null) {
+                item.getBehaviour().onSpellCast(CraftItemStack.unwrap(offHand), living, event);
+            }
+            for (ItemStack armor : living.getEquipment().getArmorContents()) {
+                if (armor == null) {
+                    continue;
+                }
+                item = ItemLibrary.getHItemFromBukkit(armor);
+                if (item != null && item.getBehaviour() != null) {
+                    item.getBehaviour().onSpellCast(CraftItemStack.unwrap(armor), living, event);
+                }
+            }
+        }
+    }
+
+    @EventHandler
+    private void onPlayerDrop(PlayerDropItemEvent event) {
+        ItemStack stack = event.getItemDrop().getItemStack();
+        HItemStack item = ItemLibrary.getHItemFromBukkit(stack);
+        if (item != null && item.getBehaviour() != null) {
+            item.getBehaviour().onPlayerDrop(CraftItemStack.unwrap(stack), event.getPlayer(), event);
+        }
+    }
+
+    @EventHandler
+    private void onEntityDrop(EntityDropItemEvent event) {
+        ItemStack stack = event.getItemDrop().getItemStack();
+        HItemStack item = ItemLibrary.getHItemFromBukkit(stack);
+        if (item != null && item.getBehaviour() != null) {
+            item.getBehaviour().onEntityDrop(CraftItemStack.unwrap(stack), event.getEntity(), event);
+        }
+    }
+
+    @EventHandler
+    private void onBlockDrop(BlockDropItemEvent event) {
+        for (Item stack : event.getItems()) {
+            HItemStack item = ItemLibrary.getHItemFromBukkit(stack.getItemStack());
+            if (item != null && item.getBehaviour() != null) {
+                item.getBehaviour().onBlockDrop(CraftItemStack.unwrap(stack.getItemStack()), event.getPlayer(), event);
+            }
+        }
+    }
+
+    @EventHandler
+    private void onPickup(PlayerAttemptPickupItemEvent event) {
+        ItemStack stack = event.getItem().getItemStack();
+        HItemStack item = ItemLibrary.getHItemFromBukkit(stack);
+        if (item != null && item.getBehaviour() != null) {
+            item.getBehaviour().onPickup(CraftItemStack.unwrap(stack), event.getPlayer(), event);
+        }
+    }
+
+    @EventHandler
+    private void onEntityPickup(EntityPickupItemEvent event) {
+        ItemStack stack = event.getItem().getItemStack();
+        HItemStack item = ItemLibrary.getHItemFromBukkit(stack);
+        if (item != null && item.getBehaviour() != null) {
+            item.getBehaviour().onEntityPickup(CraftItemStack.unwrap(stack), event.getEntity(), event);
+        }
+    }
+
+    @EventHandler
+    private void onInteract(PlayerInteractEvent event) {
+        if (!event.hasItem()) {
+            return;
+        }
+        HItemStack item = ItemLibrary.getHItemFromBukkit(event.getItem());
+        if (event.getAction().isRightClick()) {
+            if (item != null && item.getBehaviour() != null) {
+                item.getBehaviour().onRightClick(CraftItemStack.unwrap(event.getItem()), event.getPlayer(), event);
+            }
+        }
+        if (event.getAction().isLeftClick()) {
+            if (item != null && item.getBehaviour() != null) {
+                item.getBehaviour().onLeftClick(CraftItemStack.unwrap(event.getItem()), event.getPlayer(), event);
+            }
+        }
+    }
+
+    @EventHandler
+    private void onEquip(PlayerArmorChangeEvent event) {
+        if (event.getNewItem() == event.getOldItem()) {
+            return;
+        }
+        ItemStack stack = event.getNewItem();
+        HItemStack item = ItemLibrary.getHItemFromBukkit(stack);
+        if (item != null && item.getBehaviour() != null && stack.getType() != Material.AIR) {
+            item.getBehaviour().onEquip(CraftItemStack.unwrap(stack), event.getPlayer(), event);
+            return;
+        }
+        stack = event.getOldItem();
+        item = ItemLibrary.getHItemFromBukkit(stack);
+        if (item != null && item.getBehaviour() != null) {
+            item.getBehaviour().onUnequip(CraftItemStack.unwrap(stack), event.getPlayer(), event);
+        }
+    }
+
+    @EventHandler
+    private void onServerStartFinish(ServerLoadEvent event) {
+        lib.readyBehaviours();
+    }
+
+}
diff --git a/src/main/java/de/erethon/hephaestus/HItem.java b/src/main/java/de/erethon/hephaestus/HItem.java
new file mode 100644
index 0000000000000000000000000000000000000000..62cc2fc771aec42e9203e831d5a30e5da8b9c206
--- /dev/null
+++ b/src/main/java/de/erethon/hephaestus/HItem.java
@@ -0,0 +1,280 @@
+package de.erethon.hephaestus;
+
+import io.papermc.paper.adventure.PaperAdventure;
+import net.kyori.adventure.text.Component;
+import net.kyori.adventure.text.minimessage.MiniMessage;
+import net.minecraft.core.registries.BuiltInRegistries;
+import net.minecraft.resources.ResourceLocation;
+import net.minecraft.server.Main;
+import net.minecraft.world.entity.EquipmentSlot;
+import net.minecraft.world.entity.ai.attributes.AttributeModifier;
+import net.minecraft.world.item.Item;
+import net.minecraft.world.item.ItemStack;
+import org.bukkit.Bukkit;
+import org.bukkit.Material;
+import org.bukkit.configuration.InvalidConfigurationException;
+import org.bukkit.configuration.file.YamlConfiguration;
+import org.bukkit.craftbukkit.util.CraftMagicNumbers;
+import org.bukkit.plugin.Plugin;
+import org.checkerframework.checker.nullness.qual.Nullable;
+import org.jetbrains.annotations.NotNull;
+
+import java.io.File;
+import java.io.IOException;
+import java.lang.reflect.InvocationTargetException;
+import java.util.List;
+import java.util.Map;
+import java.util.logging.Level;
+
+import static net.minecraft.server.MinecraftServer.LOGGER;
+
+public class HItem extends YamlConfiguration {
+
+    private final MiniMessage mm = MiniMessage.miniMessage();
+
+    private ResourceLocation id;
+    private Item baseItem;
+    private HItemBehaviour behaviour;
+    private Component displayName;
+    private List<Component> lore;
+    private String mojangID;
+    private int modelData;
+    private Map<EquipmentSlot, AttributeModifier> modifierMap;
+    private int maxStackSize;
+    private int maxDurability;
+    private String classURL;
+    private boolean hasBehaviour = false;
+    private String pluginName;
+    private Plugin plugin;
+    private Class<? extends HItemBehaviour> behaviourClass;
+    private ClassLoader classLoader;
+
+    private boolean isEnabled = true;
+    private HItemStack replaceWith = null;
+
+    public HItem(Plugin plugin) {
+    }
+
+    public HItem(File file) {
+        super();
+        try {
+            load(file);
+        } catch (IOException | InvalidConfigurationException e) {
+            Bukkit.getLogger().log(Level.WARNING, "[Hephaestus] Failed to load " + file.getName());
+            Bukkit.getLogger().log(Level.WARNING, e.getMessage());
+            throw new RuntimeException(e);
+        }
+    }
+
+    public ResourceLocation getId() {
+        return id;
+    }
+
+    public void setId(ResourceLocation location) {
+        this.id = location;
+    }
+
+    public Component getDisplayName() {
+        return displayName;
+    }
+
+    public List<Component> getLore() {
+        return lore;
+    }
+
+    public String getMojangID() {
+        return mojangID;
+    }
+
+    public void setMojangID(String mojangID) {
+        this.mojangID = mojangID;
+    }
+
+    public Item getBaseItem() {
+        return baseItem;
+    }
+
+    public void setClassURL(String url) {
+        this.classURL = url;
+    }
+
+    public int getModelData() {
+        return modelData;
+    }
+
+    public Map<EquipmentSlot, AttributeModifier> getModifierMap() {
+        return modifierMap;
+    }
+
+    public int getMaxStackSize() {
+        return maxStackSize;
+    }
+
+    public int getMaxDurability() {
+        return maxDurability;
+    }
+
+    public boolean isEnabled() {
+        return isEnabled;
+    }
+
+    public HItemStack getReplaceWith() {
+        return replaceWith;
+    }
+
+    public @Nullable String getClassURL() {
+        return classURL;
+    }
+
+    public String getOwningPlugin() {
+        return pluginName;
+    }
+
+    public void setOwningPlugin(String name) {
+        pluginName = name;
+    }
+
+    public boolean hasBehaviour() {
+        return hasBehaviour;
+    }
+
+    public HItemBehaviour getBehaviour() {
+        return behaviour;
+    }
+
+    public void setBehaviour(HItemBehaviour behaviour) {
+        this.behaviour = behaviour;
+    }
+
+    public void setPlugin(Plugin plugin) {
+        this.plugin = plugin;
+    }
+
+    public HItemStack getItem() {
+        return getItem(1);
+    }
+
+    public HItemStack getItem(int count) {
+        HItemStack item = prepareNewStack().hItemStack;
+        item.getStack().setCount(count);
+        return item;
+    }
+
+    private ItemStack prepareNewStack() {
+        ItemStack baseItem = new ItemStack(BuiltInRegistries.ITEM.get(new ResourceLocation(getMojangID())));
+        for (EquipmentSlot slot : EquipmentSlot.values()) {
+            // baseItem.getAttributeModifiers(slot).clear(); this is immutable apparently
+        }
+        baseItem.hItemStack = new HItemStack(baseItem, this);
+        baseItem.getOrCreateTag().putInt("CustomModelData", getModelData());
+        baseItem.getOrCreateTag().putString("HephaestusID", id.toString());
+        baseItem.setHoverName(PaperAdventure.asVanilla(displayName));
+        baseItem.getBukkitStack().editMeta(m -> m.lore(lore)); // Annoying but oh well.
+        return baseItem;
+    }
+
+    public void loadBehaviour() {
+        if (plugin == null) {
+            plugin = Bukkit.getPluginManager().getPlugin(pluginName);
+            if (plugin == null) {
+                LOGGER.info("Failed to find plugin " + pluginName + " for item " + id.toString());
+            }
+        }
+        classLoader = plugin.getClass().getClassLoader();
+        try {
+            behaviourClass = ((Class<? extends HItemBehaviour>) Class.forName(getClassURL(), true, classLoader));
+        } catch (ClassNotFoundException e) {
+            throw new RuntimeException(e);
+        }
+        try {
+            behaviour = behaviourClass.getDeclaredConstructor(HItem.class).newInstance(this);
+        } catch (InstantiationException | IllegalAccessException | InvocationTargetException | NoSuchMethodException e) {
+            throw new RuntimeException(e);
+        }
+        LOGGER.info("Loaded behaviour for " + id + ": " + behaviour.getClass().getName());
+        hasBehaviour = true;
+    }
+
+    @Override
+    public void load(@NotNull File file) throws IOException, InvalidConfigurationException {
+        super.load(file);
+        id = ResourceLocation.tryParse(getString("id", "minecraft:stone"));
+        if (contains("class")) {
+            classURL = getString("class");
+            pluginName = getString("owningPlugin");
+            hasBehaviour = true;
+        }
+        isEnabled = getBoolean("enabled", true);
+        displayName = mm.deserialize(getString("displayName", ""));
+        for (String string : getStringList("lore")) {
+            lore.add(mm.deserialize(string));
+        }
+        mojangID = getString("mojangID", "minecraft:stone");
+        baseItem = BuiltInRegistries.ITEM.get(new ResourceLocation(mojangID));
+        modelData = getInt("modelData", 0);
+        maxStackSize = getInt("maxStackSize", 1);
+        maxDurability = getInt("maxDurability", 0);
+        if (contains("modifiers")) {
+            for (String slot : getConfigurationSection("modifiers").getKeys(false)) {
+                EquipmentSlot equipmentSlot = EquipmentSlot.byName(slot);
+                AttributeModifier modifier = new AttributeModifier(
+                    getString("modifiers." + slot + ".name", "HephaestusModifier"),
+                    getDouble("modifiers." + slot + ".amount", 0),
+                    AttributeModifier.Operation.valueOf(getString("modifiers." + slot + ".operation", "ADD_NUMBER")));
+                modifierMap.put(equipmentSlot, modifier);
+            }
+        }
+    }
+
+
+    public void save(File file) {
+        set("id", id.toString());
+        set("owningPlugin", pluginName);
+        set("class", classURL);
+        set("enabled", isEnabled);
+        set("mojangID", mojangID);
+        try {
+            super.save(file);
+        } catch (IOException e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    public static class Builder {
+        private final HItem item;
+        private final Plugin plugin;
+        private HItemBehaviour behaviour;
+
+        public Builder(Plugin plugin, ResourceLocation location) {
+            this.item = new HItem(plugin);
+            this.plugin = plugin;
+            item.setId(location);
+        }
+
+        public Builder behaviour(HItemBehaviour behaviour) {
+            this.behaviour = behaviour;
+            item.setClassURL(behaviour.getClass().getName());
+            item.setOwningPlugin(plugin.getName());
+            item.setPlugin(plugin);
+            return this;
+        }
+        public Builder baseItem(Material material) {
+            Item item1 = CraftMagicNumbers.getItem(material);
+            item.setMojangID(BuiltInRegistries.ITEM.getKey(item1).toString());
+            return this;
+        }
+
+        public HItem register() {
+            Main.itemLibrary.registerItem(plugin, item);
+            File folder = Main.itemLibrary.getFolder();
+            File saveFolder = new File(folder, item.getId().getNamespace());
+            if (!saveFolder.exists()) {
+                saveFolder.mkdirs();
+            }
+            File saveFile = new File(saveFolder, item.getId().getPath() + ".yml");
+            item.save(saveFile);
+            return item;
+        }
+    }
+
+}
diff --git a/src/main/java/de/erethon/hephaestus/HItemBehaviour.java b/src/main/java/de/erethon/hephaestus/HItemBehaviour.java
new file mode 100644
index 0000000000000000000000000000000000000000..f5c220a6142d60db52a9035c9c665641f02e2d82
--- /dev/null
+++ b/src/main/java/de/erethon/hephaestus/HItemBehaviour.java
@@ -0,0 +1,88 @@
+package de.erethon.hephaestus;
+
+import com.destroystokyo.paper.event.player.PlayerArmorChangeEvent;
+import de.erethon.spellbook.api.SpellCastEvent;
+import net.minecraft.core.BlockPos;
+import net.minecraft.world.InteractionHand;
+import net.minecraft.world.InteractionResult;
+import net.minecraft.world.entity.LivingEntity;
+import net.minecraft.world.entity.player.Player;
+import net.minecraft.world.item.ItemStack;
+import net.minecraft.world.level.Level;
+import net.minecraft.world.level.block.state.BlockState;
+import org.bukkit.entity.Entity;
+import org.bukkit.event.block.BlockDropItemEvent;
+import org.bukkit.event.entity.EntityDamageByEntityEvent;
+import org.bukkit.event.entity.EntityDropItemEvent;
+import org.bukkit.event.entity.EntityPickupItemEvent;
+import org.bukkit.event.inventory.ClickType;
+import org.bukkit.event.inventory.InventoryAction;
+import org.bukkit.event.inventory.InventoryClickEvent;
+import org.bukkit.event.player.PlayerAttemptPickupItemEvent;
+import org.bukkit.event.player.PlayerDropItemEvent;
+import org.bukkit.event.player.PlayerInteractEvent;
+
+public class HItemBehaviour {
+
+    protected HItem item;
+
+    public HItemBehaviour(HItem item) {
+        this.item = item;
+    }
+
+    public void onLeftClick(ItemStack stack, org.bukkit.entity.Player player, PlayerInteractEvent event) {
+    }
+
+    public void onRightClick(ItemStack stack, org.bukkit.entity.Player player, PlayerInteractEvent event) {
+    }
+
+    public void onPlayerDrop(ItemStack stack, org.bukkit.entity.Player player, PlayerDropItemEvent event) {
+    }
+
+    public void onEntityDrop(ItemStack stack, org.bukkit.entity.Entity entity, EntityDropItemEvent event) {
+    }
+
+    public void onBlockDrop(ItemStack stack, org.bukkit.entity.Player player, BlockDropItemEvent event) {
+    }
+
+    public void onPickup(ItemStack stack, org.bukkit.entity.Player player, PlayerAttemptPickupItemEvent event) {
+    }
+
+    public void onEntityPickup(ItemStack stack, org.bukkit.entity.Entity entity, EntityPickupItemEvent event) {
+    }
+
+    public void onEquip(ItemStack stack, org.bukkit.entity.Player player, PlayerArmorChangeEvent event) {
+    }
+
+    public void onUnequip(ItemStack stack, org.bukkit.entity.Player player, PlayerArmorChangeEvent event) {
+    }
+
+    public boolean onAttack(ItemStack stack, LivingEntity target, Player attacker) {
+        return true;
+    }
+
+    public void onDamaged(ItemStack stack, org.bukkit.entity.LivingEntity damaged, EntityDamageByEntityEvent event) {
+    }
+
+    public int onItemHurt(ItemStack stack, int amount) {
+        return amount;
+    }
+
+    public boolean onMine(ItemStack stack, Level world, BlockState state, BlockPos pos, Player miner) {
+        return true;
+    }
+
+    public InteractionResult onInteractLiving(ItemStack stack, Player player, LivingEntity living, InteractionHand hand) {
+        return stack.getItem().interactLivingEntity(stack, player, living, hand);
+    }
+
+    public void onSpellCast(ItemStack stack, org.bukkit.entity.LivingEntity damaged, SpellCastEvent event) {
+    }
+
+    public void onTick(ItemStack stack, Entity entity, int slot, boolean selected) {
+    }
+
+    public void onInventoryClick(ItemStack stack, org.bukkit.entity.Player player, ClickType type, InventoryAction action, InventoryClickEvent event) {
+    }
+
+}
diff --git a/src/main/java/de/erethon/hephaestus/HItemStack.java b/src/main/java/de/erethon/hephaestus/HItemStack.java
new file mode 100644
index 0000000000000000000000000000000000000000..d62a4a41b88f38d808882d55640c480a40ee6189
--- /dev/null
+++ b/src/main/java/de/erethon/hephaestus/HItemStack.java
@@ -0,0 +1,80 @@
+package de.erethon.hephaestus;
+
+import net.minecraft.resources.ResourceLocation;
+import net.minecraft.world.item.ItemStack;
+
+import java.util.ArrayList;
+import java.util.List;
+
+import static net.minecraft.server.MinecraftServer.LOGGER;
+
+public class HItemStack {
+
+    private HItem item;
+    private ItemStack stack;
+    private int level = 0;
+    private HRarity rarity = HRarity.COMMON;
+    private List<Slotable> slotted = new ArrayList<>();
+
+
+    public HItemStack(ItemStack stack, HItem config) {
+        this.stack = stack;
+        this.item = config;
+    }
+
+    public HItemStack(ItemStack stack, HItem config, int level) {
+        this.stack = stack;
+        this.item = config;
+        this.level = level;
+    }
+
+    public HItemStack(ItemStack stack, HItem config, int level, HRarity rarity) {
+        this.stack = stack;
+        this.item = config;
+        this.level = level;
+        this.rarity = rarity;
+    }
+
+    public ResourceLocation getId() {
+        return item.getId();
+    }
+
+    public boolean needsUpdating(ItemStack stack) {
+        return !ItemStack.matches(stack, this.stack);
+    }
+
+    public ItemStack getStack() {
+        return stack;
+    }
+
+    public org.bukkit.inventory.ItemStack getBukkitStack() {
+        return stack.getBukkitStack();
+    }
+
+    public HItemBehaviour getBehaviour() {
+        return item.getBehaviour();
+    }
+
+    public HItem config() {
+        return item;
+    }
+
+    public boolean isEnabled() {
+        return item.isEnabled();
+    }
+
+    public boolean isVanilla() {
+        return item.getId().getNamespace().equals("minecraft");
+    }
+
+    public HItemStack replaceWith() {
+        return item.getReplaceWith();
+    }
+
+    public void setStack(ItemStack stack) {
+        this.stack = stack;
+    }
+
+    public void validate() {
+    }
+}
diff --git a/src/main/java/de/erethon/hephaestus/HRarity.java b/src/main/java/de/erethon/hephaestus/HRarity.java
new file mode 100644
index 0000000000000000000000000000000000000000..524fae53d99642ca35837db8a7bbb34616058fa1
--- /dev/null
+++ b/src/main/java/de/erethon/hephaestus/HRarity.java
@@ -0,0 +1,20 @@
+package de.erethon.hephaestus;
+
+
+import net.kyori.adventure.text.format.NamedTextColor;
+import net.kyori.adventure.text.format.TextColor;
+
+public enum HRarity {
+    TRASH(NamedTextColor.GRAY),
+    COMMON(NamedTextColor.WHITE),
+    UNCOMMON(NamedTextColor.GREEN),
+    RARE(NamedTextColor.BLUE),
+    EPIC(NamedTextColor.DARK_PURPLE),
+    LEGENDARY(NamedTextColor.GOLD);
+
+    private TextColor color;
+
+    HRarity(TextColor color) {
+        this.color = color;
+    }
+}
diff --git a/src/main/java/de/erethon/hephaestus/ItemLibrary.java b/src/main/java/de/erethon/hephaestus/ItemLibrary.java
new file mode 100644
index 0000000000000000000000000000000000000000..8ef09800709175c8844fb6979708e56be62d03fd
--- /dev/null
+++ b/src/main/java/de/erethon/hephaestus/ItemLibrary.java
@@ -0,0 +1,171 @@
+package de.erethon.hephaestus;
+
+import de.erethon.spellbook.api.SpellLibrary;
+import net.minecraft.core.registries.BuiltInRegistries;
+import net.minecraft.resources.ResourceLocation;
+import net.minecraft.world.item.Item;
+import org.bukkit.Bukkit;
+import org.bukkit.Material;
+import org.bukkit.configuration.file.YamlConfiguration;
+import org.bukkit.craftbukkit.inventory.CraftItemStack;
+import org.bukkit.craftbukkit.util.CraftMagicNumbers;
+import org.bukkit.inventory.ItemStack;
+import org.bukkit.plugin.Plugin;
+import org.bukkit.plugin.java.JavaPlugin;
+import org.jetbrains.annotations.Nullable;
+
+import java.io.File;
+import java.io.IOException;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Set;
+import java.util.logging.Level;
+
+import static net.minecraft.server.MinecraftServer.LOGGER;
+
+public class ItemLibrary {
+
+    private final File folder;
+
+    private final HashMap<ResourceLocation, HItem> items = new HashMap<>();
+    private final HashMap<Material, HItem> vanillaMaterials = new HashMap<>();
+    private final Set<HItem> itemsWithBehaviour = new HashSet<>();
+    private HBukkitHandler bukkitHandler = null;
+
+    public ItemLibrary(File folder) {
+        this.folder = new File(folder + "/plugins/Hephaestus/items");
+        if (!folder.exists()) {
+            folder.mkdirs();
+        }
+        generateVanillaItemFiles();
+        loadItems();
+    }
+
+    public boolean has(ResourceLocation key) {
+        return items.containsKey(key);
+    }
+
+    @Nullable
+    public HItem get(ResourceLocation key) {
+        return items.get(key);
+    }
+
+    @Nullable
+    public HItem get(Material material) {
+        return vanillaMaterials.get(material);
+    }
+
+    @Nullable
+    public HItem get(ItemStack bukkitStack) {
+        CraftItemStack craftItemStack = (CraftItemStack) bukkitStack;
+        if (craftItemStack.handle.hItemStack == null) {
+            return null;
+        }
+        return craftItemStack.handle.hItemStack.config();
+    }
+
+    @Nullable
+    public HItem get(net.minecraft.world.item.ItemStack stack) {
+        return stack.hItemStack.config();
+    }
+
+    public boolean registerItem(Plugin plugin, HItem item) {
+        if (bukkitHandler == null) { // Use the first plugin we get to register the bukkit handler, because we need a Plugin for listeners.
+            bukkitHandler = new HBukkitHandler((JavaPlugin) plugin, this);
+        }
+        if (item.getId().getNamespace().equals("minecraft")) {
+            throw new RuntimeException(plugin.getName() + " tried registering a custom item in the vanilla namespace: " + item.getId().toString());
+        }
+        if (items.containsKey(item.getId())) {
+            Bukkit.getLogger().log(Level.WARNING, "[Hephaestus] Item with ID " + item.getId() + " is already registered.");
+            return false;
+        }
+        item.setPlugin(plugin);
+        items.put(item.getId(), item);
+        return true;
+    }
+
+    public boolean overrideVanillaBehaviour(Plugin plugin, HItem item) {
+        if (bukkitHandler == null) {
+            bukkitHandler = new HBukkitHandler((JavaPlugin) plugin, this);
+        }
+        item.setPlugin(plugin);
+        return true;
+    }
+
+
+    public boolean unregisterItem(HItem item) {
+        if (!items.containsKey(item.getId())) {
+            Bukkit.getLogger().log(Level.WARNING, "[Hephaestus] Item with ID " + item.getId() + " is not registered.");
+            return false;
+        }
+        if (item.getId().getNamespace().equals("minecraft")) {
+            throw new RuntimeException("Vanilla items can not be unregistered.");
+        }
+        items.remove(item.getId());
+        return true;
+    }
+
+    private void generateVanillaItemFiles() {
+        for (Item item : BuiltInRegistries.ITEM.stream().toList()) {
+            File vanillaFolder = new File(folder, "vanilla");
+            if (!vanillaFolder.exists()) {
+                vanillaFolder.mkdirs();
+            }
+            YamlConfiguration configuration = new YamlConfiguration();
+            ResourceLocation location = BuiltInRegistries.ITEM.getKey(item);
+            configuration.set("id", location.toString());
+            configuration.set("mojangID", location.toString());
+            configuration.set("maxStackSize", item.getMaxStackSize());
+            configuration.set("maxDurability", item.getMaxDamage());
+            File file = new File(vanillaFolder, location.getPath() + ".yml");
+            if (file.exists()) {
+                return;
+            }
+            try {
+                configuration.save(file);
+            } catch (IOException e) {
+                throw new RuntimeException(e);
+            }
+        }
+    }
+
+    private void loadItems() {
+        for (File file : SpellLibrary.getFilesForFolder(folder)) {
+            if (!file.getName().contains(".yml")) {
+                continue;
+            }
+            HItem config = new HItem(file);
+            items.put(config.getId(), config);
+            vanillaMaterials.put(CraftMagicNumbers.getMaterial(config.getBaseItem()), config);
+            if (config.hasBehaviour()) {
+                itemsWithBehaviour.add(config);
+            }
+        }
+        LOGGER.info("[Hephaestus] Loaded " + items.keySet().size() + " items.");
+    }
+
+    public File getFolder() {
+        return folder;
+    }
+
+    public void readyBehaviours() {
+        for (HItem item : itemsWithBehaviour) {
+            item.loadBehaviour();
+        }
+    }
+
+    public void enableHandler(Plugin plugin) {
+        if (bukkitHandler == null) { // Use the first plugin we get to register the bukkit handler, because we need a Plugin for listeners.
+            bukkitHandler = new HBukkitHandler((JavaPlugin) plugin, this);
+        }
+    }
+
+    public static @Nullable HItemStack getHItemFromBukkit(ItemStack stack) {
+        CraftItemStack craftItemStack = CraftItemStack.asCraftCopy(stack);
+        if (craftItemStack.handle == null) { // why in the fuck are there craft stacks that have no nms stack???
+            return null;
+        }
+        return craftItemStack.handle.hItemStack;
+     }
+}
diff --git a/src/main/java/de/erethon/hephaestus/Slotable.java b/src/main/java/de/erethon/hephaestus/Slotable.java
new file mode 100644
index 0000000000000000000000000000000000000000..1f9cad66d1e138f568825055509433b6c2dd903a
--- /dev/null
+++ b/src/main/java/de/erethon/hephaestus/Slotable.java
@@ -0,0 +1,8 @@
+package de.erethon.hephaestus;
+
+import java.util.Set;
+
+public class Slotable {
+
+
+}
diff --git a/src/main/java/de/erethon/hephaestus/events/MobSpawnEntityCreateEvent.java b/src/main/java/de/erethon/hephaestus/events/MobSpawnEntityCreateEvent.java
new file mode 100644
index 0000000000000000000000000000000000000000..53d1624375afb11f487bad0b52166ebe899f97a2
--- /dev/null
+++ b/src/main/java/de/erethon/hephaestus/events/MobSpawnEntityCreateEvent.java
@@ -0,0 +1,46 @@
+package de.erethon.hephaestus.events;
+
+import net.minecraft.core.BlockPos;
+import net.minecraft.core.Holder;
+import net.minecraft.server.level.ServerLevel;
+import net.minecraft.world.entity.EntityType;
+import net.minecraft.world.entity.Mob;
+import net.minecraft.world.entity.SpawnGroupData;
+import net.minecraft.world.entity.player.Player;
+import net.minecraft.world.level.biome.Biome;
+import net.minecraft.world.level.biome.MobSpawnSettings;
+import org.bukkit.event.Event;
+import org.bukkit.event.HandlerList;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+public class MobSpawnEntityCreateEvent extends Event {
+    private static final HandlerList handlers = new HandlerList();
+
+    public ServerLevel world;
+    public MobSpawnSettings.SpawnerData spawnerData;
+    public Mob mob;
+    public BlockPos.MutableBlockPos pos;
+    public Player nearestPlayer;
+    public SpawnGroupData data;
+
+    public MobSpawnEntityCreateEvent(Mob mob, ServerLevel world, MobSpawnSettings.SpawnerData spawnerData, BlockPos.MutableBlockPos pos, Player player, @Nullable SpawnGroupData data) {
+        this.mob = mob;
+        this.world = world;
+        this.spawnerData = spawnerData;
+        this.pos = pos;
+        this.nearestPlayer = player;
+        this.data = data;
+    }
+
+
+    @NotNull
+    @Override
+    public HandlerList getHandlers() {
+        return handlers;
+    }
+
+    public static HandlerList getHandlerList() {
+        return handlers;
+    }
+}
diff --git a/src/main/java/de/erethon/hephaestus/events/MobSpawnSettingsEvent.java b/src/main/java/de/erethon/hephaestus/events/MobSpawnSettingsEvent.java
new file mode 100644
index 0000000000000000000000000000000000000000..3827cd2200bfa3434555e973b9a6b7ea9d1ff12c
--- /dev/null
+++ b/src/main/java/de/erethon/hephaestus/events/MobSpawnSettingsEvent.java
@@ -0,0 +1,52 @@
+package de.erethon.hephaestus.events;
+
+import net.minecraft.core.BlockPos;
+import net.minecraft.core.Holder;
+import net.minecraft.server.level.ServerLevel;
+import net.minecraft.util.RandomSource;
+import net.minecraft.util.random.WeightedRandomList;
+import net.minecraft.world.entity.MobCategory;
+import net.minecraft.world.level.StructureManager;
+import net.minecraft.world.level.biome.Biome;
+import net.minecraft.world.level.biome.MobSpawnSettings;
+import net.minecraft.world.level.chunk.ChunkGenerator;
+import org.bukkit.event.Event;
+import org.bukkit.event.HandlerList;
+import org.jetbrains.annotations.NotNull;
+
+import javax.annotation.Nullable;
+import java.util.Optional;
+
+public class MobSpawnSettingsEvent extends Event {
+
+    private static final HandlerList handlers = new HandlerList();
+
+    public Holder<Biome> biomeHolder;
+    public WeightedRandomList<MobSpawnSettings.SpawnerData> spawnerData;
+    public ServerLevel world;
+    public StructureManager structureAccessor;
+    public ChunkGenerator chunkGenerator;
+    public MobCategory spawnGroup;
+    public RandomSource random;
+    public BlockPos pos;
+
+    public MobSpawnSettingsEvent(ServerLevel world, StructureManager structureAccessor, ChunkGenerator chunkGenerator, MobCategory spawnGroup, BlockPos pos, @Nullable Holder<Biome> biomeEntry) {
+        this.biomeHolder = biomeEntry;
+        this.world = world;
+        this.structureAccessor = structureAccessor;
+        this.chunkGenerator = chunkGenerator;
+        this.spawnGroup = spawnGroup;
+        this.pos = pos;
+        this.spawnerData = null;
+    }
+
+    @NotNull
+    @Override
+    public HandlerList getHandlers() {
+        return handlers;
+    }
+
+    public static HandlerList getHandlerList() {
+        return handlers;
+    }
+}
diff --git a/src/main/java/de/erethon/papyrus/ContainerLoadEvent.java b/src/main/java/de/erethon/papyrus/ContainerLoadEvent.java
index e672340310a8863a627c083cbea2c08a8d6049b9..b6e1158585a4a8971a2fcc5fcea55e9d467b03e6 100644
--- a/src/main/java/de/erethon/papyrus/ContainerLoadEvent.java
+++ b/src/main/java/de/erethon/papyrus/ContainerLoadEvent.java
@@ -12,7 +12,12 @@ public class ContainerLoadEvent extends Event {
 
     private NonNullList<ItemStack> itemStacks;
 
-    public ContainerLoadEvent(NonNullList<ItemStack> items) {
+    /**
+     * Note: This event may be triggered async during feature placement.
+     * This shouldn't happen on Erethon, but who knows.
+     */
+    public ContainerLoadEvent(NonNullList<ItemStack> items, boolean async) {
+        super(async);
         this.itemStacks = items;
     }
 
diff --git a/src/main/java/io/papermc/paper/chunk/system/scheduling/NewChunkHolder.java b/src/main/java/io/papermc/paper/chunk/system/scheduling/NewChunkHolder.java
index b66a7d4aab887309579154815a0d4abf9de506b0..daa036030b93fbb388f55c265cac22defae6b917 100644
--- a/src/main/java/io/papermc/paper/chunk/system/scheduling/NewChunkHolder.java
+++ b/src/main/java/io/papermc/paper/chunk/system/scheduling/NewChunkHolder.java
@@ -1943,6 +1943,7 @@ public final class NewChunkHolder {
             throw death;
         } catch (final Throwable thr) {
             LOGGER.error("Failed to save entity data (" + this.chunkX + "," + this.chunkZ + ") in world '" + this.world.getWorld().getName() + "'");
+            thr.printStackTrace(); // Papyrus - actual useful error
         }
 
         return true;
diff --git a/src/main/java/net/minecraft/network/protocol/game/ClientboundAddEntityPacket.java b/src/main/java/net/minecraft/network/protocol/game/ClientboundAddEntityPacket.java
index c43bf052b4a3e6cf693fad9b08526dc396d4dfef..041537c626b73aefb7bc0937c0f43e3d375cc0e2 100644
--- a/src/main/java/net/minecraft/network/protocol/game/ClientboundAddEntityPacket.java
+++ b/src/main/java/net/minecraft/network/protocol/game/ClientboundAddEntityPacket.java
@@ -58,7 +58,8 @@ public class ClientboundAddEntityPacket implements Packet<ClientGamePacketListen
         this.xa = (int)(Mth.clamp(velocity.x, -3.9D, 3.9D) * 8000.0D);
         this.ya = (int)(Mth.clamp(velocity.y, -3.9D, 3.9D) * 8000.0D);
         this.za = (int)(Mth.clamp(velocity.z, -3.9D, 3.9D) * 8000.0D);
-        if (this.entity != null && entity.displayEntityType != null) { // ModelEngine uses null entities because stupid
+        this.entity = entity;
+        if (entity.displayEntityType != null) { // ModelEngine uses null entities because stupid
             this.type = entity.displayEntityType;
         }
     }
diff --git a/src/main/java/net/minecraft/server/Bootstrap.java b/src/main/java/net/minecraft/server/Bootstrap.java
index 5816998cde8504d58732e63eb2179a5d828f35bb..96ac8c11b80abeaee1a218fc4ab53571f8eec995 100644
--- a/src/main/java/net/minecraft/server/Bootstrap.java
+++ b/src/main/java/net/minecraft/server/Bootstrap.java
@@ -1,7 +1,10 @@
 package net.minecraft.server;
 
 import com.mojang.logging.LogUtils;
+
+import java.io.File;
 import java.io.PrintStream;
+import java.nio.file.Paths;
 import java.time.Duration;
 import java.time.Instant;
 import java.util.Set;
@@ -9,6 +12,8 @@ import java.util.TreeSet;
 import java.util.concurrent.atomic.AtomicLong;
 import java.util.function.Function;
 import java.util.function.Supplier;
+
+import de.erethon.hephaestus.ItemLibrary;
 import net.minecraft.SharedConstants;
 import net.minecraft.commands.Commands;
 import net.minecraft.commands.arguments.selector.options.EntitySelectorOptions;
@@ -69,6 +74,7 @@ public class Bootstrap {
             } else {
                 FireBlock.bootStrap();
                 ComposterBlock.bootStrap();
+                Main.itemLibrary = new ItemLibrary(new File(Paths.get("").toUri())); // Papyrus
                 if (EntityType.getKey(EntityType.PLAYER) == null) {
                     throw new IllegalStateException("Failed loading EntityTypes");
                 } else {
diff --git a/src/main/java/net/minecraft/server/Main.java b/src/main/java/net/minecraft/server/Main.java
index 7573c12a77797146c51ef2dfe4b2a636df45e21a..e86db50efb6196db4ecae1d8b38b18cc1ae1507d 100644
--- a/src/main/java/net/minecraft/server/Main.java
+++ b/src/main/java/net/minecraft/server/Main.java
@@ -16,6 +16,8 @@ import java.nio.file.Paths;
 import java.util.Optional;
 import java.util.function.BooleanSupplier;
 import javax.annotation.Nullable;
+
+import de.erethon.hephaestus.ItemLibrary;
 import joptsimple.NonOptionArgumentSpec;
 import joptsimple.OptionParser;
 import joptsimple.OptionSet;
@@ -71,6 +73,7 @@ import org.bukkit.configuration.file.YamlConfiguration;
 public class Main {
 
     private static final Logger LOGGER = LogUtils.getLogger();
+    public static ItemLibrary itemLibrary; // Papyrus
 
     public Main() {}
 
diff --git a/src/main/java/net/minecraft/server/MinecraftServer.java b/src/main/java/net/minecraft/server/MinecraftServer.java
index c535eb33ef8e820d11d75702059613043982deee..3dac663daf000e443b141d11c1432ec2524d33de 100644
--- a/src/main/java/net/minecraft/server/MinecraftServer.java
+++ b/src/main/java/net/minecraft/server/MinecraftServer.java
@@ -14,6 +14,7 @@ import com.mojang.authlib.GameProfileRepository;
 import com.mojang.authlib.minecraft.MinecraftSessionService;
 import com.mojang.datafixers.DataFixer;
 import com.mojang.logging.LogUtils;
+import de.erethon.hephaestus.ItemLibrary;
 import it.unimi.dsi.fastutil.longs.LongIterator;
 import it.unimi.dsi.fastutil.objects.ObjectArrayList;
 import java.awt.image.BufferedImage;
diff --git a/src/main/java/net/minecraft/server/level/ServerEntity.java b/src/main/java/net/minecraft/server/level/ServerEntity.java
index d46e61640b241d32df05240dedd2c23f138725e6..89923aaaf0d8eecb279b47a0d24b8fd6d48ab591 100644
--- a/src/main/java/net/minecraft/server/level/ServerEntity.java
+++ b/src/main/java/net/minecraft/server/level/ServerEntity.java
@@ -314,7 +314,9 @@ public class ServerEntity {
             // CraftBukkit end
 
             if (!collection.isEmpty()) {
-                sender.accept(new ClientboundUpdateAttributesPacket(this.entity.getId(), collection));
+               if (((LivingEntity) this.entity).syncAttributes) { // Papyrus - custom entities
+                   sender.accept(new ClientboundUpdateAttributesPacket(this.entity.getId(), collection));
+               }
             }
 
             if (((LivingEntity) this.entity).isFallFlying()) {
@@ -388,7 +390,9 @@ public class ServerEntity {
                     ((ServerPlayer) this.entity).getBukkitEntity().injectScaledMaxHealth(set, false);
                 }
                 // CraftBukkit end
-                this.broadcastAndSend(new ClientboundUpdateAttributesPacket(this.entity.getId(), set));
+                if (((LivingEntity) this.entity).syncAttributes) { // Papyrus - custom entities
+                    this.broadcastAndSend(new ClientboundUpdateAttributesPacket(this.entity.getId(), set));
+                }
             }
 
             set.clear();
diff --git a/src/main/java/net/minecraft/world/ContainerHelper.java b/src/main/java/net/minecraft/world/ContainerHelper.java
index 01063c3bc672eecdf8fe1c5b5a89fbf576c940f3..949207f366aa0659c563fa123047112d30b6a4ec 100644
--- a/src/main/java/net/minecraft/world/ContainerHelper.java
+++ b/src/main/java/net/minecraft/world/ContainerHelper.java
@@ -53,8 +53,13 @@ public class ContainerHelper {
                 stacks.set(j, ItemStack.of(compoundTag));
             }
         }
-        ContainerLoadEvent loadEvent = new ContainerLoadEvent(stacks);
-        Bukkit.getPluginManager().callEvent(loadEvent);
+        ContainerLoadEvent loadEvent = new ContainerLoadEvent(stacks, false);
+        try {
+            Bukkit.getPluginManager().callEvent(loadEvent);
+        } catch (IllegalStateException e) { // This really only happens during feature placement which shouldn't happen in production anyways
+            loadEvent = new ContainerLoadEvent(stacks, true);
+            Bukkit.getPluginManager().callEvent(loadEvent);
+        }
 
     }
 
diff --git a/src/main/java/net/minecraft/world/entity/EntityType.java b/src/main/java/net/minecraft/world/entity/EntityType.java
index cf90ecefea3586d0ce236288ab515bfa67239814..03f7f0b2b262196bf625cd793201b22331d0d6d6 100644
--- a/src/main/java/net/minecraft/world/entity/EntityType.java
+++ b/src/main/java/net/minecraft/world/entity/EntityType.java
@@ -2,10 +2,8 @@ package net.minecraft.world.entity;
 
 import com.google.common.collect.ImmutableSet;
 import com.mojang.logging.LogUtils;
-import java.util.List;
-import java.util.Optional;
-import java.util.Spliterator;
-import java.util.UUID;
+
+import java.util.*;
 import java.util.function.Consumer;
 import java.util.function.Function;
 import java.util.stream.Stream;
@@ -161,6 +159,7 @@ import net.minecraft.world.phys.AABB;
 import net.minecraft.world.phys.shapes.Shapes;
 import net.minecraft.world.phys.shapes.VoxelShape;
 import org.bukkit.event.entity.CreatureSpawnEvent;
+import org.bukkit.plugin.Plugin;
 import org.slf4j.Logger;
 
 public class EntityType<T extends Entity> implements FeatureElement, EntityTypeTest<Entity, T> {
@@ -316,6 +315,8 @@ public class EntityType<T extends Entity> implements FeatureElement, EntityTypeT
     private final EntityDimensions dimensions;
     private final FeatureFlagSet requiredFeatures;
 
+    public static final Map<String, Map.Entry<Plugin, Class<? extends Entity>>> customEntities = new HashMap<>(); // Papyrus - Custom Entities
+
     private static <T extends Entity> EntityType<T> register(String id, EntityType.Builder type) { // CraftBukkit - decompile error
         return (EntityType) Registry.register(BuiltInRegistries.ENTITY_TYPE, id, (EntityType<T>) type.build(id)); // CraftBukkit - decompile error
     }
@@ -596,6 +597,19 @@ public class EntityType<T extends Entity> implements FeatureElement, EntityTypeT
     public static Optional<Entity> create(CompoundTag nbt, Level world, boolean generation) {
         // Paper end - add generation bool param
         return Util.ifElse(EntityType.by(nbt).map((entitytypes) -> {
+            // Papyrus start - Custom entities
+            if (nbt.contains("papyrus-entity-id")) {
+                String customEntityId = nbt.getString("papyrus-entity-id");
+                if (customEntities.containsKey(customEntityId)) {
+                    Map.Entry<Plugin, Class<? extends Entity>> entityEntry = customEntities.get(customEntityId);
+                    if (entitytypes.getBaseClass().isAssignableFrom(entityEntry.getValue())) {
+                        return EntityType.createCustomEntity(entityEntry.getValue(), entitytypes, entityEntry.getKey(), world);
+                    } else {
+                        LOGGER.warn("Entity type {} is not assignable from custom entity class {}", entitytypes, entityEntry.getValue());
+                    }
+                }
+            }
+            // Papyrus end
             return entitytypes.create(world);
         }), (entity) -> {
             if (generation) entity.generation = true; // Paper - add generation bool param
@@ -605,6 +619,18 @@ public class EntityType<T extends Entity> implements FeatureElement, EntityTypeT
         });
     }
 
+    // Papyrus start - Custom entities
+    public static Entity createCustomEntity(Class<? extends Entity> customEntityClass, EntityType<?> type, Plugin plugin, Level world) {
+        try {
+            return customEntityClass.getConstructor(EntityType.class, Level.class).newInstance(type, world);
+        } catch (Exception e) {
+            LOGGER.warn("Failed to create custom entity", e);
+            e.printStackTrace();
+            return type.create(world);
+        }
+    }
+    // Papyrus end
+
     public AABB getAABB(double feetX, double feetY, double feetZ) {
         float f = this.getWidth() / 2.0F;
 
diff --git a/src/main/java/net/minecraft/world/entity/LivingEntity.java b/src/main/java/net/minecraft/world/entity/LivingEntity.java
index 006017d58f031a863e124a2e400d870faba23800..1b7bef85f4ebaa67aa23de1de161d93df4f4f3a4 100644
--- a/src/main/java/net/minecraft/world/entity/LivingEntity.java
+++ b/src/main/java/net/minecraft/world/entity/LivingEntity.java
@@ -214,6 +214,7 @@ public abstract class LivingEntity extends Entity implements Attackable {
     public org.bukkit.craftbukkit.entity.CraftLivingEntity getBukkitLivingEntity() { return (org.bukkit.craftbukkit.entity.CraftLivingEntity) super.getBukkitEntity(); } // Paper
     public boolean silentDeath = false; // Paper - mark entity as dying silently for cancellable death event
     public net.kyori.adventure.util.TriState frictionState = net.kyori.adventure.util.TriState.NOT_SET; // Paper
+    public boolean syncAttributes = true; // Papyrus - for custom entities
 
     @Override
     public float getBukkitYaw() {
diff --git a/src/main/java/net/minecraft/world/entity/Mob.java b/src/main/java/net/minecraft/world/entity/Mob.java
index 7999b85b7df949e4dfcad2ca23489324ebd8a2c7..117809c39c7bf76581e122d92dce965af48e4358 100644
--- a/src/main/java/net/minecraft/world/entity/Mob.java
+++ b/src/main/java/net/minecraft/world/entity/Mob.java
@@ -174,7 +174,7 @@ public abstract class Mob extends LivingEntity implements Targeting {
     protected void registerGoals() {}
 
     public static AttributeSupplier.Builder createMobAttributes() {
-        return LivingEntity.createLivingAttributes().add(Attributes.FOLLOW_RANGE, 16.0D).add(Attributes.ATTACK_KNOCKBACK).add(Attributes.ATTACK_SPEED, 80.0);
+        return LivingEntity.createLivingAttributes().add(Attributes.FOLLOW_RANGE, 16.0D).add(Attributes.ATTACK_KNOCKBACK).add(Attributes.ATTACK_SPEED, 80.0).add(Attributes.ATTACK_DAMAGE, 128);
     }
 
     protected PathNavigation createNavigation(Level world) {
@@ -1272,6 +1272,12 @@ public abstract class Mob extends LivingEntity implements Targeting {
         RandomSource randomsource = world.getRandom();
 
         this.getAttribute(Attributes.FOLLOW_RANGE).addPermanentModifier(new AttributeModifier("Random spawn bonus", randomsource.triangle(0.0D, 0.11485000000000001D), AttributeModifier.Operation.MULTIPLY_BASE));
+        // Papyrus
+        this.getAttribute(Attributes.MAX_HEALTH).addPermanentModifier(new AttributeModifier("Papyrus Random spawn bonus", randomsource.triangle(0.0D, 0.33D), AttributeModifier.Operation.MULTIPLY_BASE));
+        this.getAttribute(Attributes.ATTACK_DAMAGE).addPermanentModifier(new AttributeModifier("Papyrus Random spawn bonus", randomsource.triangle(0.0D, 0.2D), AttributeModifier.Operation.MULTIPLY_BASE));
+        this.getAttribute(Attributes.ADV_PHYSICAL).addPermanentModifier(new AttributeModifier("Papyrus Random spawn bonus", randomsource.triangle(0.0D, 0.33D), AttributeModifier.Operation.MULTIPLY_BASE));
+        this.getAttribute(Attributes.RES_PHYSICAL).addPermanentModifier(new AttributeModifier("Papyrus Random spawn bonus", randomsource.triangle(0.0D, 0.33D), AttributeModifier.Operation.MULTIPLY_BASE));
+
         if (randomsource.nextFloat() < 0.05F) {
             this.setLeftHanded(true);
         } else {
diff --git a/src/main/java/net/minecraft/world/item/ItemStack.java b/src/main/java/net/minecraft/world/item/ItemStack.java
index 4ab0452af6eeae9316311d473be23de9a74d9e4b..4637149278b3ab316c589a41335a595dd2e68ee9 100644
--- a/src/main/java/net/minecraft/world/item/ItemStack.java
+++ b/src/main/java/net/minecraft/world/item/ItemStack.java
@@ -23,12 +23,15 @@ import java.util.function.Predicate;
 import java.util.stream.Collectors;
 import java.util.stream.Stream;
 import javax.annotation.Nullable;
+
+import de.erethon.hephaestus.HItemStack;
+import de.erethon.hephaestus.HItemBehaviour;
+import de.erethon.hephaestus.HItem;
 import net.minecraft.ChatFormatting;
 import net.minecraft.Util;
 import net.minecraft.advancements.CriteriaTriggers;
 import net.minecraft.commands.arguments.blocks.BlockStateParser;
 import net.minecraft.core.BlockPos;
-import net.minecraft.core.Direction;
 import net.minecraft.core.Holder;
 import net.minecraft.core.HolderSet;
 import net.minecraft.core.Registry;
@@ -45,8 +48,8 @@ import net.minecraft.network.chat.ComponentUtils;
 import net.minecraft.network.chat.HoverEvent;
 import net.minecraft.network.chat.MutableComponent;
 import net.minecraft.network.chat.Style;
-import net.minecraft.network.protocol.game.ClientboundBlockUpdatePacket;
 import net.minecraft.resources.ResourceLocation;
+import net.minecraft.server.Main;
 import net.minecraft.tags.TagKey;
 import net.minecraft.util.ExtraCodecs;
 import net.minecraft.util.RandomSource;
@@ -84,7 +87,7 @@ import org.slf4j.Logger;
 // CraftBukkit start
 import com.mojang.serialization.Dynamic;
 import java.util.Map;
-import java.util.Objects;
+
 import net.minecraft.server.MinecraftServer;
 import net.minecraft.server.level.ServerLevel;
 import net.minecraft.server.level.ServerPlayer;
@@ -234,6 +237,9 @@ public final class ItemStack {
         return net.minecraft.nbt.StringTag.valueOf(org.bukkit.craftbukkit.util.CraftChatMessage.toJSON(component));
     }
     // Paper end
+    // Papyrus start
+    public HItemStack hItemStack;
+    public HItemBehaviour behaviour;
 
     public ItemStack(ItemLike item) {
         this(item, 1);
@@ -258,7 +264,6 @@ public final class ItemStack {
         if (this.item.canBeDepleted()) {
             this.setDamageValue(this.getDamageValue());
         }
-
     }
 
     private ItemStack(@Nullable Void v) {
@@ -289,7 +294,33 @@ public final class ItemStack {
         if (this.getItem().canBeDepleted()) {
             this.setDamageValue(this.getDamageValue());
         }
-
+        if (is(Items.AIR)) {
+            return;
+        }
+        if (tag != null && tag.contains("HephaestusID")) { // Papyrus - Hephaestus
+            String id = tag.getString("HephaestusID");
+            HItem hConfig = Main.itemLibrary.get(new ResourceLocation(id));
+            if (hConfig != null) {
+                hItemStack = new HItemStack(this, hConfig);
+                hItemStack.setStack(this);
+                if (hItemStack.getBehaviour() != null) {
+                    behaviour = hItemStack.getBehaviour();
+                }
+            } else {
+                LOGGER.info("[Hephaestus] Unable to find hItem for " + id);
+            }
+        } else {
+            HItem hConfig = Main.itemLibrary.get(BuiltInRegistries.ITEM.getKey(item.asItem()));
+            if (hConfig != null) {
+                hItemStack = new HItemStack(this, hConfig);
+                hItemStack.setStack(this);
+                if (hItemStack.getBehaviour() != null) {
+                    behaviour = hItemStack.getBehaviour();
+                }
+            } else {
+                LOGGER.info("[Hephaestus] Unable to find vanilla hItem for " + item.asItem().toString());
+            }
+        }
     }
 
     private ItemStack(CompoundTag nbt) {
@@ -311,7 +342,10 @@ public final class ItemStack {
     }
 
     public boolean isItemEnabled(FeatureFlagSet enabledFeatures) {
-        return this.isEmpty() || this.getItem().isEnabled(enabledFeatures);
+        if (hItemStack != null) {
+            return hItemStack.isEnabled() || isEmpty();
+        }
+        return this.isEmpty() || (this.getItem().isEnabled(enabledFeatures));
     }
 
     public ItemStack split(int amount) {
@@ -584,14 +618,20 @@ public final class ItemStack {
         nbt.putString("id", minecraftkey == null ? "minecraft:air" : minecraftkey.toString());
         nbt.putByte("Count", (byte) this.count);
         if (this.tag != null) {
+            if (hItemStack != null && !hItemStack.isVanilla()) { // Papyrus
+                tag.putString("HephaestusID", hItemStack.getId().toString());
+                LOGGER.info("Saved item." + tag.toString());
+            }
             nbt.put("tag", this.tag.copy());
         }
-
         return nbt;
     }
 
     public int getMaxStackSize() {
-        return this.getItem().getMaxStackSize();
+        if (hItemStack == null) {
+            return getItem().getMaxStackSize();
+        }
+        return hItemStack.config().getMaxStackSize(); // Papyrus
     }
 
     public boolean isStackable() {
@@ -621,7 +661,10 @@ public final class ItemStack {
     }
 
     public int getMaxDamage() {
-        return this.getItem().getMaxDamage();
+        if (hItemStack == null) {
+            return getItem().getMaxDamage();
+        }
+        return hItemStack.config().getMaxDurability();
     }
 
     public boolean hurt(int amount, RandomSource random, @Nullable LivingEntity player) { // Paper - allow any living entity instead of only ServerPlayers
@@ -675,6 +718,9 @@ public final class ItemStack {
             }
 
             j = this.getDamageValue() + amount;
+            if (behaviour != null) {
+                j = behaviour.onItemHurt(this, j);
+            }
             this.setDamageValue(j);
             return j >= this.getMaxDamage();
         }
@@ -727,7 +773,7 @@ public final class ItemStack {
     public void hurtEnemy(LivingEntity target, net.minecraft.world.entity.player.Player attacker) {
         Item item = this.getItem();
 
-        if (item.hurtEnemy(this, target, attacker)) {
+        if (item.hurtEnemy(this, target, attacker) && (behaviour != null && behaviour.onAttack(this, target, attacker)) ) {
             attacker.awardStat(Stats.ITEM_USED.get(item));
         }
 
@@ -736,7 +782,7 @@ public final class ItemStack {
     public void mineBlock(Level world, net.minecraft.world.level.block.state.BlockState state, BlockPos pos, net.minecraft.world.entity.player.Player miner) {
         Item item = this.getItem();
 
-        if (item.mineBlock(this, world, state, pos, miner)) {
+        if (item.mineBlock(this, world, state, pos, miner) && (behaviour != null && behaviour.onMine(this, world, state, pos, miner))) {
             miner.awardStat(Stats.ITEM_USED.get(item));
         }
 
@@ -747,6 +793,9 @@ public final class ItemStack {
     }
 
     public InteractionResult interactLivingEntity(net.minecraft.world.entity.player.Player user, LivingEntity entity, InteractionHand hand) {
+        if (behaviour != null) {
+            return behaviour.onInteractLiving(this, user, entity, hand);
+        }
         return this.getItem().interactLivingEntity(this, user, entity, hand);
     }
 
@@ -766,6 +815,9 @@ public final class ItemStack {
             if (this.tag != null) {
                 itemstack.tag = this.tag.copy();
             }
+            if (hItemStack != null) {
+                itemstack.hItemStack = hItemStack;
+            }
 
             return itemstack;
         }
@@ -811,6 +863,9 @@ public final class ItemStack {
 
         if (this.getItem() != null) {
             this.getItem().inventoryTick(this, world, entity, slot, selected);
+            if (behaviour != null) {
+                behaviour.onTick(this, entity.getBukkitEntity(), slot, selected);
+            }
         }
 
     }
@@ -926,6 +981,33 @@ public final class ItemStack {
 
         if (nbt != null) {
             this.getItem().verifyTagAfterLoad(nbt);
+            if (is(Items.AIR)) {
+                return;
+            }
+            if (tag != null && tag.contains("HephaestusID")) { // Papyrus - Hephaestus
+                String id = tag.getString("HephaestusID");
+                HItem hConfig = Main.itemLibrary.get(new ResourceLocation(id));
+                if (hConfig != null) {
+                    hItemStack = new HItemStack(this, hConfig);
+                    hItemStack.setStack(this);
+                    if (hItemStack.getBehaviour() != null) {
+                        behaviour = hItemStack.getBehaviour();
+                    }
+                } else {
+                    LOGGER.info("[Hephaestus] Unable to find hItem for " + id);
+                }
+            } else {
+                HItem hConfig = Main.itemLibrary.get(BuiltInRegistries.ITEM.getKey(item.asItem()));
+                if (hConfig != null) {
+                    hItemStack = new HItemStack(this, hConfig);
+                    hItemStack.setStack(this);
+                    if (hItemStack.getBehaviour() != null) {
+                        behaviour = hItemStack.getBehaviour();
+                    }
+                } else {
+                    LOGGER.info("[Hephaestus] Unable to find vanilla hItem for " + item.asItem().toString());
+                }
+            }
         }
 
     }
diff --git a/src/main/java/net/minecraft/world/level/NaturalSpawner.java b/src/main/java/net/minecraft/world/level/NaturalSpawner.java
index 9c2d62feff1816f5729060c6192269a5b2d34153..2c3d440bb59707f1708eba1261d0b8f8c554b77f 100644
--- a/src/main/java/net/minecraft/world/level/NaturalSpawner.java
+++ b/src/main/java/net/minecraft/world/level/NaturalSpawner.java
@@ -1,6 +1,8 @@
 package net.minecraft.world.level;
 
 import com.mojang.logging.LogUtils;
+import de.erethon.hephaestus.events.MobSpawnEntityCreateEvent;
+import de.erethon.hephaestus.events.MobSpawnSettingsEvent;
 import it.unimi.dsi.fastutil.objects.Object2IntMap;
 import it.unimi.dsi.fastutil.objects.Object2IntMaps;
 import it.unimi.dsi.fastutil.objects.Object2IntOpenHashMap;
@@ -25,13 +27,7 @@ import net.minecraft.util.Mth;
 import net.minecraft.util.RandomSource;
 import net.minecraft.util.VisibleForDebug;
 import net.minecraft.util.random.WeightedRandomList;
-import net.minecraft.world.entity.Entity;
-import net.minecraft.world.entity.EntityType;
-import net.minecraft.world.entity.Mob;
-import net.minecraft.world.entity.MobCategory;
-import net.minecraft.world.entity.MobSpawnType;
-import net.minecraft.world.entity.SpawnGroupData;
-import net.minecraft.world.entity.SpawnPlacements;
+import net.minecraft.world.entity.*;
 import net.minecraft.world.entity.player.Player;
 import net.minecraft.world.level.biome.Biome;
 import net.minecraft.world.level.biome.MobSpawnSettings;
@@ -287,8 +283,13 @@ public final class NaturalSpawner {
                                     }
                                     if (doSpawning == PreSpawnStatus.SUCCESS && checker.test(biomesettingsmobs_c.type, blockposition_mutableblockposition, chunk)) {
                                         // Paper end
-                                        Mob entityinsentient = NaturalSpawner.getMobForSpawn(world, biomesettingsmobs_c.type);
-
+                                        // Papyrus start - use our spawn method
+                                        MobSpawnEntityCreateEvent event = getPapyrusMobForSpawn(world, biomesettingsmobs_c, blockposition_mutableblockposition, entityhuman, groupdataentity);
+                                        Mob entityinsentient = event.mob;
+                                        biomesettingsmobs_c = event.spawnerData;
+                                        groupdataentity = event.data;
+                                        blockposition_mutableblockposition = event.pos;
+                                        // Papyrus end
                                         if (entityinsentient == null) {
                                             return j; // Paper
                                         }
@@ -401,14 +402,30 @@ public final class NaturalSpawner {
         return null;
     }
 
+    private static MobSpawnEntityCreateEvent getPapyrusMobForSpawn(ServerLevel level, MobSpawnSettings.SpawnerData spawnerData, BlockPos.MutableBlockPos pos, Player player, @Nullable SpawnGroupData groupData) {
+        try {
+            Entity entity = spawnerData.type.create(level);
+            if (entity instanceof Mob mob) {
+                MobSpawnEntityCreateEvent event = new MobSpawnEntityCreateEvent(mob, level, spawnerData, pos, player, groupData);
+                event.callEvent();
+                return event;
+            }
+            NaturalSpawner.LOGGER.warn("Can't spawn entity of type: {}", BuiltInRegistries.ENTITY_TYPE.getKey(spawnerData.type));
+        } catch (Exception exception) {
+            NaturalSpawner.LOGGER.warn("Failed to create mob", exception);
+            com.destroystokyo.paper.exception.ServerInternalException.reportInternalException(exception); // Paper
+        }
+        return null;
+    }
+
     private static boolean isValidPositionForMob(ServerLevel world, Mob entity, double squaredDistance) {
         return squaredDistance > (double) (entity.getType().getCategory().getDespawnDistance() * entity.getType().getCategory().getDespawnDistance()) && entity.removeWhenFarAway(squaredDistance) ? false : entity.checkSpawnRules(world, MobSpawnType.NATURAL) && entity.checkSpawnObstruction(world);
     }
 
     private static Optional<MobSpawnSettings.SpawnerData> getRandomSpawnMobAt(ServerLevel world, StructureManager structureAccessor, ChunkGenerator chunkGenerator, MobCategory spawnGroup, RandomSource random, BlockPos pos) {
         Holder<Biome> holder = world.getBiome(pos);
-
-        return spawnGroup == MobCategory.WATER_AMBIENT && holder.is(BiomeTags.REDUCED_WATER_AMBIENT_SPAWNS) && random.nextFloat() < 0.98F ? Optional.empty() : NaturalSpawner.mobsAt(world, structureAccessor, chunkGenerator, spawnGroup, pos, holder).getRandom(random);
+        Optional<MobSpawnSettings.SpawnerData> data = spawnGroup == MobCategory.WATER_AMBIENT && holder.is(BiomeTags.REDUCED_WATER_AMBIENT_SPAWNS) && random.nextFloat() < 0.98F ? Optional.empty() : NaturalSpawner.mobsAt(world, structureAccessor, chunkGenerator, spawnGroup, pos, holder).getRandom(random);
+        return data;
     }
 
     private static boolean canSpawnMobAt(ServerLevel world, StructureManager structureAccessor, ChunkGenerator chunkGenerator, MobCategory spawnGroup, MobSpawnSettings.SpawnerData spawnEntry, BlockPos pos) {
@@ -416,6 +433,12 @@ public final class NaturalSpawner {
     }
 
     private static WeightedRandomList<MobSpawnSettings.SpawnerData> mobsAt(ServerLevel world, StructureManager structureAccessor, ChunkGenerator chunkGenerator, MobCategory spawnGroup, BlockPos pos, @Nullable Holder<Biome> biomeEntry) {
+        // Papyrus start
+        MobSpawnSettingsEvent event = new MobSpawnSettingsEvent(world, structureAccessor, chunkGenerator, spawnGroup, pos, biomeEntry);
+        event.callEvent();
+        if (event.spawnerData != null) {
+            return event.spawnerData;
+        }
         return NaturalSpawner.isInNetherFortressBounds(pos, world, spawnGroup, structureAccessor) ? NetherFortressStructure.FORTRESS_ENEMIES : chunkGenerator.getMobsAt(biomeEntry != null ? biomeEntry : world.getBiome(pos), structureAccessor, spawnGroup, pos);
     }
 
diff --git a/src/main/java/net/minecraft/world/level/biome/Biome.java b/src/main/java/net/minecraft/world/level/biome/Biome.java
index efca73d4de33028cf9df944f36e51b7b50f7a4c5..e28347cf823578ac9e002332c3ac9ffd59a4bf6f 100644
--- a/src/main/java/net/minecraft/world/level/biome/Biome.java
+++ b/src/main/java/net/minecraft/world/level/biome/Biome.java
@@ -4,6 +4,7 @@ import com.google.common.collect.ImmutableList;
 import com.mojang.serialization.Codec;
 import com.mojang.serialization.MapCodec;
 import com.mojang.serialization.codecs.RecordCodecBuilder;
+import de.erethon.hephaestus.events.MobSpawnSettingsEvent;
 import it.unimi.dsi.fastutil.longs.Long2FloatLinkedOpenHashMap;
 import java.util.Optional;
 import javax.annotation.Nullable;
diff --git a/src/main/java/net/minecraft/world/level/block/Block.java b/src/main/java/net/minecraft/world/level/block/Block.java
index e21867d3956078bb0db4ceed45e5811e9acd7377..9e1ae79ad84045933e435cc5cb08e7137dfa0790 100644
--- a/src/main/java/net/minecraft/world/level/block/Block.java
+++ b/src/main/java/net/minecraft/world/level/block/Block.java
@@ -304,7 +304,7 @@ public class Block extends BlockBehaviour implements ItemLike {
     public static List<ItemStack> getDrops(BlockState state, ServerLevel world, BlockPos pos, @Nullable BlockEntity blockEntity, @Nullable Entity entity, ItemStack stack) {
         LootParams.Builder lootparams_a = (new LootParams.Builder(world)).withParameter(LootContextParams.ORIGIN, Vec3.atCenterOf(pos)).withParameter(LootContextParams.TOOL, stack).withOptionalParameter(LootContextParams.THIS_ENTITY, entity).withOptionalParameter(LootContextParams.BLOCK_ENTITY, blockEntity);
 
-        return state.getDrops(lootparams_a);
+        return state.   getDrops(lootparams_a);
     }
 
     public static void dropResources(BlockState state, Level world, BlockPos pos) {
diff --git a/src/main/java/net/minecraft/world/level/chunk/ChunkStatus.java b/src/main/java/net/minecraft/world/level/chunk/ChunkStatus.java
index a907b79fd8291a0e92db138f37239d17424188a1..aeaf6a32a06d375c3e5fd0f88c23d9503d78758a 100644
--- a/src/main/java/net/minecraft/world/level/chunk/ChunkStatus.java
+++ b/src/main/java/net/minecraft/world/level/chunk/ChunkStatus.java
@@ -139,7 +139,7 @@ public class ChunkStatus {
     });
     public static final ChunkStatus SPAWN = ChunkStatus.registerSimple("spawn", ChunkStatus.LIGHT, 0, ChunkStatus.POST_FEATURES, ChunkStatus.ChunkType.PROTOCHUNK, (chunkstatus, worldserver, chunkgenerator, list, ichunkaccess) -> {
         if (!ichunkaccess.isUpgrading()) {
-            chunkgenerator.spawnOriginalMobs(new WorldGenRegion(worldserver, list, chunkstatus, -1));
+            // chunkgenerator.spawnOriginalMobs(new WorldGenRegion(worldserver, list, chunkstatus, -1)); Papyrus - disable mob spawn during chunk gen
         }
 
     });
diff --git a/src/main/java/org/bukkit/craftbukkit/CraftServer.java b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
index c0cb4358e013372aa880aac9e40639f865246e6b..07a373f7ccbe298e579e7db268b1662a0df5d393 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftServer.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
@@ -14,6 +14,7 @@ import com.mojang.brigadier.tree.CommandNode;
 import com.mojang.brigadier.tree.LiteralCommandNode;
 import com.mojang.serialization.Dynamic;
 import com.mojang.serialization.Lifecycle;
+import de.erethon.hephaestus.ItemLibrary;
 import de.erethon.spellbook.api.SpellbookAPI;
 import it.unimi.dsi.fastutil.objects.Object2IntOpenHashMap;
 import java.awt.image.BufferedImage;
