From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Malfrador <malfrador@gmail.com>
Date: Sat, 19 Mar 2022 17:48:55 +0100
Subject: [PATCH] allow disabling in block check


diff --git a/src/main/java/net/minecraft/world/entity/Entity.java b/src/main/java/net/minecraft/world/entity/Entity.java
index 84fce7dccf9232209f939a32acfc3131e62eb27c..1ab45f09acbe80290f0a90413b853137e26fdbee 100644
--- a/src/main/java/net/minecraft/world/entity/Entity.java
+++ b/src/main/java/net/minecraft/world/entity/Entity.java
@@ -336,6 +336,7 @@ public abstract class Entity implements Nameable, EntityAccess, CommandSource, i
     @javax.annotation.Nullable
     private UUID originWorld;
     public boolean freezeLocked = false; // Paper - Freeze Tick Lock API
+    public boolean noWallCollision; // Papyrus - allow disabling collision with blocks
 
     public void setOrigin(@javax.annotation.Nonnull Location location) {
         this.origin = location.toVector();
@@ -1415,6 +1416,9 @@ public abstract class Entity implements Nameable, EntityAccess, CommandSource, i
     }
 
     protected void checkInsideBlocks() {
+        if (noWallCollision) { // Papyrus - allow disabling wall checks
+            return;
+        }
         AABB axisalignedbb = this.getBoundingBox();
         BlockPos blockposition = new BlockPos(axisalignedbb.minX + 0.001D, axisalignedbb.minY + 0.001D, axisalignedbb.minZ + 0.001D);
         BlockPos blockposition1 = new BlockPos(axisalignedbb.maxX - 0.001D, axisalignedbb.maxY - 0.001D, axisalignedbb.maxZ - 0.001D);
@@ -2462,6 +2466,9 @@ public abstract class Entity implements Nameable, EntityAccess, CommandSource, i
     }
 
     public boolean isInWall() {
+        if (noWallCollision) { // Papyrus - allow disabling collision
+            return false;
+        }
         if (this.noPhysics) {
             return false;
         } else {
