From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Jason Penilla <11360596+jpenilla@users.noreply.github.com>
Date: Wed, 11 Dec 2024 19:23:25 -0700
Subject: [PATCH] 'feature patch'


diff --git a/net/minecraft/DefaultUncaughtExceptionHandler.java b/net/minecraft/DefaultUncaughtExceptionHandler.java
index 6341da454460aa6c00075f82163b97286f760eee..a8bc94121a0209faa8dd293ff3c0862dac280cdb 100644
--- a/net/minecraft/DefaultUncaughtExceptionHandler.java
+++ b/net/minecraft/DefaultUncaughtExceptionHandler.java
@@ -12,6 +12,7 @@ public class DefaultUncaughtExceptionHandler implements UncaughtExceptionHandler
 
     @Override
     public void uncaughtException(Thread thread, Throwable exception) {
+        System.out.println("im a feature");
         this.logger.error("Caught previously unhandled exception :", exception);
     }
 }
